// @inworld/runtime/benchmarks/voice_agent_guinea_pig/server/components/event_factory.ts
import { v4 } from 'uuid';

const toMB = (b: number) => +(b / 1024 / 1024).toFixed(2);

interface CpuUsage {
  user: number;
  system: number;
}

interface MemoryUsage {
  rss: number;
  heapTotal: number;
  heapUsed: number;
  external: number;
}

export class EventFactory {
  static text(
    text: string,
    interactionId: string,
    source: { isAgent?: boolean; isUser?: boolean; name?: string },
  ) {
    const date = new Date();

    return {
      type: 'TEXT',
      text: { text, final: true },
      date,
      packetId: { utteranceId: v4(), interactionId },
      routing: { source },
    };
  }

  static error(error: Error, code: string, interactionId: string) {
    return {
      type: 'ERROR',
      error: error.toString(),
      code: code,
      packetId: { interactionId },
    };
  }

  static interactionEnd(interactionId: string) {
    return {
      type: 'INTERACTION_END',
      date: new Date(),
      packetId: { interactionId },
    };
  }

  static audio(audio: string, interactionId: string, utteranceId: string) {
    return {
      type: 'AUDIO',
      audio: { chunk: audio },
      date: new Date(),
      packetId: { utteranceId, interactionId },
      routing: { source: { isAgent: true } },
    };
  }

  static newInteraction(interactionId: string, interruptionEnabled: boolean) {
    return {
      type: 'NEW_INTERACTION',
      date: new Date(),
      packetId: { interactionId },
      interruptionEnabled,
    };
  }

  static memorySnapshot(): { data: MemoryUsage; type: string } {
    const mem = process.memoryUsage();
    const record = {
      timestamp: new Date().toISOString(),
      rss: toMB(mem.rss),
      heapTotal: toMB(mem.heapTotal),
      heapUsed: toMB(mem.heapUsed),
      external: toMB(mem.external),
    };
    return {
      type: 'MEMORY_SNAPSHOT',
      data: record,
    };
  }

  static cpuSnapshot(): {
    data: CpuUsage;
    type: string;
  } {
    const cpuUsage = process.cpuUsage();
    const record = {
      system: cpuUsage.system,
      user: cpuUsage.user,
    };
    return {
      type: 'CPU_SNAPSHOT',
      data: record,
    };
  }
}
