// @inworld/runtime/build/common/stream/abstract_stream_type.js
"use strict";
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractStreamType = void 0;
const internal_1 = __importDefault(require("../../internal/decorators/internal"));
/**
 * Abstract base class for all API-level data types.
 */
let AbstractStreamType = (() => {
    var _a;
    let _instanceExtraInitializers = [];
    let _stream_decorators;
    let _stream_initializers = [];
    let _stream_extraInitializers = [];
    let _onDestroy_decorators;
    let _onDestroy_initializers = [];
    let _onDestroy_extraInitializers = [];
    let _getStream_decorators;
    return _a = class AbstractStreamType {
            constructor(stream, onDestroy) {
                /**
                 * @internal
                 */
                this.stream = (__runInitializers(this, _instanceExtraInitializers), __runInitializers(this, _stream_initializers, void 0));
                /**
                 * @internal
                 */
                this.onDestroy = (__runInitializers(this, _stream_extraInitializers), __runInitializers(this, _onDestroy_initializers, void 0));
                __runInitializers(this, _onDestroy_extraInitializers);
                this.stream = stream;
                this.onDestroy = onDestroy;
            }
            static isStreamType(value) {
                return value instanceof _a;
            }
            /**
             * @internal
             */
            getStream() {
                return this.stream;
            }
            /**
             * Gets metadata from the underlying stream.
             * Returns an empty object if metadata is not available.
             *
             * @returns {Record<string, any>} Metadata object from the stream
             */
            getMetadata() {
                var _b;
                if (typeof ((_b = this.stream) === null || _b === void 0 ? void 0 : _b.getMetadata) === 'function') {
                    return this.stream.getMetadata();
                }
                return {};
            }
        },
        (() => {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _stream_decorators = [internal_1.default];
            _onDestroy_decorators = [internal_1.default];
            _getStream_decorators = [internal_1.default];
            __esDecorate(_a, null, _getStream_decorators, { kind: "method", name: "getStream", static: false, private: false, access: { has: obj => "getStream" in obj, get: obj => obj.getStream }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(null, null, _stream_decorators, { kind: "field", name: "stream", static: false, private: false, access: { has: obj => "stream" in obj, get: obj => obj.stream, set: (obj, value) => { obj.stream = value; } }, metadata: _metadata }, _stream_initializers, _stream_extraInitializers);
            __esDecorate(null, null, _onDestroy_decorators, { kind: "field", name: "onDestroy", static: false, private: false, access: { has: obj => "onDestroy" in obj, get: obj => obj.onDestroy, set: (obj, value) => { obj.onDestroy = value; } }, metadata: _metadata }, _onDestroy_initializers, _onDestroy_extraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        _a;
})();
exports.AbstractStreamType = AbstractStreamType;
