// @inworld/runtime/build/common/stream/multimodal_content.js
"use strict";
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultimodalContentStream = void 0;
const internal_1 = __importDefault(require("../../internal/decorators/internal"));
const inworld_1 = require("../../internal/error/inworld");
const pb_1 = require("../../internal/pb");
const audio_1 = require("../data_types/api/audio");
const multimodal_content_1 = require("../data_types/api/multimodal_content");
const abstract_stream_type_1 = require("./abstract_stream_type");
/**
 * Class for handling multimodal content streams using addon async iterator protocol.
 * Provides functionality to read multimodal content (text or audio) from native stream implementations.
 */
let MultimodalContentStream = (() => {
    var _a;
    let _classSuper = abstract_stream_type_1.AbstractStreamType;
    let _staticExtraInitializers = [];
    let _static_fromExternal_decorators;
    return _a = class MultimodalContentStream extends _classSuper {
            /**
             * Creates a new MultimodalContentStream instance from a addon stream.
             *
             * @param {any} stream - addon stream object that implements async iterator protocol
             * @param {() => void} [onDestroy] - Optional callback to call when the stream is destroyed
             */
            constructor(stream, onDestroy) {
                super(stream, onDestroy);
            }
            /**
             * Retrieves the next multimodal content item from the stream using addon async iterator protocol.
             *
             * @returns {Promise<MultimodalContentStreamIterationResult>} A promise that resolves to the next multimodal content result.
             */
            async next() {
                var _b, _c;
                try {
                    const iteratorResult = await this.stream.next();
                    if (iteratorResult.done) {
                        (_b = this.onDestroy) === null || _b === void 0 ? void 0 : _b.call(this);
                        return {
                            done: true,
                        };
                    }
                    const { text, audio: audioData } = pb_1.PbType.unpackMultimodalContent(iteratorResult.value);
                    const multimodalContent = new multimodal_content_1.MultimodalContent({
                        text: text,
                        audio: audioData
                            ? new audio_1.Audio({
                                data: Array.from(audioData.data),
                                sampleRate: audioData.sampleRate,
                            })
                            : undefined,
                    });
                    return {
                        value: multimodalContent,
                        done: false,
                    };
                }
                catch (error) {
                    (_c = this.onDestroy) === null || _c === void 0 ? void 0 : _c.call(this);
                    throw new inworld_1.InworldError('Failed to read multimodal content stream', (error === null || error === void 0 ? void 0 : error.message) || 'Unknown stream error');
                }
            }
            /**
             * Implements the AsyncIterator interface for the MultimodalContentStream.
             *
             * @returns {AsyncIterator<MultimodalContentStreamIterationResult>} An async iterator for the multimodal content stream.
             */
            [Symbol.asyncIterator]() {
                return {
                    next: async () => {
                        const result = await this.next();
                        if (result.done) {
                            return { done: true, value: undefined };
                        }
                        return { done: false, value: result.value };
                    },
                };
            }
            /**
             * Creates a new MultimodalContentStream instance from a addon stream.
             *
             * @param {RuntimeInputStream<any>} stream - addon stream object or external reference
             * @returns {MultimodalContentStream} A new MultimodalContentStream instance
             * @internal
             */
            static fromExternal(stream) {
                return new _a(stream);
            }
            /**
             * Converts the stream to external format for addon consumption.
             *
             * @returns {StreamingTaggedTypeBase<'MultimodalContentStream', { stream: true }, any>} Tagged type with stream metadata
             * @internal
             */
            toExternal() {
                return {
                    type: 'MultimodalContentStream',
                    data: { stream: true },
                    _stream: this.getStream(),
                };
            }
        },
        (() => {
            var _b;
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create((_b = _classSuper[Symbol.metadata]) !== null && _b !== void 0 ? _b : null) : void 0;
            _static_fromExternal_decorators = [internal_1.default];
            __esDecorate(_a, null, _static_fromExternal_decorators, { kind: "method", name: "fromExternal", static: true, private: false, access: { has: obj => "fromExternal" in obj, get: obj => obj.fromExternal }, metadata: _metadata }, null, _staticExtraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            __runInitializers(_a, _staticExtraInitializers);
        })(),
        _a;
})();
exports.MultimodalContentStream = MultimodalContentStream;
