// @inworld/runtime/build/common/stream/tts_output.js
"use strict";
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TTSOutputStream = void 0;
const error_handling_1 = require("../../graph/dsl/error_handling");
const internal_1 = __importDefault(require("../../internal/decorators/internal"));
const inworld_1 = require("../../internal/error/inworld");
const pb_1 = require("../../internal/pb");
const audio_1 = require("../data_types/api/audio");
const abstract_stream_type_1 = require("./abstract_stream_type");
/**
 * Class for handling TTS output streams using addon async iterator protocol.
 * Provides functionality to read audio and text data from native stream implementations.
 * @internal
 */
let TTSOutputStream = (() => {
    var _a;
    let _classSuper = abstract_stream_type_1.AbstractStreamType;
    let _staticExtraInitializers = [];
    let _static_fromExternal_decorators;
    return _a = class TTSOutputStream extends _classSuper {
            /**
             * Creates a new TTSOutputStream instance from a addon stream.
             *
             * @param stream - addon stream object that implements async iterator protocol
             * @param onDestroy - Optional callback to execute when stream is consumed
             */
            constructor(stream, onDestroy) {
                super(stream, onDestroy);
            }
            /**
             * Aborts the TTS stream, canceling any pending operations.
             * This can be used to interrupt ongoing TTS generation.
             */
            abort() {
                var _b;
                if (!this.stream || typeof this.stream.abort !== 'function') {
                    return;
                }
                this.stream.abort();
                (_b = this.onDestroy) === null || _b === void 0 ? void 0 : _b.call(this);
            }
            /**
             * Asynchronously reads the next item from the stream using addon async iterator protocol.
             *
             * @returns Promise resolving to a result object with audio/text content or done flag
             * @throws InworldError if reading from the stream fails
             */
            async next() {
                var _b, _c, _d, _e;
                try {
                    const iteratorResult = await this.stream.next();
                    if (iteratorResult.done) {
                        (_b = this.onDestroy) === null || _b === void 0 ? void 0 : _b.call(this);
                        return { done: true };
                    }
                    const { text, audio } = pb_1.PbType.unpackTTSOutput(iteratorResult.value);
                    let result = {
                        done: false,
                    };
                    if (audio === null || audio === void 0 ? void 0 : audio.data) {
                        const numberArray = Array.from(audio === null || audio === void 0 ? void 0 : audio.data);
                        result.audio = new audio_1.Audio({
                            data: numberArray,
                            sampleRate: audio.sampleRate,
                        });
                    }
                    if (text) {
                        result.text = text;
                    }
                    return result;
                }
                catch (error) {
                    (_c = this.onDestroy) === null || _c === void 0 ? void 0 : _c.call(this);
                    if ((error === null || error === void 0 ? void 0 : error.status) === error_handling_1.ErrorStatusCode.Cancelled) {
                        return { done: true };
                    }
                    // temporary hack to support realtime demo. Should be removed once new binaries arrived https://github.com/inworld-ai/inworld-framework-cpp/pull/2180
                    if (/CANCELLED/i.test((_d = error === null || error === void 0 ? void 0 : error.message) !== null && _d !== void 0 ? _d : '') ||
                        /cancelled|canceled/i.test((_e = error === null || error === void 0 ? void 0 : error.message) !== null && _e !== void 0 ? _e : '')) {
                        return { done: true };
                    }
                    throw new inworld_1.InworldError('Failed to read TTS output stream', (error === null || error === void 0 ? void 0 : error.message) || 'Unknown stream error');
                }
            }
            /**
             * Creates a new TTSOutputStream instance from a addon stream.
             *
             * @param stream - addon stream object or external reference
             * @returns A new TTSOutputStream instance
             * @internal
             */
            static fromExternal(stream) {
                return new _a(stream);
            }
            /**
             * Converts the stream to external format for addon consumption.
             *
             * @returns {StreamingTaggedTypeBase<'TTSOutputStream', { stream: true }, any>} Tagged type with stream metadata
             * @internal
             */
            toExternal() {
                return {
                    type: 'TTSOutputStream',
                    data: { stream: true },
                    _stream: this.getStream(),
                };
            }
            [(_static_fromExternal_decorators = [internal_1.default], Symbol.asyncIterator)]() {
                return {
                    next: async () => {
                        const result = await this.next();
                        if (result.done) {
                            return { done: true, value: undefined };
                        }
                        return { done: false, value: result };
                    },
                };
            }
        },
        (() => {
            var _b;
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create((_b = _classSuper[Symbol.metadata]) !== null && _b !== void 0 ? _b : null) : void 0;
            __esDecorate(_a, null, _static_fromExternal_decorators, { kind: "method", name: "fromExternal", static: true, private: false, access: { has: obj => "fromExternal" in obj, get: obj => obj.fromExternal }, metadata: _metadata }, null, _staticExtraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            __runInitializers(_a, _staticExtraInitializers);
        })(),
        _a;
})();
exports.TTSOutputStream = TTSOutputStream;
