// @inworld/runtime/build/graph/dsl/components/fake_tts_component.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FakeTTSComponent = void 0;
const snakify_1 = __importDefault(require("../utils/snakify"));
const abstract_component_1 = require("./abstract_component");
/**
 * Fake TTS (Text-to-Speech) component for load testing.
 * Provides a simulated text-to-speech service for testing purposes
 * with configurable delays, error rates, and chunk generation.
 *
 * @example
 * ```typescript
 * const fakeTtsComponent = new FakeTTSComponent({
 *   loadTestConfig: {
 *     firstChunkDelay: 50,
 *     sampleRate: 16000,
 *     errorProbability: 0.1,
 *     chunksPerRequest: 3,
 *     interChunkDelay: 10,
 *     collectMetrics: true
 *   }
 * });
 * ```
 */
class FakeTTSComponent extends abstract_component_1.AbstractComponent {
    /**
     * Gets the default load test configuration values.
     *
     * @returns Default configuration matching C++ LoadTestTtsConfig defaults.
     */
    static getDefaultLoadTestConfig() {
        return {
            firstChunkDelay: 100, // 100ms
            sampleRate: 16000, // 16kHz
            errorProbability: 0.0, // 0% error rate
            chunksPerRequest: 5, // 5 chunks
            interChunkDelay: 20, // 20ms
            collectMetrics: true, // enabled
        };
    }
    /**
     * Creates a new FakeTTSComponent instance.
     *
     * @param props - Configuration for the fake TTS component.
     */
    constructor(props) {
        super(props);
        this.loadTestConfig = Object.assign(Object.assign({}, FakeTTSComponent.getDefaultLoadTestConfig()), props.loadTestConfig);
    }
    /**
     * Converts the fake TTS component to a graph configuration component.
     *
     * @returns The configured fake TTS component for graph execution.
     */
    toGraphConfigComponent() {
        return {
            id: this.id,
            type: 'FakeTTSInterface',
            creation_config: {
                type: 'library::LoadTestTtsConfig',
                properties: (0, snakify_1.default)(this.loadTestConfig),
            },
        };
    }
}
exports.FakeTTSComponent = FakeTTSComponent;
