// @inworld/runtime/build/graph/dsl/components/local_vad_component.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalVADComponent = void 0;
const snakify_1 = __importDefault(require("../utils/snakify"));
const abstract_component_1 = require("./abstract_component");
/**
 * Local VAD component that wraps `LocalVADConfig`.
 *
 * @remarks
 * Mirrors the realtime-agent blueprint defaults while allowing callers to
 * override the model path, device information, or speech threshold.
 *
 * @example
 * ```typescript
 * const vad = new LocalVADComponent({
 *   id: 'local_vad',
 *   modelPath: '{{VAD_MODEL_PATH}}',
 *   speechThreshold: 0.6,
 * });
 * ```
 */
class LocalVADComponent extends abstract_component_1.AbstractComponent {
    /**
     * Creates a new {@link LocalVADComponent}.
     *
     * @param {LocalVADComponentProps} props - Component configuration.
     */
    constructor(props) {
        super(props);
        this.modelPath = props.modelPath;
        this.device = props.device;
        this.speechThreshold = props.speechThreshold;
    }
    toGraphConfigComponent() {
        var _a, _b;
        const config = {
            modelPath: this.modelPath,
            device: (_a = this.device) !== null && _a !== void 0 ? _a : DEFAULT_DEVICE,
            defaultConfig: {
                speechThreshold: (_b = this.speechThreshold) !== null && _b !== void 0 ? _b : 0.5,
            },
        };
        return {
            id: this.id,
            type: 'VADInterface',
            creation_config: {
                type: 'LocalVADConfig',
                properties: (0, snakify_1.default)(config),
            },
        };
    }
}
exports.LocalVADComponent = LocalVADComponent;
const DEFAULT_DEVICE = {
    type: 'CPU',
    index: -1,
    info: {
        name: 'CPU',
        timestamp: 0,
        freeMemoryBytes: 0,
        totalMemoryBytes: 0,
    },
};
