// @inworld/runtime/build/graph/dsl/components/local_turn_detector_component.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalTurnDetectorComponent = void 0;
const snakify_1 = __importDefault(require("../utils/snakify"));
const abstract_component_1 = require("./abstract_component");
/**
 * Local turn detector component that wraps `LocalTurnDetectorConfig`.
 *
 * @example
 * ```typescript
 * const turnDetector = new LocalTurnDetectorComponent({
 *   id: 'local_turn_detector',
 *   modelPath: '{{TURN_DETECTOR_MODEL_PATH}}',
 *   threshold: 0.55,
 * });
 * ```
 */
class LocalTurnDetectorComponent extends abstract_component_1.AbstractComponent {
    /**
     * Creates a new {@link LocalTurnDetectorComponent}.
     *
     * @param {LocalTurnDetectorComponentProps} props - Component configuration.
     */
    constructor(props) {
        super(props);
        this.modelPath = props.modelPath;
        this.device = props.device;
        this.threshold = props.threshold;
    }
    toGraphConfigComponent() {
        var _a, _b;
        const config = {
            modelPath: this.modelPath,
            device: (_a = this.device) !== null && _a !== void 0 ? _a : DEFAULT_DEVICE,
            defaultConfig: {
                threshold: (_b = this.threshold) !== null && _b !== void 0 ? _b : 0.5,
            },
        };
        return {
            id: this.id,
            type: 'TurnDetectorInterface',
            creation_config: {
                type: 'LocalTurnDetectorConfig',
                properties: (0, snakify_1.default)(config),
            },
        };
    }
}
exports.LocalTurnDetectorComponent = LocalTurnDetectorComponent;
const DEFAULT_DEVICE = {
    type: 'CPU',
    index: -1,
    info: {
        name: 'CPU',
        timestamp: 0,
        freeMemoryBytes: 0,
        totalMemoryBytes: 0,
    },
};
