// @inworld/runtime/build/graph/dsl/components/remote_stt_component.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteSTTComponent = void 0;
const constants_1 = require("../constants");
const snakify_1 = __importDefault(require("../utils/snakify"));
const abstract_component_1 = require("./abstract_component");
/**
 * Remote STT (Speech-to-Text) component for audio transcription services.
 * Provides access to external speech recognition services to convert audio
 * input into text transcriptions.
 *
 * @remarks
 * API key injection is handled by the graph builder at assembly time. You do not
 * need to include secrets in this component definition.
 *
 * @example
 * ```typescript
 * const sttComponent = new RemoteSTTComponent({
 *   id: 'my-stt-component',
 *   sttConfig: {
 *     languageCode: 'en-US'
 *   },
 *   service: 'inworld',
 *   modelId: 'groq/whisper-large-v3'
 * });
 * ```
 */
class RemoteSTTComponent extends abstract_component_1.AbstractComponent {
    /**
     * Creates a new RemoteSTTComponent instance.
     *
     * @param props - Configuration for the remote STT component.
     */
    constructor(props) {
        var _a, _b;
        super(props);
        this.sttConfig = props.sttConfig;
        this.service = (_a = props.service) !== null && _a !== void 0 ? _a : 'inworld';
        this.modelId = (_b = props.modelId) !== null && _b !== void 0 ? _b : constants_1.DEFAULT_STT_MODEL_ID;
        this.configOverrides = props.configOverrides;
    }
    /**
     * Converts the remote STT component to a graph configuration component.
     *
     * @returns The configured remote STT component for graph execution.
     */
    toGraphConfigComponent() {
        const config = this.buildServiceConfig();
        return {
            id: this.id,
            type: 'STTInterface',
            creation_config: {
                type: 'RemoteSTTConfig',
                properties: {
                    service: this.service,
                    config: (0, snakify_1.default)(config),
                },
            },
        };
    }
    buildServiceConfig() {
        var _a, _b, _c, _d;
        const overrides = (_a = this.configOverrides) !== null && _a !== void 0 ? _a : {};
        const mergedDefaultConfig = Object.assign(Object.assign({}, this.sttConfig), ((_b = overrides === null || overrides === void 0 ? void 0 : overrides.defaultConfig) !== null && _b !== void 0 ? _b : {}));
        const base = Object.assign(Object.assign({}, overrides), { apiKey: (_c = overrides === null || overrides === void 0 ? void 0 : overrides.apiKey) !== null && _c !== void 0 ? _c : this.getDefaultApiKeyPlaceholder(), modelId: (_d = overrides === null || overrides === void 0 ? void 0 : overrides.modelId) !== null && _d !== void 0 ? _d : this.modelId, defaultConfig: mergedDefaultConfig });
        return base;
    }
    getDefaultApiKeyPlaceholder() {
        return this.service === 'groq' ? '{{GROQ_API_KEY}}' : '{{INWORLD_API_KEY}}';
    }
}
exports.RemoteSTTComponent = RemoteSTTComponent;
