// @inworld/runtime/build/graph/dsl/components/remote_streaming_stt_component.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteStreamingSTTComponent = void 0;
const snakify_1 = __importDefault(require("../utils/snakify"));
const abstract_component_1 = require("./abstract_component");
/**
 * Remote streaming STT component for bidirectional transcription services.
 *
 * @example
 * ```typescript
 * const streamingStt = new RemoteStreamingSTTComponent({
 *   id: 'assembly_streaming_stt',
 *   modelId: 'assemblyai/universal-streaming-multilingual',
 *   defaultConfig: {
 *     silenceThresholdMs: 1500,
 *   },
 * });
 * ```
 */
class RemoteStreamingSTTComponent extends abstract_component_1.AbstractComponent {
    constructor(props) {
        super(props);
        this.defaultConfig = props.defaultConfig;
        this.modelId = props.modelId;
        this.endpointOverride = props.endpointOverride;
        this.defaultTimeout = props.defaultTimeout;
        this.apiKey = props.apiKey;
    }
    toGraphConfigComponent() {
        var _a;
        const properties = {
            defaultConfig: (_a = this.defaultConfig) !== null && _a !== void 0 ? _a : {},
        };
        if (this.apiKey) {
            properties.apiKey = this.apiKey;
        }
        if (this.modelId) {
            properties.modelId = this.modelId;
        }
        if (this.endpointOverride) {
            properties.endpointOverride = this.endpointOverride;
        }
        if (this.defaultTimeout !== undefined) {
            properties.defaultTimeout = this.defaultTimeout;
        }
        return {
            id: this.id,
            type: 'StreamingSTTInterface',
            creation_config: {
                type: 'RemoteStreamingSTTConfig',
                properties: (0, snakify_1.default)(properties),
            },
        };
    }
}
exports.RemoteStreamingSTTComponent = RemoteStreamingSTTComponent;
