// @inworld/runtime/build/graph/dsl/components/remote_knowledge_component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteKnowledgeComponent = void 0;
const abstract_component_1 = require("./abstract_component");
/**
 * Remote knowledge component for knowledge base services.
 * Provides access to external knowledge storage and retrieval services,
 * handling document parsing, chunking, and embedding for semantic search.
 *
 * @example
 * ```typescript
 * const knowledgeComponent = new RemoteKnowledgeComponent({
 *   maxCharsPerChunk: 1500,
 *   maxChunksPerDocument: 20
 * });
 * ```
 */
class RemoteKnowledgeComponent extends abstract_component_1.AbstractComponent {
    /**
     * Creates a new RemoteKnowledgeComponent instance.
     *
     * @param props - Configuration for the remote knowledge component.
     */
    constructor(props) {
        var _a, _b;
        super(props);
        this.maxCharsPerChunk = (_a = props.maxCharsPerChunk) !== null && _a !== void 0 ? _a : 1000;
        this.maxChunksPerDocument = (_b = props.maxChunksPerDocument) !== null && _b !== void 0 ? _b : 10;
    }
    /**
     * Converts the remote knowledge component to a graph configuration component.
     *
     * @returns The configured remote knowledge component for graph execution.
     */
    toGraphConfigComponent() {
        return {
            id: this.id,
            type: 'KnowledgeInterface',
            creation_config: {
                type: 'RemoteKnowledgeConfig',
                properties: {
                    knowledge_compile_config: {
                        parsing_config: {
                            max_chars_per_chunk: this.maxCharsPerChunk,
                            max_chunks_per_document: this.maxChunksPerDocument,
                        },
                    },
                },
            },
        };
    }
}
exports.RemoteKnowledgeComponent = RemoteKnowledgeComponent;
