// @inworld/runtime/build/graph/dsl/components/safety_checker_component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SafetyCheckerComponent = void 0;
const abstract_component_1 = require("./abstract_component");
class SafetyCheckerComponent extends abstract_component_1.AbstractComponent {
    /**
     * Creates a new SafetyCheckerComponent instance.
     *
     * @param props - Configuration for the safety checker component.
     */
    constructor(props) {
        super(props);
        this.modelWeightsPath = props.modelWeightsPath;
        this.embedderComponentId = props.embedderComponentId;
    }
    /**
     * Converts the safety checker component to a graph configuration component.
     *
     * @returns The configured safety checker component for graph execution.
     */
    toGraphConfigComponent() {
        const component = {
            id: this.id,
            type: 'SafetyCheckerInterface',
            creation_config: {
                type: 'SafetyCheckerCreationConfig',
                properties: {
                    model_weights_path: this.modelWeightsPath,
                },
            },
        };
        return component;
    }
}
exports.SafetyCheckerComponent = SafetyCheckerComponent;
