// @inworld/runtime/build/graph/dsl/components/text_classifier_component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextClassifierComponent = void 0;
const abstract_component_1 = require("./abstract_component");
/**
 * Text classifier component for content classification and topic detection.
 *
 * This component provides a reusable text classification interface that can be shared
 * across multiple TextClassifierNode instances. Use this when you need the same
 * classification model and configuration in multiple parts of your graph.
 *
 * @example
 * ```typescript
 * // Create reusable text classifier component
 * const classifierComponent = new TextClassifierComponent({
 *   id: 'shared-classifier',
 *   modelWeightsPath: '/models/text_classifier.json',
 *   embedderComponentId: 'embedder_component',
 *   supportedClasses: ['hategroup', 'selfharm', 'sexual', 'sexualminors', 'substance']
 * });
 *
 * // Use in multiple nodes
 * const spamDetector = new TextClassifierNode({
 *   id: 'spam-detector',
 *   textClassifierComponent: classifierComponent
 * });
 *
 * const sentimentAnalyzer = new TextClassifierNode({
 *   id: 'sentiment-analyzer',
 *   textClassifierComponent: classifierComponent
 * });
 * ```
 */
class TextClassifierComponent extends abstract_component_1.AbstractComponent {
    /**
     * Creates a new TextClassifierComponent instance.
     *
     * @param props - Configuration for the text classifier component.
     */
    constructor(props) {
        super(props);
        this.modelWeightsPath = props.modelWeightsPath;
        this.embedderComponentId = props.embedderComponentId;
        this.supportedClasses = props.supportedClasses;
    }
    /**
     * Converts the text classifier component to a graph configuration component.
     *
     * @returns The configured text classifier component for graph execution
     */
    toGraphConfigComponent() {
        const component = {
            id: this.id,
            type: 'TextClassifierInterface',
            creation_config: {
                type: 'TextClassifierCreationConfig',
                properties: {
                    embedder_component_id: this.embedderComponentId,
                    classifier_config: {
                        model_weights_path: this.modelWeightsPath,
                        supported_classes: this.supportedClasses || [],
                    },
                },
            },
        };
        return component;
    }
}
exports.TextClassifierComponent = TextClassifierComponent;
