// @inworld/runtime/build/graph/dsl/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_TTS_VOICE = exports.DEFAULT_SPEECH_SYNTHESIS_CONFIG = exports.DEFAULT_TTS_MODEL_ID = exports.DEFAULT_SPEECH_RECOGNITION_CONFIG = exports.DEFAULT_STT_MODEL_ID = exports.DEFAULT_INTENT_MATCHER_CONFIG = exports.DEFAULT_RETRIEVAL_CONFIG = exports.DEFAULT_TEXT_CONFIG = void 0;
exports.DEFAULT_TEXT_CONFIG = {
    max_new_tokens: 2500,
    max_prompt_length: 100,
    repetition_penalty: 1,
    top_p: 1,
    temperature: 1,
    frequency_penalty: 0,
    presence_penalty: 0,
    stop_sequences: [],
};
exports.DEFAULT_RETRIEVAL_CONFIG = {
    threshold: 0.7,
    top_k: 5,
};
exports.DEFAULT_INTENT_MATCHER_CONFIG = {
    embedding: {
        similarity_threshold: 0.88,
    },
    llm: {
        generation_config: exports.DEFAULT_TEXT_CONFIG,
        prompt_template: '',
        embedding_similarity_threshold: 0.7,
        max_embedding_matches_for_llm: 5,
    },
    top_n_intents: 3,
};
exports.DEFAULT_STT_MODEL_ID = 'groq/whisper-large-v3';
exports.DEFAULT_SPEECH_RECOGNITION_CONFIG = {
    languageCode: 'en-US',
};
exports.DEFAULT_TTS_MODEL_ID = 'inworld-tts-1-max';
exports.DEFAULT_SPEECH_SYNTHESIS_CONFIG = {
    inworldConfig: {
        modelId: exports.DEFAULT_TTS_MODEL_ID,
        inference: {
            speakingRate: 1,
            temperature: 0.8,
        },
        postprocessing: {
            sampleRate: 48000,
        },
    },
};
exports.DEFAULT_TTS_VOICE = {
    id: 'Ashley',
    language_code: 'en-US',
};
