// @inworld/runtime/build/graph/dsl/nodes/realtime_agent/realtime_agent_tts_first_chunk_checking_node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RealtimeAgentTtsFirstChunkCheckingNode = void 0;
const abstract_node_1 = require("../abstract_node");
/**
 * Built-in realtime agent node that reads the first chunk from a TTS stream and
 * signals TTS readiness (used for interruption-aware playback handling).
 *
 * @input {GraphTypes.TTSOutputStream} {@link GraphTypes.TTSOutputStream} - Incoming TTS stream
 * @output {GraphTypes.TTSOutputStream} {@link GraphTypes.TTSOutputStream} - The same stream (or a cancelled stream)
 */
class RealtimeAgentTtsFirstChunkCheckingNode extends abstract_node_1.AbstractNode {
    /**
     * Creates a new {@link RealtimeAgentTtsFirstChunkCheckingNode}.
     *
     * @param {AbstractNodeProps} [props] - Optional node configuration.
     */
    constructor(props = {}) {
        super(props);
    }
    toGraphConfigNode() {
        return Object.assign({ id: this.id, type: 'RealtimeAgent.TtsFirstChunkCheckingNode' }, (this.reportToClient
            ? {
                execution_config: {
                    type: 'NodeExecutionConfig',
                    properties: {
                        report_to_client: true,
                    },
                },
            }
            : {}));
    }
}
exports.RealtimeAgentTtsFirstChunkCheckingNode = RealtimeAgentTtsFirstChunkCheckingNode;
