// @inworld/runtime/build/graph/dsl/nodes/realtime_agent/realtime_agent_interaction_info_node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RealtimeAgentInteractionInfoNode = void 0;
const abstract_node_1 = require("../abstract_node");
/**
 * Built-in realtime agent node that combines slicer state (interaction id) and text
 * into an `InteractionInfo` JSON object.
 *
 * @remarks
 * The underlying C++ type serializes to JSON:
 * `{ type: \"InteractionInfo\", data: { interaction_id, text, isInterrupted, ... } }`.
 *
 * @input {GraphTypes.Custom<any>} {@link GraphTypes.Custom} - Slicer state (first input)
 * @input {String} {@link String} - Text content (second input; from STT or text extractor)
 * @output {GraphTypes.Custom<any>} {@link GraphTypes.Custom} - JSON object with `type: \"InteractionInfo\"`
 */
class RealtimeAgentInteractionInfoNode extends abstract_node_1.AbstractNode {
    /**
     * Creates a new {@link RealtimeAgentInteractionInfoNode}.
     *
     * @param {AbstractNodeProps} [props] - Optional node configuration.
     */
    constructor(props = {}) {
        super(props);
    }
    toGraphConfigNode() {
        return Object.assign({ id: this.id, type: 'RealtimeAgent.InteractionInfoNode' }, (this.reportToClient
            ? {
                execution_config: {
                    type: 'NodeExecutionConfig',
                    properties: {
                        report_to_client: true,
                    },
                },
            }
            : {}));
    }
}
exports.RealtimeAgentInteractionInfoNode = RealtimeAgentInteractionInfoNode;
