// @inworld/runtime/build/graph/dsl/nodes/realtime_agent/realtime_agent_prompt_variables_node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RealtimeAgentPromptVariablesNode = void 0;
const abstract_node_1 = require("../abstract_node");
/**
 * Built-in realtime agent node that prepares a JSON object with variables for
 * rendering an LLM prompt template (e.g., user query + conversation history).
 *
 * @input {String} {@link String} - User query text
 * @output {GraphTypes.Custom<any>} {@link GraphTypes.Custom} - JSON prompt variables
 */
class RealtimeAgentPromptVariablesNode extends abstract_node_1.AbstractNode {
    /**
     * Creates a new {@link RealtimeAgentPromptVariablesNode}.
     *
     * @param {AbstractNodeProps} [props] - Optional node configuration.
     */
    constructor(props = {}) {
        super(props);
    }
    toGraphConfigNode() {
        return Object.assign({ id: this.id, type: 'RealtimeAgent.PromptVariablesNode' }, (this.reportToClient
            ? {
                execution_config: {
                    type: 'NodeExecutionConfig',
                    properties: {
                        report_to_client: true,
                    },
                },
            }
            : {}));
    }
}
exports.RealtimeAgentPromptVariablesNode = RealtimeAgentPromptVariablesNode;
