// @inworld/runtime/build/graph/dsl/nodes/realtime_agent/realtime_agent_interaction_queue_node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RealtimeAgentInteractionQueueNode = void 0;
const abstract_node_1 = require("../abstract_node");
/**
 * Built-in realtime agent node that serializes interaction processing:
 * it ensures only one interaction proceeds to LLM at a time.
 *
 * @remarks
 * This node is typically used with a loop edge from `RealtimeAgent.StateUpdaterNode`
 * to trigger re-evaluation when the current interaction completes.
 *
 * @input {GraphTypes.Custom<any>} {@link GraphTypes.Custom} - `InteractionInfo` JSON object
 * @input {String} {@link String} - Optional completion trigger (loop feedback)
 * @output {String} {@link String} - Text to send to the LLM (or empty string when queued)
 */
class RealtimeAgentInteractionQueueNode extends abstract_node_1.AbstractNode {
    /**
     * Creates a new {@link RealtimeAgentInteractionQueueNode}.
     *
     * @param {AbstractNodeProps} [props] - Optional node configuration.
     */
    constructor(props = {}) {
        super(props);
    }
    toGraphConfigNode() {
        return Object.assign({ id: this.id, type: 'RealtimeAgent.InteractionQueueNode' }, (this.reportToClient
            ? {
                execution_config: {
                    type: 'NodeExecutionConfig',
                    properties: {
                        report_to_client: true,
                    },
                },
            }
            : {}));
    }
}
exports.RealtimeAgentInteractionQueueNode = RealtimeAgentInteractionQueueNode;
