// @inworld/runtime/build/graph/dsl/nodes/realtime_agent/realtime_agent_input_slicer_node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RealtimeAgentInputSlicerNode = void 0;
const abstract_node_1 = require("../abstract_node");
/**
 * Built-in realtime agent node that slices an incoming multimodal stream
 * (audio or text) into interaction turns and detects interruptions.
 *
 * @remarks
 * This node is designed to be used with a self-loop edge to keep consuming
 * streaming input until the stream completes.
 *
 * @input {GraphTypes.MultimodalContentStream} {@link GraphTypes.MultimodalContentStream} - Incoming multimodal stream (audio or text chunks)
 * @output {GraphTypes.Custom<any>} {@link GraphTypes.Custom} - Internal slicer state (not typically consumed directly; used for routing)
 */
class RealtimeAgentInputSlicerNode extends abstract_node_1.AbstractNode {
    /**
     * Creates a new {@link RealtimeAgentInputSlicerNode}.
     *
     * @param {RealtimeAgentInputSlicerNodeProps} props - Node configuration.
     */
    constructor(props) {
        var _a, _b, _c;
        super(props);
        this.executionConfig = {
            vadComponentId: props.vadComponentId,
            turnDetectorComponentId: props.turnDetectorComponentId,
            maxSpeechDurationPerTurnMs: (_a = props.maxSpeechDurationPerTurnMs) !== null && _a !== void 0 ? _a : 30000,
            maxSilenceToFinishTurnMs: (_b = props.maxSilenceToFinishTurnMs) !== null && _b !== void 0 ? _b : 200,
            sampleRate: (_c = props.sampleRate) !== null && _c !== void 0 ? _c : 16000,
        };
    }
    toGraphConfigNode() {
        return {
            id: this.id,
            type: 'RealtimeAgent.InputSlicerNode',
            execution_config: {
                type: 'RealtimeAgent.InputSlicerNodeExecutionConfig',
                properties: Object.assign(Object.assign({ vad_component_id: this.executionConfig.vadComponentId }, (this.executionConfig.turnDetectorComponentId
                    ? {
                        turn_detector_component_id: this.executionConfig.turnDetectorComponentId,
                    }
                    : {})), { max_speech_duration_per_turn_ms: this.executionConfig.maxSpeechDurationPerTurnMs, max_silence_to_finish_turn_ms: this.executionConfig.maxSilenceToFinishTurnMs, sample_rate: this.executionConfig.sampleRate }),
            },
        };
    }
}
exports.RealtimeAgentInputSlicerNode = RealtimeAgentInputSlicerNode;
