// @inworld/runtime/build/graph/dsl/nodes/realtime_agent/realtime_agent_state_updater_node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RealtimeAgentStateUpdaterNode = void 0;
const abstract_node_1 = require("../abstract_node");
/**
 * Built-in realtime agent node that consumes the text chunking stream to aggregate
 * the final text and stores the running TTS stream for cancellation / interruption handling.
 *
 * @input {GraphTypes.TextStream} {@link GraphTypes.TextStream} - Text chunks stream (from TextChunkingNode)
 * @input {GraphTypes.TTSOutputStream} {@link GraphTypes.TTSOutputStream} - TTS output stream (from TTSNode)
 * @output {String} {@link String} - Aggregated assistant text
 */
class RealtimeAgentStateUpdaterNode extends abstract_node_1.AbstractNode {
    /**
     * Creates a new {@link RealtimeAgentStateUpdaterNode}.
     *
     * @param {AbstractNodeProps} [props] - Optional node configuration.
     */
    constructor(props = {}) {
        super(props);
    }
    toGraphConfigNode() {
        return Object.assign({ id: this.id, type: 'RealtimeAgent.StateUpdaterNode' }, (this.reportToClient
            ? {
                execution_config: {
                    type: 'NodeExecutionConfig',
                    properties: {
                        report_to_client: true,
                    },
                },
            }
            : {}));
    }
}
exports.RealtimeAgentStateUpdaterNode = RealtimeAgentStateUpdaterNode;
