// @inworld/runtime/build/graph/dsl/nodes/realtime_agent/realtime_agent_audio_extractor_node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RealtimeAgentAudioExtractorNode = void 0;
const abstract_node_1 = require("../abstract_node");
/**
 * Built-in realtime agent node that extracts completed audio from the slicer state.
 *
 * @input {GraphTypes.Custom<any>} {@link GraphTypes.Custom} - Internal slicer state
 * @output {GraphTypes.Audio} {@link GraphTypes.Audio} - Extracted audio turn
 */
class RealtimeAgentAudioExtractorNode extends abstract_node_1.AbstractNode {
    /**
     * Creates a new {@link RealtimeAgentAudioExtractorNode}.
     *
     * @param {AbstractNodeProps} [props] - Optional node configuration.
     */
    constructor(props = {}) {
        super(props);
    }
    toGraphConfigNode() {
        return Object.assign({ id: this.id, type: 'RealtimeAgent.AudioExtractorNode' }, (this.reportToClient
            ? {
                execution_config: {
                    type: 'NodeExecutionConfig',
                    properties: {
                        report_to_client: true,
                    },
                },
            }
            : {}));
    }
}
exports.RealtimeAgentAudioExtractorNode = RealtimeAgentAudioExtractorNode;
