// @inworld/runtime/build/internal/default.js
"use strict";
/**
 * Defaults for different Inworld Runtime services.
 * @internal
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Default = void 0;
class Default {
}
exports.Default = Default;
Default.telemetry = {
    /**
     * Default app version used for telemetry.
     * @internal
     */
    telemetryAppVersion: 'unspecified_version',
    /**
     * Default app name used for telemetry.
     * @internal
     */
    telemetryAppName: 'unspecified_app',
    /**
     * Default endpoint used for auth.
     *
     * You can override this by setting the INWORLD_ENV environment variable:
     *   DEV   -> api-engine.dev.inworld.ai:443
     *   STAGE -> api-engine.stage.inworld.ai:443
     *   PROD  -> api-engine.inworld.ai:443
     * Any other value or unset will use this default.
     * @internal
     */
    authEndpoint: 'api-engine.inworld.ai:443',
    flushTimeoutMs: 1000,
    shutdownTimeoutMs: 2000,
    logger: {
        sinkAbslLogs: false,
    },
    metrics: {
        exportInterval: 10000,
        exportTimeout: 5000,
        metricCacheSize: 1000,
    },
    tracer: {
        maxSpanCaptureQueueSize: 50,
        enableSpanCapture: false,
        metricCacheSize: 1000,
        samplingRate: 1,
    },
};
Default.embedder = {
    provider: 'inworld',
    modelName: 'BAAI/bge-large-en-v1.5',
};
Default.llm = {
    provider: 'inworld',
    modelName: 'meta-llama/Llama-3.1-70b-Instruct',
};
Default.featureFlag = {
    evaluationMode: 'EVALUATION_MODE_IN_PROCESS_ONLY',
    pollIntervalMs: 30000,
};
