// @inworld/runtime/build/internal/node_api.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpanWrapper = exports.telemetryNamespace = exports.getInworldAddonEnv = exports.stopInworldRuntime = exports.registerCallbacks = exports.createGraph = exports.__resetInworldNodeAddonModule = void 0;
const path_1 = __importDefault(require("path"));
const pb_1 = require("./pb");
const platform_detection_1 = require("./utils/platform_detection");
/**
 * Default thread pool size for the native addon.
 * Can be overridden via INWORLD_ADDON_POOL_SIZE environment variable.
 */
const DEFAULT_ADDON_POOL_SIZE = 64;
/**
 * Get the addon pool size from environment variable or use default.
 */
const getAddonPoolSize = () => {
    const envValue = process.env.INWORLD_ADDON_POOL_SIZE;
    if (envValue) {
        const parsed = parseInt(envValue, 10);
        if (!isNaN(parsed) && parsed > 0) {
            return parsed;
        }
    }
    return DEFAULT_ADDON_POOL_SIZE;
};
/**
 * Path to libinworld.node
 */
const inworldNodeAddonPath = process.env.INWORLD_NODE_ADDON_PATH ||
    (0, platform_detection_1.getBinaryPath)(path_1.default.join(__dirname, '..', '..', 'bin'), true);
let inworldAddon = require(inworldNodeAddonPath);
let inworldAddonEnv = inworldAddon.createEnv(pb_1.graphMethods.createEnvRequest({
    addonPoolSize: getAddonPoolSize(),
}));
const __resetInworldNodeAddonModule = () => {
    inworldAddon = require(inworldNodeAddonPath);
    inworldAddonEnv = inworldAddon.createEnv(pb_1.graphMethods.createEnvRequest({
        addonPoolSize: getAddonPoolSize(),
    }));
};
exports.__resetInworldNodeAddonModule = __resetInworldNodeAddonModule;
const createGraph = (request) => {
    return inworldAddonEnv.createGraph(request);
};
exports.createGraph = createGraph;
/**
 * Register custom callback functions for nodes and edge conditions
 * @param callbacks - Object containing callback definitions
 * @param callbacks.nodes - Map of node type names to their callback functions
 * @param callbacks.conditions - Map of condition type names to their callback
 *     functions
 */
const registerCallbacks = (callbacks) => {
    return inworldAddonEnv.registerCallbacks(callbacks);
};
exports.registerCallbacks = registerCallbacks;
const stopInworldRuntime = async () => inworldAddonEnv.close();
exports.stopInworldRuntime = stopInworldRuntime;
/**
 * Access to the addon environment for primitives2.
 * Provides access to all primitive namespaces (speech, llm, embeddings, etc.)
 */
const getInworldAddonEnv = () => inworldAddonEnv;
exports.getInworldAddonEnv = getInworldAddonEnv;
exports.telemetryNamespace = inworldAddon.telemetry;
exports.SpanWrapper = inworldAddon.SpanWrapper;
