// @inworld/runtime/build/primitives/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeviceError = exports.MCPError = exports.KeywordMatcherError = exports.TextClassifierError = exports.NERError = exports.KnowledgeError = exports.EmbeddingError = exports.PromptBuilderError = exports.LLMError = exports.AECFilterError = exports.TurnDetectorError = exports.VADError = exports.TTSError = exports.STTError = exports.PrimitiveError = void 0;
const inworld_1 = require("../internal/error/inworld");
/**
 * Base error class for all primitives-related errors.
 * Extends InworldError to maintain consistency with the existing error handling.
 */
class PrimitiveError extends inworld_1.InworldError {
    constructor(message, stack) {
        super(message, stack);
    }
}
exports.PrimitiveError = PrimitiveError;
/**
 * Error thrown by Speech-to-Text (STT) operations.
 */
class STTError extends PrimitiveError {
    constructor(message, stack) {
        super(message, stack);
    }
}
exports.STTError = STTError;
/**
 * Error thrown by Text-to-Speech (TTS) operations.
 */
class TTSError extends PrimitiveError {
    constructor(message, stack) {
        super(message, stack);
    }
}
exports.TTSError = TTSError;
/**
 * Error thrown by Voice Activity Detection (VAD) operations.
 */
class VADError extends PrimitiveError {
    constructor(message, stack) {
        super(message, stack);
    }
}
exports.VADError = VADError;
/**
 * Error thrown by Turn Detector operations.
 */
class TurnDetectorError extends PrimitiveError {
    constructor(message, stack) {
        super(message, stack);
    }
}
exports.TurnDetectorError = TurnDetectorError;
/**
 * Error thrown by Acoustic Echo Cancellation (AEC) Filter operations.
 */
class AECFilterError extends PrimitiveError {
    constructor(message, stack) {
        super(message, stack);
    }
}
exports.AECFilterError = AECFilterError;
/**
 * Error thrown by Large Language Model (LLM) operations.
 */
class LLMError extends PrimitiveError {
    constructor(message, stack) {
        super(message, stack);
    }
}
exports.LLMError = LLMError;
/**
 * Error thrown by Prompt Builder operations.
 */
class PromptBuilderError extends PrimitiveError {
    constructor(message, stack) {
        super(message, stack);
    }
}
exports.PromptBuilderError = PromptBuilderError;
/**
 * Error thrown by Text Embedder operations.
 */
class EmbeddingError extends PrimitiveError {
    constructor(message, stack) {
        super(message, stack);
    }
}
exports.EmbeddingError = EmbeddingError;
/**
 * Error thrown by Knowledge operations (RAG).
 */
class KnowledgeError extends PrimitiveError {
    constructor(message, stack) {
        super(message, stack);
    }
}
exports.KnowledgeError = KnowledgeError;
/**
 * Error thrown by Named Entity Recognition (NER) operations.
 */
class NERError extends PrimitiveError {
    constructor(message, stack) {
        super(message, stack);
    }
}
exports.NERError = NERError;
/**
 * Error thrown by Text Classifier operations.
 */
class TextClassifierError extends PrimitiveError {
    constructor(message, stack) {
        super(message, stack);
    }
}
exports.TextClassifierError = TextClassifierError;
/**
 * Error thrown by Keyword Matcher operations.
 */
class KeywordMatcherError extends PrimitiveError {
    constructor(message, stack) {
        super(message, stack);
    }
}
exports.KeywordMatcherError = KeywordMatcherError;
/**
 * Error thrown by MCP Client operations.
 */
class MCPError extends PrimitiveError {
    constructor(message, stack) {
        super(message, stack);
    }
}
exports.MCPError = MCPError;
/**
 * Error thrown by Device Registry operations.
 */
class DeviceError extends PrimitiveError {
    constructor(message, stack) {
        super(message, stack);
    }
}
exports.DeviceError = DeviceError;
