// @inworld/runtime/build/primitives/index.js
"use strict";
/**
 * Inworld Primitives2 - New N-API addon-based primitives implementation.
 *
 * This module provides TypeScript wrappers around the N-API addon primitives,
 * offering a more ergonomic and type-safe API compared to direct addon usage.
 *
 * @module primitives2
 * @see {@link ../internal/node-api.ts} for addon access
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PbHelper = exports.speech = exports.platform = exports.nlu = exports.mcp = exports.llm = exports.knowledge = exports.embeddings = void 0;
// Export error classes
__exportStar(require("./errors"), exports);
// Export common types
__exportStar(require("./types"), exports);
// Export category-specific modules
exports.embeddings = __importStar(require("./embeddings"));
exports.knowledge = __importStar(require("./knowledge"));
exports.llm = __importStar(require("./llm"));
exports.mcp = __importStar(require("./mcp"));
exports.nlu = __importStar(require("./nlu"));
exports.platform = __importStar(require("./platform"));
exports.speech = __importStar(require("./speech"));
// Re-export protobuf helpers for convenience
var pb_helpers_1 = require("../internal/pb_helpers");
Object.defineProperty(exports, "PbHelper", { enumerable: true, get: function () { return pb_helpers_1.PbHelper; } });
