// @inworld/runtime/build/telemetry/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExporterType = exports.LogLevel = exports.MetricType = void 0;
/**
 * Metric types for telemetry data collection
 * @description Defines the different types of metrics that can be recorded
 */
var MetricType;
(function (MetricType) {
    /**
     * Counter with unsigned integer values
     * @description Monotonically increasing counter for counting events
     */
    MetricType["CounterUInt"] = "CounterUInt";
    /**
     * Counter with double values
     * @description Monotonically increasing counter with floating-point precision
     */
    MetricType["CounterDouble"] = "CounterDouble";
    /**
     * Gauge with integer values
     * @description Point-in-time measurement that can go up or down
     */
    MetricType["GaugeInt"] = "GaugeInt";
    /**
     * Gauge with double values
     * @description Point-in-time measurement with floating-point precision
     */
    MetricType["GaugeDouble"] = "GaugeDouble";
    /**
     * Histogram with unsigned integer values
     * @description Distribution of values over time for integer data
     */
    MetricType["HistogramUInt"] = "HistogramUInt";
    /**
     * Histogram with double values
     * @description Distribution of values over time for floating-point data
     */
    MetricType["HistogramDouble"] = "HistogramDouble";
})(MetricType || (exports.MetricType = MetricType = {}));
/**
 * Log levels for telemetry logging
 * @description Maps to VERBOSITY values in the underlying telemetry system
 */
var LogLevel;
(function (LogLevel) {
    /**
     * Trace level - most verbose logging
     * @description Verbosity = 2, captures all possible log messages
     */
    LogLevel["Trace"] = "Trace";
    /**
     * Debug level - detailed debugging information
     * @description Verbosity = 1, useful for development and troubleshooting
     */
    LogLevel["Debug"] = "Debug";
    /**
     * Info level - general information messages
     * @description Verbosity = 0, standard operational information
     */
    LogLevel["Info"] = "Info";
    /**
     * Warning level - potential issues
     * @description Important messages about potential problems
     */
    LogLevel["Warn"] = "Warn";
    /**
     * Error level - error conditions
     * @description Error messages for actual problems
     */
    LogLevel["Error"] = "Error";
})(LogLevel || (exports.LogLevel = LogLevel = {}));
/**
 * Exporter types for telemetry data
 * @description Defines where telemetry data is sent
 */
var ExporterType;
(function (ExporterType) {
    /**
     * Local exporter - data stays on the local machine
     * @description Useful for development, debugging, or local analysis
     */
    ExporterType["Local"] = "Local";
    /**
     * Remote exporter - data sent to remote telemetry service
     * @description Production-ready, sends data to configured endpoint
     */
    ExporterType["Remote"] = "Remote";
})(ExporterType || (exports.ExporterType = ExporterType = {}));
