// @inworld/runtime/build/graph/GraphOutputStream.js
"use strict";
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphOutputStream = void 0;
const internal_1 = __importDefault(require("../internal/decorators/internal"));
const inworld_1 = require("../internal/error/inworld");
const GraphOutputStreamResponse_1 = require("./GraphOutputStreamResponse");
const from_external_1 = require("./utils/from_external");
/**
 * Represents a handler for a graph execution using tagged values (V2).
 * This version works with native JS objects instead of BaseData instances.
 */
let GraphOutputStream = (() => {
    var _a;
    let _instanceExtraInitializers = [];
    let _jsExecution_decorators;
    let _jsExecution_initializers = [];
    let _jsExecution_extraInitializers = [];
    let _onStreamFinished_decorators;
    let _onStreamFinished_initializers = [];
    let _onStreamFinished_extraInitializers = [];
    let _checkAndCallStreamFinished_decorators;
    let _setOnFinished_decorators;
    return _a = class GraphOutputStream {
            /**
             * Creates a new GraphOutputStreamV2 instance.
             * @param jsExecution - The NAPI JsExecution object from graph.start()
             * @internal
             */
            constructor(jsExecution) {
                /**
                 * The index of the execution (maintained for backward compatibility).
                 */
                this.index = (__runInitializers(this, _instanceExtraInitializers), 0);
                /**
                 * @internal
                 */
                this.jsExecution = __runInitializers(this, _jsExecution_initializers, void 0);
                /**
                 * Optional callback to call when the stream is finished
                 * @internal
                 */
                this.onStreamFinished = (__runInitializers(this, _jsExecution_extraInitializers), __runInitializers(this, _onStreamFinished_initializers, void 0));
                /**
                 * Counter for tracking pending responses that haven't completed processing yet.
                 */
                this.pendingResponses = (__runInitializers(this, _onStreamFinished_extraInitializers), 0);
                /**
                 * Counter for tracking completed responses.
                 */
                this.completedResponses = 0;
                /**
                 * Flag to indicate that no more responses will be created (stream has ended).
                 */
                this.noMoreResponses = false;
                this.jsExecution = jsExecution;
            }
            /**
             * Gets the next result from the execution.
             * @returns A promise that resolves to the response from the node
             */
            async next() {
                if (this.noMoreResponses) {
                    const finalResponse = new GraphOutputStreamResponse_1.GraphOutputStreamResponse(undefined, true);
                    this.checkAndCallStreamFinished();
                    return finalResponse;
                }
                try {
                    const iteratorResult = await this.jsExecution.next();
                    if (iteratorResult.done) {
                        this.noMoreResponses = true;
                        this.checkAndCallStreamFinished();
                        const finalResponse = new GraphOutputStreamResponse_1.GraphOutputStreamResponse(undefined, true);
                        if (this.onStreamFinished) {
                            finalResponse.setOnProcessingComplete(() => {
                                this.completedResponses++;
                                this.checkAndCallStreamFinished();
                            });
                        }
                        return finalResponse;
                    }
                    this.pendingResponses++;
                    const response = new GraphOutputStreamResponse_1.GraphOutputStreamResponse((0, from_external_1.fromExternal)(iteratorResult.value));
                    response.setOnProcessingComplete(() => {
                        this.completedResponses++;
                        this.checkAndCallStreamFinished();
                    });
                    return response;
                }
                catch (error) {
                    this.noMoreResponses = true;
                    if (error instanceof inworld_1.InworldError) {
                        throw error;
                    }
                    throw new inworld_1.InworldError('Graph Output Stream error', error.message);
                }
            }
            abort() {
                this.noMoreResponses = true;
                this.jsExecution.abort();
                this.checkAndCallStreamFinished();
            }
            /**
             * Checks if all responses have been completed and calls onStreamFinished if so.
             * This should be called when a response completes or when the stream ends.
             * @internal
             */
            checkAndCallStreamFinished() {
                if (this.noMoreResponses &&
                    this.completedResponses === this.pendingResponses) {
                    if (this.onStreamFinished) {
                        this.onStreamFinished();
                    }
                }
            }
            /**
             *
             * @internal
             */
            setOnFinished(callback) {
                this.onStreamFinished = callback;
            }
            /**
             * Enables async iteration over graph output responses.
             */
            [(_jsExecution_decorators = [internal_1.default], _onStreamFinished_decorators = [internal_1.default], _checkAndCallStreamFinished_decorators = [internal_1.default], _setOnFinished_decorators = [internal_1.default], Symbol.asyncIterator)]() {
                return {
                    next: async () => {
                        const result = await this.next();
                        if (result.done || this.noMoreResponses) {
                            return { done: true, value: undefined };
                        }
                        return { done: false, value: result };
                    },
                    return: async () => {
                        this.abort();
                        return { done: true, value: undefined };
                    },
                    throw: async (error) => {
                        this.abort();
                        throw error;
                    },
                };
            }
        },
        (() => {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            __esDecorate(_a, null, _checkAndCallStreamFinished_decorators, { kind: "method", name: "checkAndCallStreamFinished", static: false, private: false, access: { has: obj => "checkAndCallStreamFinished" in obj, get: obj => obj.checkAndCallStreamFinished }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(_a, null, _setOnFinished_decorators, { kind: "method", name: "setOnFinished", static: false, private: false, access: { has: obj => "setOnFinished" in obj, get: obj => obj.setOnFinished }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(null, null, _jsExecution_decorators, { kind: "field", name: "jsExecution", static: false, private: false, access: { has: obj => "jsExecution" in obj, get: obj => obj.jsExecution, set: (obj, value) => { obj.jsExecution = value; } }, metadata: _metadata }, _jsExecution_initializers, _jsExecution_extraInitializers);
            __esDecorate(null, null, _onStreamFinished_decorators, { kind: "field", name: "onStreamFinished", static: false, private: false, access: { has: obj => "onStreamFinished" in obj, get: obj => obj.onStreamFinished, set: (obj, value) => { obj.onStreamFinished = value; } }, metadata: _metadata }, _onStreamFinished_initializers, _onStreamFinished_extraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        _a;
})();
exports.GraphOutputStream = GraphOutputStream;
