// @inworld/runtime/build/graph/GraphOutputStreamResponse.js
"use strict";
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphOutputStreamResponse = void 0;
const type_mapping_1 = require("../common/data_types/type_mapping");
const internal_1 = __importDefault(require("../internal/decorators/internal"));
/**
 * Represents a response from a graph output stream.
 * This class contains all possible data types that can be returned
 * from graph nodes during execution. Each type serves a specific purpose
 * in the graph processing pipeline.
 */
let GraphOutputStreamResponse = (() => {
    var _a;
    let _onProcessingComplete_decorators;
    let _onProcessingComplete_initializers = [];
    let _onProcessingComplete_extraInitializers = [];
    return _a = class GraphOutputStreamResponse {
            /**
             * Creates a new GraphOutputStreamResponse instance.
             *
             * @param data - The graph output stream data, if available
             * @param done - Whether the graph has finished executing. When true, indicates that
             *               the graph execution has completed and no more data will be streamed.
             */
            constructor(data, done = false) {
                this.data = data;
                this.done = done;
                /**
                 * Optional callback to call when response processing is complete
                 * @internal
                 */
                this.onProcessingComplete = __runInitializers(this, _onProcessingComplete_initializers, void 0);
                __runInitializers(this, _onProcessingComplete_extraInitializers);
                this.data = data;
                this.done = done;
            }
            get typeName() {
                return (0, type_mapping_1.getTypeName)(this.data);
            }
            /**
             * Sets a callback to be called when response processing is complete.
             * This is called after all handlers in processResponse have finished executing.
             *
             * @param callback - Function to call when processing is complete
             * @internal
             */
            setOnProcessingComplete(callback) {
                this.onProcessingComplete = callback;
            }
            isString() {
                return type_mapping_1.GraphTypesTypeGuards.isString(this.data);
            }
            isCustom() {
                return type_mapping_1.GraphTypesTypeGuards.isCustom(this.data);
            }
            isClassificationResult() {
                return type_mapping_1.GraphTypesTypeGuards.isClassificationResult(this.data);
            }
            isContent() {
                return type_mapping_1.GraphTypesTypeGuards.isContent(this.data);
            }
            isGoalAdvancement() {
                return type_mapping_1.GraphTypesTypeGuards.isGoalAdvancement(this.data);
            }
            isAudio() {
                return type_mapping_1.GraphTypesTypeGuards.isAudio(this.data);
            }
            isKnowledgeRecords() {
                return type_mapping_1.GraphTypesTypeGuards.isKnowledgeRecords(this.data);
            }
            isListToolsResponse() {
                return type_mapping_1.GraphTypesTypeGuards.isListToolsResponse(this.data);
            }
            isLLMChatRequest() {
                return type_mapping_1.GraphTypesTypeGuards.isLLMChatRequest(this.data);
            }
            isMatchedIntents() {
                return type_mapping_1.GraphTypesTypeGuards.isMatchedIntents(this.data);
            }
            isMatchedKeywords() {
                return type_mapping_1.GraphTypesTypeGuards.isMatchedKeywords(this.data);
            }
            isSafetyResult() {
                return type_mapping_1.GraphTypesTypeGuards.isSafetyResult(this.data);
            }
            isToolCallResponse() {
                return type_mapping_1.GraphTypesTypeGuards.isToolCallResponse(this.data);
            }
            isTextStream() {
                return type_mapping_1.GraphTypesTypeGuards.isTextStream(this.data);
            }
            isContentStream() {
                return type_mapping_1.GraphTypesTypeGuards.isContentStream(this.data);
            }
            isTTSOutputStream() {
                return type_mapping_1.GraphTypesTypeGuards.isTTSOutputStream(this.data);
            }
            isSpeechChunkStream() {
                return type_mapping_1.GraphTypesTypeGuards.isSpeechChunkStream(this.data);
            }
            isGraphError() {
                return type_mapping_1.GraphTypesTypeGuards.isGraphError(this.data);
            }
            isLLMChatResponse() {
                return type_mapping_1.GraphTypesTypeGuards.isLLMChatResponse(this.data);
            }
            /**
             * Processes the graph output stream response with type safety using a visitor pattern.
             * This function makes it easy to handle different response types in a switch-like pattern.
             *
             * @param handlers - Object containing handler functions for each type
             * @returns The result from the appropriate handler
             *
             * @example
             * ```typescript
             * const result = response.visit({
             *   string: (text) => `Got text: ${text}`,
             *   Custom: (custom) => `Got custom data: ${JSON.stringify(custom)}`,
             *   TextStream: async (stream) => {
             *     // Handle stream...
             *     return "Processed stream";
             *   },
             *   // Add other handlers as needed
             *   default: (value) => {
             *     console.log(`Unknown type received from stream.`);
             *   },
             * });
             * ```
             */
            async processResponse(handlers) {
                let result;
                if (this.done && handlers.done) {
                    // do not break on done, there might be data present;
                    result = handlers.done();
                    if (this.onProcessingComplete) {
                        this.onProcessingComplete();
                    }
                }
                if (this.done && !this.data) {
                    if (this.onProcessingComplete) {
                        this.onProcessingComplete();
                    }
                    return Promise.resolve();
                }
                if (this.isString() && handlers.string) {
                    result = handlers.string(this.data);
                }
                else if (this.isCustom() && handlers.Custom) {
                    result = handlers.Custom(this.data);
                }
                else if (this.isClassificationResult() && handlers.ClassificationResult) {
                    result = handlers.ClassificationResult(this.data);
                }
                else if (this.isContent() && handlers.Content) {
                    result = handlers.Content(this.data);
                }
                else if (this.isGoalAdvancement() && handlers.GoalAdvancement) {
                    result = handlers.GoalAdvancement(this.data);
                }
                else if (this.isKnowledgeRecords() && handlers.KnowledgeRecords) {
                    result = handlers.KnowledgeRecords(this.data);
                }
                else if (this.isListToolsResponse() && handlers.ListToolsResponse) {
                    result = handlers.ListToolsResponse(this.data);
                }
                else if (this.isLLMChatRequest() && handlers.LLMChatRequest) {
                    result = handlers.LLMChatRequest(this.data);
                }
                else if (this.isMatchedIntents() && handlers.MatchedIntents) {
                    result = handlers.MatchedIntents(this.data);
                }
                else if (this.isMatchedKeywords() && handlers.MatchedKeywords) {
                    result = handlers.MatchedKeywords(this.data);
                }
                else if (this.isSafetyResult() && handlers.SafetyResult) {
                    result = handlers.SafetyResult(this.data);
                }
                else if (this.isToolCallResponse() && handlers.ToolCallResponse) {
                    result = handlers.ToolCallResponse(this.data);
                }
                else if (this.isTextStream() && handlers.TextStream) {
                    result = handlers.TextStream(this.data);
                }
                else if (this.isContentStream() && handlers.ContentStream) {
                    result = handlers.ContentStream(this.data);
                }
                else if (this.isTTSOutputStream() && handlers.TTSOutputStream) {
                    result = handlers.TTSOutputStream(this.data);
                }
                else if (this.isAudio() && handlers.Audio) {
                    result = handlers.Audio(this.data);
                }
                else if (this.isSpeechChunkStream() && handlers.SpeechChunkStream) {
                    result = handlers.SpeechChunkStream(this.data);
                }
                else if (this.isLLMChatResponse() &&
                    (handlers.LLMChatResponse || handlers.Content)) {
                    if (handlers.LLMChatResponse) {
                        result = handlers.LLMChatResponse(this.data);
                    }
                    else {
                        result = handlers.Content(this.data);
                    }
                }
                else if (this.isGraphError() && handlers.error) {
                    result = handlers.error(this.data);
                }
                else if (handlers.default) {
                    result = handlers.default(this.data);
                }
                if (this.onProcessingComplete) {
                    this.onProcessingComplete();
                }
                if (result) {
                    return Promise.resolve(result);
                }
            }
        },
        (() => {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _onProcessingComplete_decorators = [internal_1.default];
            __esDecorate(null, null, _onProcessingComplete_decorators, { kind: "field", name: "onProcessingComplete", static: false, private: false, access: { has: obj => "onProcessingComplete" in obj, get: obj => obj.onProcessingComplete, set: (obj, value) => { obj.onProcessingComplete = value; } }, metadata: _metadata }, _onProcessingComplete_initializers, _onProcessingComplete_extraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        _a;
})();
exports.GraphOutputStreamResponse = GraphOutputStreamResponse;
