// @inworld/runtime/build/graph/dsl/nodes/custom_node.js
"use strict";
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomNode = void 0;
const internal_1 = __importDefault(require("../../../internal/decorators/internal"));
const nodes_1 = require("../../nodes");
const abstract_node_1 = require("./abstract_node");
/**
 * Base class for creating custom (user-defined) nodes.
 *
 * Subclasses must implement the `process` method. A unique node type is
 * automatically registered once per subclass.
 *
 * @example
 * ```typescript
 * import { CustomNode, ProcessContext } from '@inworld/runtime/graph';
 *
 * // Define a custom node that processes the input text
 * class CustomTextNode extends CustomNode {
 *   async process(
 *     context: ProcessContext,
 *     input: string,
 *   ): Promise<{ processedText: string }> {
 *     return {
 *       processedText: input.toUpperCase(),
 *     };
 *   }
 * }
 *
 * // Create an instance of the custom node
 * const customTextNode = new CustomTextNode();
 * ```
 */
let CustomNode = (() => {
    var _a;
    let _classSuper = abstract_node_1.AbstractNode;
    let _executionConfig_decorators;
    let _executionConfig_initializers = [];
    let _executionConfig_extraInitializers = [];
    return _a = class CustomNode extends _classSuper {
            /**
             * Override this method in subclasses to specify the output stream type
             * for generator-based process methods.
             * @returns The stream type (e.g., 'Audio', 'Text', 'BaseData')
             */
            static getStreamType() {
                return 'BaseData';
            }
            /**
             * Creates a new `CustomNode`.
             *
             * @remarks
             * Subclasses can pass an `executionConfig` object to surface additional
             * execution-time properties. Execution config would be available to the
             * process function in the {@link ProcessContext}.
             *
             * @typeParam InputType - The type of the inputs to the node, should be one of {@link CustomNodeInputTypes} or any for custom processing.
             * @typeParam OutputType - The type of the outputs from the node, should be one of {@link CustomNodeOutputTypes} or any for custom processing.
             *
             * @param props - Custom node options including optional `executionConfig`.
             */
            constructor(props = {}) {
                var _b;
                super(props);
                this.executionConfig = __runInitializers(this, _executionConfig_initializers, void 0);
                this.schema = __runInitializers(this, _executionConfig_extraInitializers);
                if (props === null || props === void 0 ? void 0 : props.executionConfig) {
                    const executionConfigProperties = Object.assign({}, props.executionConfig);
                    this.executionConfig = {
                        reportToClient: (_b = props === null || props === void 0 ? void 0 : props.reportToClient) !== null && _b !== void 0 ? _b : false,
                        properties: Object.assign({}, executionConfigProperties),
                    };
                }
                this.deferSchemaRegistration();
                const registry = nodes_1.CallbackRegistry.getInstance();
                const typeName = this.constructor.name + 'Type';
                const streamType = this.constructor.getStreamType();
                if (!registry.hasRegisteredNode(typeName)) {
                    _a.typeReferences[this.constructor.name] = (0, nodes_1.registerCustomNodeType)(typeName, this.id, this.process.bind(this), streamType, this.executionConfig);
                }
                else {
                    registry.registerNodeExecutionConfig(this.id, this.executionConfig);
                }
            }
            /**
             * Converts this custom node to a graph configuration node, including the
             * registered type and the provided execution configuration.
             */
            toGraphConfigNode() {
                var _b, _c;
                const executionConfigProperties = Object.assign(Object.assign({}, this.executionConfig), ((_c = (_b = this.executionConfig) === null || _b === void 0 ? void 0 : _b.properties) !== null && _c !== void 0 ? _c : {}));
                delete executionConfigProperties['properties'];
                delete executionConfigProperties['reportToClient'];
                return {
                    type: this.constructor.name + 'Type',
                    id: this.id,
                    execution_config: {
                        type: 'NodeExecutionConfig',
                        properties: Object.assign({ report_to_client: this.reportToClient }, executionConfigProperties),
                    },
                };
            }
            getSchemaDefinition() {
                return this.schema;
            }
            deferSchemaRegistration() {
                const schedule = () => {
                    nodes_1.CallbackRegistry.getInstance().registerNodeSchema(this.id, this.getSchemaDefinition());
                };
                if (typeof queueMicrotask === 'function') {
                    queueMicrotask(schedule);
                    return;
                }
                Promise.resolve().then(schedule);
            }
        },
        (() => {
            var _b;
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create((_b = _classSuper[Symbol.metadata]) !== null && _b !== void 0 ? _b : null) : void 0;
            _executionConfig_decorators = [internal_1.default];
            __esDecorate(null, null, _executionConfig_decorators, { kind: "field", name: "executionConfig", static: false, private: false, access: { has: obj => "executionConfig" in obj, get: obj => obj.executionConfig, set: (obj, value) => { obj.executionConfig = value; } }, metadata: _metadata }, _executionConfig_initializers, _executionConfig_extraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        _a.typeReferences = {},
        _a;
})();
exports.CustomNode = CustomNode;
