// @inworld/runtime/build/graph/dsl/nodes/goal_advancement_node.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoalAdvancementNode = void 0;
const snakify_1 = __importDefault(require("../utils/snakify"));
const abstract_node_1 = require("./abstract_node");
/**
 * Goal Advancement node that tracks and manages conversational goals.
 * This node evaluates the current state of an agent's goals and determines if
 * any goals should be activated or completed based on recent events, belief state,
 * and user interactions.
 *
 * Goals can have activation and completion conditions that are evaluated using an LLM
 * to analyze the conversation context. Goals can depend on other goals being completed,
 * creating complex goal-driven conversation flows.
 *
 * @input {GoalAdvancementRequest} {@link GraphTypes.GoalAdvancementRequest} - The data type that GoalAdvancementNode accepts as input
 * @output {GoalAdvancementResponse} {@link GraphTypes.GoalAdvancementResponse} - The data type that GoalAdvancementNode outputs
 *
 * @example
 * ```typescript
 * // Using with existing LLM component
 * const llmComponent = new RemoteLLMComponent({
 *   id: 'llm_component',
 *   provider: 'openai',
 *   modelName: 'gpt-4o-mini'
 * });
 *
 * const goalNode = new GoalAdvancementNode({
 *   id: 'goal_advancement',
 *   creationConfig: {
 *     goals: [
 *       {
 *         name: 'greet_user',
 *         motivation: 'Greet the user warmly',
 *         repeatable: false,
 *         activationCondition: {
 *           intents: [],
 *           requiredGoals: []
 *         },
 *         completionCondition: {
 *           intents: [],
 *           requiredGoals: [],
 *           detect: 'User has been greeted'
 *         }
 *       },
 *       {
 *         name: 'answer_question',
 *         motivation: 'Answer the user\'s question',
 *         repeatable: false,
 *         activationCondition: {
 *           intents: [],
 *           requiredGoals: ['greet_user']
 *         },
 *         completionCondition: {
 *           intents: [],
 *           requiredGoals: [],
 *           detect: 'Question has been answered'
 *         }
 *       }
 *     ]
 *   },
 *   executionConfig: {
 *     llm_component_id: llmComponent,
 *     text_generation_config: {
 *       temperature: 0.7,
 *       maxNewTokens: 160
 *     }
 *   }
 * });
 *
 * // Using with component ID string
 * const goalNode2 = new GoalAdvancementNode({
 *   id: 'goal_advancement',
 *   creationConfig: { goals: [...] },
 *   executionConfig: {
 *     llm_component_id: 'my_llm_component'
 *   }
 * });
 * ```
 */
class GoalAdvancementNode extends abstract_node_1.AbstractNode {
    /**
     * Creates a new GoalAdvancementNode instance.
     *
     * @param {GoalAdvancementNodeProps} props - Configuration for the goal advancement node
     */
    constructor(props) {
        var _a;
        super(props);
        this.creationConfig = {
            goals: props.creationConfig.goals,
        };
        // Handle LLM component - either use component instance or ID string
        let llmComponentId;
        if (typeof props.executionConfig.llm_component_id === 'string') {
            llmComponentId = props.executionConfig.llm_component_id;
        }
        else {
            const llmComponent = props.executionConfig.llm_component_id;
            llmComponentId = llmComponent.id;
            this.addComponent(llmComponent);
        }
        this.executionConfig = {
            llmComponentId,
            textGenerationConfig: props.executionConfig.text_generation_config,
            reportToClient: (_a = props.reportToClient) !== null && _a !== void 0 ? _a : false,
        };
    }
    /**
     * Converts the goal advancement node to a graph configuration node.
     *
     * @returns {GraphConfigNode} The configured goal advancement node for graph execution
     */
    toGraphConfigNode() {
        return {
            id: this.id,
            type: 'GoalAdvancementNode',
            creation_config: {
                type: 'GoalAdvancementNodeCreationConfig',
                properties: (0, snakify_1.default)(this.creationConfig),
            },
            execution_config: {
                type: 'GoalAdvancementNodeExecutionConfig',
                properties: (0, snakify_1.default)(this.executionConfig),
            },
        };
    }
}
exports.GoalAdvancementNode = GoalAdvancementNode;
