// @inworld/runtime/build/graph/dsl/nodes/keyword_matcher_node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeywordMatcherNode = void 0;
const abstract_node_1 = require("./abstract_node");
/**
 * Keyword matcher node that matches keywords in text input.
 * This node processes text input and identifies matching keywords from a predefined list.
 *
 * @input {String} {@link String} - The data type that KeywordMatcherNode accepts as input
 * @output {GraphTypes.MatchedKeywords} {@link GraphTypes.MatchedKeywords} - The data type that KeywordMatcherNode outputs
 *
 * @example
 * ```typescript
 * const keywordNode = new KeywordMatcherNode({
 *   keywords: ['urgent', 'important', 'priority', 'asap'],
 *   reportToClient: true
 * });
 * ```
 */
class KeywordMatcherNode extends abstract_node_1.AbstractNode {
    /**
     * Creates a new KeywordMatcherNode instance.
     *
     * @param props - Configuration for the keyword matcher node.
     */
    constructor(props) {
        super(props);
        this.keywords = props.keywords;
    }
    /**
     * Converts the keyword matcher node to a graph configuration node.
     *
     * @returns The configured keyword matcher node for graph execution
     */
    toGraphConfigNode() {
        return {
            id: this.id,
            type: 'KeywordMatcherNode',
            creation_config: {
                type: 'KeywordMatcherNodeCreationConfig',
                properties: {
                    keyword_groups: this.keywords, // Changed to match CPP format
                },
            },
            execution_config: {
                type: 'NodeExecutionConfig',
                properties: {
                    report_to_client: this.reportToClient,
                },
            },
        };
    }
}
exports.KeywordMatcherNode = KeywordMatcherNode;
