// @inworld/runtime/build/graph/dsl/nodes/mcp_call_tool_node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MCPCallToolNode = void 0;
const mcp_client_component_1 = require("../components/mcp_client_component");
const abstract_node_1 = require("./abstract_node");
/**
 * MCP call tool node that calls multiple tools on an MCP server in parallel.
 * This node processes tool call requests and executes them against the configured
 * MCP server, returning the results of all tool executions.
 *
 * @input {GraphTypes.ToolCallRequest} {@link GraphTypes.ToolCallRequest} - The data type that MCPCallToolNode accepts as input
 * @output {GraphTypes.ToolCallResponse} {@link GraphTypes.ToolCallResponse} - The data type that MCPCallToolNode outputs
 *
 * @example
 * ```typescript
 * const mcpComponent = new MCPClientComponent({ sessionConfig: { ... } });
 * const mcpCallToolNode = new MCPCallToolNode({
 *   mcpComponent,
 *   reportToClient: true
 * });
 * ```
 */
class MCPCallToolNode extends abstract_node_1.AbstractNode {
    /**
     * Creates a new MCPCallToolNode instance.
     *
     * @remarks
     * Provide `sessionConfig` to create a new internalcomponent, or pass
     * `mcpComponent` to reuse an existing one.
     *
     * @param props - Configuration for the MCP call tool node.
     */
    constructor(props) {
        super(props);
        let mcpComponent;
        if ('mcpComponent' in props) {
            mcpComponent = props.mcpComponent;
        }
        else {
            mcpComponent = new mcp_client_component_1.MCPClientComponent({
                id: `${this.id}_mcp_component`,
                sessionConfig: props.sessionConfig,
            });
        }
        this.addComponent(mcpComponent);
    }
    /**
     * Converts the MCP call tool node to a graph configuration node.
     *
     * @returns The configured MCP call tool node for graph execution
     */
    toGraphConfigNode() {
        return {
            id: this.id,
            type: 'MCPCallToolNode',
            execution_config: {
                type: 'MCPCallToolNodeExecutionConfig',
                properties: {
                    mcp_component_id: this.getFirstComponent().id,
                    report_to_client: this.reportToClient,
                },
            },
        };
    }
}
exports.MCPCallToolNode = MCPCallToolNode;
