// @inworld/runtime/build/graph/dsl/nodes/mcp_list_tools_node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MCPListToolsNode = void 0;
const mcp_client_component_1 = require("../components/mcp_client_component");
const abstract_node_1 = require("./abstract_node");
/**
 * MCP list tools node that lists available tools from an MCP server.
 * This node connects to an MCP server and retrieves the list of available
 * tools. It takes no inputs and returns a ListToolsResponseData containing all
 * available tools.
 *
 * @input {any} any - The data type that MCPListToolsNode accepts as input
 * @output {GraphTypes.ListToolsResponse} {@link GraphTypes.ListToolsResponse} - The data type that MCPListToolsNode outputs
 *
 * @example
 * ```typescript
 * const mcpComponent = new MCPClientComponent({ id: 'my-mcp-component', sessionConfig: { ... } });
 * const mcpListToolsNode = new MCPListToolsNode({
 *   mcpComponent,
 *   reportToClient: true
 * });
 * ```
 */
class MCPListToolsNode extends abstract_node_1.AbstractNode {
    /**
     * Creates a new MCPListToolsNode instance.
     *
     * @remarks
     * Provide `sessionConfig` to create a new internal component, or pass
     * `mcpComponent` to reuse an existing one.
     *
     * @param props - Configuration for the MCP list tools node.
     */
    constructor(props) {
        super(props);
        let mcpComponent;
        if ('mcpComponent' in props) {
            mcpComponent = props.mcpComponent;
        }
        else {
            mcpComponent = new mcp_client_component_1.MCPClientComponent({
                id: `${this.id}_llm_component`,
                sessionConfig: props.sessionConfig,
            });
        }
        this.addComponent(mcpComponent);
    }
    /**
     * Converts the MCP list tools node to a graph configuration node.
     *
     * @returns The configured MCP list tools node for graph execution
     */
    toGraphConfigNode() {
        return {
            id: this.id,
            type: 'MCPListToolsNode',
            execution_config: {
                type: 'MCPListToolsNodeExecutionConfig',
                properties: {
                    mcp_component_id: this.getFirstComponent().id,
                    report_to_client: this.reportToClient,
                },
            },
        };
    }
}
exports.MCPListToolsNode = MCPListToolsNode;
