// @inworld/runtime/build/graph/dsl/nodes/remote_llm_completion_node.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteLLMCompletionNode = void 0;
const default_1 = require("../../../internal/default");
const remote_llm_component_1 = require("../components/remote_llm_component");
const error_handling_1 = require("../error_handling");
const snakify_1 = __importDefault(require("../utils/snakify"));
const abstract_node_1 = require("./abstract_node");
/**
 * Remote LLM completion node for text generation.
 * Use a pre-configured LLM component (reusable across nodes), or provide provider/model
 * details and the node will create the component for you.
 *
 * @input {String} {@link String} - The data type that LLMCompletionNode accepts as input
 * @output {String} {@link String} - The data type that LLMCompletionNode outputs
 *
 * @example
 * ```typescript
 * // Using LLM provider configuration
 * const completionNode = new RemoteLLMCompletionNode({
 *   id: 'my-completion-node',
 *   provider: 'openai',
 *   modelName: 'gpt-4o-mini',
 *   stream: true
 * });
 *
 * // Using existing LLM component
 * const completionNodeWithComponent = new RemoteLLMCompletionNode({
 *   id: 'my-completion-node',
 *   llmComponent: existingLLMComponent
 * });
 *
 * // Using default settings
 * const defaultCompletionNode = new RemoteLLMCompletionNode();
 * ```
 */
class RemoteLLMCompletionNode extends abstract_node_1.AbstractNode {
    /**
     * Creates a new RemoteLLMCompletionNode instance.
     *
     * @remarks
     * Provide LLM provider settings to create a new internal component, or pass
     * `llmComponent` to reuse an existing one. If omitted, defaults are applied.
     *
     * @param props - Optional configuration for the completion node.
     */
    constructor(props = {}) {
        var _a, _b, _c, _d, _e;
        super(props);
        this.executionConfig = {
            llmComponentId: '',
            textGenerationConfig: Object.assign({}, (0, snakify_1.default)((_a = props.textGenerationConfig) !== null && _a !== void 0 ? _a : {})),
            stream: (_b = props.stream) !== null && _b !== void 0 ? _b : false,
            reportToClient: (_c = props.reportToClient) !== null && _c !== void 0 ? _c : false,
        };
        if (props.errorHandling) {
            this.executionConfig.errorHandling = (0, error_handling_1.toGraphConfig)(props.errorHandling);
        }
        let llmComponent;
        if ('llmComponent' in props) {
            llmComponent = props.llmComponent;
        }
        else {
            llmComponent = new remote_llm_component_1.RemoteLLMComponent({
                id: `${this.id}_llm_component`,
                provider: (_d = props.provider) !== null && _d !== void 0 ? _d : default_1.Default.llm.provider,
                modelName: (_e = props.modelName) !== null && _e !== void 0 ? _e : default_1.Default.llm.modelName,
            });
        }
        this.addComponent(llmComponent);
        this.executionConfig.llmComponentId = llmComponent.id;
    }
    toGraphConfigNode() {
        return {
            id: this.id,
            type: 'LLMCompletionNode',
            execution_config: {
                type: 'LLMCompletionNodeExecutionConfig',
                properties: (0, snakify_1.default)(this.executionConfig),
            },
        };
    }
}
exports.RemoteLLMCompletionNode = RemoteLLMCompletionNode;
