// @inworld/runtime/build/graph/dsl/nodes/text_classifier_node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextClassifierNode = void 0;
const abstract_node_1 = require("./abstract_node");
/**
 * Text classifier node for content classification and topic detection.
 * Analyzes text input and classifies it into predefined categories using ML models.
 * Returns classification results with confidence scores.
 *
 * You can either use a pre-configured text classifier component that could be reused across
 * multiple nodes or node will create a new component for you.
 *
 * @input {String} {@link String} - The data type that TextClassifierNode accepts as input
 * @output {GraphTypes.ClassificationResult} {@link GraphTypes.ClassificationResult} - The data type that TextClassifierNode outputs
 *
 * @example
 * ```typescript
 * // Using component configuration
 * const classifierNode = new TextClassifierNode({
 *   id: 'content-classifier',
 *   modelWeightsPath: '/models/classifier_weights.json',
 *   embedderComponentId: 'text_embedder_component',
 *   supportedClasses: ['hategroup', 'selfharm', 'sexual', 'sexualminors', 'substance'],
 *   classifierConfig: {
 *     classes: [
 *       { label: 'hategroup', threshold: 0.8 },
 *       { label: 'selfharm', threshold: 0.9 }
 *     ]
 *   }
 * });
 *
 * // Using existing text classifier component
 * const classifierComponent = new TextClassifierComponent({
 *   id: 'existing-classifier-component',
 *   modelWeightsPath: '/models/classifier_weights.json',
 *   embedderComponentId: 'text_embedder_component',
 *   supportedClasses: ['hategroup', 'selfharm']
 * });
 * const classifierNodeWithComponent = new TextClassifierNode({
 *   id: 'content-classifier',
 *   textClassifierComponent: classifierComponent
 * });
 * ```
 */
class TextClassifierNode extends abstract_node_1.AbstractNode {
    /**
     * Creates a new TextClassifierNode instance.
     *
     * @remarks
     * The TextClassifierNode creates its own text classifier using the factory pattern
     * and doesn't need a separate TextClassifierComponent.
     *
     * @param props - Configuration for the text classifier node.
     */
    constructor(props) {
        // Only pass id and a boolean reportToClient to super
        super({
            id: props.id,
            reportToClient: typeof props.reportToClient === 'boolean'
                ? props.reportToClient
                : false,
        });
        this.classifierConfig = props.classifierConfig;
        this.errorHandling = props.errorHandling;
        // Extract configuration based on props type
        if ('textClassifierComponent' in props) {
            // If using existing component, we don't create our own classifier
            this.addComponent(props.textClassifierComponent);
            this.modelWeightsPath = ''; // Will be handled by component
            this.embedderComponentId = '';
            this.supportedClasses = [];
        }
        else {
            // Use direct configuration - no component needed
            this.modelWeightsPath = props.modelWeightsPath;
            this.embedderComponentId = props.embedderComponentId;
            this.supportedClasses = props.supportedClasses;
        }
    }
    /**
     * Converts the text classifier node to a graph configuration node.
     *
     * @returns The configured text classifier node for graph execution
     */
    toGraphConfigNode() {
        const node = {
            id: this.id,
            type: 'TextClassifierNode',
            creation_config: {
                type: 'TextClassifierNodeCreationConfig',
                properties: {
                    embedder_component_id: this.embedderComponentId || 'text_embedder_component',
                    classifier_config: {
                        model_weights_path: this.modelWeightsPath,
                        supported_classes: this.supportedClasses || [],
                    },
                },
            },
            execution_config: {
                type: 'TextClassifierNodeExecutionConfig',
                properties: {
                    report_to_client: this.reportToClient,
                    classifier_config: this.classifierConfig || { classes: [] },
                },
            },
        };
        // Add error_handling if provided
        if (this.errorHandling) {
            node.execution_config.properties.error_handling =
                this.errorHandling;
        }
        return node;
    }
}
exports.TextClassifierNode = TextClassifierNode;
