// @inworld/runtime/build/graph/dsl/nodes/transform_node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransformNode = void 0;
const inworld_1 = require("../../../internal/error/inworld");
const abstract_node_1 = require("./abstract_node");
/**
 * Declarative data transformation node that converts any BaseData payload
 * into another BaseData type using CEL expressions.
 *
 * @input {GraphTypes.Custom<any>} {@link GraphTypes.Custom} - Accepts any BaseData-compatible input payload
 * @output {GraphTypes.Custom<any>} {@link GraphTypes.Custom} - Outputs the BaseData type specified by {@link TransformNodeProps.outputType}
 *
 * @example
 * ```typescript
 * const transformNode = new TransformNode({
 *   id: 'extract_llm_text',
 *   reportToClient: true,
 *   outputType: 'Text',
 *   outputTemplate: {
 *     value: 'input.responses[0].candidates[0].text',
 *   },
 * });
 * ```
 */
class TransformNode extends abstract_node_1.AbstractNode {
    /**
     * Creates a new {@link TransformNode}.
     *
     * @param {TransformNodeProps} props - Node configuration including the target type and template.
     */
    constructor(props) {
        var _a;
        super(props);
        if (!((_a = props.outputType) === null || _a === void 0 ? void 0 : _a.trim())) {
            throw new inworld_1.InworldError('TransformNode requires a non-empty outputType');
        }
        this.outputType = props.outputType;
        this.outputTemplate = props.outputTemplate;
    }
    /**
     * Serializes the node into graph configuration format understood by the runtime.
     *
     * @returns The graph configuration node that represents this transform node.
     */
    toGraphConfigNode() {
        const creationProperties = {
            output_type: this.outputType,
        };
        if (this.outputTemplate !== undefined) {
            creationProperties.output_template = this.outputTemplate;
        }
        return {
            id: this.id,
            type: 'TransformNode',
            creation_config: {
                type: 'TransformNodeCreationConfig',
                properties: creationProperties,
            },
            execution_config: {
                type: 'NodeExecutionConfig',
                properties: {
                    report_to_client: this.reportToClient,
                },
            },
        };
    }
}
exports.TransformNode = TransformNode;
