// @inworld/runtime/build/index.js
"use strict";
/**
 * Main entry point for the Inworld Runtime Node.js SDK
 *
 * This file exports all the public interfaces, types and utilities of the
 * Inworld Runtime SDK, organized into namespaced modules for easier
 * consumption. It serves as the primary interface for developers to interact
 * with the SDK.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphOutputStream = exports.Graph = exports.stopInworldRuntime = exports.DataStreamWithMetadata = exports.telemetry = exports.core = exports.common = void 0;
const utils_1 = require("./common/utils");
const node_api_1 = require("./internal/node_api");
Object.defineProperty(exports, "stopInworldRuntime", { enumerable: true, get: function () { return node_api_1.stopInworldRuntime; } });
const telemetry_1 = require("./telemetry");
const types_1 = require("./telemetry/types");
exports.common = {
    generateExecutionId: utils_1.generateExecutionId,
};
exports.core = {
    generateExecutionId: utils_1.generateExecutionId,
};
exports.telemetry = {
    init: telemetry_1.init,
    shutdown: telemetry_1.shutdown,
    ExporterType: types_1.ExporterType,
    LogLevel: types_1.LogLevel,
    logger: telemetry_1.logger,
    metric: telemetry_1.metric,
    configureMetric: telemetry_1.configureMetric,
    MetricType: types_1.MetricType,
    Span: telemetry_1.Span,
    startSpan: telemetry_1.startSpan,
    startSpanWithAttributes: telemetry_1.startSpanWithAttributes,
    startSpanWithParent: telemetry_1.startSpanWithParent,
};
var data_stream_with_metadata_1 = require("./common/data_types/api/data_stream_with_metadata");
Object.defineProperty(exports, "DataStreamWithMetadata", { enumerable: true, get: function () { return data_stream_with_metadata_1.DataStreamWithMetadata; } });
var graph_1 = require("./graph");
Object.defineProperty(exports, "Graph", { enumerable: true, get: function () { return graph_1.Graph; } });
Object.defineProperty(exports, "GraphOutputStream", { enumerable: true, get: function () { return graph_1.GraphOutputStream; } });
