// @inworld/runtime/build/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stopRuntime = stopRuntime;
exports.initializeInworldRuntime = initializeInworldRuntime;
/**
 * @internal
 */
require("dotenv/config");
const graph_1 = require("./graph");
const greeter_1 = require("./internal/greeter");
const telemetry_1 = require("./telemetry");
let initialized = false;
if (process.env.INWORLD_DEBUG) {
    // eslint-disable-next-line no-console
    const originalDebug = console.debug.bind(console);
    // eslint-disable-next-line no-console
    console.debug = (...args) => {
        originalDebug(...args);
    };
}
else {
    // Muted when is not debug mode;
    // eslint-disable-next-line no-console
    console.debug = () => { };
}
function stopRuntime() {
    // eslint-disable-next-line no-console
    console.debug('Starting Inworld runtime cleanup...');
    try {
        graph_1.CallbackRegistry.destroy();
        initialized = false;
        // eslint-disable-next-line no-console
        console.debug('Inworld runtime cleanup completed successfully');
    }
    catch (error) {
        // eslint-disable-next-line no-console
        console.error('Error during Inworld runtime cleanup:', error);
    }
}
['SIGINT', 'SIGSEGV', 'SIGTERM', 'SIGQUIT', 'SIGBUS', 'exit'].forEach((signal) => {
    process.on(signal, () => {
        // eslint-disable-next-line no-console
        console.debug('exit with code:', signal);
        stopRuntime();
    });
});
function initializeInworldRuntime(opts) {
    if (initialized) {
        return;
    }
    initialized = true;
    if (['true', '1'].includes(process.env.SHOW_NODEJS_WELCOME_SCREEN)) {
        (0, greeter_1.showWelcomeMessage)();
    }
    const isTelemetryDisabled = process.env.DISABLE_TELEMETRY === 'true' ||
        process.env.DISABLE_TELEMETRY === '1';
    if (!isTelemetryDisabled && opts.apiKey) {
        (0, telemetry_1.init)({
            appName: opts.appName,
            appVersion: opts.appVersion,
            apiKey: opts.apiKey,
            sdkName: 'Inworld Runtime Node.js SDK',
            sdkVersion: require('../package.json').version,
        });
        ['SIGINT', 'SIGSEGV', 'SIGTERM', 'SIGQUIT', 'SIGBUS', 'exit'].forEach((signal) => {
            process.on(signal, () => {
                (0, telemetry_1.shutdown)();
            });
        });
    }
    else if (isTelemetryDisabled) {
        // eslint-disable-next-line no-console
        console.log('Telemetry is disabled via DISABLE_TELEMETRY environment variable');
    }
    else {
        // eslint-disable-next-line no-console
        console.warn('INWORLD_API_KEY is not set, telemetry will not be initialized');
    }
}
