// @inworld/runtime/build/graph/nodes/custom/process_context_Impl.js
"use strict";
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProcessContextImpl = void 0;
const internal_1 = __importDefault(require("../../../internal/decorators/internal"));
const interfaces_1 = require("../../../primitives/interfaces");
const from_external_1 = require("../../utils/from_external");
const to_external_1 = require("../../utils/to_external");
let DataStoreImpl = (() => {
    var _a;
    let _datastore_decorators;
    let _datastore_initializers = [];
    let _datastore_extraInitializers = [];
    return _a = class DataStoreImpl {
            constructor(external) {
                this.datastore = __runInitializers(this, _datastore_initializers, void 0);
                __runInitializers(this, _datastore_extraInitializers);
                this.datastore = external;
            }
            add(key, data) {
                return this.set(key, data);
            }
            set(key, data) {
                try {
                    this.datastore.set(key, (0, to_external_1.toExternal)(data));
                }
                catch (error) {
                    // eslint-disable-next-line no-console
                    console.error(error);
                    return false;
                }
                return true;
            }
            get(key) {
                return (0, from_external_1.fromExternal)(this.datastore.get(key));
            }
            has(key) {
                return this.datastore.has(key);
            }
            keys() {
                return this.datastore.keys();
            }
            size() {
                return this.datastore.size();
            }
            clear() {
                return this.datastore.clear();
            }
        },
        (() => {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _datastore_decorators = [internal_1.default];
            __esDecorate(null, null, _datastore_decorators, { kind: "field", name: "datastore", static: false, private: false, access: { has: obj => "datastore" in obj, get: obj => obj.datastore, set: (obj, value) => { obj.datastore = value; } }, metadata: _metadata }, _datastore_initializers, _datastore_extraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        _a;
})();
/**
 * @internal
 * Provides access to execution context information for custom nodes.
 */
let ProcessContextImpl = (() => {
    var _a;
    let _external_decorators;
    let _external_initializers = [];
    let _external_extraInitializers = [];
    let _datastore_decorators;
    let _datastore_initializers = [];
    let _datastore_extraInitializers = [];
    let _executionConfig_decorators;
    let _executionConfig_initializers = [];
    let _executionConfig_extraInitializers = [];
    let _llmInterface_decorators;
    let _llmInterface_initializers = [];
    let _llmInterface_extraInitializers = [];
    let _textEmbedderInterface_decorators;
    let _textEmbedderInterface_initializers = [];
    let _textEmbedderInterface_extraInitializers = [];
    return _a = class ProcessContextImpl {
            constructor(external, executionConfig) {
                this.external = __runInitializers(this, _external_initializers, void 0);
                this.datastore = (__runInitializers(this, _external_extraInitializers), __runInitializers(this, _datastore_initializers, void 0));
                this.executionConfig = (__runInitializers(this, _datastore_extraInitializers), __runInitializers(this, _executionConfig_initializers, void 0));
                this.llmInterface = (__runInitializers(this, _executionConfig_extraInitializers), __runInitializers(this, _llmInterface_initializers, void 0));
                this.textEmbedderInterface = (__runInitializers(this, _llmInterface_extraInitializers), __runInitializers(this, _textEmbedderInterface_initializers, void 0));
                __runInitializers(this, _textEmbedderInterface_extraInitializers);
                this.external = external;
                this.datastore = new DataStoreImpl(external.processContext.datastore);
                this.executionConfig = executionConfig;
            }
            /**
             * Gets the datastore for the execution.
             * @returns The datastore
             */
            getDatastore() {
                return this.datastore;
            }
            /**
             * Gets the execution ID from the context.
             * @returns The execution ID
             */
            getExecutionId() {
                return this.external.processContext.executionId;
            }
            /**
             * Gets the variant string from the context.
             * @returns The variant string
             */
            getVariant() {
                return this.external.processContext.variant;
            }
            /**
             * Checks if the execution has been cancelled.
             * @returns True if cancelled, false otherwise
             */
            get isCancelled() {
                return this.external.processContext.isCancelled;
            }
            /**
             * Gets the configuration helper from the context.
             * @returns The configuration object
             */
            getExecutionConfig() {
                var _b, _c, _d, _e;
                return Object.assign(Object.assign({}, this.external.processContext.executionConfig), { properties: Object.assign(Object.assign({}, ((_c = (_b = this.executionConfig) === null || _b === void 0 ? void 0 : _b.properties) !== null && _c !== void 0 ? _c : {})), ((_e = (_d = this.external.processContext.executionConfig) === null || _d === void 0 ? void 0 : _d.properties) !== null && _e !== void 0 ? _e : {})) });
            }
            get nodeId() {
                return this.external.nodeId;
            }
            getLLMInterface(componentId) {
                if (this.llmInterface) {
                    return this.llmInterface;
                }
                this.llmInterface = new interfaces_1.LLMInterfaceWrapper(this.external.processContext.componentRegistry, componentId);
                return this.llmInterface;
            }
            getEmbedderInterface(componentId) {
                if (this.textEmbedderInterface) {
                    return this.textEmbedderInterface;
                }
                this.textEmbedderInterface = new interfaces_1.TextEmbedderInterfaceWrapper(this.external.processContext.componentRegistry, componentId);
                return this.textEmbedderInterface;
            }
        },
        (() => {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
            _external_decorators = [internal_1.default];
            _datastore_decorators = [internal_1.default];
            _executionConfig_decorators = [internal_1.default];
            _llmInterface_decorators = [internal_1.default];
            _textEmbedderInterface_decorators = [internal_1.default];
            __esDecorate(null, null, _external_decorators, { kind: "field", name: "external", static: false, private: false, access: { has: obj => "external" in obj, get: obj => obj.external, set: (obj, value) => { obj.external = value; } }, metadata: _metadata }, _external_initializers, _external_extraInitializers);
            __esDecorate(null, null, _datastore_decorators, { kind: "field", name: "datastore", static: false, private: false, access: { has: obj => "datastore" in obj, get: obj => obj.datastore, set: (obj, value) => { obj.datastore = value; } }, metadata: _metadata }, _datastore_initializers, _datastore_extraInitializers);
            __esDecorate(null, null, _executionConfig_decorators, { kind: "field", name: "executionConfig", static: false, private: false, access: { has: obj => "executionConfig" in obj, get: obj => obj.executionConfig, set: (obj, value) => { obj.executionConfig = value; } }, metadata: _metadata }, _executionConfig_initializers, _executionConfig_extraInitializers);
            __esDecorate(null, null, _llmInterface_decorators, { kind: "field", name: "llmInterface", static: false, private: false, access: { has: obj => "llmInterface" in obj, get: obj => obj.llmInterface, set: (obj, value) => { obj.llmInterface = value; } }, metadata: _metadata }, _llmInterface_initializers, _llmInterface_extraInitializers);
            __esDecorate(null, null, _textEmbedderInterface_decorators, { kind: "field", name: "textEmbedderInterface", static: false, private: false, access: { has: obj => "textEmbedderInterface" in obj, get: obj => obj.textEmbedderInterface, set: (obj, value) => { obj.textEmbedderInterface = value; } }, metadata: _metadata }, _textEmbedderInterface_initializers, _textEmbedderInterface_extraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        _a;
})();
exports.ProcessContextImpl = ProcessContextImpl;
