// @inworld/runtime/build/common/data_types/api/custom.js
"use strict";
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomData = void 0;
exports.CustomCreate = CustomCreate;
const internal_1 = __importDefault(require("../../../internal/decorators/internal"));
const pb_1 = require("../../../internal/pb");
const abstract_api_data_type_1 = require("./abstract_api_data_type");
let CustomData = (() => {
    var _a;
    let _classSuper = abstract_api_data_type_1.AbstractApiDataType;
    let _instanceExtraInitializers = [];
    let ___inw_value_decorators;
    let ___inw_value_initializers = [];
    let ___inw_value_extraInitializers = [];
    let ___inw_schema_decorators;
    let ___inw_schema_initializers = [];
    let ___inw_schema_extraInitializers = [];
    let _getValue_decorators;
    let _getSchema_decorators;
    let _toTaggedValue_decorators;
    let _toProto_decorators;
    return _a = class CustomData extends _classSuper {
            constructor(params) {
                super();
                /** @internal */
                this.__inw_value = (__runInitializers(this, _instanceExtraInitializers), __runInitializers(this, ___inw_value_initializers, void 0));
                /** @internal */
                this.__inw_schema = (__runInitializers(this, ___inw_value_extraInitializers), __runInitializers(this, ___inw_schema_initializers, void 0));
                __runInitializers(this, ___inw_schema_extraInitializers);
                Object.defineProperty(this, '__inw_value', {
                    value: params.value,
                    enumerable: false,
                    configurable: false,
                    writable: false,
                });
                Object.defineProperty(this, '__inw_schema', {
                    value: params.schema,
                    enumerable: false,
                    configurable: false,
                    writable: false,
                });
                Object.assign(this, params.value);
            }
            /** @internal */
            getValue() {
                return this.__inw_value;
            }
            /** @internal */
            getSchema() {
                return this.__inw_schema;
            }
            /** @internal */
            toTaggedValue() {
                return {
                    type: 'Json',
                    data: {
                        value: this.__inw_value,
                    },
                };
            }
            /** @internal */
            toProto() {
                return pb_1.PbType.customData(this.__inw_value, this.__inw_schema);
            }
        },
        (() => {
            var _b;
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create((_b = _classSuper[Symbol.metadata]) !== null && _b !== void 0 ? _b : null) : void 0;
            ___inw_value_decorators = [internal_1.default];
            ___inw_schema_decorators = [internal_1.default];
            _getValue_decorators = [internal_1.default];
            _getSchema_decorators = [internal_1.default];
            _toTaggedValue_decorators = [internal_1.default];
            _toProto_decorators = [internal_1.default];
            __esDecorate(_a, null, _getValue_decorators, { kind: "method", name: "getValue", static: false, private: false, access: { has: obj => "getValue" in obj, get: obj => obj.getValue }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(_a, null, _getSchema_decorators, { kind: "method", name: "getSchema", static: false, private: false, access: { has: obj => "getSchema" in obj, get: obj => obj.getSchema }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(_a, null, _toTaggedValue_decorators, { kind: "method", name: "toTaggedValue", static: false, private: false, access: { has: obj => "toTaggedValue" in obj, get: obj => obj.toTaggedValue }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(_a, null, _toProto_decorators, { kind: "method", name: "toProto", static: false, private: false, access: { has: obj => "toProto" in obj, get: obj => obj.toProto }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(null, null, ___inw_value_decorators, { kind: "field", name: "__inw_value", static: false, private: false, access: { has: obj => "__inw_value" in obj, get: obj => obj.__inw_value, set: (obj, value) => { obj.__inw_value = value; } }, metadata: _metadata }, ___inw_value_initializers, ___inw_value_extraInitializers);
            __esDecorate(null, null, ___inw_schema_decorators, { kind: "field", name: "__inw_schema", static: false, private: false, access: { has: obj => "__inw_schema" in obj, get: obj => obj.__inw_schema, set: (obj, value) => { obj.__inw_schema = value; } }, metadata: _metadata }, ___inw_schema_initializers, ___inw_schema_extraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        _a;
})();
exports.CustomData = CustomData;
function CustomCreate(members) {
    return class extends CustomData {
        constructor() {
            super({ value: members });
        }
    };
}
