// @inworld/runtime/build/common/data_types/api/data_stream_with_metadata.js
"use strict";
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataStreamWithMetadata = void 0;
const internal_1 = __importDefault(require("../../../internal/decorators/internal"));
const pb_1 = require("../../../internal/pb");
const audio_chunk_1 = require("../../stream/audio_chunk");
const content_1 = require("../../stream/content");
const multimodal_content_1 = require("../../stream/multimodal_content");
const text_1 = require("../../stream/text");
const tts_output_1 = require("../../stream/tts_output");
const abstract_api_data_type_1 = require("./abstract_api_data_type");
/**
 * Class for wrapping streaming BaseData with arbitrary metadata.
 *
 * This class combines:
 * - A RuntimeInputStream of BaseData
 * - Arbitrary JSON metadata fields
 *
 * The stream's element type is automatically captured in metadata.elementType
 * when converted from C++ (e.g., 'Text', 'Audio', 'Content', 'SimpleBaseData').
 *
 * Example usage:
 * ```typescript
 * const wrapper = new DataStreamWithMetadata(
 *   runtimeInputStream,
 *   { language: "en", source: "llm", model: "gpt-4" }
 * );
 *
 * // Check what type of elements the stream contains
 * if (wrapper.getElementType() === 'Audio') {
 *   // Handle audio stream
 * }
 *
 * // Access stream - automatically detects type from runtime
 * const stream = wrapper.toStream();  // Auto-detects: AudioChunkStream | TextStream | etc.
 * for await (const item of stream) {
 *   console.log(item);
 * }
 *
 * // Runtime will return the correct type based on stream.type property
 * // No need to specify type parameters!
 * ```
 */
let DataStreamWithMetadata = (() => {
    var _a;
    let _classSuper = abstract_api_data_type_1.AbstractApiDataType;
    let _instanceExtraInitializers = [];
    let _toTaggedValue_decorators;
    let _toProto_decorators;
    return _a = class DataStreamWithMetadata extends _classSuper {
            /**
             * Creates a new DataStreamWithMetadata instance.
             * @param streamOrWrapper - RuntimeInputStream of BaseData, TextStream, or any object with getStream() method
             * @param metadata - Arbitrary metadata fields. Can include 'elementType' to
             *                   document what type of elements the stream contains.
             */
            constructor(streamOrWrapper, metadata = {}) {
                super();
                /**
                 * Arbitrary metadata fields.
                 * By convention, metadata.elementType contains the stream's element type
                 * (e.g., 'Text', 'Audio', 'Content', 'SimpleBaseData').
                 */
                this.metadata = __runInitializers(this, _instanceExtraInitializers);
                // Extract the underlying NAPI stream if wrapped
                if (streamOrWrapper &&
                    typeof streamOrWrapper.getStream === 'function') {
                    this.stream = streamOrWrapper.getStream();
                }
                else {
                    this.stream = streamOrWrapper;
                }
                this.metadata = metadata;
            }
            /**
             * Gets the stream.
             *
             * @returns The RuntimeInputStream
             */
            getStream() {
                return this.stream;
            }
            /**
             * Gets the metadata.
             *
             * @returns The metadata object
             */
            getMetadata() {
                return this.metadata;
            }
            /**
             * Gets the element type of the stream if specified in metadata.
             * This indicates what type of BaseData elements the stream contains
             * (e.g., 'Text', 'Audio', 'Content', 'SimpleBaseData').
             *
             * @returns The element type string, or undefined if not specified
             */
            getElementType() {
                return this.metadata.elementType;
            }
            /**
             * Converts the underlying NAPI stream to the appropriate typed stream wrapper.
             *
             * Auto-detects the stream type from:
             * 1. metadata.elementType (set by C++ when creating DataStreamWithMetadata)
             * 2. stream.type property (fallback for direct streams)
             *
             * Returns the correct wrapper (TextStream, AudioChunkStream, ContentStream, or TTSOutputStream).
             *
             * @returns The appropriate stream wrapper based on runtime type detection
             *
             * @example
             * // Auto-detect from runtime - works for any stream type:
             * const stream = wrapper.toStream();
             * for await (const item of stream) {
             *   // TypeScript knows this is one of the stream types
             *   console.log(item);
             * }
             */
            toStream() {
                // Prefer metadata.elementType (more specific) over stream.type
                // When DataStreamWithMetadata wraps BaseData streams, the elementType tells us
                // what kind of BaseData it contains (e.g., 'Audio', 'Text')
                const elementType = this.metadata.elementType || this.stream.type;
                const streamType = elementType;
                switch (streamType) {
                    case 'TTSOutput':
                        return tts_output_1.TTSOutputStream.fromExternal(this.stream);
                    case 'Content':
                        return content_1.ContentStream.fromExternal(this.stream);
                    case 'MultimodalContent':
                        return multimodal_content_1.MultimodalContentStream.fromExternal(this.stream);
                    case 'Audio':
                        return audio_chunk_1.AudioChunkStream.fromExternal(this.stream);
                    case 'Text':
                    case 'String':
                    case 'BaseData':
                    default:
                        // For Text, String, and unknown types, default to TextStream
                        return text_1.TextStream.fromExternal(this.stream);
                }
            }
            /**
             * @deprecated Use toStream() instead. This method will be removed in a future version.
             * Reconstructs a TextStream from the underlying NAPI stream.
             */
            toTextStream() {
                return text_1.TextStream.fromExternal(this.stream);
            }
            /**
             * @deprecated Use toStream() instead. This method will be removed in a future version.
             * Reconstructs an AudioChunkStream from the underlying NAPI stream.
             */
            toAudioChunkStream() {
                return audio_chunk_1.AudioChunkStream.fromExternal(this.stream);
            }
            /**
             * @deprecated Use toStream() instead. This method will be removed in a future version.
             * Reconstructs a ContentStream from the underlying NAPI stream.
             */
            toContentStream() {
                return content_1.ContentStream.fromExternal(this.stream);
            }
            /**
             * @deprecated Use toStream() instead. This method will be removed in a future version.
             * Reconstructs a TTSOutputStream from the underlying NAPI stream.
             */
            toTTSOutputStream() {
                return tts_output_1.TTSOutputStream.fromExternal(this.stream);
            }
            /** @internal */
            toTaggedValue() {
                // Serialize metadata to protobuf format with _pb field
                const pbData = pb_1.PbType.dataStreamWithMetadata(this.metadata);
                return {
                    _pb: pbData._pb,
                    _stream: this.stream,
                };
            }
            /** @internal */
            toProto() {
                // Return the serialized protobuf data
                return pb_1.PbType.dataStreamWithMetadata(this.metadata);
            }
        },
        (() => {
            var _b;
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create((_b = _classSuper[Symbol.metadata]) !== null && _b !== void 0 ? _b : null) : void 0;
            _toTaggedValue_decorators = [internal_1.default];
            _toProto_decorators = [internal_1.default];
            __esDecorate(_a, null, _toTaggedValue_decorators, { kind: "method", name: "toTaggedValue", static: false, private: false, access: { has: obj => "toTaggedValue" in obj, get: obj => obj.toTaggedValue }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(_a, null, _toProto_decorators, { kind: "method", name: "toProto", static: false, private: false, access: { has: obj => "toProto" in obj, get: obj => obj.toProto }, metadata: _metadata }, null, _instanceExtraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        _a;
})();
exports.DataStreamWithMetadata = DataStreamWithMetadata;
