// @inworld/runtime/build/common/data_types/api/llm_chat_routing_request.js
"use strict";
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __setFunctionName = (this && this.__setFunctionName) || function (f, name, prefix) {
    if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
    return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LLMChatRoutingRequest = void 0;
const internal_1 = __importDefault(require("../../../internal/decorators/internal"));
const pb_1 = require("../../../internal/pb");
const validate_1 = require("../../../internal/utils/validate");
const graphs_1 = require("../../api/graphs");
const abstract_api_data_type_1 = require("./abstract_api_data_type");
/**
 * Represents a complete chat routing request for LLM processing with
 * intelligent fallback.
 *
 * This class stores messages, model configuration, and routing configuration
 * for LLM requests, enabling automatic fallback to alternative models when the
 * primary model fails.
 */
let LLMChatRoutingRequest = (() => {
    let _classDecorators = [(0, validate_1.ValidateInput)(graphs_1.LLMChatRoutingRequestSchema, true)];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _classSuper = abstract_api_data_type_1.AbstractApiDataType;
    let _instanceExtraInitializers = [];
    let _toTaggedValue_decorators;
    let _toProto_decorators;
    var LLMChatRoutingRequest = _classThis = class extends _classSuper {
        /**
         * Creates a new LLMChatRoutingRequest instance.
         * @param {LLMChatRoutingRequestInterface} llmChatRoutingRequest - The chat
         *     routing request to wrap.
         */
        constructor(llmChatRoutingRequest) {
            var _a;
            super();
            /**
             * The messages to be sent to the LLM.
             */
            this.messages = __runInitializers(this, _instanceExtraInitializers);
            this.messages = (_a = llmChatRoutingRequest.messages) !== null && _a !== void 0 ? _a : [];
            this.modelId = llmChatRoutingRequest.modelId;
            this.modelSelection = llmChatRoutingRequest.modelSelection;
            this.promptVariables = llmChatRoutingRequest.promptVariables;
            this.responseFormat = llmChatRoutingRequest.responseFormat;
            this.strategy = llmChatRoutingRequest.strategy;
            this.stream = llmChatRoutingRequest.stream;
            this.textGenerationConfig = llmChatRoutingRequest.textGenerationConfig;
            this.toolChoice = llmChatRoutingRequest.toolChoice;
            this.tools = llmChatRoutingRequest.tools;
        }
        /** @internal */
        toTaggedValue() {
            throw new Error('Not implemented');
        }
        /** @internal */
        toProto() {
            return pb_1.PbType.llmChatRoutingRequest({
                messages: this.messages,
                modelId: this.modelId,
                modelSelection: this.modelSelection,
                promptVariables: this.promptVariables,
                responseFormat: this.responseFormat,
                strategy: this.strategy,
                stream: this.stream,
                textGenerationConfig: this.textGenerationConfig,
                toolChoice: this.toolChoice,
                tools: this.tools,
            });
        }
    };
    __setFunctionName(_classThis, "LLMChatRoutingRequest");
    (() => {
        var _a;
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create((_a = _classSuper[Symbol.metadata]) !== null && _a !== void 0 ? _a : null) : void 0;
        _toTaggedValue_decorators = [internal_1.default];
        _toProto_decorators = [internal_1.default];
        __esDecorate(_classThis, null, _toTaggedValue_decorators, { kind: "method", name: "toTaggedValue", static: false, private: false, access: { has: obj => "toTaggedValue" in obj, get: obj => obj.toTaggedValue }, metadata: _metadata }, null, _instanceExtraInitializers);
        __esDecorate(_classThis, null, _toProto_decorators, { kind: "method", name: "toProto", static: false, private: false, access: { has: obj => "toProto" in obj, get: obj => obj.toProto }, metadata: _metadata }, null, _instanceExtraInitializers);
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        LLMChatRoutingRequest = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return LLMChatRoutingRequest = _classThis;
})();
exports.LLMChatRoutingRequest = LLMChatRoutingRequest;
