// @inworld/runtime/build/common/data_types/api/matched_keywords.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __setFunctionName = (this && this.__setFunctionName) || function (f, name, prefix) {
    if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
    return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MatchedKeywords = void 0;
const z = __importStar(require("zod"));
const internal_1 = __importDefault(require("../../../internal/decorators/internal"));
const pb_1 = require("../../../internal/pb");
const validate_1 = require("../../../internal/utils/validate");
const library_1 = require("../../api/library");
const abstract_api_data_type_1 = require("./abstract_api_data_type");
/**
 * Represents the result of keyword matching operations.
 *
 * This class encapsulates the result from keyword matching, containing
 * information about matched keywords and their associated groups.
 */
let MatchedKeywords = (() => {
    let _classDecorators = [(0, validate_1.ValidateInput)(z.array(library_1.KeywordMatchSchema), true)];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _classSuper = abstract_api_data_type_1.AbstractApiDataType;
    let _instanceExtraInitializers = [];
    let _toTaggedValue_decorators;
    let _toProto_decorators;
    var MatchedKeywords = _classThis = class extends _classSuper {
        /**
         * Creates a new MatchedKeywords instance.
         * @param keywords The list of matched keywords.
         */
        constructor(keywords) {
            super();
            /**
             * The matched keywords and related metadata.
             */
            this.keywords = __runInitializers(this, _instanceExtraInitializers);
            this.keywords = keywords;
        }
        /** @internal */
        toTaggedValue() {
            return {
                type: 'MatchedKeywords',
                data: {
                    matchedKeywords: this.keywords,
                },
            };
        }
        /** @internal */
        toProto() {
            return pb_1.PbType.matchedKeywords(this.keywords);
        }
    };
    __setFunctionName(_classThis, "MatchedKeywords");
    (() => {
        var _a;
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create((_a = _classSuper[Symbol.metadata]) !== null && _a !== void 0 ? _a : null) : void 0;
        _toTaggedValue_decorators = [internal_1.default];
        _toProto_decorators = [internal_1.default];
        __esDecorate(_classThis, null, _toTaggedValue_decorators, { kind: "method", name: "toTaggedValue", static: false, private: false, access: { has: obj => "toTaggedValue" in obj, get: obj => obj.toTaggedValue }, metadata: _metadata }, null, _instanceExtraInitializers);
        __esDecorate(_classThis, null, _toProto_decorators, { kind: "method", name: "toProto", static: false, private: false, access: { has: obj => "toProto" in obj, get: obj => obj.toProto }, metadata: _metadata }, null, _instanceExtraInitializers);
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        MatchedKeywords = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return MatchedKeywords = _classThis;
})();
exports.MatchedKeywords = MatchedKeywords;
