// @inworld/runtime/build/common/data_types/api/tts_request.js
"use strict";
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TTSRequest = void 0;
const internal_1 = __importDefault(require("../../../internal/decorators/internal"));
const pb_1 = require("../../../internal/pb");
const stream_1 = require("../../stream");
const abstract_api_data_type_1 = require("./abstract_api_data_type");
/**
 * This class encapsulates a request for text-to-speech synthesis, containing
 * either text content or a data stream of text, along with an optional voice
 * configuration that can override the default voice settings.
 */
let TTSRequest = (() => {
    var _a;
    let _classSuper = abstract_api_data_type_1.AbstractApiDataType;
    let _instanceExtraInitializers = [];
    let _toTaggedValue_decorators;
    let _toProto_decorators;
    return _a = class TTSRequest extends _classSuper {
            /**
             * Creates a new TTSRequest instance.
             * @param request The TTS request to wrap.
             * @param stream
             */
            constructor(request, stream) {
                super();
                /**
                 * The type of content to synthesize.
                 */
                this.contentType = __runInitializers(this, _instanceExtraInitializers);
                this.contentType = request.stream ? 'stream' : 'text';
                this.content = request.content;
                this.stream = request.stream;
                this._stream = stream ? stream_1.TextStream.fromExternal(stream) : void 0;
                this.voice = request.voice;
            }
            /**
             * Creates a TTSRequest with text content.
             *
             * @param text - The text content to synthesize
             * @param voice - Optional voice configuration
             * @returns A new TTSRequest instance
             */
            static withText(text, voice) {
                return new _a({
                    stream: false,
                    content: { value: text },
                    voice,
                });
            }
            /**
             * Creates a TTSRequest with stream content.
             *
             * @param stream - The stream content to synthesize
             * @param voice - Optional voice configuration
             * @param synthesisConfig
             * @returns A new TTSRequest instance
             */
            static withStream(stream, voice, synthesisConfig) {
                return new _a({
                    stream: true,
                    voice,
                    synthesisConfig,
                }, stream.getStream());
            }
            /**
             * Checks if the request has text content.
             *
             * @returns True if the request contains text content
             */
            hasTextContent() {
                var _b;
                return this.contentType === 'text' && !!((_b = this.content) === null || _b === void 0 ? void 0 : _b.value);
            }
            /**
             * Checks if the request has stream content.
             *
             * @returns True if the request contains stream content
             */
            hasStreamContent() {
                return this.contentType === 'stream' && !!this.stream;
            }
            /**
             * Gets the text content if available.
             *
             * @returns The text content or undefined
             */
            getText() {
                var _b;
                return (_b = this.content) === null || _b === void 0 ? void 0 : _b.value;
            }
            /**
             * Gets the voice configuration if available.
             *
             * @returns The voice configuration or undefined
             */
            getVoice() {
                return this.voice;
            }
            /** @internal */
            toTaggedValue() {
                if (this.contentType === 'text') {
                    throw new Error('Not implement');
                }
                return {
                    type: 'TTSRequest',
                    _stream: this._stream.getStream(),
                };
            }
            /** @internal */
            toProto() {
                if (this.contentType === 'text') {
                    throw new Error('Text content type not implemented for protobuf');
                }
                return pb_1.PbType.ttsRequest({
                    content: this.content,
                    stream: this.stream,
                    synthesisConfig: this.synthesisConfig,
                    voice: this.voice,
                }, this._stream.getStream());
            }
        },
        (() => {
            var _b;
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create((_b = _classSuper[Symbol.metadata]) !== null && _b !== void 0 ? _b : null) : void 0;
            _toTaggedValue_decorators = [internal_1.default];
            _toProto_decorators = [internal_1.default];
            __esDecorate(_a, null, _toTaggedValue_decorators, { kind: "method", name: "toTaggedValue", static: false, private: false, access: { has: obj => "toTaggedValue" in obj, get: obj => obj.toTaggedValue }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(_a, null, _toProto_decorators, { kind: "method", name: "toProto", static: false, private: false, access: { has: obj => "toProto" in obj, get: obj => obj.toProto }, metadata: _metadata }, null, _instanceExtraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        _a;
})();
exports.TTSRequest = TTSRequest;
