// @inworld/runtime/build/common/api/graphs.js
"use strict";
/**
 *
 * Note: This file is autogenerated and not intended to be modified manually.
 *
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.BeliefStateSchema = exports.AgentInfoSchema = exports.PurpleErrorSchema = exports.DataStreamWithMetadataSchema = exports.CustomDataSchema = exports.UsageSchema = exports.ClassificationResultSchema = exports.AudioSchema = exports.AnySchema = exports.TextChunkingNodeExecutionConfigSchema = exports.RandomCannedTextNodeExecutionConfigSchema = exports.IntentsOutputNodeExecutionConfigSchema = exports.MCPListToolsNodeExecutionConfigSchema = exports.MCPCallToolNodeExecutionConfigSchema = exports.IntentsLlmPromptVariablesBuilderNodeExecutionConfigSchema = exports.SortCriteriaSchema = exports.RealtimeAgentInputSlicerNodeExecutionConfigSchema = exports.GoogleSerperApiSearchNodeExecutionConfigSchema = exports.RetryConfigSchema = exports.FallbackConfigSchema = exports.CooldownConfigSchema = exports.IntentsEmbeddingMatcherNodeExecutionConfigSchema = exports.EmbeddingMatcherNodeExecutionConfigSchema = exports.CustomNodeExecutionConfigSchema = exports.ParameterTypeSchema = exports.EvaluationModeSchema = exports.WorkerCountStrategySchema = exports.TelemetryLogLevelSchema = exports.TelemetryExporterTypeSchema = exports.FinishReasonSchema = exports.SortMetricSchema = exports.SortDirectionSchema = exports.StatusCodeSchema = exports.isTimestampType = exports.isResponseFormat = exports.isRemoteTtsConfig = exports.isRemoteSttConfig = exports.isRemoteStreamingSttConfig = exports.isRemoteLlmConfig = exports.isNerCreationConfig = exports.isLocalVadConfig = exports.isLocalSttConfig = exports.isLocalStreamingSttConfig = exports.isLocalLlmConfig = exports.isLocalAecFilterConfig = exports.isImageDetail = exports.isGoogleSerperApiClientCreationConfig = exports.isElevenLabsTextNormalization = exports.isElevenLabsOutputFormat = exports.isDeviceType = void 0;
exports.ToolCallResponseSchema = exports.MultimodalContentSchema = exports.MatchedPhrasesSchema = exports.MatchedKeywordsSchema = exports.LLMCompletionRequestSchema = exports.LLMChatMetadataSchema = exports.QueryEventSchema = exports.GoalForEvaluationSchema = exports.InteractionEventSchema = exports.ContentDataSchema = exports.STTNodeExecutionConfigSchema = exports.LLMCompletionNodeExecutionConfigSchema = exports.LLMSelectionConfigSchema = exports.KnowledgeNodeExecutionConfigSchema = exports.NodeErrorConfigSchema = exports.GraphParameterSchema = exports.SubgraphNodeSchema = exports.TransformNodeCreationConfigSchema = exports.TextClassifierNodeCreationConfigSchema = exports.IntentsStrictMatcherNodeCreationConfigSchema = exports.IntentsLlmPromptVariablesBuilderNodeCreationConfigSchema = exports.LLMChatRoutingNodeCreationConfigSchema = exports.KnowledgeNodeCreationConfigSchema = exports.IntentNodeCreationConfigSchema = exports.EmbeddingMatcherNodeCreationConfigSchema = exports.GraphEdgeSchema = exports.CustomComponentCreationConfigSchema = exports.FeatureFlagConfigSchema = exports.ThreadPoolConfigSchema = exports.AddonConfigSchema = exports.UserContextSchema = exports.TimestampSchema = exports.ToolCallRequestSchema = exports.STTRequestSchema = exports.SafetyResultSchema = exports.TextSchema = exports.EventSchema = exports.PhraseMatchSchema = exports.MatchedIntentsSchema = exports.LLMCompletionResponseSchema = exports.LLMChatAttemptSchema = exports.ListToolsResponseSchema = exports.JsonSchema = exports.GoalAdvancementResponseSchema = exports.TriggerEventSchema = exports.ResponseEventSchema = exports.IntentMatchSchema = exports.ConditionSchema = exports.AgentSpeechSchema = exports.AgentActionSchema = void 0;
exports.CreateGraphRequestSchema = exports.GraphSpecSchema = exports.MainGraphSchema = exports.SubgraphSchema = exports.GraphNodeSchema = exports.StartExecutionRequestSchema = exports.ProcessContextSchema = exports.RegularNodeSchema = exports.ExecutionContextSchema = exports.NodeExecutionConfigSchema = exports.DataContainerSchema = exports.LLMChatRoutingNodeExecutionConfigSchema = exports.LLMChatRoutingRequestSchema = exports.LLMChatRoutingStrategySchema = exports.ComponentSchema = exports.LLMChatWeightedLoadBalancingStrategySchema = exports.ComponentCreationConfigSchema = exports.TTSNodeExecutionConfigSchema = exports.LLMChatWeightedRouteSchema = exports.LLMChatNodeExecutionConfigSchema = exports.IntentNodeExecutionConfigSchema = exports.NodeCreationConfigSchema = exports.TTSRequestSchema = exports.LLMChatResponseSchema = exports.LLMChatRequestSchema = exports.GoalAdvancementRequestSchema = exports.TextClassifierNodeExecutionConfigSchema = exports.LLMChatRouteSchema = exports.LLMChatDefaultsSchema = exports.GoalAdvancementNodeExecutionConfigSchema = exports.KeywordMatcherNodeCreationConfigSchema = exports.GoalAdvancementNodeCreationConfigSchema = exports.IntentsEmbeddingMatcherNodeCreationConfigSchema = exports.GraphCreationConfigSchema = exports.CreateEnvRequestSchema = exports.TTSOutputSchema = void 0;
exports.isStatusCode = isStatusCode;
exports.isSortDirection = isSortDirection;
exports.isSortMetric = isSortMetric;
exports.isFinishReason = isFinishReason;
exports.isTelemetryExporterType = isTelemetryExporterType;
exports.isTelemetryLogLevel = isTelemetryLogLevel;
exports.isWorkerCountStrategy = isWorkerCountStrategy;
exports.isEvaluationMode = isEvaluationMode;
exports.isParameterType = isParameterType;
exports.isCustomNodeExecutionConfig = isCustomNodeExecutionConfig;
exports.isEmbeddingMatcherNodeExecutionConfig = isEmbeddingMatcherNodeExecutionConfig;
exports.isIntentsEmbeddingMatcherNodeExecutionConfig = isIntentsEmbeddingMatcherNodeExecutionConfig;
exports.isCooldownConfig = isCooldownConfig;
exports.isFallbackConfig = isFallbackConfig;
exports.isRetryConfig = isRetryConfig;
exports.isLogitBias = isLogitBias;
exports.isGoogleSerperApiSearchNodeExecutionConfig = isGoogleSerperApiSearchNodeExecutionConfig;
exports.isRealtimeAgentInputSlicerNodeExecutionConfig = isRealtimeAgentInputSlicerNodeExecutionConfig;
exports.isEmbeddingMatcherConfig = isEmbeddingMatcherConfig;
exports.isRetrievalConfig = isRetrievalConfig;
exports.isImageUrl = isImageUrl;
exports.isToolCall = isToolCall;
exports.isFunctionChoice = isFunctionChoice;
exports.isTool = isTool;
exports.isModelId = isModelId;
exports.isSortCriteria = isSortCriteria;
exports.isIntentsLlmPromptVariablesBuilderNodeExecutionConfig = isIntentsLlmPromptVariablesBuilderNodeExecutionConfig;
exports.isMCPCallToolNodeExecutionConfig = isMCPCallToolNodeExecutionConfig;
exports.isMCPListToolsNodeExecutionConfig = isMCPListToolsNodeExecutionConfig;
exports.isIntentsOutputNodeExecutionConfig = isIntentsOutputNodeExecutionConfig;
exports.isRandomCannedTextNodeExecutionConfig = isRandomCannedTextNodeExecutionConfig;
exports.isSpeechRecognitionConfig = isSpeechRecognitionConfig;
exports.isTextChunkingNodeExecutionConfig = isTextChunkingNodeExecutionConfig;
exports.isClassThreshold = isClassThreshold;
exports.isElevenLabsQueryParams = isElevenLabsQueryParams;
exports.isElevenLabsRequestParams = isElevenLabsRequestParams;
exports.isSpeechSynthesisInferenceConfig = isSpeechSynthesisInferenceConfig;
exports.isSpeechSynthesisPostprocessingConfig = isSpeechSynthesisPostprocessingConfig;
exports.isVoice = isVoice;
exports.isAny = isAny;
exports.isAudio = isAudio;
exports.isClassificationResult = isClassificationResult;
exports.isUsage = isUsage;
exports.isCustomData = isCustomData;
exports.isDataStreamWithMetadata = isDataStreamWithMetadata;
exports.isPurpleError = isPurpleError;
exports.isAgentInfo = isAgentInfo;
exports.isBeliefState = isBeliefState;
exports.isAgentAction = isAgentAction;
exports.isAgentSpeech = isAgentSpeech;
exports.isCondition = isCondition;
exports.isIntentMatch = isIntentMatch;
exports.isResponseEvent = isResponseEvent;
exports.isTriggerEvent = isTriggerEvent;
exports.isGoalAdvancementResponse = isGoalAdvancementResponse;
exports.isGoogleSerperApiSearchRequest = isGoogleSerperApiSearchRequest;
exports.isKnowledgeGraph = isKnowledgeGraph;
exports.isSitelink = isSitelink;
exports.isPeopleAlsoAsk = isPeopleAlsoAsk;
exports.isTopStory = isTopStory;
exports.isJson = isJson;
exports.isKnowledgeRecords = isKnowledgeRecords;
exports.isListToolsResponse = isListToolsResponse;
exports.isLLMChatAttempt = isLLMChatAttempt;
exports.isLLMCompletionResponse = isLLMCompletionResponse;
exports.isMatchedIntents = isMatchedIntents;
exports.isKeywordMatch = isKeywordMatch;
exports.isPhraseMatch = isPhraseMatch;
exports.isEvent = isEvent;
exports.isText = isText;
exports.isSafetyResult = isSafetyResult;
exports.isSTTRequest = isSTTRequest;
exports.isToolCallRequest = isToolCallRequest;
exports.isToolCallResult = isToolCallResult;
exports.isTimestamp = isTimestamp;
exports.isUserContext = isUserContext;
exports.isCredentials = isCredentials;
exports.isAddonConfig = isAddonConfig;
exports.isLoggerConfig = isLoggerConfig;
exports.isTelemetryConfig = isTelemetryConfig;
exports.isThreadPoolConfig = isThreadPoolConfig;
exports.isFeatureFlagConfig = isFeatureFlagConfig;
exports.isCustomComponentCreationConfig = isCustomComponentCreationConfig;
exports.isAcousticEchoCancellationConfig = isAcousticEchoCancellationConfig;
exports.isDeviceSelection = isDeviceSelection;
exports.isParsingConfig = isParsingConfig;
exports.isStreamSpeechRecognitionConfig = isStreamSpeechRecognitionConfig;
exports.isLocalTextEmbedderConfig = isLocalTextEmbedderConfig;
exports.isVoiceActivityDetectionConfig = isVoiceActivityDetectionConfig;
exports.isMCPHttpConfig = isMCPHttpConfig;
exports.isParagonApiConfig = isParagonApiConfig;
exports.isMCPStdioConfig = isMCPStdioConfig;
exports.isDictionaryRule = isDictionaryRule;
exports.isLanguage = isLanguage;
exports.isRemoteTextEmbedderConfig = isRemoteTextEmbedderConfig;
exports.isTextClassifierCreationConfig = isTextClassifierCreationConfig;
exports.isTurnDetectionConfig = isTurnDetectionConfig;
exports.isGraphEdge = isGraphEdge;
exports.isIntent = isIntent;
exports.isEmbeddingMatcherNodeCreationConfig = isEmbeddingMatcherNodeCreationConfig;
exports.isIntentNodeCreationConfig = isIntentNodeCreationConfig;
exports.isKeywordGroup = isKeywordGroup;
exports.isKnowledgeNodeCreationConfig = isKnowledgeNodeCreationConfig;
exports.isLLMChatRoutingNodeCreationConfig = isLLMChatRoutingNodeCreationConfig;
exports.isIntentsLlmPromptVariablesBuilderNodeCreationConfig = isIntentsLlmPromptVariablesBuilderNodeCreationConfig;
exports.isIntentsStrictMatcherNodeCreationConfig = isIntentsStrictMatcherNodeCreationConfig;
exports.isTextClassifierNodeCreationConfig = isTextClassifierNodeCreationConfig;
exports.isTransformNodeCreationConfig = isTransformNodeCreationConfig;
exports.isSubgraphNode = isSubgraphNode;
exports.isGraphParameter = isGraphParameter;
exports.isNodeErrorConfig = isNodeErrorConfig;
exports.isTextGenerationConfig = isTextGenerationConfig;
exports.isLLMMatcherConfig = isLLMMatcherConfig;
exports.isKnowledgeNodeExecutionConfig = isKnowledgeNodeExecutionConfig;
exports.isContentItem = isContentItem;
exports.isToolChoice = isToolChoice;
exports.isLLMSelectionConfig = isLLMSelectionConfig;
exports.isLLMCompletionNodeExecutionConfig = isLLMCompletionNodeExecutionConfig;
exports.isSTTNodeExecutionConfig = isSTTNodeExecutionConfig;
exports.isTextClassifierConfig = isTextClassifierConfig;
exports.isElevenLabsSpeechSynthesisConfig = isElevenLabsSpeechSynthesisConfig;
exports.isInworldSpeechSynthesisConfig = isInworldSpeechSynthesisConfig;
exports.isContentData = isContentData;
exports.isInteractionEvent = isInteractionEvent;
exports.isGoalForEvaluation = isGoalForEvaluation;
exports.isQueryEvent = isQueryEvent;
exports.isOrganicResult = isOrganicResult;
exports.isLLMChatMetadata = isLLMChatMetadata;
exports.isLLMCompletionRequest = isLLMCompletionRequest;
exports.isMatchedKeywords = isMatchedKeywords;
exports.isMatchedPhrases = isMatchedPhrases;
exports.isMultimodalContent = isMultimodalContent;
exports.isToolCallResponse = isToolCallResponse;
exports.isTTSOutput = isTTSOutput;
exports.isCreateEnvRequest = isCreateEnvRequest;
exports.isGraphCreationConfig = isGraphCreationConfig;
exports.isKnowledgeCompileConfig = isKnowledgeCompileConfig;
exports.isMCPSessionCreationConfig = isMCPSessionCreationConfig;
exports.isEntity = isEntity;
exports.isRemoteKnowledgeConfig = isRemoteKnowledgeConfig;
exports.isLocalTurnDetectorConfig = isLocalTurnDetectorConfig;
exports.isIntentsEmbeddingMatcherNodeCreationConfig = isIntentsEmbeddingMatcherNodeCreationConfig;
exports.isGoalAdvancementNodeCreationConfig = isGoalAdvancementNodeCreationConfig;
exports.isKeywordMatcherNodeCreationConfig = isKeywordMatcherNodeCreationConfig;
exports.isGoalAdvancementNodeExecutionConfig = isGoalAdvancementNodeExecutionConfig;
exports.isIntentMatcherConfig = isIntentMatcherConfig;
exports.isMessage = isMessage;
exports.isLLMChatDefaults = isLLMChatDefaults;
exports.isLLMChatRoute = isLLMChatRoute;
exports.isTextClassifierNodeExecutionConfig = isTextClassifierNodeExecutionConfig;
exports.isSpeechSynthesisConfig = isSpeechSynthesisConfig;
exports.isGoalAdvancementRequest = isGoalAdvancementRequest;
exports.isGoogleSerperApiSearchResponse = isGoogleSerperApiSearchResponse;
exports.isLLMChatRequest = isLLMChatRequest;
exports.isLLMChatResponse = isLLMChatResponse;
exports.isTTSRequest = isTTSRequest;
exports.isLocalKnowledgeConfig = isLocalKnowledgeConfig;
exports.isMCPClientCreationConfig = isMCPClientCreationConfig;
exports.isTurnDetectorCreationConfig = isTurnDetectorCreationConfig;
exports.isNodeCreationConfig = isNodeCreationConfig;
exports.isIntentNodeExecutionConfig = isIntentNodeExecutionConfig;
exports.isLLMChatNodeExecutionConfig = isLLMChatNodeExecutionConfig;
exports.isLLMChatWeightedRoute = isLLMChatWeightedRoute;
exports.isTTSNodeExecutionConfig = isTTSNodeExecutionConfig;
exports.isComponentCreationConfig = isComponentCreationConfig;
exports.isLLMChatWeightedLoadBalancingStrategy = isLLMChatWeightedLoadBalancingStrategy;
exports.isComponent = isComponent;
exports.isLLMChatRoutingStrategy = isLLMChatRoutingStrategy;
exports.isLLMChatRoutingRequest = isLLMChatRoutingRequest;
exports.isLLMChatRoutingNodeExecutionConfig = isLLMChatRoutingNodeExecutionConfig;
exports.isDataContainer = isDataContainer;
exports.isNodeExecutionConfig = isNodeExecutionConfig;
exports.isExecutionContext = isExecutionContext;
exports.isRegularNode = isRegularNode;
exports.isProcessContext = isProcessContext;
exports.isStartExecutionRequest = isStartExecutionRequest;
exports.isGraphNode = isGraphNode;
exports.isSubgraph = isSubgraph;
exports.isMainGraph = isMainGraph;
exports.isGraphSpec = isGraphSpec;
exports.isCreateGraphRequest = isCreateGraphRequest;
/* eslint-disable @typescript-eslint/no-unused-vars */
const z = __importStar(require("zod"));
const library_1 = require("./library");
var library_2 = require("./library");
Object.defineProperty(exports, "isDeviceType", { enumerable: true, get: function () { return library_2.isDeviceType; } });
Object.defineProperty(exports, "isElevenLabsOutputFormat", { enumerable: true, get: function () { return library_2.isElevenLabsOutputFormat; } });
Object.defineProperty(exports, "isElevenLabsTextNormalization", { enumerable: true, get: function () { return library_2.isElevenLabsTextNormalization; } });
Object.defineProperty(exports, "isGoogleSerperApiClientCreationConfig", { enumerable: true, get: function () { return library_2.isGoogleSerperApiClientCreationConfig; } });
Object.defineProperty(exports, "isImageDetail", { enumerable: true, get: function () { return library_2.isImageDetail; } });
Object.defineProperty(exports, "isLocalAecFilterConfig", { enumerable: true, get: function () { return library_2.isLocalAecFilterConfig; } });
Object.defineProperty(exports, "isLocalLlmConfig", { enumerable: true, get: function () { return library_2.isLocalLlmConfig; } });
Object.defineProperty(exports, "isLocalStreamingSttConfig", { enumerable: true, get: function () { return library_2.isLocalStreamingSttConfig; } });
Object.defineProperty(exports, "isLocalSttConfig", { enumerable: true, get: function () { return library_2.isLocalSttConfig; } });
Object.defineProperty(exports, "isLocalVadConfig", { enumerable: true, get: function () { return library_2.isLocalVadConfig; } });
Object.defineProperty(exports, "isNerCreationConfig", { enumerable: true, get: function () { return library_2.isNerCreationConfig; } });
Object.defineProperty(exports, "isRemoteLlmConfig", { enumerable: true, get: function () { return library_2.isRemoteLlmConfig; } });
Object.defineProperty(exports, "isRemoteStreamingSttConfig", { enumerable: true, get: function () { return library_2.isRemoteStreamingSttConfig; } });
Object.defineProperty(exports, "isRemoteSttConfig", { enumerable: true, get: function () { return library_2.isRemoteSttConfig; } });
Object.defineProperty(exports, "isRemoteTtsConfig", { enumerable: true, get: function () { return library_2.isRemoteTtsConfig; } });
Object.defineProperty(exports, "isResponseFormat", { enumerable: true, get: function () { return library_2.isResponseFormat; } });
Object.defineProperty(exports, "isTimestampType", { enumerable: true, get: function () { return library_2.isTimestampType; } });
exports.StatusCodeSchema = z.enum([
    'STATUS_CODE_ABORTED',
    'STATUS_CODE_ALREADY_EXISTS',
    'STATUS_CODE_CANCELLED',
    'STATUS_CODE_DATA_LOSS',
    'STATUS_CODE_DEADLINE_EXCEEDED',
    'STATUS_CODE_FAILED_PRECONDITION',
    'STATUS_CODE_INTERNAL',
    'STATUS_CODE_INVALID_ARGUMENT',
    'STATUS_CODE_NOT_FOUND',
    'STATUS_CODE_OK',
    'STATUS_CODE_OUT_OF_RANGE',
    'STATUS_CODE_PERMISSION_DENIED',
    'STATUS_CODE_RESOURCE_EXHAUSTED',
    'STATUS_CODE_UNAUTHENTICATED',
    'STATUS_CODE_UNAVAILABLE',
    'STATUS_CODE_UNIMPLEMENTED',
    'STATUS_CODE_UNKNOWN',
]);
/**
 * Sort Direction
 *
 * The direction for ordering. If not specified, uses intelligent defaults
 * (e.g., Price/Latency=Ascending, Intelligence/Throughput=Descending).
 *
 * Allowed values: "SORT_DIRECTION_UNSPECIFIED", "SORT_DIRECTION_ASCENDING",
 * "SORT_DIRECTION_DESCENDING"
 *
 * Type: 'SORT_DIRECTION_UNSPECIFIED' | 'SORT_DIRECTION_ASCENDING' |
 * 'SORT_DIRECTION_DESCENDING'.
 */
exports.SortDirectionSchema = z.enum([
    'SORT_DIRECTION_ASCENDING',
    'SORT_DIRECTION_DESCENDING',
    'SORT_DIRECTION_UNSPECIFIED',
]);
/**
 * Sort Metric
 *
 * The metric to sort by.
 *
 * Allowed values: "SORT_METRIC_UNSPECIFIED", "SORT_METRIC_PRICE",
 * "SORT_METRIC_LATENCY", "SORT_METRIC_THROUGHPUT", "SORT_METRIC_INTELLIGENCE",
 * "SORT_METRIC_MATH", "SORT_METRIC_CODING"
 *
 * Type: 'SORT_METRIC_UNSPECIFIED' | 'SORT_METRIC_PRICE' | 'SORT_METRIC_LATENCY'
 * | 'SORT_METRIC_THROUGHPUT' | 'SORT_METRIC_INTELLIGENCE' | 'SORT_METRIC_MATH'
 * | 'SORT_METRIC_CODING'.
 */
exports.SortMetricSchema = z.enum([
    'SORT_METRIC_CODING',
    'SORT_METRIC_INTELLIGENCE',
    'SORT_METRIC_LATENCY',
    'SORT_METRIC_MATH',
    'SORT_METRIC_PRICE',
    'SORT_METRIC_THROUGHPUT',
    'SORT_METRIC_UNSPECIFIED',
]);
/**
 * Finish Reason
 *
 * The reason the model stopped generating tokens
 *
 * Allowed values: "unspecified", "stop", "length", "content_filter",
 * "tool_call"
 *
 * Type: 'unspecified' | 'stop' | 'length' | 'content_filter' | 'tool_call'.
 */
exports.FinishReasonSchema = z.enum([
    'content_filter',
    'length',
    'stop',
    'tool_call',
    'unspecified',
]);
/**
 * Telemetry Exporter Type
 *
 * Where to send telemetry: console output or Inworld's monitoring service
 * Default: LOCAL
 *
 * Allowed values: "TELEMETRY_EXPORTER_TYPE_UNSPECIFIED",
 * "TELEMETRY_EXPORTER_TYPE_LOCAL", "TELEMETRY_EXPORTER_TYPE_REMOTE"
 *
 * Type: 'TELEMETRY_EXPORTER_TYPE_UNSPECIFIED' | 'TELEMETRY_EXPORTER_TYPE_LOCAL'
 * | 'TELEMETRY_EXPORTER_TYPE_REMOTE'.
 */
exports.TelemetryExporterTypeSchema = z.enum([
    'TELEMETRY_EXPORTER_TYPE_LOCAL',
    'TELEMETRY_EXPORTER_TYPE_REMOTE',
    'TELEMETRY_EXPORTER_TYPE_UNSPECIFIED',
]);
/**
 * Telemetry Log Level
 *
 * Minimum logging level for telemetry messages. Messages below this level are
 * filtered out. Default: INFO
 *
 * Allowed values: "TELEMETRY_LOG_LEVEL_UNSPECIFIED",
 * "TELEMETRY_LOG_LEVEL_TRACE", "TELEMETRY_LOG_LEVEL_DEBUG",
 * "TELEMETRY_LOG_LEVEL_INFO", "TELEMETRY_LOG_LEVEL_WARN",
 * "TELEMETRY_LOG_LEVEL_ERROR"
 *
 * Type: 'TELEMETRY_LOG_LEVEL_UNSPECIFIED' | 'TELEMETRY_LOG_LEVEL_TRACE' |
 * 'TELEMETRY_LOG_LEVEL_DEBUG' | 'TELEMETRY_LOG_LEVEL_INFO' |
 * 'TELEMETRY_LOG_LEVEL_WARN' | 'TELEMETRY_LOG_LEVEL_ERROR'.
 */
exports.TelemetryLogLevelSchema = z.enum([
    'TELEMETRY_LOG_LEVEL_DEBUG',
    'TELEMETRY_LOG_LEVEL_ERROR',
    'TELEMETRY_LOG_LEVEL_INFO',
    'TELEMETRY_LOG_LEVEL_TRACE',
    'TELEMETRY_LOG_LEVEL_UNSPECIFIED',
    'TELEMETRY_LOG_LEVEL_WARN',
]);
/**
 * Worker Count Strategy
 *
 * Strategy for determining worker count
 *
 * Allowed values: "WORKER_COUNT_STRATEGY_UNSPECIFIED",
 * "WORKER_COUNT_STRATEGY_FIXED", "WORKER_COUNT_STRATEGY_SYSTEM_CORES"
 *
 * Type: 'WORKER_COUNT_STRATEGY_UNSPECIFIED' | 'WORKER_COUNT_STRATEGY_FIXED' |
 * 'WORKER_COUNT_STRATEGY_SYSTEM_CORES'.
 */
exports.WorkerCountStrategySchema = z.enum([
    'WORKER_COUNT_STRATEGY_FIXED',
    'WORKER_COUNT_STRATEGY_SYSTEM_CORES',
    'WORKER_COUNT_STRATEGY_UNSPECIFIED',
]);
/**
 * Evaluation Mode
 *
 * Evaluation mode for handling user contexts and graph instances
 *
 * Allowed values: "EVALUATION_MODE_UNSPECIFIED",
 * "EVALUATION_MODE_IN_PROCESS_ONLY", "EVALUATION_MODE_REMOTE_ONLY"
 *
 * Type: 'EVALUATION_MODE_UNSPECIFIED' | 'EVALUATION_MODE_IN_PROCESS_ONLY' |
 * 'EVALUATION_MODE_REMOTE_ONLY'.
 */
exports.EvaluationModeSchema = z.enum([
    'EVALUATION_MODE_IN_PROCESS_ONLY',
    'EVALUATION_MODE_REMOTE_ONLY',
    'EVALUATION_MODE_UNSPECIFIED',
]);
/**
 * Parameter Type
 *
 * Type of the parameter
 *
 * Allowed values: "PARAMETER_TYPE_UNSPECIFIED", "PARAMETER_TYPE_STRING",
 * "PARAMETER_TYPE_NUMBER", "PARAMETER_TYPE_INTEGER", "PARAMETER_TYPE_BOOLEAN",
 * "PARAMETER_TYPE_JSON", "PARAMETER_TYPE_ARRAY", "PARAMETER_TYPE_OBJECT"
 *
 * Type: 'PARAMETER_TYPE_UNSPECIFIED' | 'PARAMETER_TYPE_STRING' |
 * 'PARAMETER_TYPE_NUMBER' | 'PARAMETER_TYPE_INTEGER' | 'PARAMETER_TYPE_BOOLEAN'
 * | 'PARAMETER_TYPE_JSON' | 'PARAMETER_TYPE_ARRAY' | 'PARAMETER_TYPE_OBJECT'.
 */
exports.ParameterTypeSchema = z.enum([
    'PARAMETER_TYPE_ARRAY',
    'PARAMETER_TYPE_BOOLEAN',
    'PARAMETER_TYPE_INTEGER',
    'PARAMETER_TYPE_JSON',
    'PARAMETER_TYPE_NUMBER',
    'PARAMETER_TYPE_OBJECT',
    'PARAMETER_TYPE_STRING',
    'PARAMETER_TYPE_UNSPECIFIED',
]);
/**
 * Custom Node Execution Config
 *
 * CustomNodeExecutionConfig provides additional properties for custom node
 * execution.
 */
exports.CustomNodeExecutionConfigSchema = z.object({
    properties: z.record(z.string(), z.any()).optional(),
});
/**
 * Embedding Matcher Node Execution Config
 *
 * EmbeddingMatcherNodeExecutionConfig runtime configuration for embedding
 * matcher nodes.
 */
exports.EmbeddingMatcherNodeExecutionConfigSchema = z.object({
    similarityThreshold: z.number().optional(),
    topN: z.number().optional(),
});
/**
 * Intents Embedding Matcher Node Execution Config
 *
 * IntentsEmbeddingMatcherNodeExecutionConfig runtime configuration for the
 * embedding-based intent matcher node.
 */
exports.IntentsEmbeddingMatcherNodeExecutionConfigSchema = z.object({
    similarityThreshold: z.number().optional(),
});
/**
 * Cooldown Config
 *
 * CooldownConfig defines cooldown behavior after consecutive failures. When a
 * node fails min_consecutive_failures times in a row, it enters a cooldown
 * period for the specified duration.
 */
exports.CooldownConfigSchema = z.object({
    cooldownDuration: z.string().optional(),
    minConsecutiveFailures: z.number().optional(),
});
/**
 * Fallback Config
 *
 * FallbackConfig defines fallback behavior for node execution. When a node
 * fails with a status code in handle_errors, execution can fall back to an
 * alternative node. Status codes in ignore_errors will not trigger fallback.
 */
exports.FallbackConfigSchema = z.object({
    cooldown: exports.CooldownConfigSchema.optional(),
    handleErrors: z.array(exports.StatusCodeSchema).optional(),
    ignoreErrors: z.array(exports.StatusCodeSchema).optional(),
    nodeId: z.string().optional(),
});
/**
 * Retry Config
 *
 * RetryConfig defines retry behavior for node execution. When a node fails with
 * a status code in handle_errors, it will be retried up to max_attempts times.
 * Status codes in ignore_errors will not trigger retries.
 */
exports.RetryConfigSchema = z.object({
    handleErrors: z.array(exports.StatusCodeSchema).optional(),
    ignoreErrors: z.array(exports.StatusCodeSchema).optional(),
    maxAttempts: z.number().optional(),
});
/**
 * Google Serper Api Search Node Execution Config
 *
 * GoogleSerperApiSearchNodeExecutionConfig runtime configuration for the Google
 * Serper API search node.
 */
exports.GoogleSerperApiSearchNodeExecutionConfigSchema = z.object({
    googleSerperApiComponentId: z.string(),
});
/**
 * Realtime Agent Input Slicer Node Execution Config
 *
 * RealtimeAgentInputSlicerNodeExecutionConfig runtime configuration for the
 * InputSlicerNode. This node processes continuous input streams (audio or text)
 * and detects interaction boundaries.
 */
exports.RealtimeAgentInputSlicerNodeExecutionConfigSchema = z.object({
    maxSilenceToFinishTurnMs: z.number().optional(),
    maxSpeechDurationPerTurnMs: z.number().optional(),
    sampleRate: z.number().optional(),
    turnDetectorComponentId: z.string().optional(),
    vadComponentId: z.string(),
});
/**
 * Sort Criteria
 *
 * SortCriteria specifies a single ordering criterion with metric and direction.
 */
exports.SortCriteriaSchema = z.object({
    direction: exports.SortDirectionSchema.optional(),
    metric: exports.SortMetricSchema,
});
exports.IntentsLlmPromptVariablesBuilderNodeExecutionConfigSchema = z.object({
    maxIntents: z.number().optional(),
    maxPhrasesPerIntent: z.number().optional(),
});
/**
 * MCP Call Tool Node Execution Config
 *
 * MCPCallToolNodeExecutionConfig runtime configuration for MCP call tool nodes.
 */
exports.MCPCallToolNodeExecutionConfigSchema = z.object({
    mcpComponentId: z.string(),
});
/**
 * MCP List Tools Node Execution Config
 *
 * MCPListToolsNodeExecutionConfig runtime configuration for MCP list tools
 * nodes.
 */
exports.MCPListToolsNodeExecutionConfigSchema = z.object({
    mcpComponentId: z.string(),
});
/**
 * Intents Output Node Execution Config
 *
 * IntentsOutputNodeExecutionConfig runtime configuration for the output node
 * that filters matched intents.
 */
exports.IntentsOutputNodeExecutionConfigSchema = z.object({
    topNIntents: z.number().optional(),
});
/**
 * Random Canned Text Node Execution Config
 *
 * RandomCannedTextNodeExecutionConfig runtime configuration for random canned
 * text nodes.
 */
exports.RandomCannedTextNodeExecutionConfigSchema = z.object({
    cannedPhrases: z.array(z.string()),
});
/**
 * Text Chunking Node Execution Config
 *
 * TextChunkingNodeExecutionConfig runtime configuration for text chunking
 * nodes.
 */
exports.TextChunkingNodeExecutionConfigSchema = z.object({
    minChunkLength: z.number().optional(),
    splitByConjunctions: z.array(z.string()).optional(),
});
/**
 * Any
 */
exports.AnySchema = z.object({
    '@type': z.string().optional(),
});
/**
 * Audio
 *
 * Audio data type containing audio samples and metadata
 */
exports.AudioSchema = z.object({
    data: z.string().optional(),
    sampleRate: z.number(),
});
/**
 * Classification Result
 *
 * Result of text classification operations containing information about the
 * classes that were detected
 */
exports.ClassificationResultSchema = z.object({
    classes: z.array(z.string()).optional(),
});
/**
 * Usage
 *
 * Token usage information
 */
exports.UsageSchema = z.object({
    completionTokens: z.number().optional(),
    promptTokens: z.number().optional(),
});
/**
 * Custom Data
 *
 * Custom data with JSON data and optional JSON schema
 */
exports.CustomDataSchema = z.object({
    schema: z.record(z.string(), z.any()).optional(),
    value: z.record(z.string(), z.any()),
});
/**
 * Data Stream With Metadata
 *
 * Data stream container with associated metadata
 */
exports.DataStreamWithMetadataSchema = z.object({
    metadata: z.record(z.string(), z.any()).optional(),
});
exports.PurpleErrorSchema = z.object({
    code: z.number(),
    message: z.string(),
});
/**
 * Agent Info
 *
 * AgentInfo contains information about an agent or player.
 */
exports.AgentInfoSchema = z.object({
    description: z.string().optional(),
    name: z.string(),
});
/**
 * Belief State
 *
 * BeliefState represents the agent's understanding of the world state.
 */
exports.BeliefStateSchema = z.object({
    values: z.record(z.string(), z.any()).optional(),
});
/**
 * Agent Action
 *
 * Agent action event.
 */
exports.AgentActionSchema = z.object({
    agentName: z.string(),
    name: z.string(),
    parameters: z.record(z.string(), z.string()).optional(),
});
/**
 * Agent Speech
 *
 * Agent speech event.
 */
exports.AgentSpeechSchema = z.object({
    agentName: z.string(),
    utterance: z.string(),
});
/**
 * Condition
 *
 * Condition for goal activation or completion.
 */
exports.ConditionSchema = z.object({
    detect: z.string().optional(),
    intents: z.array(z.string()).optional(),
    requiredGoals: z.array(z.string()).optional(),
});
/**
 * Intent Match
 *
 * Intent match information
 */
exports.IntentMatchSchema = z.object({
    intentName: z.string(),
    score: z.number(),
});
/**
 * Response Event
 *
 * ResponseEvent represents an agent response event (post-generation).
 */
exports.ResponseEventSchema = z.object({
    agentResponse: z.string(),
    playerQuery: z.string(),
});
/**
 * Trigger Event
 *
 * TriggerEvent represents an external trigger event.
 */
exports.TriggerEventSchema = z.object({
    triggerName: z.string(),
});
/**
 * Goal Advancement Response
 *
 * Result of goal advancement evaluation.
 */
exports.GoalAdvancementResponseSchema = z.object({
    activatedGoals: z.array(z.string()).optional(),
    beliefState: exports.BeliefStateSchema.optional(),
    completedGoals: z.array(z.string()).optional(),
    currentGoals: z.array(z.string()).optional(),
});
/**
 * Json
 *
 * JSON data container
 */
exports.JsonSchema = z.object({
    value: z.string(),
});
/**
 * List Tools Response
 *
 * Response containing list of available tools
 */
exports.ListToolsResponseSchema = z.object({
    tools: z.array(library_1.ToolSchema).optional(),
});
/**
 * LLM Chat Attempt
 *
 * Information about a single model attempt during LLM routing.
 */
exports.LLMChatAttemptSchema = z.object({
    errorMessage: z.string().optional(),
    modelId: library_1.ModelIdSchema,
    success: z.boolean().optional(),
    timeToFirstTokenMs: z.number().optional(),
    warnings: z.array(z.string()).optional(),
});
/**
 * LLM Completion Response
 *
 * Response from LLM text completion processing
 */
exports.LLMCompletionResponseSchema = z.object({
    content: z.string().optional(),
    finishReason: exports.FinishReasonSchema.optional(),
    stream: z.boolean().optional(),
});
/**
 * Matched Intents
 *
 * Result of intent matching operations including a collection of matched
 * intents with their confidence scores
 */
exports.MatchedIntentsSchema = z.object({
    intentMatches: z.array(exports.IntentMatchSchema).optional(),
});
/**
 * Phrase Match
 *
 * Information about a matched phrase with similarity score
 */
exports.PhraseMatchSchema = z.object({
    phrase: z.string(),
    score: z.number(),
});
/**
 * Event
 *
 * Arbitrary event with optional payload.
 */
exports.EventSchema = z.object({
    name: z.string(),
    payload: z.record(z.string(), z.any()).optional(),
});
/**
 * Text
 *
 * Text data type for simple string content
 */
exports.TextSchema = z.object({
    value: z.string(),
});
/**
 * Safety Result
 *
 * Safety check result indicating whether certain content is considered safe for
 * processing
 */
exports.SafetyResultSchema = z.object({
    isSafe: z.boolean().optional(),
    text: z.string(),
});
/**
 * STT Request
 *
 * Request for speech-to-text transcription
 */
exports.STTRequestSchema = z.object({
    content: exports.AudioSchema.optional(),
    speechRecognitionConfig: library_1.SpeechRecognitionConfigSchema.optional(),
    stream: z.boolean().optional(),
});
/**
 * Tool Call Request
 *
 * Collection of tool call data
 */
exports.ToolCallRequestSchema = z.object({
    toolCalls: z.array(library_1.ToolCallSchema).optional(),
});
/**
 * Timestamp
 *
 * Timestamp information for a token/word in synthesized speech
 */
exports.TimestampSchema = z.object({
    endTime: z.number().optional(),
    startTime: z.number().optional(),
    token: z.string().optional(),
});
/**
 * User Context
 *
 * UserContext contains user-specific metadata for graph variant selection and
 * targeting.
 */
exports.UserContextSchema = z.object({
    attributes: z.record(z.string(), z.string()).optional(),
    targetingKey: z.string().optional(),
});
/**
 * Addon Config
 *
 * AddonConfig provides configuration for the Node.js addon-specific settings.
 * This is separate from the C++ Environment thread pool configuration.
 */
exports.AddonConfigSchema = z.object({
    addonPoolSize: z.number().optional(),
});
/**
 * ThreadPoolConfig defines how many worker threads should be allocated for
 * graph execution. By default, Environment uses a shared thread pool that is
 * reused across all graphs for better resource efficiency.
 *
 * Shared vs Dedicated Thread Pools: - Pass ThreadPoolConfig to
 * Environment::Create() for a shared pool (default, recommended) - Individual
 * graphs can override with dedicated_thread_pool_config in GraphCreationConfig
 */
exports.ThreadPoolConfigSchema = z.object({
    coreMultiplier: z.number().optional(),
    workerCount: z.number().optional(),
    workerStrategy: exports.WorkerCountStrategySchema.optional(),
});
/**
 * Feature Flag Config
 *
 * FeatureFlagConfig controls how feature flags and A/B testing are handled for
 * graph selection. This enables dynamic graph configuration based on user
 * attributes without code changes.
 */
exports.FeatureFlagConfigSchema = z.object({
    enabled: z.boolean().optional(),
    evaluationMode: exports.EvaluationModeSchema.optional(),
    flagdAddress: z.string().optional(),
    pollInterval: z.string().optional(),
});
/**
 * Custom Component Creation Config
 *
 * CustomComponentCreationConfig provides additional properties for custom
 * component creation.
 */
exports.CustomComponentCreationConfigSchema = z.object({
    properties: z.record(z.string(), z.any()).optional(),
});
/**
 * Graph Edge
 *
 * GraphEdge represents an edge connecting nodes in the graph
 */
exports.GraphEdgeSchema = z.object({
    conditionExpression: z.string().optional(),
    conditionId: z.string().optional(),
    fromNode: z.string(),
    loopBool: z.boolean().optional(),
    loopTemplate: z.string().optional(),
    optionalBool: z.boolean().optional(),
    optionalTemplate: z.string().optional(),
    toNode: z.string().optional(),
    toNodeExpression: z.string().optional(),
    toNodes: z.array(z.string()).optional(),
});
/**
 * Embedding Matcher Node Creation Config
 *
 * EmbeddingMatcherNodeCreationConfig configuration for creating embedding
 * matcher nodes.
 */
exports.EmbeddingMatcherNodeCreationConfigSchema = z.object({
    embedderComponentId: z.string(),
    phrases: z.array(z.string()),
});
/**
 * Intent Node Creation Config
 *
 * IntentNodeCreationConfig configuration for creating intent nodes.
 */
exports.IntentNodeCreationConfigSchema = z.object({
    embedderComponentId: z.string().optional(),
    intents: z.array(library_1.IntentSchema),
    llmComponentId: z.string().optional(),
});
/**
 * Knowledge Node Creation Config
 *
 * KnowledgeNodeCreationConfig configuration for creating knowledge nodes.
 */
exports.KnowledgeNodeCreationConfigSchema = z.object({
    knowledgeComponentId: z.string(),
    knowledgeId: z.string(),
    knowledgeRecords: z.array(z.string()),
});
/**
 * LLM Chat Routing Node Creation Config
 *
 * LLMChatRoutingNodeCreationConfig configuration for creating LLM chat routing
 * nodes.
 */
exports.LLMChatRoutingNodeCreationConfigSchema = z.object({
    apiKey: z.string(),
    defaultTimeout: z.string().optional(),
});
exports.IntentsLlmPromptVariablesBuilderNodeCreationConfigSchema = z.object({
    intents: z.array(library_1.IntentSchema),
});
/**
 * Intents Strict Matcher Node Creation Config
 *
 * IntentsStrictMatcherNodeCreationConfig configuration for creating a strict
 * intent matcher node. This node performs exact matching against normalized
 * intent phrases.
 */
exports.IntentsStrictMatcherNodeCreationConfigSchema = z.object({
    intents: z.array(library_1.IntentSchema),
});
/**
 * Text Classifier Node Creation Config
 *
 * TextClassifierNodeCreationConfig configuration for creating text classifier
 * nodes.
 */
exports.TextClassifierNodeCreationConfigSchema = z.object({
    classifierConfig: library_1.TextClassifierCreationConfigSchema,
    embedderComponentId: z.string(),
});
/**
 * Transform Node Creation Config
 *
 * TransformNodeCreationConfig configuration for creating transform nodes.
 */
exports.TransformNodeCreationConfigSchema = z.object({
    outputTemplate: z.record(z.string(), z.any()).optional(),
    outputType: z.string(),
});
/**
 * Subgraph Node
 *
 * SubgraphNode represents a node that references a subgraph
 */
exports.SubgraphNodeSchema = z.object({
    id: z.string(),
    parameters: z.record(z.string(), z.any()).optional(),
    subgraphId: z.string(),
});
/**
 * Graph Parameter
 *
 * GraphParameter defines a parameter that can be passed to a graph
 */
exports.GraphParameterSchema = z.object({
    name: z.string(),
    type: exports.ParameterTypeSchema,
});
/**
 * Node Error Config
 *
 * NodeErrorConfig defines error handling configuration for a node. This
 * includes retry policies, fallback nodes, and cooldown behavior.
 */
exports.NodeErrorConfigSchema = z.object({
    cooldown: exports.CooldownConfigSchema.optional(),
    fallbacks: z.array(exports.FallbackConfigSchema).optional(),
    retries: z.array(exports.RetryConfigSchema).optional(),
});
/**
 * Knowledge Node Execution Config
 *
 * KnowledgeNodeExecutionConfig runtime configuration for knowledge nodes.
 */
exports.KnowledgeNodeExecutionConfigSchema = z.object({
    knowledgeIds: z.array(z.string()),
    retrievalConfig: library_1.RetrievalConfigSchema.optional(),
});
/**
 * LLM Selection Config
 *
 * LLM routing configuration for model selection and fallback
 */
exports.LLMSelectionConfigSchema = z.object({
    ignore: z.array(library_1.ModelIdSchema).optional(),
    models: z.array(library_1.ModelIdSchema).optional(),
    sort: z.array(exports.SortCriteriaSchema).optional(),
});
/**
 * LLM Completion Node Execution Config
 *
 * LLMCompletionNodeExecutionConfig runtime configuration for LLM completion
 * nodes.
 */
exports.LLMCompletionNodeExecutionConfigSchema = z.object({
    llmComponentId: z.string(),
    promptTemplate: z.string().optional(),
    stream: z.boolean().optional(),
    textGenerationConfig: library_1.TextGenerationConfigSchema.optional(),
});
/**
 * STT Node Execution Config
 *
 * STTNodeExecutionConfig runtime configuration for STT nodes.
 */
exports.STTNodeExecutionConfigSchema = z.object({
    speechRecognitionConfig: library_1.SpeechRecognitionConfigSchema.optional(),
    sttComponentId: z.string(),
});
/**
 * Content Data
 *
 * Content data containing text and optional tool calls
 */
exports.ContentDataSchema = z.object({
    content: z.string(),
    finishReason: exports.FinishReasonSchema.optional(),
    modelName: z.string().optional(),
    toolCalls: z.array(library_1.ToolCallSchema).optional(),
    usage: exports.UsageSchema.optional(),
});
/**
 * Interaction Event
 *
 * Event in the interaction history.
 */
exports.InteractionEventSchema = z.object({
    agentAction: exports.AgentActionSchema.optional(),
    agentSpeech: exports.AgentSpeechSchema.optional(),
});
/**
 * Goal For Evaluation
 *
 * GoalForEvaluation represents a goal for evaluation purposes.
 */
exports.GoalForEvaluationSchema = z.object({
    activationCondition: exports.ConditionSchema.optional(),
    completionCondition: exports.ConditionSchema.optional(),
    motivation: z.string().optional(),
    name: z.string(),
    repeatable: z.boolean().optional(),
});
/**
 * Query Event
 *
 * QueryEvent represents a player query event (pre-generation).
 */
exports.QueryEventSchema = z.object({
    intents: z.array(exports.IntentMatchSchema).optional(),
    playerQuery: z.string(),
});
/**
 * LLM Chat Metadata
 *
 * Metadata about the routing decision and model selection process.
 */
exports.LLMChatMetadataSchema = z.object({
    attempts: z.array(exports.LLMChatAttemptSchema).optional(),
    generationId: z.string().optional(),
    reasoning: z.string().optional(),
    totalDurationMs: z.number().optional(),
});
/**
 * LLM Completion Request
 *
 * Text completion request for LLM processing
 */
exports.LLMCompletionRequestSchema = z.object({
    text: z.string(),
    textGenerationConfig: library_1.TextGenerationConfigSchema.optional(),
});
/**
 * Matched Keywords
 *
 * Result of keyword matching operations containing information about matched
 * keywords and their associated groups
 */
exports.MatchedKeywordsSchema = z.object({
    keywordMatches: z.array(library_1.KeywordMatchSchema).optional(),
});
/**
 * Matched Phrases
 *
 * Result of phrase matching using embeddings including a collection of matched
 * phrases with their similarity scores
 */
exports.MatchedPhrasesSchema = z.object({
    phraseMatches: z.array(exports.PhraseMatchSchema).optional(),
});
/**
 * Multimodal Content
 *
 * Multi-modal content container supporting text and audio data types
 */
exports.MultimodalContentSchema = z.object({
    audio: exports.AudioSchema.optional(),
    event: exports.EventSchema.optional(),
    text: z
        .string()
        .transform((val) => {
        if (isText(val)) {
            return val.value;
        }
        return exports.TextSchema.parse(val);
    })
        .optional(),
});
/**
 * Tool Call Response
 *
 * Collection of tool call results
 */
exports.ToolCallResponseSchema = z.object({
    toolCallResults: z.array(library_1.ToolCallResultSchema).optional(),
});
/**
 * TTS Output
 *
 * Text-to-speech output containing both original text and synthesized audio
 */
exports.TTSOutputSchema = z.object({
    audio: exports.AudioSchema,
    text: z.string(),
    timestamps: z.array(exports.TimestampSchema).optional(),
});
/**
 * CreateEnvRequest represents the request to create a JsEnv instance. This
 * message contains all configuration options for the Environment.
 *
 * Used by: createEnv(config?) or new JsEnv(config?)
 */
exports.CreateEnvRequestSchema = z.object({
    addonConfig: exports.AddonConfigSchema.optional(),
    initializeFromEnvVars: z.boolean().optional(),
    loggerConfig: library_1.LoggerConfigSchema.optional(),
    substitutions: z.record(z.string(), z.string()).optional(),
    telemetry: library_1.TelemetryConfigSchema.optional(),
    threadPool: exports.ThreadPoolConfigSchema.optional(),
});
/**
 * GraphCreationConfig provides options for controlling how graphs are created,
 * including feature flag behavior and thread pool settings.
 *
 * Usage Patterns: - Local mode: For local development and testing - Feature
 * flags disabled - Simple, predictable graph creation - Remote mode: For
 * production environments (recommended default) - Feature flags enabled with
 * Inworld's production flagd service - Dynamic graph variant selection based on
 * user attributes - A/B testing and gradual rollouts supported - InProcess
 * mode: For high-throughput production - Feature flags enabled with in-process
 * evaluation - Lower latency than remote evaluation
 */
exports.GraphCreationConfigSchema = z.object({
    dedicatedThreadPoolConfig: exports.ThreadPoolConfigSchema.optional(),
    featureFlagConfig: exports.FeatureFlagConfigSchema.optional(),
});
/**
 * Intents Embedding Matcher Node Creation Config
 *
 * IntentsEmbeddingMatcherNodeCreationConfig configuration for creating an
 * embedding-based intent matcher node. This node performs similarity matching
 * using text embeddings.
 */
exports.IntentsEmbeddingMatcherNodeCreationConfigSchema = z.object({
    embedderComponentId: z.string(),
    intents: z.array(library_1.IntentSchema),
});
/**
 * Goal Advancement Node Creation Config
 *
 * GoalAdvancementNodeCreationConfig configuration for creating goal advancement
 * nodes.
 */
exports.GoalAdvancementNodeCreationConfigSchema = z.object({
    goals: z.array(exports.GoalForEvaluationSchema),
});
/**
 * Keyword Matcher Node Creation Config
 *
 * KeywordMatcherNodeCreationConfig configuration for creating keyword matcher
 * nodes.
 */
exports.KeywordMatcherNodeCreationConfigSchema = z.object({
    keywordGroups: z.array(library_1.KeywordGroupSchema),
});
/**
 * Goal Advancement Node Execution Config
 *
 * GoalAdvancementNodeExecutionConfig runtime configuration for goal advancement
 * nodes.
 */
exports.GoalAdvancementNodeExecutionConfigSchema = z.object({
    llmComponentId: z.string(),
    textGenerationConfig: library_1.TextGenerationConfigSchema.optional(),
});
/**
 * LLM Chat Defaults
 *
 * Default configuration values for LLM chat routing. Used when a route doesn't
 * specify its own values.
 */
exports.LLMChatDefaultsSchema = z.object({
    messageTemplates: z.array(library_1.MessageSchema).optional(),
    textGenerationConfig: library_1.TextGenerationConfigSchema.optional(),
});
/**
 * LLM Chat Route
 *
 * Configuration for a single model route endpoint. Defines which model to use,
 * how to handle fallback/auto-selection, text generation parameters, and
 * message templates.
 */
exports.LLMChatRouteSchema = z.object({
    messageTemplates: z.array(library_1.MessageSchema).optional(),
    modelId: library_1.ModelIdSchema.optional(),
    modelSelection: exports.LLMSelectionConfigSchema.optional(),
    textGenerationConfig: library_1.TextGenerationConfigSchema.optional(),
});
/**
 * Text Classifier Node Execution Config
 *
 * TextClassifierNodeExecutionConfig runtime configuration for text classifier
 * nodes.
 */
exports.TextClassifierNodeExecutionConfigSchema = z.object({
    classifierConfig: library_1.TextClassifierConfigSchema.optional(),
});
/**
 * Goal Advancement Request
 *
 * Goal advancement evaluation request.
 */
exports.GoalAdvancementRequestSchema = z.object({
    agentInfo: exports.AgentInfoSchema,
    beliefState: exports.BeliefStateSchema.optional(),
    eventHistory: z.array(exports.InteractionEventSchema).optional(),
    goalsForEvaluationOverride: z.array(exports.GoalForEvaluationSchema).optional(),
    playerInfo: exports.AgentInfoSchema,
    queryEvent: exports.QueryEventSchema.optional(),
    responseEvent: exports.ResponseEventSchema.optional(),
    triggerEvent: exports.TriggerEventSchema.optional(),
});
/**
 * LLM Chat Request
 *
 * Complete chat request for LLM processing
 */
exports.LLMChatRequestSchema = z.object({
    messages: z.array(library_1.MessageSchema),
    responseFormat: library_1.ResponseFormatSchema.default('RESPONSE_FORMAT_TEXT').optional(),
    textGenerationConfig: library_1.TextGenerationConfigSchema.optional(),
    toolChoice: library_1.ToolChoiceSchema.optional(),
    tools: z.array(library_1.ToolSchema).optional(),
});
/**
 * LLM Chat Response
 *
 * LLM response with content and tool calls
 */
exports.LLMChatResponseSchema = z.object({
    content: z.string().optional(),
    finishReason: exports.FinishReasonSchema.optional(),
    metadata: exports.LLMChatMetadataSchema.optional(),
    modelName: z.string().optional(),
    stream: z.boolean().optional(),
    toolCalls: z.array(library_1.ToolCallSchema).optional(),
    usage: exports.UsageSchema.optional(),
});
/**
 * TTS Request
 *
 * TTS request with content and optional voice and synthesis configurations
 */
exports.TTSRequestSchema = z.object({
    content: exports.TextSchema.optional(),
    stream: z.boolean().optional(),
    synthesisConfig: library_1.SpeechSynthesisConfigSchema.optional(),
    voice: library_1.VoiceSchema.optional(),
});
/**
 * Node Creation Config
 *
 * NodeCreationConfig wraps node creation configurations
 */
exports.NodeCreationConfigSchema = z.object({
    embeddingMatcherNodeCreationConfig: exports.IntentsEmbeddingMatcherNodeCreationConfigSchema.optional(),
    embeddingMatcherNodeCreationConfigEmb: exports.EmbeddingMatcherNodeCreationConfigSchema.optional(),
    goalAdvancementNodeCreationConfig: exports.GoalAdvancementNodeCreationConfigSchema.optional(),
    intentNodeCreationConfig: exports.IntentNodeCreationConfigSchema.optional(),
    keywordMatcherNodeCreationConfig: exports.KeywordMatcherNodeCreationConfigSchema.optional(),
    knowledgeNodeCreationConfig: exports.KnowledgeNodeCreationConfigSchema.optional(),
    llmChatRoutingNodeCreationConfig: exports.LLMChatRoutingNodeCreationConfigSchema.optional(),
    llmPromptVariablesBuilderNodeCreationConfig: exports.IntentsLlmPromptVariablesBuilderNodeCreationConfigSchema.optional(),
    strictMatcherNodeCreationConfig: exports.IntentsStrictMatcherNodeCreationConfigSchema.optional(),
    textClassifierNodeCreationConfig: exports.TextClassifierNodeCreationConfigSchema.optional(),
    transformNodeCreationConfig: exports.TransformNodeCreationConfigSchema.optional(),
});
/**
 * Intent Node Execution Config
 *
 * IntentNodeExecutionConfig runtime configuration for intent nodes.
 */
exports.IntentNodeExecutionConfigSchema = z.object({
    matcherConfig: library_1.IntentMatcherConfigSchema,
});
/**
 * LLM Chat Node Execution Config
 *
 * LLMChatNodeExecutionConfig runtime configuration for LLM chat nodes.
 */
exports.LLMChatNodeExecutionConfigSchema = z.object({
    llmComponentId: z.string(),
    messageTemplates: z.array(library_1.MessageSchema).optional(),
    responseFormat: library_1.ResponseFormatSchema.optional(),
    stream: z.boolean().optional(),
    textGenerationConfig: library_1.TextGenerationConfigSchema.optional(),
    toolChoice: library_1.ToolChoiceSchema.optional(),
    tools: z.array(library_1.ToolSchema).optional(),
});
/**
 * LLM Chat Weighted Route
 *
 * A model route with an associated weight for load balancing. Extends
 * LLMChatRoute with a weight for traffic distribution.
 */
exports.LLMChatWeightedRouteSchema = z.object({
    route: exports.LLMChatRouteSchema.optional(),
    weight: z.number(),
});
/**
 * TTS Node Execution Config
 *
 * TTSNodeExecutionConfig runtime configuration for TTS nodes.
 */
exports.TTSNodeExecutionConfigSchema = z.object({
    synthesisConfig: library_1.SpeechSynthesisConfigSchema.optional(),
    ttsComponentId: z.string(),
    voice: library_1.VoiceSchema,
});
/**
 * Component Creation Config
 *
 * ComponentCreationConfig wraps component creation configurations Supports
 * strongly-typed configs from various domains (embeddings, llm, speech,
 * knowledge, mcp, nlu, serper, etc.)
 */
exports.ComponentCreationConfigSchema = z.object({
    customComponentCreationConfig: exports.CustomComponentCreationConfigSchema.optional(),
    googleSerperApiClientCreationConfig: library_1.GoogleSerperApiClientCreationConfigSchema.optional(),
    localAecFilterConfig: library_1.LocalAecFilterConfigSchema.optional(),
    localKnowledgeConfig: library_1.LocalKnowledgeConfigSchema.optional(),
    localLlmConfig: library_1.LocalLlmConfigSchema.optional(),
    localStreamingSttConfig: library_1.LocalStreamingSttConfigSchema.optional(),
    localSttConfig: library_1.LocalSttConfigSchema.optional(),
    localTextEmbedderConfig: library_1.LocalTextEmbedderConfigSchema.optional(),
    localVadConfig: library_1.LocalVadConfigSchema.optional(),
    mcpClientCreationConfig: library_1.MCPClientCreationConfigSchema.optional(),
    nerCreationConfig: library_1.NerCreationConfigSchema.optional(),
    remoteKnowledgeConfig: library_1.RemoteKnowledgeConfigSchema.optional(),
    remoteLlmConfig: library_1.RemoteLlmConfigSchema.optional(),
    remoteStreamingSttConfig: library_1.RemoteStreamingSttConfigSchema.optional(),
    remoteSttConfig: library_1.RemoteSttConfigSchema.optional(),
    remoteTextEmbedderConfig: library_1.RemoteTextEmbedderConfigSchema.optional(),
    remoteTtsConfig: library_1.RemoteTtsConfigSchema.optional(),
    textClassifierCreationConfig: library_1.TextClassifierCreationConfigSchema.optional(),
    turnDetectorCreationConfig: library_1.TurnDetectorCreationConfigSchema.optional(),
});
/**
 * LLM Chat Weighted Load Balancing Strategy
 *
 * Routing strategy that distributes requests across multiple routes based on
 * configured weights.
 */
exports.LLMChatWeightedLoadBalancingStrategySchema = z.object({
    routes: z.array(exports.LLMChatWeightedRouteSchema).optional(),
});
/**
 * Component
 *
 * Component represents a component that can be used by graph nodes
 */
exports.ComponentSchema = z.object({
    creationConfig: exports.ComponentCreationConfigSchema.optional(),
    id: z.string(),
    type: z.string(),
});
/**
 * LLM Chat Routing Strategy
 *
 * Strategy for routing requests to model endpoints.
 */
exports.LLMChatRoutingStrategySchema = z.object({
    weightedLoadBalancing: exports.LLMChatWeightedLoadBalancingStrategySchema.optional(),
});
/**
 * LLM Chat Routing Request
 *
 * Complete chat request for LLM routing with intelligent model selection
 */
exports.LLMChatRoutingRequestSchema = z.object({
    messages: z.array(library_1.MessageSchema),
    modelId: library_1.ModelIdSchema.optional(),
    modelSelection: exports.LLMSelectionConfigSchema.optional(),
    promptVariables: z.record(z.string(), z.any()).optional(),
    responseFormat: library_1.ResponseFormatSchema.optional(),
    strategy: exports.LLMChatRoutingStrategySchema.optional(),
    stream: z.boolean().optional(),
    textGenerationConfig: library_1.TextGenerationConfigSchema.optional(),
    toolChoice: library_1.ToolChoiceSchema.optional(),
    tools: z.array(library_1.ToolSchema).optional(),
});
/**
 * LLM Chat Routing Node Execution Config
 *
 * Runtime configuration for the LLMChatRoutingNode. Provides static (default)
 * configuration for the node. Request-level configuration always takes
 * precedence.
 */
exports.LLMChatRoutingNodeExecutionConfigSchema = z.object({
    defaults: exports.LLMChatDefaultsSchema.optional(),
    strategy: exports.LLMChatRoutingStrategySchema.optional(),
});
/**
 * Data Container
 *
 * Universal data container containing all runtime data types
 */
exports.DataContainerSchema = z.object({
    any: exports.AnySchema.optional(),
    audio: exports.AudioSchema.optional(),
    classificationResult: exports.ClassificationResultSchema.optional(),
    contentData: exports.ContentDataSchema.optional(),
    customData: exports.CustomDataSchema.optional(),
    dataStreamWithMetadata: exports.DataStreamWithMetadataSchema.optional(),
    error: exports.PurpleErrorSchema.optional(),
    goalAdvancementRequest: exports.GoalAdvancementRequestSchema.optional(),
    goalAdvancementResponse: exports.GoalAdvancementResponseSchema.optional(),
    googleSerperApiSearchRequest: library_1.GoogleSerperApiSearchRequestSchema.optional(),
    googleSerperApiSearchResponse: library_1.GoogleSerperApiSearchResponseSchema.optional(),
    json: exports.JsonSchema.optional(),
    knowledgeRecords: library_1.KnowledgeRecordsSchema.optional(),
    listToolsResponse: exports.ListToolsResponseSchema.optional(),
    llmChatRequest: exports.LLMChatRequestSchema.optional(),
    llmChatResponse: exports.LLMChatResponseSchema.optional(),
    llmChatRoutingRequest: exports.LLMChatRoutingRequestSchema.optional(),
    llmCompletionRequest: exports.LLMCompletionRequestSchema.optional(),
    llmCompletionResponse: exports.LLMCompletionResponseSchema.optional(),
    matchedIntents: exports.MatchedIntentsSchema.optional(),
    matchedKeywords: exports.MatchedKeywordsSchema.optional(),
    matchedPhrases: exports.MatchedPhrasesSchema.optional(),
    multiModalContent: exports.MultimodalContentSchema.optional(),
    safetyResult: exports.SafetyResultSchema.optional(),
    sttRequest: exports.STTRequestSchema.optional(),
    text: exports.TextSchema.optional(),
    toolCallRequest: exports.ToolCallRequestSchema.optional(),
    toolCallResponse: exports.ToolCallResponseSchema.optional(),
    toolCallResult: library_1.ToolCallResultSchema.optional(),
    ttsOutput: exports.TTSOutputSchema.optional(),
    ttsRequest: exports.TTSRequestSchema.optional(),
});
/**
 * Node Execution Config
 *
 * NodeExecutionConfig wraps node execution configurations. This message
 * combines base fields shared by all node types with subtype-specific
 * configuration via a oneof field.
 */
exports.NodeExecutionConfigSchema = z.object({
    customNodeExecutionConfig: exports.CustomNodeExecutionConfigSchema.optional(),
    embeddingMatcherNodeExecutionConfig: exports.EmbeddingMatcherNodeExecutionConfigSchema.optional(),
    embeddingMatcherNodeExecutionConfigIntent: exports.IntentsEmbeddingMatcherNodeExecutionConfigSchema.optional(),
    errorHandling: exports.NodeErrorConfigSchema.optional(),
    goalAdvancementNodeExecutionConfig: exports.GoalAdvancementNodeExecutionConfigSchema.optional(),
    googleSerperApiSearchNodeExecutionConfig: exports.GoogleSerperApiSearchNodeExecutionConfigSchema.optional(),
    inputSlicerNodeExecutionConfig: exports.RealtimeAgentInputSlicerNodeExecutionConfigSchema.optional(),
    intentNodeExecutionConfig: exports.IntentNodeExecutionConfigSchema.optional(),
    knowledgeNodeExecutionConfig: exports.KnowledgeNodeExecutionConfigSchema.optional(),
    llmChatNodeExecutionConfig: exports.LLMChatNodeExecutionConfigSchema.optional(),
    llmChatRoutingNodeExecutionConfig: exports.LLMChatRoutingNodeExecutionConfigSchema.optional(),
    llmCompletionNodeExecutionConfig: exports.LLMCompletionNodeExecutionConfigSchema.optional(),
    llmPromptVariablesBuilderNodeExecutionConfig: exports.IntentsLlmPromptVariablesBuilderNodeExecutionConfigSchema.optional(),
    mcpCallToolNodeExecutionConfig: exports.MCPCallToolNodeExecutionConfigSchema.optional(),
    mcpListToolsNodeExecutionConfig: exports.MCPListToolsNodeExecutionConfigSchema.optional(),
    outputNodeExecutionConfig: exports.IntentsOutputNodeExecutionConfigSchema.optional(),
    randomCannedTextNodeExecutionConfig: exports.RandomCannedTextNodeExecutionConfigSchema.optional(),
    reportToClient: z.boolean().optional(),
    sttNodeExecutionConfig: exports.STTNodeExecutionConfigSchema.optional(),
    textChunkingNodeExecutionConfig: exports.TextChunkingNodeExecutionConfigSchema.optional(),
    textClassifierNodeExecutionConfig: exports.TextClassifierNodeExecutionConfigSchema.optional(),
    ttsNodeExecutionConfig: exports.TTSNodeExecutionConfigSchema.optional(),
});
/**
 * Execution Context
 *
 * ExecutionContext contains optional execution metadata including execution ID,
 * user context, and data store prefill that can be passed to the Execute
 * method.
 */
exports.ExecutionContextSchema = z.object({
    dataToStore: z.record(z.string(), exports.DataContainerSchema).optional(),
    executionId: z.string().optional(),
    userContext: exports.UserContextSchema.optional(),
    userCredentials: library_1.CredentialsSchema.optional(),
});
/**
 * Regular Node
 *
 * RegularNode represents a regular node in the graph
 */
exports.RegularNodeSchema = z.object({
    creationConfig: exports.NodeCreationConfigSchema.optional(),
    executionConfig: exports.NodeExecutionConfigSchema.optional(),
    id: z.string(),
    type: z.string(),
});
/**
 * ProcessContext contains the data fields from ProcessContext that are exposed
 * to JavaScript. Wrapper objects (datastore, cancellationContext, span,
 * componentRegistry) are not included as they are JavaScript interfaces to C++
 * objects, not serializable data.
 *
 * Note: isCancelled is accessed via cancellationContext.isCancelled() method,
 * not as serialized data, so it's not included here.
 */
exports.ProcessContextSchema = z.object({
    executionConfig: exports.NodeExecutionConfigSchema.optional(),
    executionId: z.string().optional(),
    hasTelemetryContext: z.boolean().optional(),
    variant: z.string().optional(),
});
/**
 * StartExecutionRequest represents the request to start a graph execution.
 * Input data is passed separately; this message contains execution context
 * options.
 *
 * Used by: JsExecutor.start(data, opts?)
 */
exports.StartExecutionRequestSchema = z.object({
    executionContext: exports.ExecutionContextSchema.optional(),
});
/**
 * Graph Node
 *
 * GraphNode represents a node in the graph (oneOf RegularNode | SubgraphNode)
 */
exports.GraphNodeSchema = z.object({
    regularNode: exports.RegularNodeSchema.optional(),
    subgraphNode: exports.SubgraphNodeSchema.optional(),
});
/**
 * Subgraph
 *
 * Subgraph represents a reusable graph that can be referenced by SubgraphNode
 */
exports.SubgraphSchema = z.object({
    edges: z.array(exports.GraphEdgeSchema).optional(),
    endNodes: z.array(z.string()),
    id: z.string(),
    nodes: z.array(exports.GraphNodeSchema),
    parameters: z.array(exports.GraphParameterSchema).optional(),
    startNodes: z.array(z.string()),
});
/**
 * Main Graph
 *
 * MainGraph represents the main graph that serves as the entry point
 */
exports.MainGraphSchema = z.object({
    edges: z.array(exports.GraphEdgeSchema).optional(),
    endNodes: z.array(z.string()).optional(),
    id: z.string(),
    nodes: z.array(exports.GraphNodeSchema).optional(),
    startNodes: z.array(z.string()),
});
/**
 * Graph Spec
 *
 * GraphSpec represents the complete graph configuration that replaces the JSON
 * schema. This is the top-level message for graph specifications.
 */
exports.GraphSpecSchema = z.object({
    components: z.array(exports.ComponentSchema).optional(),
    main: exports.MainGraphSchema,
    schemaVersion: z.string(),
    subgraphs: z.array(exports.SubgraphSchema).optional(),
});
/**
 * CreateGraphRequest represents the request to create a graph executor. This
 * message combines the graph specification with creation options.
 *
 * Used by: JsEnv.createGraph(spec, opts?)
 */
exports.CreateGraphRequestSchema = z.object({
    config: exports.GraphCreationConfigSchema.optional(),
    spec: exports.GraphSpecSchema.optional(),
    specJson: z.string().optional(),
    substitutions: z.record(z.string(), z.string()).optional(),
});
function isStatusCode(obj) {
    return exports.StatusCodeSchema.safeParse(obj).success;
}
function isSortDirection(obj) {
    return exports.SortDirectionSchema.safeParse(obj).success;
}
function isSortMetric(obj) {
    return exports.SortMetricSchema.safeParse(obj).success;
}
function isFinishReason(obj) {
    return exports.FinishReasonSchema.safeParse(obj).success;
}
function isTelemetryExporterType(obj) {
    return exports.TelemetryExporterTypeSchema.safeParse(obj).success;
}
function isTelemetryLogLevel(obj) {
    return exports.TelemetryLogLevelSchema.safeParse(obj).success;
}
function isWorkerCountStrategy(obj) {
    return exports.WorkerCountStrategySchema.safeParse(obj).success;
}
function isEvaluationMode(obj) {
    return exports.EvaluationModeSchema.safeParse(obj).success;
}
function isParameterType(obj) {
    return exports.ParameterTypeSchema.safeParse(obj).success;
}
function isCustomNodeExecutionConfig(obj) {
    return exports.CustomNodeExecutionConfigSchema.safeParse(obj).success;
}
function isEmbeddingMatcherNodeExecutionConfig(obj) {
    return exports.EmbeddingMatcherNodeExecutionConfigSchema.safeParse(obj).success;
}
function isIntentsEmbeddingMatcherNodeExecutionConfig(obj) {
    return exports.IntentsEmbeddingMatcherNodeExecutionConfigSchema.safeParse(obj)
        .success;
}
function isCooldownConfig(obj) {
    return exports.CooldownConfigSchema.safeParse(obj).success;
}
function isFallbackConfig(obj) {
    return exports.FallbackConfigSchema.safeParse(obj).success;
}
function isRetryConfig(obj) {
    return exports.RetryConfigSchema.safeParse(obj).success;
}
function isLogitBias(obj) {
    return library_1.LogitBiasSchema.safeParse(obj).success;
}
function isGoogleSerperApiSearchNodeExecutionConfig(obj) {
    return exports.GoogleSerperApiSearchNodeExecutionConfigSchema.safeParse(obj).success;
}
function isRealtimeAgentInputSlicerNodeExecutionConfig(obj) {
    return exports.RealtimeAgentInputSlicerNodeExecutionConfigSchema.safeParse(obj)
        .success;
}
function isEmbeddingMatcherConfig(obj) {
    return library_1.EmbeddingMatcherConfigSchema.safeParse(obj).success;
}
function isRetrievalConfig(obj) {
    return library_1.RetrievalConfigSchema.safeParse(obj).success;
}
function isImageUrl(obj) {
    return library_1.ImageUrlSchema.safeParse(obj).success;
}
function isToolCall(obj) {
    return library_1.ToolCallSchema.safeParse(obj).success;
}
function isFunctionChoice(obj) {
    return library_1.FunctionChoiceSchema.safeParse(obj).success;
}
function isTool(obj) {
    return library_1.ToolSchema.safeParse(obj).success;
}
function isModelId(obj) {
    return library_1.ModelIdSchema.safeParse(obj).success;
}
function isSortCriteria(obj) {
    return exports.SortCriteriaSchema.safeParse(obj).success;
}
function isIntentsLlmPromptVariablesBuilderNodeExecutionConfig(obj) {
    return exports.IntentsLlmPromptVariablesBuilderNodeExecutionConfigSchema.safeParse(obj).success;
}
function isMCPCallToolNodeExecutionConfig(obj) {
    return exports.MCPCallToolNodeExecutionConfigSchema.safeParse(obj).success;
}
function isMCPListToolsNodeExecutionConfig(obj) {
    return exports.MCPListToolsNodeExecutionConfigSchema.safeParse(obj).success;
}
function isIntentsOutputNodeExecutionConfig(obj) {
    return exports.IntentsOutputNodeExecutionConfigSchema.safeParse(obj).success;
}
function isRandomCannedTextNodeExecutionConfig(obj) {
    return exports.RandomCannedTextNodeExecutionConfigSchema.safeParse(obj).success;
}
function isSpeechRecognitionConfig(obj) {
    return library_1.SpeechRecognitionConfigSchema.safeParse(obj).success;
}
function isTextChunkingNodeExecutionConfig(obj) {
    return exports.TextChunkingNodeExecutionConfigSchema.safeParse(obj).success;
}
function isClassThreshold(obj) {
    return library_1.ClassThresholdSchema.safeParse(obj).success;
}
function isElevenLabsQueryParams(obj) {
    return library_1.ElevenLabsQueryParamsSchema.safeParse(obj).success;
}
function isElevenLabsRequestParams(obj) {
    return library_1.ElevenLabsRequestParamsSchema.safeParse(obj).success;
}
function isSpeechSynthesisInferenceConfig(obj) {
    return library_1.SpeechSynthesisInferenceConfigSchema.safeParse(obj).success;
}
function isSpeechSynthesisPostprocessingConfig(obj) {
    return library_1.SpeechSynthesisPostprocessingConfigSchema.safeParse(obj).success;
}
function isVoice(obj) {
    return library_1.VoiceSchema.safeParse(obj).success;
}
function isAny(obj) {
    return exports.AnySchema.safeParse(obj).success;
}
function isAudio(obj) {
    return exports.AudioSchema.safeParse(obj).success;
}
function isClassificationResult(obj) {
    return exports.ClassificationResultSchema.safeParse(obj).success;
}
function isUsage(obj) {
    return exports.UsageSchema.safeParse(obj).success;
}
function isCustomData(obj) {
    return exports.CustomDataSchema.safeParse(obj).success;
}
function isDataStreamWithMetadata(obj) {
    return exports.DataStreamWithMetadataSchema.safeParse(obj).success;
}
function isPurpleError(obj) {
    return exports.PurpleErrorSchema.safeParse(obj).success;
}
function isAgentInfo(obj) {
    return exports.AgentInfoSchema.safeParse(obj).success;
}
function isBeliefState(obj) {
    return exports.BeliefStateSchema.safeParse(obj).success;
}
function isAgentAction(obj) {
    return exports.AgentActionSchema.safeParse(obj).success;
}
function isAgentSpeech(obj) {
    return exports.AgentSpeechSchema.safeParse(obj).success;
}
function isCondition(obj) {
    return exports.ConditionSchema.safeParse(obj).success;
}
function isIntentMatch(obj) {
    return exports.IntentMatchSchema.safeParse(obj).success;
}
function isResponseEvent(obj) {
    return exports.ResponseEventSchema.safeParse(obj).success;
}
function isTriggerEvent(obj) {
    return exports.TriggerEventSchema.safeParse(obj).success;
}
function isGoalAdvancementResponse(obj) {
    return exports.GoalAdvancementResponseSchema.safeParse(obj).success;
}
function isGoogleSerperApiSearchRequest(obj) {
    return library_1.GoogleSerperApiSearchRequestSchema.safeParse(obj).success;
}
function isKnowledgeGraph(obj) {
    return library_1.KnowledgeGraphSchema.safeParse(obj).success;
}
function isSitelink(obj) {
    return library_1.SitelinkSchema.safeParse(obj).success;
}
function isPeopleAlsoAsk(obj) {
    return library_1.PeopleAlsoAskSchema.safeParse(obj).success;
}
function isTopStory(obj) {
    return library_1.TopStorySchema.safeParse(obj).success;
}
function isJson(obj) {
    return exports.JsonSchema.safeParse(obj).success;
}
function isKnowledgeRecords(obj) {
    return library_1.KnowledgeRecordsSchema.safeParse(obj).success;
}
function isListToolsResponse(obj) {
    return exports.ListToolsResponseSchema.safeParse(obj).success;
}
function isLLMChatAttempt(obj) {
    return exports.LLMChatAttemptSchema.safeParse(obj).success;
}
function isLLMCompletionResponse(obj) {
    return exports.LLMCompletionResponseSchema.safeParse(obj).success;
}
function isMatchedIntents(obj) {
    return exports.MatchedIntentsSchema.safeParse(obj).success;
}
function isKeywordMatch(obj) {
    return library_1.KeywordMatchSchema.safeParse(obj).success;
}
function isPhraseMatch(obj) {
    return exports.PhraseMatchSchema.safeParse(obj).success;
}
function isEvent(obj) {
    return exports.EventSchema.safeParse(obj).success;
}
function isText(obj) {
    return exports.TextSchema.safeParse(obj).success;
}
function isSafetyResult(obj) {
    return exports.SafetyResultSchema.safeParse(obj).success;
}
function isSTTRequest(obj) {
    return exports.STTRequestSchema.safeParse(obj).success;
}
function isToolCallRequest(obj) {
    return exports.ToolCallRequestSchema.safeParse(obj).success;
}
function isToolCallResult(obj) {
    return library_1.ToolCallResultSchema.safeParse(obj).success;
}
function isTimestamp(obj) {
    return exports.TimestampSchema.safeParse(obj).success;
}
function isUserContext(obj) {
    return exports.UserContextSchema.safeParse(obj).success;
}
function isCredentials(obj) {
    return library_1.CredentialsSchema.safeParse(obj).success;
}
function isAddonConfig(obj) {
    return exports.AddonConfigSchema.safeParse(obj).success;
}
function isLoggerConfig(obj) {
    return library_1.LoggerConfigSchema.safeParse(obj).success;
}
function isTelemetryConfig(obj) {
    return library_1.TelemetryConfigSchema.safeParse(obj).success;
}
function isThreadPoolConfig(obj) {
    return exports.ThreadPoolConfigSchema.safeParse(obj).success;
}
function isFeatureFlagConfig(obj) {
    return exports.FeatureFlagConfigSchema.safeParse(obj).success;
}
function isCustomComponentCreationConfig(obj) {
    return exports.CustomComponentCreationConfigSchema.safeParse(obj).success;
}
function isAcousticEchoCancellationConfig(obj) {
    return library_1.AcousticEchoCancellationConfigSchema.safeParse(obj).success;
}
function isDeviceSelection(obj) {
    return library_1.DeviceSelectionSchema.safeParse(obj).success;
}
function isParsingConfig(obj) {
    return library_1.ParsingConfigSchema.safeParse(obj).success;
}
function isStreamSpeechRecognitionConfig(obj) {
    return library_1.StreamSpeechRecognitionConfigSchema.safeParse(obj).success;
}
function isLocalTextEmbedderConfig(obj) {
    return library_1.LocalTextEmbedderConfigSchema.safeParse(obj).success;
}
function isVoiceActivityDetectionConfig(obj) {
    return library_1.VoiceActivityDetectionConfigSchema.safeParse(obj).success;
}
function isMCPHttpConfig(obj) {
    return library_1.MCPHttpConfigSchema.safeParse(obj).success;
}
function isParagonApiConfig(obj) {
    return library_1.ParagonApiConfigSchema.safeParse(obj).success;
}
function isMCPStdioConfig(obj) {
    return library_1.MCPStdioConfigSchema.safeParse(obj).success;
}
function isDictionaryRule(obj) {
    return library_1.DictionaryRuleSchema.safeParse(obj).success;
}
function isLanguage(obj) {
    return library_1.LanguageSchema.safeParse(obj).success;
}
function isRemoteTextEmbedderConfig(obj) {
    return library_1.RemoteTextEmbedderConfigSchema.safeParse(obj).success;
}
function isTextClassifierCreationConfig(obj) {
    return library_1.TextClassifierCreationConfigSchema.safeParse(obj).success;
}
function isTurnDetectionConfig(obj) {
    return library_1.TurnDetectionConfigSchema.safeParse(obj).success;
}
function isGraphEdge(obj) {
    return exports.GraphEdgeSchema.safeParse(obj).success;
}
function isIntent(obj) {
    return library_1.IntentSchema.safeParse(obj).success;
}
function isEmbeddingMatcherNodeCreationConfig(obj) {
    return exports.EmbeddingMatcherNodeCreationConfigSchema.safeParse(obj).success;
}
function isIntentNodeCreationConfig(obj) {
    return exports.IntentNodeCreationConfigSchema.safeParse(obj).success;
}
function isKeywordGroup(obj) {
    return library_1.KeywordGroupSchema.safeParse(obj).success;
}
function isKnowledgeNodeCreationConfig(obj) {
    return exports.KnowledgeNodeCreationConfigSchema.safeParse(obj).success;
}
function isLLMChatRoutingNodeCreationConfig(obj) {
    return exports.LLMChatRoutingNodeCreationConfigSchema.safeParse(obj).success;
}
function isIntentsLlmPromptVariablesBuilderNodeCreationConfig(obj) {
    return exports.IntentsLlmPromptVariablesBuilderNodeCreationConfigSchema.safeParse(obj)
        .success;
}
function isIntentsStrictMatcherNodeCreationConfig(obj) {
    return exports.IntentsStrictMatcherNodeCreationConfigSchema.safeParse(obj).success;
}
function isTextClassifierNodeCreationConfig(obj) {
    return exports.TextClassifierNodeCreationConfigSchema.safeParse(obj).success;
}
function isTransformNodeCreationConfig(obj) {
    return exports.TransformNodeCreationConfigSchema.safeParse(obj).success;
}
function isSubgraphNode(obj) {
    return exports.SubgraphNodeSchema.safeParse(obj).success;
}
function isGraphParameter(obj) {
    return exports.GraphParameterSchema.safeParse(obj).success;
}
function isNodeErrorConfig(obj) {
    return exports.NodeErrorConfigSchema.safeParse(obj).success;
}
function isTextGenerationConfig(obj) {
    return library_1.TextGenerationConfigSchema.safeParse(obj).success;
}
function isLLMMatcherConfig(obj) {
    return library_1.LLMMatcherConfigSchema.safeParse(obj).success;
}
function isKnowledgeNodeExecutionConfig(obj) {
    return exports.KnowledgeNodeExecutionConfigSchema.safeParse(obj).success;
}
function isContentItem(obj) {
    return library_1.ContentItemSchema.safeParse(obj).success;
}
function isToolChoice(obj) {
    return library_1.ToolChoiceSchema.safeParse(obj).success;
}
function isLLMSelectionConfig(obj) {
    return exports.LLMSelectionConfigSchema.safeParse(obj).success;
}
function isLLMCompletionNodeExecutionConfig(obj) {
    return exports.LLMCompletionNodeExecutionConfigSchema.safeParse(obj).success;
}
function isSTTNodeExecutionConfig(obj) {
    return exports.STTNodeExecutionConfigSchema.safeParse(obj).success;
}
function isTextClassifierConfig(obj) {
    return library_1.TextClassifierConfigSchema.safeParse(obj).success;
}
function isElevenLabsSpeechSynthesisConfig(obj) {
    return library_1.ElevenLabsSpeechSynthesisConfigSchema.safeParse(obj).success;
}
function isInworldSpeechSynthesisConfig(obj) {
    return library_1.InworldSpeechSynthesisConfigSchema.safeParse(obj).success;
}
function isContentData(obj) {
    return exports.ContentDataSchema.safeParse(obj).success;
}
function isInteractionEvent(obj) {
    return exports.InteractionEventSchema.safeParse(obj).success;
}
function isGoalForEvaluation(obj) {
    return exports.GoalForEvaluationSchema.safeParse(obj).success;
}
function isQueryEvent(obj) {
    return exports.QueryEventSchema.safeParse(obj).success;
}
function isOrganicResult(obj) {
    return library_1.OrganicResultSchema.safeParse(obj).success;
}
function isLLMChatMetadata(obj) {
    return exports.LLMChatMetadataSchema.safeParse(obj).success;
}
function isLLMCompletionRequest(obj) {
    return exports.LLMCompletionRequestSchema.safeParse(obj).success;
}
function isMatchedKeywords(obj) {
    return exports.MatchedKeywordsSchema.safeParse(obj).success;
}
function isMatchedPhrases(obj) {
    return exports.MatchedPhrasesSchema.safeParse(obj).success;
}
function isMultimodalContent(obj) {
    return exports.MultimodalContentSchema.safeParse(obj).success;
}
function isToolCallResponse(obj) {
    return exports.ToolCallResponseSchema.safeParse(obj).success;
}
function isTTSOutput(obj) {
    return exports.TTSOutputSchema.safeParse(obj).success;
}
function isCreateEnvRequest(obj) {
    return exports.CreateEnvRequestSchema.safeParse(obj).success;
}
function isGraphCreationConfig(obj) {
    return exports.GraphCreationConfigSchema.safeParse(obj).success;
}
function isKnowledgeCompileConfig(obj) {
    return library_1.KnowledgeCompileConfigSchema.safeParse(obj).success;
}
function isMCPSessionCreationConfig(obj) {
    return library_1.MCPSessionCreationConfigSchema.safeParse(obj).success;
}
function isEntity(obj) {
    return library_1.EntitySchema.safeParse(obj).success;
}
function isRemoteKnowledgeConfig(obj) {
    return library_1.RemoteKnowledgeConfigSchema.safeParse(obj).success;
}
function isLocalTurnDetectorConfig(obj) {
    return library_1.LocalTurnDetectorConfigSchema.safeParse(obj).success;
}
function isIntentsEmbeddingMatcherNodeCreationConfig(obj) {
    return exports.IntentsEmbeddingMatcherNodeCreationConfigSchema.safeParse(obj).success;
}
function isGoalAdvancementNodeCreationConfig(obj) {
    return exports.GoalAdvancementNodeCreationConfigSchema.safeParse(obj).success;
}
function isKeywordMatcherNodeCreationConfig(obj) {
    return exports.KeywordMatcherNodeCreationConfigSchema.safeParse(obj).success;
}
function isGoalAdvancementNodeExecutionConfig(obj) {
    return exports.GoalAdvancementNodeExecutionConfigSchema.safeParse(obj).success;
}
function isIntentMatcherConfig(obj) {
    return library_1.IntentMatcherConfigSchema.safeParse(obj).success;
}
function isMessage(obj) {
    return library_1.MessageSchema.safeParse(obj).success;
}
function isLLMChatDefaults(obj) {
    return exports.LLMChatDefaultsSchema.safeParse(obj).success;
}
function isLLMChatRoute(obj) {
    return exports.LLMChatRouteSchema.safeParse(obj).success;
}
function isTextClassifierNodeExecutionConfig(obj) {
    return exports.TextClassifierNodeExecutionConfigSchema.safeParse(obj).success;
}
function isSpeechSynthesisConfig(obj) {
    return library_1.SpeechSynthesisConfigSchema.safeParse(obj).success;
}
function isGoalAdvancementRequest(obj) {
    return exports.GoalAdvancementRequestSchema.safeParse(obj).success;
}
function isGoogleSerperApiSearchResponse(obj) {
    return library_1.GoogleSerperApiSearchResponseSchema.safeParse(obj).success;
}
function isLLMChatRequest(obj) {
    return exports.LLMChatRequestSchema.safeParse(obj).success;
}
function isLLMChatResponse(obj) {
    return exports.LLMChatResponseSchema.safeParse(obj).success;
}
function isTTSRequest(obj) {
    return exports.TTSRequestSchema.safeParse(obj).success;
}
function isLocalKnowledgeConfig(obj) {
    return library_1.LocalKnowledgeConfigSchema.safeParse(obj).success;
}
function isMCPClientCreationConfig(obj) {
    return library_1.MCPClientCreationConfigSchema.safeParse(obj).success;
}
function isTurnDetectorCreationConfig(obj) {
    return library_1.TurnDetectorCreationConfigSchema.safeParse(obj).success;
}
function isNodeCreationConfig(obj) {
    return exports.NodeCreationConfigSchema.safeParse(obj).success;
}
function isIntentNodeExecutionConfig(obj) {
    return exports.IntentNodeExecutionConfigSchema.safeParse(obj).success;
}
function isLLMChatNodeExecutionConfig(obj) {
    return exports.LLMChatNodeExecutionConfigSchema.safeParse(obj).success;
}
function isLLMChatWeightedRoute(obj) {
    return exports.LLMChatWeightedRouteSchema.safeParse(obj).success;
}
function isTTSNodeExecutionConfig(obj) {
    return exports.TTSNodeExecutionConfigSchema.safeParse(obj).success;
}
function isComponentCreationConfig(obj) {
    return exports.ComponentCreationConfigSchema.safeParse(obj).success;
}
function isLLMChatWeightedLoadBalancingStrategy(obj) {
    return exports.LLMChatWeightedLoadBalancingStrategySchema.safeParse(obj).success;
}
function isComponent(obj) {
    return exports.ComponentSchema.safeParse(obj).success;
}
function isLLMChatRoutingStrategy(obj) {
    return exports.LLMChatRoutingStrategySchema.safeParse(obj).success;
}
function isLLMChatRoutingRequest(obj) {
    return exports.LLMChatRoutingRequestSchema.safeParse(obj).success;
}
function isLLMChatRoutingNodeExecutionConfig(obj) {
    return exports.LLMChatRoutingNodeExecutionConfigSchema.safeParse(obj).success;
}
function isDataContainer(obj) {
    return exports.DataContainerSchema.safeParse(obj).success;
}
function isNodeExecutionConfig(obj) {
    return exports.NodeExecutionConfigSchema.safeParse(obj).success;
}
function isExecutionContext(obj) {
    return exports.ExecutionContextSchema.safeParse(obj).success;
}
function isRegularNode(obj) {
    return exports.RegularNodeSchema.safeParse(obj).success;
}
function isProcessContext(obj) {
    return exports.ProcessContextSchema.safeParse(obj).success;
}
function isStartExecutionRequest(obj) {
    return exports.StartExecutionRequestSchema.safeParse(obj).success;
}
function isGraphNode(obj) {
    return exports.GraphNodeSchema.safeParse(obj).success;
}
function isSubgraph(obj) {
    return exports.SubgraphSchema.safeParse(obj).success;
}
function isMainGraph(obj) {
    return exports.MainGraphSchema.safeParse(obj).success;
}
function isGraphSpec(obj) {
    return exports.GraphSpecSchema.safeParse(obj).success;
}
function isCreateGraphRequest(obj) {
    return exports.CreateGraphRequestSchema.safeParse(obj).success;
}
