// @inworld/runtime/build/common/api/library.js
"use strict";
/**
 *
 * Note: This file is autogenerated and not intended to be modified manually.
 *
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmbeddingMatcherConfigSchema = exports.IntentSchema = exports.EntityMatchSchema = exports.ClassifyTextResponseSchema = exports.ClassThresholdSchema = exports.MCPStdioConfigSchema = exports.ParagonApiConfigSchema = exports.MCPHttpConfigSchema = exports.ToolListSchema = exports.ToolCallResultSchema = exports.LLMSpecSchema = exports.LLMPricingSchema = exports.LLMPerformanceSchema = exports.ModelIdSchema = exports.LLMBenchmarksSchema = exports.GetModelsRequestSchema = exports.ToolSchema = exports.FunctionChoiceSchema = exports.ReasoningConfigSchema = exports.ToolCallSchema = exports.ImageUrlSchema = exports.LogitBiasSchema = exports.RemoveKnowledgeRequestSchema = exports.LanguageSchema = exports.ParsingConfigSchema = exports.GetKnowledgeResponseSchema = exports.RetrievalConfigSchema = exports.CompileKnowledgeResponseSchema = exports.KnowledgeRecordsSchema = exports.FileSchema = exports.RemoteTextEmbedderConfigSchema = exports.DeviceSelectionSchema = exports.EmbedRequestSchema = exports.EmbedResponseSchema = exports.CredentialsSchema = exports.SpanStatusSchema = exports.SpanKindSchema = exports.MetricValueTypeSchema = exports.MetricKindSchema = exports.LogLevelSchema = exports.ExporterTypeSchema = exports.TurnStatusSchema = exports.TimestampTypeSchema = exports.ElevenLabsTextNormalizationSchema = exports.ElevenLabsOutputFormatSchema = exports.ResponseFormatSchema = exports.EffortSchema = exports.ImageDetailSchema = exports.FileTypeSchema = exports.DeviceTypeSchema = void 0;
exports.GenerateContentRequestSchema = exports.ToolChoiceSchema = exports.ContentItemSchema = exports.TextGenerationConfigSchema = exports.RemoteKnowledgeConfigSchema = exports.KnowledgeCompileConfigSchema = exports.GetKnowledgeRequestSchema = exports.CompileKnowledgeRequestSchema = exports.LocalTextEmbedderConfigSchema = exports.EmbedBatchResponseSchema = exports.EmbedBatchRequestSchema = exports.HistogramMetricSchema = exports.GaugeMetricSchema = exports.CounterMetricSchema = exports.LogMessageRequestSchema = exports.SpanEventSchema = exports.SimpleTelemetryConfigSchema = exports.TracerConfigSchema = exports.MetricConfigSchema = exports.BatchProcessingOptionsSchema = exports.LocalVadConfigSchema = exports.TurnPredictionSchema = exports.LocalTurnDetectorConfigSchema = exports.VoiceSchema = exports.SpeechSynthesisPostprocessingConfigSchema = exports.SpeechSynthesisInferenceConfigSchema = exports.ElevenLabsRequestParamsSchema = exports.ElevenLabsQueryParamsSchema = exports.RemoteStreamingSttConfigSchema = exports.LocalStreamingSttConfigSchema = exports.StreamSpeechRecognitionResponseSchema = exports.StreamSpeechRecognitionConfigSchema = exports.RemoteSttConfigSchema = exports.LocalSttConfigSchema = exports.SpeechRecognitionConfigSchema = exports.TurnDetectionConfigSchema = exports.AudioChunkTimestampSchema = exports.VoiceActivityDetectionConfigSchema = exports.AcousticEchoCancellationConfigSchema = exports.TopStorySchema = exports.PeopleAlsoAskSchema = exports.SitelinkSchema = exports.KnowledgeGraphSchema = exports.GoogleSerperApiSearchRequestSchema = exports.GoogleSerperApiClientCreationConfigSchema = exports.DeviceInfoSchema = exports.TextClassifierCreationConfigSchema = exports.DictionaryRuleSchema = exports.KeywordMatchSchema = exports.KeywordGroupSchema = void 0;
exports.ConfigureTelemetryRequestSchema = exports.TTSCreationConfigSchema = exports.SynthesizeSpeechRequestSchema = exports.GenerateContentChatRequestSchema = exports.KnowledgeCreationConfigSchema = exports.DequeueSpanTracesResponseSchema = exports.TelemetryConfigSchema = exports.RemoteTtsConfigSchema = exports.SpeechSynthesisConfigSchema = exports.DetectTurnCompletionRequestSchema = exports.AECFilterCreationConfigSchema = exports.GoogleSerperApiSearchResponseSchema = exports.NerCreationConfigSchema = exports.IntentMatcherConfigSchema = exports.ClassifyTextRequestSchema = exports.MCPClientCreationConfigSchema = exports.LLMCreationConfigSchema = exports.GetModelsResponseSchema = exports.MessageSchema = exports.LocalKnowledgeConfigSchema = exports.TextEmbedderCreationConfigSchema = exports.RecordMetricRequestSchema = exports.SpanDataSchema = exports.MetricsConfigSchema = exports.LoggerConfigSchema = exports.VADCreationConfigSchema = exports.TurnDetectorCreationConfigSchema = exports.InworldSpeechSynthesisConfigSchema = exports.ElevenLabsSpeechSynthesisConfigSchema = exports.StreamingSTTCreationConfigSchema = exports.StreamRecognizeSpeechRequestSchema = exports.STTCreationConfigSchema = exports.RecognizeSpeechRequestSchema = exports.FilterAudioRequestSchema = exports.DetectVoiceActivityRequestSchema = exports.AudioChunkSchema = exports.DetectSilenceRequestSchema = exports.LocalAecFilterConfigSchema = exports.OrganicResultSchema = exports.DeviceSchema = exports.EntitySchema = exports.KeywordMatcherResponseSchema = exports.LLMMatcherConfigSchema = exports.ExtractEntitiesResponseSchema = exports.TextClassifierConfigSchema = exports.MCPSessionCreationConfigSchema = exports.RemoteLlmConfigSchema = exports.LocalLlmConfigSchema = exports.ModelInfoSchema = void 0;
exports.isDeviceType = isDeviceType;
exports.isFileType = isFileType;
exports.isImageDetail = isImageDetail;
exports.isEffort = isEffort;
exports.isResponseFormat = isResponseFormat;
exports.isElevenLabsOutputFormat = isElevenLabsOutputFormat;
exports.isElevenLabsTextNormalization = isElevenLabsTextNormalization;
exports.isTimestampType = isTimestampType;
exports.isTurnStatus = isTurnStatus;
exports.isExporterType = isExporterType;
exports.isLogLevel = isLogLevel;
exports.isMetricKind = isMetricKind;
exports.isMetricValueType = isMetricValueType;
exports.isSpanKind = isSpanKind;
exports.isSpanStatus = isSpanStatus;
exports.isCredentials = isCredentials;
exports.isEmbedResponse = isEmbedResponse;
exports.isEmbedRequest = isEmbedRequest;
exports.isDeviceSelection = isDeviceSelection;
exports.isRemoteTextEmbedderConfig = isRemoteTextEmbedderConfig;
exports.isFile = isFile;
exports.isKnowledgeRecords = isKnowledgeRecords;
exports.isCompileKnowledgeResponse = isCompileKnowledgeResponse;
exports.isRetrievalConfig = isRetrievalConfig;
exports.isGetKnowledgeResponse = isGetKnowledgeResponse;
exports.isParsingConfig = isParsingConfig;
exports.isLanguage = isLanguage;
exports.isRemoveKnowledgeRequest = isRemoveKnowledgeRequest;
exports.isLogitBias = isLogitBias;
exports.isImageUrl = isImageUrl;
exports.isToolCall = isToolCall;
exports.isReasoningConfig = isReasoningConfig;
exports.isFunctionChoice = isFunctionChoice;
exports.isTool = isTool;
exports.isGetModelsRequest = isGetModelsRequest;
exports.isLLMBenchmarks = isLLMBenchmarks;
exports.isModelId = isModelId;
exports.isLLMPerformance = isLLMPerformance;
exports.isLLMPricing = isLLMPricing;
exports.isLLMSpec = isLLMSpec;
exports.isToolCallResult = isToolCallResult;
exports.isToolList = isToolList;
exports.isMCPHttpConfig = isMCPHttpConfig;
exports.isParagonApiConfig = isParagonApiConfig;
exports.isMCPStdioConfig = isMCPStdioConfig;
exports.isClassThreshold = isClassThreshold;
exports.isClassifyTextResponse = isClassifyTextResponse;
exports.isEntityMatch = isEntityMatch;
exports.isIntent = isIntent;
exports.isEmbeddingMatcherConfig = isEmbeddingMatcherConfig;
exports.isKeywordGroup = isKeywordGroup;
exports.isKeywordMatch = isKeywordMatch;
exports.isDictionaryRule = isDictionaryRule;
exports.isTextClassifierCreationConfig = isTextClassifierCreationConfig;
exports.isDeviceInfo = isDeviceInfo;
exports.isGoogleSerperApiClientCreationConfig = isGoogleSerperApiClientCreationConfig;
exports.isGoogleSerperApiSearchRequest = isGoogleSerperApiSearchRequest;
exports.isKnowledgeGraph = isKnowledgeGraph;
exports.isSitelink = isSitelink;
exports.isPeopleAlsoAsk = isPeopleAlsoAsk;
exports.isTopStory = isTopStory;
exports.isAcousticEchoCancellationConfig = isAcousticEchoCancellationConfig;
exports.isVoiceActivityDetectionConfig = isVoiceActivityDetectionConfig;
exports.isAudioChunkTimestamp = isAudioChunkTimestamp;
exports.isTurnDetectionConfig = isTurnDetectionConfig;
exports.isSpeechRecognitionConfig = isSpeechRecognitionConfig;
exports.isLocalSttConfig = isLocalSttConfig;
exports.isRemoteSttConfig = isRemoteSttConfig;
exports.isStreamSpeechRecognitionConfig = isStreamSpeechRecognitionConfig;
exports.isStreamSpeechRecognitionResponse = isStreamSpeechRecognitionResponse;
exports.isLocalStreamingSttConfig = isLocalStreamingSttConfig;
exports.isRemoteStreamingSttConfig = isRemoteStreamingSttConfig;
exports.isElevenLabsQueryParams = isElevenLabsQueryParams;
exports.isElevenLabsRequestParams = isElevenLabsRequestParams;
exports.isSpeechSynthesisInferenceConfig = isSpeechSynthesisInferenceConfig;
exports.isSpeechSynthesisPostprocessingConfig = isSpeechSynthesisPostprocessingConfig;
exports.isVoice = isVoice;
exports.isLocalTurnDetectorConfig = isLocalTurnDetectorConfig;
exports.isTurnPrediction = isTurnPrediction;
exports.isLocalVadConfig = isLocalVadConfig;
exports.isBatchProcessingOptions = isBatchProcessingOptions;
exports.isMetricConfig = isMetricConfig;
exports.isTracerConfig = isTracerConfig;
exports.isSimpleTelemetryConfig = isSimpleTelemetryConfig;
exports.isSpanEvent = isSpanEvent;
exports.isLogMessageRequest = isLogMessageRequest;
exports.isCounterMetric = isCounterMetric;
exports.isGaugeMetric = isGaugeMetric;
exports.isHistogramMetric = isHistogramMetric;
exports.isEmbedBatchRequest = isEmbedBatchRequest;
exports.isEmbedBatchResponse = isEmbedBatchResponse;
exports.isLocalTextEmbedderConfig = isLocalTextEmbedderConfig;
exports.isCompileKnowledgeRequest = isCompileKnowledgeRequest;
exports.isGetKnowledgeRequest = isGetKnowledgeRequest;
exports.isKnowledgeCompileConfig = isKnowledgeCompileConfig;
exports.isRemoteKnowledgeConfig = isRemoteKnowledgeConfig;
exports.isTextGenerationConfig = isTextGenerationConfig;
exports.isContentItem = isContentItem;
exports.isToolChoice = isToolChoice;
exports.isGenerateContentRequest = isGenerateContentRequest;
exports.isModelInfo = isModelInfo;
exports.isLocalLlmConfig = isLocalLlmConfig;
exports.isRemoteLlmConfig = isRemoteLlmConfig;
exports.isMCPSessionCreationConfig = isMCPSessionCreationConfig;
exports.isTextClassifierConfig = isTextClassifierConfig;
exports.isExtractEntitiesResponse = isExtractEntitiesResponse;
exports.isLLMMatcherConfig = isLLMMatcherConfig;
exports.isKeywordMatcherResponse = isKeywordMatcherResponse;
exports.isEntity = isEntity;
exports.isDevice = isDevice;
exports.isOrganicResult = isOrganicResult;
exports.isLocalAecFilterConfig = isLocalAecFilterConfig;
exports.isDetectSilenceRequest = isDetectSilenceRequest;
exports.isAudioChunk = isAudioChunk;
exports.isDetectVoiceActivityRequest = isDetectVoiceActivityRequest;
exports.isFilterAudioRequest = isFilterAudioRequest;
exports.isRecognizeSpeechRequest = isRecognizeSpeechRequest;
exports.isSTTCreationConfig = isSTTCreationConfig;
exports.isStreamRecognizeSpeechRequest = isStreamRecognizeSpeechRequest;
exports.isStreamingSTTCreationConfig = isStreamingSTTCreationConfig;
exports.isElevenLabsSpeechSynthesisConfig = isElevenLabsSpeechSynthesisConfig;
exports.isInworldSpeechSynthesisConfig = isInworldSpeechSynthesisConfig;
exports.isTurnDetectorCreationConfig = isTurnDetectorCreationConfig;
exports.isVADCreationConfig = isVADCreationConfig;
exports.isLoggerConfig = isLoggerConfig;
exports.isMetricsConfig = isMetricsConfig;
exports.isSpanData = isSpanData;
exports.isRecordMetricRequest = isRecordMetricRequest;
exports.isTextEmbedderCreationConfig = isTextEmbedderCreationConfig;
exports.isLocalKnowledgeConfig = isLocalKnowledgeConfig;
exports.isMessage = isMessage;
exports.isGetModelsResponse = isGetModelsResponse;
exports.isLLMCreationConfig = isLLMCreationConfig;
exports.isMCPClientCreationConfig = isMCPClientCreationConfig;
exports.isClassifyTextRequest = isClassifyTextRequest;
exports.isIntentMatcherConfig = isIntentMatcherConfig;
exports.isNerCreationConfig = isNerCreationConfig;
exports.isGoogleSerperApiSearchResponse = isGoogleSerperApiSearchResponse;
exports.isAECFilterCreationConfig = isAECFilterCreationConfig;
exports.isDetectTurnCompletionRequest = isDetectTurnCompletionRequest;
exports.isSpeechSynthesisConfig = isSpeechSynthesisConfig;
exports.isRemoteTtsConfig = isRemoteTtsConfig;
exports.isTelemetryConfig = isTelemetryConfig;
exports.isDequeueSpanTracesResponse = isDequeueSpanTracesResponse;
exports.isKnowledgeCreationConfig = isKnowledgeCreationConfig;
exports.isGenerateContentChatRequest = isGenerateContentChatRequest;
exports.isSynthesizeSpeechRequest = isSynthesizeSpeechRequest;
exports.isTTSCreationConfig = isTTSCreationConfig;
exports.isConfigureTelemetryRequest = isConfigureTelemetryRequest;
const z = __importStar(require("zod"));
/**
 * Device Type
 *
 * Type of the device.
 *
 * Allowed values: "CPU", "CUDA", "Metal", "ROCm"
 *
 * Type: 'CPU' | 'CUDA' | 'Metal' | 'ROCm'.
 */
exports.DeviceTypeSchema = z
    .union([z.enum(['CPU', 'CUDA', 'Metal', 'ROCm'])])
    .transform((val) => {
    const mapping = {
        CPU: 'DEVICE_TYPE_CPU',
        CUDA: 'DEVICE_TYPE_CUDA',
        Metal: 'DEVICE_TYPE_METAL',
        ROCm: 'DEVICE_TYPE_ROCM',
    };
    return mapping[val] || val;
});
/**
 * File Type
 *
 * Type of the file.
 *
 * Allowed values: "TYPE_UNSPECIFIED", "TYPE_PDF", "TYPE_TXT"
 *
 * Type: 'TYPE_UNSPECIFIED' | 'TYPE_PDF' | 'TYPE_TXT'.
 */
exports.FileTypeSchema = z.enum([
    'TYPE_PDF',
    'TYPE_TXT',
    'TYPE_UNSPECIFIED',
]);
/**
 * Image Detail
 *
 * Allowed values: "auto", "low", "high"
 *
 * Type: 'auto' | 'low' | 'high'.
 */
exports.ImageDetailSchema = z
    .union([z.enum(['auto', 'low', 'high'])])
    .transform((val) => {
    const mapping = {
        auto: 'IMAGE_DETAIL_AUTO',
        low: 'IMAGE_DETAIL_LOW',
        high: 'IMAGE_DETAIL_HIGH',
    };
    return mapping[val] || val;
});
/**
 * Effort
 *
 * Allowed values: "EFFORT_UNSPECIFIED", "EFFORT_NONE", "EFFORT_MINIMAL", "EFFORT_LOW", "EFFORT_MEDIUM", "EFFORT_HIGH", "EFFORT_XHIGH"
 *
 * Type: 'EFFORT_UNSPECIFIED' | 'EFFORT_NONE' | 'EFFORT_MINIMAL' | 'EFFORT_LOW' | 'EFFORT_MEDIUM' | 'EFFORT_HIGH' | 'EFFORT_XHIGH'.
 */
exports.EffortSchema = z.enum([
    'EFFORT_HIGH',
    'EFFORT_LOW',
    'EFFORT_MEDIUM',
    'EFFORT_MINIMAL',
    'EFFORT_NONE',
    'EFFORT_UNSPECIFIED',
    'EFFORT_XHIGH',
]);
/**
 * Response Format
 *
 * Optional response format specification
 *
 * Allowed values: "text", "json", "json_schema"
 *
 * Type: 'text' | 'json' | 'json_schema'.
 */
exports.ResponseFormatSchema = z
    .union([z.enum(['text', 'json', 'json_schema'])])
    .transform((val) => {
    const mapping = {
        text: 'RESPONSE_FORMAT_TEXT',
        json: 'RESPONSE_FORMAT_JSON',
        json_schema: 'RESPONSE_FORMAT_JSON_SCHEMA',
    };
    return mapping[val] || val;
});
/**
 * Eleven Labs Output Format
 *
 * Output format.
 *
 * Allowed values: "ELEVENLABS_OUTPUT_FORMAT_UNSPECIFIED", "ELEVENLABS_OUTPUT_FORMAT_MP3_22050_32", "ELEVENLABS_OUTPUT_FORMAT_MP3_44100_32", "ELEVENLABS_OUTPUT_FORMAT_MP3_44100_64", "ELEVENLABS_OUTPUT_FORMAT_MP3_44100_96", "ELEVENLABS_OUTPUT_FORMAT_MP3_44100_128", "ELEVENLABS_OUTPUT_FORMAT_MP3_44100_192"
 *
 * Type: 'ELEVENLABS_OUTPUT_FORMAT_UNSPECIFIED' | 'ELEVENLABS_OUTPUT_FORMAT_MP3_22050_32' | 'ELEVENLABS_OUTPUT_FORMAT_MP3_44100_32' | 'ELEVENLABS_OUTPUT_FORMAT_MP3_44100_64' | 'ELEVENLABS_OUTPUT_FORMAT_MP3_44100_96' | 'ELEVENLABS_OUTPUT_FORMAT_MP3_44100_128' | 'ELEVENLABS_OUTPUT_FORMAT_MP3_44100_192'.
 */
exports.ElevenLabsOutputFormatSchema = z.enum([
    'ELEVENLABS_OUTPUT_FORMAT_MP3_22050_32',
    'ELEVENLABS_OUTPUT_FORMAT_MP3_44100_128',
    'ELEVENLABS_OUTPUT_FORMAT_MP3_44100_192',
    'ELEVENLABS_OUTPUT_FORMAT_MP3_44100_32',
    'ELEVENLABS_OUTPUT_FORMAT_MP3_44100_64',
    'ELEVENLABS_OUTPUT_FORMAT_MP3_44100_96',
    'ELEVENLABS_OUTPUT_FORMAT_UNSPECIFIED',
]);
/**
 * Eleven Labs Text Normalization
 *
 * Text normalization setting.
 *
 * Allowed values: "ELEVENLABS_TEXT_NORMALIZATION_UNSPECIFIED", "ELEVENLABS_TEXT_NORMALIZATION_AUTO", "ELEVENLABS_TEXT_NORMALIZATION_ON", "ELEVENLABS_TEXT_NORMALIZATION_OFF"
 *
 * Type: 'ELEVENLABS_TEXT_NORMALIZATION_UNSPECIFIED' | 'ELEVENLABS_TEXT_NORMALIZATION_AUTO' | 'ELEVENLABS_TEXT_NORMALIZATION_ON' | 'ELEVENLABS_TEXT_NORMALIZATION_OFF'.
 */
exports.ElevenLabsTextNormalizationSchema = z.enum([
    'ELEVENLABS_TEXT_NORMALIZATION_AUTO',
    'ELEVENLABS_TEXT_NORMALIZATION_OFF',
    'ELEVENLABS_TEXT_NORMALIZATION_ON',
    'ELEVENLABS_TEXT_NORMALIZATION_UNSPECIFIED',
]);
/**
 * Timestamp Type
 *
 * Optional type of timestamp alignment to return.
 *
 * Allowed values: "TIMESTAMP_TYPE_UNSPECIFIED", "TIMESTAMP_TYPE_WORD", "TIMESTAMP_TYPE_CHARACTER"
 *
 * Type: 'TIMESTAMP_TYPE_UNSPECIFIED' | 'TIMESTAMP_TYPE_WORD' | 'TIMESTAMP_TYPE_CHARACTER'.
 */
exports.TimestampTypeSchema = z.enum([
    'TIMESTAMP_TYPE_CHARACTER',
    'TIMESTAMP_TYPE_UNSPECIFIED',
    'TIMESTAMP_TYPE_WORD',
]);
/**
 * Turn Status
 *
 * Detection result indicating whether the turn is complete or not.
 *
 * Allowed values: "Incomplete", "Complete"
 *
 * Type: 'Incomplete' | 'Complete'.
 */
exports.TurnStatusSchema = z
    .union([z.enum(['Incomplete', 'Complete'])])
    .transform((val) => {
    const mapping = {
        Incomplete: 'TURN_INCOMPLETE',
        Complete: 'TURN_COMPLETE',
    };
    return mapping[val] || val;
});
/**
 * Exporter Type
 *
 * Type of exporter to use.
 *
 * Allowed values: "EXPORTER_TYPE_UNSPECIFIED", "EXPORTER_TYPE_LOCAL", "EXPORTER_TYPE_REMOTE"
 *
 * Type: 'EXPORTER_TYPE_UNSPECIFIED' | 'EXPORTER_TYPE_LOCAL' | 'EXPORTER_TYPE_REMOTE'.
 */
exports.ExporterTypeSchema = z.enum([
    'EXPORTER_TYPE_LOCAL',
    'EXPORTER_TYPE_REMOTE',
    'EXPORTER_TYPE_UNSPECIFIED',
]);
/**
 * Log Level
 *
 * Minimum log level to record.
 *
 * Allowed values: "LOG_LEVEL_UNSPECIFIED", "LOG_LEVEL_TRACE", "LOG_LEVEL_DEBUG", "LOG_LEVEL_INFO", "LOG_LEVEL_WARN", "LOG_LEVEL_ERROR"
 *
 * Type: 'LOG_LEVEL_UNSPECIFIED' | 'LOG_LEVEL_TRACE' | 'LOG_LEVEL_DEBUG' | 'LOG_LEVEL_INFO' | 'LOG_LEVEL_WARN' | 'LOG_LEVEL_ERROR'.
 */
exports.LogLevelSchema = z.enum([
    'LOG_LEVEL_DEBUG',
    'LOG_LEVEL_ERROR',
    'LOG_LEVEL_INFO',
    'LOG_LEVEL_TRACE',
    'LOG_LEVEL_UNSPECIFIED',
    'LOG_LEVEL_WARN',
]);
/**
 * Metric Kind
 *
 * Kind of the metric (counter, gauge, or histogram).
 *
 * Allowed values: "METRIC_KIND_UNSPECIFIED", "METRIC_KIND_COUNTER", "METRIC_KIND_GAUGE", "METRIC_KIND_HISTOGRAM"
 *
 * Type: 'METRIC_KIND_UNSPECIFIED' | 'METRIC_KIND_COUNTER' | 'METRIC_KIND_GAUGE' | 'METRIC_KIND_HISTOGRAM'.
 */
exports.MetricKindSchema = z.enum([
    'METRIC_KIND_COUNTER',
    'METRIC_KIND_GAUGE',
    'METRIC_KIND_HISTOGRAM',
    'METRIC_KIND_UNSPECIFIED',
]);
/**
 * Metric Value Type
 *
 * Value type of the metric (integer or double).
 *
 * Allowed values: "METRIC_VALUE_TYPE_UNSPECIFIED", "METRIC_VALUE_TYPE_INTEGER", "METRIC_VALUE_TYPE_DOUBLE"
 *
 * Type: 'METRIC_VALUE_TYPE_UNSPECIFIED' | 'METRIC_VALUE_TYPE_INTEGER' | 'METRIC_VALUE_TYPE_DOUBLE'.
 */
exports.MetricValueTypeSchema = z.enum([
    'METRIC_VALUE_TYPE_DOUBLE',
    'METRIC_VALUE_TYPE_INTEGER',
    'METRIC_VALUE_TYPE_UNSPECIFIED',
]);
/**
 * Span Kind
 *
 * Span kind
 *
 * Allowed values: "SPAN_KIND_UNSPECIFIED", "SPAN_KIND_INTERNAL", "SPAN_KIND_SERVER", "SPAN_KIND_CLIENT", "SPAN_KIND_PRODUCER", "SPAN_KIND_CONSUMER"
 *
 * Type: 'SPAN_KIND_UNSPECIFIED' | 'SPAN_KIND_INTERNAL' | 'SPAN_KIND_SERVER' | 'SPAN_KIND_CLIENT' | 'SPAN_KIND_PRODUCER' | 'SPAN_KIND_CONSUMER'.
 */
exports.SpanKindSchema = z.enum([
    'SPAN_KIND_CLIENT',
    'SPAN_KIND_CONSUMER',
    'SPAN_KIND_INTERNAL',
    'SPAN_KIND_PRODUCER',
    'SPAN_KIND_SERVER',
    'SPAN_KIND_UNSPECIFIED',
]);
/**
 * Span Status
 *
 * Span status
 *
 * Allowed values: "SPAN_STATUS_UNSET", "SPAN_STATUS_OK", "SPAN_STATUS_ERROR"
 *
 * Type: 'SPAN_STATUS_UNSET' | 'SPAN_STATUS_OK' | 'SPAN_STATUS_ERROR'.
 */
exports.SpanStatusSchema = z.enum([
    'SPAN_STATUS_ERROR',
    'SPAN_STATUS_OK',
    'SPAN_STATUS_UNSET',
]);
/**
 * Credentials
 *
 * Credentials for authenticating with Inworld and third-party services.
 */
exports.CredentialsSchema = z.object({
    inworldApiKey: z.string().optional(),
    inworldJwt: z.string().optional(),
});
/**
 * Embed Response
 *
 * TextEmbedderInterface Response EmbedResponse returned by TextEmbedderInterface.Embed
 */
exports.EmbedResponseSchema = z.object({
    embedding: z.array(z.number()),
});
/**
 * Embed Request
 *
 * TextEmbedderInterface Request EmbedRequest for TextEmbedderInterface.Embed
 */
exports.EmbedRequestSchema = z.object({
    credentials: exports.CredentialsSchema.optional(),
    text: z.string(),
});
/**
 * Device Selection
 *
 * DeviceSelection specifies a device by type and index.
 */
exports.DeviceSelectionSchema = z.object({
    index: z.number(),
    type: exports.DeviceTypeSchema,
});
/**
 * Remote Text Embedder Config
 *
 * RemoteTextEmbedderConfig specifies the parameters required to initialize a remote embedding model that interacts with the Inworld API.
 */
exports.RemoteTextEmbedderConfigSchema = z.object({
    apiKey: z.string(),
    defaultTimeout: z.string().optional(),
    endpointOverride: z.string().optional(),
    modelName: z.string(),
    provider: z.string(),
});
/**
 * File
 *
 * File represents a file with its content and type.
 */
exports.FileSchema = z.object({
    content: z.string(),
    type: exports.FileTypeSchema,
});
/**
 * Knowledge Records
 *
 * Wrapper for a list of knowledge records.
 */
exports.KnowledgeRecordsSchema = z.object({
    records: z.array(z.string()).optional(),
});
/**
 * Compile Knowledge Response
 *
 * CompileKnowledgeResponse containing compiled knowledge records.
 */
exports.CompileKnowledgeResponseSchema = z.object({
    records: z.array(z.string()).optional(),
});
/**
 * Retrieval Config
 *
 * RetrievalConfig represents the configuration for retrieving knowledge.
 */
exports.RetrievalConfigSchema = z.object({
    threshold: z.number(),
    topK: z.number(),
});
/**
 * Get Knowledge Response
 *
 * GetKnowledgeResponse containing retrieved knowledge records.
 */
exports.GetKnowledgeResponseSchema = z.object({
    records: z.array(z.string()).optional(),
});
/**
 * Parsing Config
 *
 * ParsingConfig defines the configuration for parsing documents into chunks.
 */
exports.ParsingConfigSchema = z.object({
    maxCharsPerChunk: z.number(),
    maxChunksPerDocument: z.number(),
});
/**
 * Language
 *
 * Language represents a language with optional locale.
 */
exports.LanguageSchema = z.object({
    code: z.string(),
    locale: z.string().optional(),
});
/**
 * Remove Knowledge Request
 *
 * RemoveKnowledgeRequest for KnowledgeInterface.RemoveKnowledge
 */
exports.RemoveKnowledgeRequestSchema = z.object({
    credentials: exports.CredentialsSchema.optional(),
    knowledgeId: z.string(),
});
/**
 * Logit Bias
 *
 * LogitBias modifies the likelihood of specified tokens appearing in the completion.
 */
exports.LogitBiasSchema = z.object({
    biasValue: z.number(),
    tokenId: z.string(),
});
/**
 * Image Url
 *
 * ImageUrl represents an image URL with detail level specification.
 */
exports.ImageUrlSchema = z.object({
    detail: exports.ImageDetailSchema.optional(),
    url: z.string(),
});
/**
 * Tool Call
 *
 * ToolCall represents a tool call made by the model.
 */
exports.ToolCallSchema = z.object({
    args: z.string().optional(),
    id: z.string(),
    name: z.string(),
});
/**
 * Reasoning Config
 *
 * ReasoningConfig for models that support chain-of-thought reasoning. Provides a unified interface across different providers (OpenAI, Anthropic, Google, Groq, etc.).
 */
exports.ReasoningConfigSchema = z.object({
    effort: exports.EffortSchema.optional(),
    exclude: z.boolean().optional(),
    maxTokens: z.number().optional(),
});
/**
 * Function Choice
 *
 * FunctionChoice represents a specific function to be called.
 */
exports.FunctionChoiceSchema = z.object({
    name: z.string(),
    type: z.string(),
});
/**
 * Tool
 *
 * Tool represents a tool that the model may call.
 */
exports.ToolSchema = z.object({
    description: z.string(),
    name: z.string(),
    properties: z
        .union([z.string(), z.record(z.string(), z.any())])
        .transform((val) => {
        if (typeof val === 'object') {
            return JSON.stringify(val);
        }
        return val;
    }),
});
/**
 * Get Models Request
 *
 * GetModelsRequest for LLMInterface.GetModels
 */
exports.GetModelsRequestSchema = z.object({
    credentials: exports.CredentialsSchema.optional(),
    modelName: z.string().optional(),
    provider: z.string().optional(),
});
/**
 * LLM Benchmarks
 *
 * Benchmark scores for LLM model capabilities
 */
exports.LLMBenchmarksSchema = z.object({
    codingIndex: z.number().optional(),
    intelligenceIndex: z.number().optional(),
    mathIndex: z.number().optional(),
});
/**
 * Model Id
 *
 * ModelId represents the identifier for an LLM model.
 */
exports.ModelIdSchema = z.object({
    modelName: z.string().optional(),
    provider: z.string(),
});
/**
 * LLM Performance
 *
 * Performance metrics for LLM model latency and throughput
 */
exports.LLMPerformanceSchema = z.object({
    medianOutputTokensPerSecond: z.number().optional(),
    medianTimeToFirstAnswerToken: z.number().optional(),
    medianTimeToFirstTokenSeconds: z.number().optional(),
});
/**
 * LLM Pricing
 *
 * Pricing information for LLM model usage
 */
exports.LLMPricingSchema = z.object({
    completion: z.number().optional(),
    image: z.number().optional(),
    internalReasoning: z.number().optional(),
    prompt: z.number().optional(),
    request: z.number().optional(),
    webSearch: z.number().optional(),
});
/**
 * LLM Spec
 *
 * LLM model specifications and capabilities
 */
exports.LLMSpecSchema = z.object({
    contextLength: z.number().optional(),
    description: z.string().optional(),
    inputModalities: z.array(z.string()).optional(),
    maxCompletionTokens: z.number().optional(),
    outputModalities: z.array(z.string()).optional(),
    supportedParameters: z.array(z.string()).optional(),
});
/**
 * Tool Call Result
 *
 * ToolCallResult represents the result of a tool call.
 */
exports.ToolCallResultSchema = z.object({
    result: z.string(),
});
/**
 * Tool List
 *
 * ToolList represents a list of tools.
 */
exports.ToolListSchema = z.object({
    tools: z.array(exports.ToolSchema),
});
/**
 * MCP Http Config
 *
 * MCPHttpConfig configuration for creating an MCP client that communicates over HTTP.
 */
exports.MCPHttpConfigSchema = z.object({
    apiKey: z.string(),
    defaultTimeout: z.string().optional(),
});
/**
 * Paragon Api Config
 *
 * ParagonApiConfig configuration for creating an MCP client that communicates with the Paragon API.
 */
exports.ParagonApiConfigSchema = z.object({
    defaultTimeout: z.string().optional(),
    jwtToken: z.string(),
});
/**
 * MCP Stdio Config
 *
 * MCPStdioConfig configuration for creating an MCP client that communicates over stdio.
 */
exports.MCPStdioConfigSchema = z.object({
    env: z.record(z.string(), z.string()),
});
/**
 * Class Threshold
 *
 * ClassThreshold represents a class with its detection threshold.
 */
exports.ClassThresholdSchema = z.object({
    classLabel: z.string(),
    threshold: z.number(),
});
/**
 * Classify Text Response
 *
 * ClassifyTextResponse response from a text classification operation.
 */
exports.ClassifyTextResponseSchema = z.object({
    classes: z.array(z.string()),
});
/**
 * Entity Match
 *
 * EntityMatch a structure that describes a single occurrence of an entity in text.
 */
exports.EntityMatchSchema = z.object({
    entityName: z.string(),
    ruleName: z.string(),
    text: z.string(),
});
/**
 * Intent
 *
 * Intent represents an intent with a name and a list of training phrases.
 */
exports.IntentSchema = z.object({
    name: z.string(),
    phrases: z.array(z.string()),
});
/**
 * Embedding Matcher Config
 *
 * EmbeddingMatcherConfig configuration for embedding-based intent matching.
 */
exports.EmbeddingMatcherConfigSchema = z.object({
    similarityThreshold: z.number(),
});
/**
 * Keyword Group
 *
 * KeywordGroup represents a named group of keywords.
 */
exports.KeywordGroupSchema = z.object({
    keywords: z.array(z.string()),
    name: z.string(),
});
/**
 * Keyword Match
 *
 * KeywordMatch information about a matched keyword.
 */
exports.KeywordMatchSchema = z.object({
    groupName: z.string(),
    keyword: z.string(),
});
/**
 * Dictionary Rule
 *
 * DictionaryRule a rule that defines a custom entity and its variations.
 */
exports.DictionaryRuleSchema = z.object({
    displayName: z.string(),
    name: z.string(),
    synonyms: z.array(z.string()).optional(),
});
/**
 * Text Classifier Creation Config
 *
 * TextClassifierCreationConfig configuration for creating a text classifier.
 */
exports.TextClassifierCreationConfigSchema = z.object({
    modelWeightsPath: z.string(),
    supportedClasses: z.array(z.string()),
});
/**
 * Device Info
 *
 * DeviceInfo contains information about a device at a given point in time.
 */
exports.DeviceInfoSchema = z.object({
    freeMemoryBytes: z.number(),
    name: z.string(),
    timestamp: z.coerce.date(),
    totalMemoryBytes: z.number(),
});
exports.GoogleSerperApiClientCreationConfigSchema = z.object({
    apiEndpoint: z.string().optional(),
    apiKey: z.string(),
    defaultTimeout: z.string().optional(),
});
/**
 * Google Serper Api Search Request
 *
 * GoogleSerperApiSearchRequest for GoogleSerperAPIClientInterface.Search
 */
exports.GoogleSerperApiSearchRequestSchema = z.object({
    autocorrection: z.boolean().optional(),
    country: z.string().optional(),
    dateRange: z.string().optional(),
    language: z.string().optional(),
    numResults: z.number().optional(),
    page: z.number().optional(),
    query: z.string(),
});
/**
 * Knowledge Graph
 *
 * Represents knowledge graph information from search results
 */
exports.KnowledgeGraphSchema = z.object({
    attributes: z.record(z.string(), z.string()).optional(),
    description: z.string().optional(),
    descriptionLink: z.string().optional(),
    descriptionSource: z.string().optional(),
    imageUrl: z.string().optional(),
    title: z.string().optional(),
    type: z.string().optional(),
    website: z.string().optional(),
});
/**
 * Sitelink
 *
 * Represents a sitelink with title and URL
 */
exports.SitelinkSchema = z.object({
    link: z.string().optional(),
    title: z.string().optional(),
});
/**
 * People Also Ask
 *
 * Represents a "People Also Ask" question and answer
 */
exports.PeopleAlsoAskSchema = z.object({
    link: z.string().optional(),
    question: z.string().optional(),
    snippet: z.string().optional(),
    title: z.string().optional(),
});
/**
 * Top Story
 *
 * Represents a top story
 */
exports.TopStorySchema = z.object({
    date: z.string().optional(),
    link: z.string().optional(),
    source: z.string().optional(),
    title: z.string().optional(),
});
/**
 * Acoustic Echo Cancellation Config
 *
 * AcousticEchoCancellationConfig for AEC filtering behavior.
 *
 * Type: Record<string, unknown>.
 */
exports.AcousticEchoCancellationConfigSchema = z.object({});
/**
 * Voice Activity Detection Config
 *
 * VoiceActivityDetectionConfig specifies VAD model parameters.
 */
exports.VoiceActivityDetectionConfigSchema = z.object({
    speechThreshold: z.number(),
});
/**
 * Audio Chunk Timestamp
 *
 * AudioChunkTimestamp represents timing information for a token in audio.
 */
exports.AudioChunkTimestampSchema = z.object({
    endTime: z.number(),
    startTime: z.number(),
    token: z.string(),
});
/**
 * Turn Detection Config
 *
 * TurnDetectionConfig specifies turn detection model parameters.
 */
exports.TurnDetectionConfigSchema = z.object({
    threshold: z.number(),
});
/**
 * Speech Recognition Config
 *
 * SpeechRecognitionConfig specifies STT model parameters.
 */
exports.SpeechRecognitionConfigSchema = z.object({
    language: z.string().optional(),
});
exports.LocalSttConfigSchema = z.object({
    defaultConfig: exports.SpeechRecognitionConfigSchema,
    device: exports.DeviceSelectionSchema,
    modelPath: z.string(),
});
exports.RemoteSttConfigSchema = z.object({
    apiKey: z.string(),
    defaultConfig: exports.SpeechRecognitionConfigSchema,
    defaultTimeout: z.string().optional(),
    endpointOverride: z.string().optional(),
});
/**
 * Stream Speech Recognition Config
 *
 * StreamSpeechRecognitionConfig specifies streaming STT parameters.
 */
exports.StreamSpeechRecognitionConfigSchema = z.object({
    silenceThresholdMs: z.number(),
    speechConfig: exports.SpeechRecognitionConfigSchema,
});
/**
 * Stream Speech Recognition Response
 *
 * Streaming STT Interface Response Stream Element StreamSpeechRecognitionResponse for StreamingSTTInterface.StreamRecognizeSpeech
 */
exports.StreamSpeechRecognitionResponseSchema = z.object({
    isFinal: z.boolean(),
    text: z.string(),
});
exports.LocalStreamingSttConfigSchema = z.object({
    defaultConfig: exports.StreamSpeechRecognitionConfigSchema,
    sttDevice: exports.DeviceSelectionSchema,
    sttModelPath: z.string(),
    vadDevice: exports.DeviceSelectionSchema,
    vadModelPath: z.string(),
});
exports.RemoteStreamingSttConfigSchema = z.object({
    apiKey: z.string(),
    defaultConfig: exports.StreamSpeechRecognitionConfigSchema,
    defaultTimeout: z.string().optional(),
    endpointOverride: z.string().optional(),
    modelId: z.string().optional(),
});
/**
 * Eleven Labs Query Params
 *
 * ElevenLabsQueryParams for ElevenLabs TTS requests.
 */
exports.ElevenLabsQueryParamsSchema = z.object({
    enableLogging: z.boolean().optional(),
    outputFormat: exports.ElevenLabsOutputFormatSchema,
});
/**
 * Eleven Labs Request Params
 *
 * ElevenLabsRequestParams for ElevenLabs TTS requests.
 */
exports.ElevenLabsRequestParamsSchema = z.object({
    applyTextNormalization: exports.ElevenLabsTextNormalizationSchema,
    modelId: z.string(),
});
/**
 * Speech Synthesis Inference Config
 *
 * SpeechSynthesisInferenceConfig specifies model inference parameters for TTS.
 */
exports.SpeechSynthesisInferenceConfigSchema = z.object({
    speakingRate: z.number(),
    temperature: z.number(),
});
/**
 * Speech Synthesis Postprocessing Config
 *
 * SpeechSynthesisPostprocessingConfig specifies postprocessing for speech waveform.
 */
exports.SpeechSynthesisPostprocessingConfigSchema = z.object({
    sampleRate: z.number(),
});
/**
 * Voice
 *
 * Voice represents a voice for TTS synthesis.
 */
exports.VoiceSchema = z.object({
    id: z.string(),
    language: z.string().optional(),
});
/**
 * Local Turn Detector Config
 *
 * LocalTurnDetectorConfig specifies the parameters for initializing a local turn detector.
 */
exports.LocalTurnDetectorConfigSchema = z.object({
    defaultConfig: exports.TurnDetectionConfigSchema,
    device: exports.DeviceSelectionSchema,
    modelPath: z.string(),
});
/**
 * Turn Prediction
 *
 * TurnPrediction represents the result of turn detection analysis.
 */
exports.TurnPredictionSchema = z.object({
    probability: z.number(),
    status: exports.TurnStatusSchema,
});
exports.LocalVadConfigSchema = z.object({
    defaultConfig: exports.VoiceActivityDetectionConfigSchema,
    device: exports.DeviceSelectionSchema,
    modelPath: z.string(),
});
/**
 * Batch Processing Options
 *
 * BatchProcessingOptions configures batch processing for telemetry data.
 */
exports.BatchProcessingOptionsSchema = z.object({
    maxExportBatchSize: z.number().optional(),
    maxQueueSize: z.number().optional(),
    scheduleDelayMillis: z.number().optional(),
});
/**
 * Metric Config
 *
 * MetricConfig defines a single metric configuration.
 */
exports.MetricConfigSchema = z.object({
    description: z.string().optional(),
    metricKind: exports.MetricKindSchema.optional(),
    name: z.string().optional(),
    unit: z.string().optional(),
    valueType: exports.MetricValueTypeSchema.optional(),
});
/**
 * Tracer Config
 *
 * TracerConfig configures distributed tracing.
 */
exports.TracerConfigSchema = z.object({
    batchProcessingOpts: exports.BatchProcessingOptionsSchema.optional(),
    enableSpanCapture: z.boolean().optional(),
    maxSpanCaptureQueueSize: z.number().optional(),
    samplingRate: z.number().optional(),
    spanDumpPath: z.string().optional(),
});
/**
 * Simple Telemetry Config
 *
 * SimpleTelemetryConfig provides a minimal telemetry configuration.
 */
exports.SimpleTelemetryConfigSchema = z.object({
    serviceName: z.string().optional(),
    serviceVersion: z.string().optional(),
});
/**
 * Span Event
 *
 * SpanEvent - Represents an event within a span
 */
exports.SpanEventSchema = z.object({
    attributes: z.record(z.string(), z.string()).optional(),
    name: z.string().optional(),
    timestampInNanos: z.number().optional(),
});
/**
 * Log Message Request
 *
 * Telemetry Interface Request Wrapper LogMessageRequest for JsSpan.logMessage
 */
exports.LogMessageRequestSchema = z.object({
    attributes: z.record(z.string(), z.string()).optional(),
    logLevel: exports.LogLevelSchema,
    message: z.string(),
});
/**
 * Counter Metric
 *
 * CounterMetric represents a counter metric value.
 */
exports.CounterMetricSchema = z.object({
    doubleValue: z.number().optional(),
    uint64Value: z.number().optional(),
});
/**
 * Gauge Metric
 *
 * GaugeMetric represents a gauge metric value.
 */
exports.GaugeMetricSchema = z.object({
    doubleValue: z.number().optional(),
    int64Value: z.number().optional(),
});
/**
 * Histogram Metric
 *
 * HistogramMetric represents a histogram metric value.
 */
exports.HistogramMetricSchema = z.object({
    doubleValue: z.number().optional(),
    uint64Value: z.number().optional(),
});
/**
 * Embed Batch Request
 *
 * TextEmbedderInterface Request EmbedBatchRequest for TextEmbedderInterface.EmbedBatch
 */
exports.EmbedBatchRequestSchema = z.object({
    credentials: exports.CredentialsSchema.optional(),
    texts: z.array(z.string()),
});
/**
 * Embed Batch Response
 *
 * TextEmbedderInterface Response EmbedBatchResponse returned by TextEmbedderInterface.EmbedBatch
 */
exports.EmbedBatchResponseSchema = z.object({
    embeddings: z.array(exports.EmbedResponseSchema),
});
/**
 * Local Text Embedder Config
 *
 * LocalTextEmbedderConfig specifies the parameters required to initialize a local embedding model.
 */
exports.LocalTextEmbedderConfigSchema = z.object({
    device: exports.DeviceSelectionSchema,
    modelPath: z.string(),
});
/**
 * Compile Knowledge Request
 *
 * CompileKnowledgeRequest for KnowledgeInterface.CompileKnowledge
 */
exports.CompileKnowledgeRequestSchema = z.object({
    credentials: exports.CredentialsSchema.optional(),
    file: exports.FileSchema.optional(),
    knowledgeId: z.string(),
    records: exports.KnowledgeRecordsSchema.optional(),
});
/**
 * Get Knowledge Request
 *
 * GetKnowledgeRequest for KnowledgeInterface.GetKnowledge
 */
exports.GetKnowledgeRequestSchema = z.object({
    credentials: exports.CredentialsSchema.optional(),
    ids: z.array(z.string()),
    retrievalConfig: exports.RetrievalConfigSchema.optional(),
    searchQuery: z.string(),
});
/**
 * Knowledge Compile Config
 *
 * KnowledgeCompileConfig represents the configuration for compiling knowledge.
 */
exports.KnowledgeCompileConfigSchema = z.object({
    parsingConfig: exports.ParsingConfigSchema,
});
/**
 * Remote Knowledge Config
 *
 * RemoteKnowledgeConfig specifies the parameters for initializing a remote knowledge system.
 */
exports.RemoteKnowledgeConfigSchema = z.object({
    apiKey: z.string(),
    defaultTimeout: z.string().optional(),
    endpointOverride: z.string().optional(),
    knowledgeCompileConfig: exports.KnowledgeCompileConfigSchema,
    language: exports.LanguageSchema.optional(),
});
/**
 * Text Generation Config
 *
 * TextGenerationConfig represents parameters to control text generation.
 */
exports.TextGenerationConfigSchema = z.object({
    frequencyPenalty: z.number().optional(),
    logitBias: z.array(exports.LogitBiasSchema).optional(),
    maxNewTokens: z.number().optional(),
    maxPromptLength: z.number().optional(),
    presencePenalty: z.number().optional(),
    repetitionPenalty: z.number().optional(),
    seed: z.number().optional(),
    stopSequences: z.array(z.string()).optional(),
    temperature: z.number().optional(),
    topP: z.number().optional(),
});
/**
 * Content Item
 *
 * ContentItem represents a single content item in a message.
 */
exports.ContentItemSchema = z.object({
    image: exports.ImageUrlSchema.optional(),
    text: z.string().optional(),
});
/**
 * Tool Choice
 *
 * ToolChoice controls which (if any) tool is called by the model.
 */
exports.ToolChoiceSchema = z.object({
    function: exports.FunctionChoiceSchema.optional(),
    type: z.string(),
    value: z.string().optional(),
});
/**
 * Generate Content Request
 *
 * GenerateContentRequest for LLMInterface.GenerateContent
 */
exports.GenerateContentRequestSchema = z.object({
    config: exports.TextGenerationConfigSchema.optional(),
    credentials: exports.CredentialsSchema.optional(),
    prompt: z.string(),
});
/**
 * Model Info
 *
 * Complete information about an LLM model
 */
exports.ModelInfoSchema = z.object({
    benchmarks: exports.LLMBenchmarksSchema.optional(),
    modelId: exports.ModelIdSchema,
    performance: exports.LLMPerformanceSchema.optional(),
    pricing: exports.LLMPricingSchema.optional(),
    spec: exports.LLMSpecSchema.optional(),
});
exports.LocalLlmConfigSchema = z.object({
    defaultConfig: exports.TextGenerationConfigSchema,
    device: exports.DeviceSelectionSchema,
    modelPath: z.string(),
});
exports.RemoteLlmConfigSchema = z.object({
    apiKey: z.string(),
    defaultConfig: exports.TextGenerationConfigSchema,
    defaultTimeout: z.string().optional(),
    endpointOverride: z.string().optional(),
    modelName: z.string(),
    provider: z.string(),
});
/**
 * MCP Session Creation Config
 *
 * MCPSessionCreationConfig configuration for creating an MCP session.
 */
exports.MCPSessionCreationConfigSchema = z.object({
    endpoint: z.string(),
    httpConfig: exports.MCPHttpConfigSchema.optional(),
    paragonConfig: exports.ParagonApiConfigSchema.optional(),
    stdioConfig: exports.MCPStdioConfigSchema.optional(),
    transport: z.string(),
});
/**
 * Text Classifier Config
 *
 * TextClassifierConfig configuration for text classification.
 */
exports.TextClassifierConfigSchema = z.object({
    classes: z.array(exports.ClassThresholdSchema),
});
/**
 * Extract Entities Response
 *
 * Entity Extractor Interface Response Wrapper ExtractEntitiesResponse for EntityExtractorInterface.ExtractEntities
 */
exports.ExtractEntitiesResponseSchema = z.object({
    matches: z.array(exports.EntityMatchSchema),
});
/**
 * Llm Matcher Config
 *
 * LlmMatcherConfig configuration for LLM-based intent matching.
 */
exports.LLMMatcherConfigSchema = z.object({
    embeddingSimilarityThreshold: z.number().optional(),
    generationConfig: exports.TextGenerationConfigSchema.optional(),
    maxEmbeddingMatchesForLlm: z.number().optional(),
    promptTemplate: z.string(),
});
/**
 * Keyword Matcher Response
 *
 * KeywordMatcherResponse response from a keyword matching operation.
 */
exports.KeywordMatcherResponseSchema = z.object({
    matches: z.array(exports.KeywordMatchSchema),
});
/**
 * Entity
 *
 * Entity a custom entity category that can be extracted from text.
 */
exports.EntitySchema = z.object({
    name: z.string(),
    rules: z.array(exports.DictionaryRuleSchema),
});
/**
 * Device
 *
 * Device represents a local computation resource that can be used for model inference.
 */
exports.DeviceSchema = z.object({
    index: z.number(),
    info: exports.DeviceInfoSchema,
    type: exports.DeviceTypeSchema,
});
/**
 * Organic Result
 *
 * Represents a single organic search result
 */
exports.OrganicResultSchema = z.object({
    attributes: z.record(z.string(), z.string()).optional(),
    date: z.string().optional(),
    link: z.string().optional(),
    position: z.number().optional(),
    sitelinks: z.array(exports.SitelinkSchema).optional(),
    snippet: z.string().optional(),
    title: z.string().optional(),
});
exports.LocalAecFilterConfigSchema = z.object({
    defaultConfig: exports.AcousticEchoCancellationConfigSchema,
    device: exports.DeviceSelectionSchema,
});
/**
 * Detect Silence Request
 *
 * VAD Interface Request Wrapper DetectSilenceRequest for VADInterface.DetectSilence
 */
exports.DetectSilenceRequestSchema = z.object({
    audioBytes: z.string(),
    config: exports.VoiceActivityDetectionConfigSchema.optional(),
    sampleRate: z.number(),
});
/**
 * Audio Chunk
 *
 * AudioChunk represents a single audio chunk with timing information.
 */
exports.AudioChunkSchema = z.object({
    data: z.instanceof(Float32Array).optional(),
    sampleRate: z.number(),
    timestamps: z.array(exports.AudioChunkTimestampSchema).optional(),
});
/**
 * Detect Voice Activity Request
 *
 * VAD Interface Request Wrapper DetectVoiceActivityRequest for VADInterface.DetectVoiceActivity
 */
exports.DetectVoiceActivityRequestSchema = z.object({
    audioChunk: exports.AudioChunkSchema,
    config: exports.VoiceActivityDetectionConfigSchema.optional(),
});
/**
 * Filter Audio Request
 *
 * AEC Interface Request Wrapper FilterAudioRequest for AECInterface.FilterAudio
 */
exports.FilterAudioRequestSchema = z.object({
    config: exports.AcousticEchoCancellationConfigSchema.optional(),
    microphoneChunk: exports.AudioChunkSchema,
    speakersChunk: exports.AudioChunkSchema,
});
/**
 * Recognize Speech Request
 *
 * STT Interface Request Wrapper RecognizeSpeechRequest for STTInterface.RecognizeSpeech
 */
exports.RecognizeSpeechRequestSchema = z.object({
    audioChunk: exports.AudioChunkSchema,
    config: exports.SpeechRecognitionConfigSchema.optional(),
    credentials: exports.CredentialsSchema.optional(),
});
/**
 * STT Creation Config
 *
 * STTCreationConfig encapsulates the configuration for creating an STT model.
 */
exports.STTCreationConfigSchema = z.object({
    localConfig: exports.LocalSttConfigSchema.optional(),
    remoteConfig: exports.RemoteSttConfigSchema.optional(),
});
/**
 * Stream Recognize Speech Request
 *
 * Streaming STT Interface Request Wrapper StreamRecognizeSpeechRequest for StreamingSTTInterface.StreamRecognizeSpeech
 */
exports.StreamRecognizeSpeechRequestSchema = z.object({
    config: exports.StreamSpeechRecognitionConfigSchema.optional(),
    credentials: exports.CredentialsSchema.optional(),
});
/**
 * StreamingSTT Creation Config
 *
 * StreamingSTTCreationConfig encapsulates the configuration for creating a streaming STT.
 */
exports.StreamingSTTCreationConfigSchema = z.object({
    localConfig: exports.LocalStreamingSttConfigSchema.optional(),
    remoteConfig: exports.RemoteStreamingSttConfigSchema.optional(),
});
/**
 * Eleven Labs Speech Synthesis Config
 *
 * ElevenLabsSpeechSynthesisConfig for ElevenLabs TTS model.
 */
exports.ElevenLabsSpeechSynthesisConfigSchema = z.object({
    queryParams: exports.ElevenLabsQueryParamsSchema,
    requestParams: exports.ElevenLabsRequestParamsSchema,
});
/**
 * Inworld Speech Synthesis Config
 *
 * InworldSpeechSynthesisConfig for Inworld TTS model.
 */
exports.InworldSpeechSynthesisConfigSchema = z.object({
    inference: exports.SpeechSynthesisInferenceConfigSchema,
    modelId: z.string().optional(),
    postprocessing: exports.SpeechSynthesisPostprocessingConfigSchema,
    timestampType: exports.TimestampTypeSchema.optional(),
});
/**
 * Turn Detector Creation Config
 *
 * TurnDetectorCreationConfig encapsulates the configuration for creating a turn detector.
 */
exports.TurnDetectorCreationConfigSchema = z.object({
    localConfig: exports.LocalTurnDetectorConfigSchema.optional(),
});
/**
 * VAD Creation Config
 *
 * VADCreationConfig encapsulates the configuration for creating a VAD model.
 */
exports.VADCreationConfigSchema = z.object({
    localConfig: exports.LocalVadConfigSchema.optional(),
});
/**
 * Logger Config
 *
 * LoggerConfig configures logging behavior.
 */
exports.LoggerConfigSchema = z.object({
    batchProcessingOpts: exports.BatchProcessingOptionsSchema.optional(),
    logLevel: exports.LogLevelSchema.optional(),
    sinkAbslLogs: z.boolean().optional(),
});
/**
 * Metrics Config
 *
 * MetricsConfig configures metrics collection and export.
 */
exports.MetricsConfigSchema = z.object({
    exportInterval: z.number().optional(),
    exportTimeout: z.number().optional(),
    metricCacheSize: z.number().optional(),
    metrics: z.array(exports.MetricConfigSchema).optional(),
});
/**
 * Span Data
 *
 * SpanData - Represents a single span trace
 */
exports.SpanDataSchema = z.object({
    attributes: z.record(z.string(), z.string()).optional(),
    description: z.string().optional(),
    durationInNanos: z.number().optional(),
    events: z.array(exports.SpanEventSchema).optional(),
    kind: exports.SpanKindSchema.optional(),
    name: z.string().optional(),
    parentSpanId: z.string().optional(),
    spanId: z.string().optional(),
    startTimeInNanos: z.number().optional(),
    status: exports.SpanStatusSchema.optional(),
    traceId: z.string().optional(),
});
/**
 * Record Metric Request
 *
 * Telemetry Interface Request Wrapper RecordMetricRequest for JsSpan.recordMetric
 */
exports.RecordMetricRequestSchema = z.object({
    attributes: z.record(z.string(), z.string()).optional(),
    counter: exports.CounterMetricSchema.optional(),
    gauge: exports.GaugeMetricSchema.optional(),
    histogram: exports.HistogramMetricSchema.optional(),
    name: z.string(),
});
/**
 * Text Embedder Creation Config
 *
 * TextEmbedderCreationConfig encapsulates the configuration details for creating either a local or remote embedding model.
 */
exports.TextEmbedderCreationConfigSchema = z.object({
    localConfig: exports.LocalTextEmbedderConfigSchema.optional(),
    remoteConfig: exports.RemoteTextEmbedderConfigSchema.optional(),
});
/**
 * Local Knowledge Config
 *
 * LocalKnowledgeConfig specifies the parameters for initializing a local knowledge system.
 */
exports.LocalKnowledgeConfigSchema = z.object({
    embedderComponentId: z.string(),
    knowledgeCompileConfig: exports.KnowledgeCompileConfigSchema,
});
/**
 * Message
 *
 * Message represents a chat message in a conversation. For text-only messages, use `content`. For multimodal, use `content_items`.
 */
exports.MessageSchema = z.object({
    content: z.string().optional(),
    contentItems: z.array(exports.ContentItemSchema).optional(),
    role: z.string(),
    toolCallId: z.string().optional(),
    toolCalls: z.array(exports.ToolCallSchema).optional(),
});
/**
 * Get Models Response
 *
 * GetModelsResponse for LLMInterface.GetModels
 */
exports.GetModelsResponseSchema = z.object({
    models: z.array(exports.ModelInfoSchema).optional(),
});
/**
 * LLM Creation Config
 *
 * LLMCreationConfig represents a configuration for creating new LLM.
 */
exports.LLMCreationConfigSchema = z.object({
    localConfig: exports.LocalLlmConfigSchema.optional(),
    remoteConfig: exports.RemoteLlmConfigSchema.optional(),
});
/**
 * MCP Client Creation Config
 *
 * MCPClientCreationConfig configuration for creating an MCP client.
 */
exports.MCPClientCreationConfigSchema = z.object({
    sessionConfig: exports.MCPSessionCreationConfigSchema,
});
/**
 * Classify Text Request
 *
 * ClassifyTextRequest for TextClassifierInterface.ClassifyText
 */
exports.ClassifyTextRequestSchema = z.object({
    config: exports.TextClassifierConfigSchema.optional(),
    text: z.string(),
});
/**
 * Intent Matcher Config
 *
 * IntentMatcherConfig configuration for intent matching.
 */
exports.IntentMatcherConfigSchema = z.object({
    embedding: exports.EmbeddingMatcherConfigSchema.optional(),
    llm: exports.LLMMatcherConfigSchema.optional(),
    topNIntents: z.number().optional(),
});
exports.NerCreationConfigSchema = z.object({
    entities: z.array(exports.EntitySchema),
});
/**
 * Google Serper Api Search Response
 *
 * GoogleSerperApiSearchResponse from GoogleSerperAPIClientInterface.Search
 */
exports.GoogleSerperApiSearchResponseSchema = z.object({
    credits: z.number().optional(),
    knowledgeGraph: exports.KnowledgeGraphSchema.optional(),
    organicResults: z.array(exports.OrganicResultSchema).optional(),
    peopleAlsoAsk: z.array(exports.PeopleAlsoAskSchema).optional(),
    relatedSearches: z.array(z.string()).optional(),
    topStories: z.array(exports.TopStorySchema).optional(),
});
/**
 * AEC Filter Creation Config
 *
 * AECFilterCreationConfig encapsulates the configuration for creating an AEC filter.
 */
exports.AECFilterCreationConfigSchema = z.object({
    localConfig: exports.LocalAecFilterConfigSchema.optional(),
});
/**
 * Detect Turn Completion Request
 *
 * Turn Detector Interface Request Wrapper DetectTurnCompletionRequest for TurnDetectorInterface.DetectTurnCompletion
 */
exports.DetectTurnCompletionRequestSchema = z.object({
    audioChunk: exports.AudioChunkSchema,
    config: exports.TurnDetectionConfigSchema.optional(),
});
/**
 * Speech Synthesis Config
 *
 * SpeechSynthesisConfig specifies the configuration for speech synthesis.
 */
exports.SpeechSynthesisConfigSchema = z.object({
    elevenlabsConfig: exports.ElevenLabsSpeechSynthesisConfigSchema.optional(),
    inworldConfig: exports.InworldSpeechSynthesisConfigSchema.optional(),
});
exports.RemoteTtsConfigSchema = z.object({
    apiKey: z.string(),
    defaultTimeout: z.string().optional(),
    endpointOverride: z.string().optional(),
    synthesisConfig: exports.SpeechSynthesisConfigSchema,
});
/**
 * Telemetry Config
 *
 * TelemetryConfig provides comprehensive telemetry configuration.
 */
exports.TelemetryConfigSchema = z.object({
    apiKey: z.string().optional(),
    endpoint: z.string().optional(),
    exporterType: exports.ExporterTypeSchema.optional(),
    flushTimeoutMs: z.number().optional(),
    logger: exports.LoggerConfigSchema.optional(),
    metrics: exports.MetricsConfigSchema.optional(),
    sdkName: z.string().optional(),
    sdkVersion: z.string().optional(),
    serviceName: z.string().optional(),
    serviceVersion: z.string().optional(),
    shutdownTimeoutMs: z.number().optional(),
    tracer: exports.TracerConfigSchema.optional(),
});
/**
 * Dequeue Span Traces Response
 *
 * DequeueSpanTracesResponse - Response for dequeueSpanTraces method
 */
exports.DequeueSpanTracesResponseSchema = z.object({
    spans: z.array(exports.SpanDataSchema).optional(),
});
/**
 * Knowledge Creation Config
 *
 * KnowledgeCreationConfig encapsulates the configuration for creating a knowledge system.
 */
exports.KnowledgeCreationConfigSchema = z.object({
    localConfig: exports.LocalKnowledgeConfigSchema.optional(),
    remoteConfig: exports.RemoteKnowledgeConfigSchema.optional(),
});
/**
 * Generate Content Chat Request
 *
 * GenerateContentChatRequest for LLMInterface.GenerateContentChat
 */
exports.GenerateContentChatRequestSchema = z.object({
    config: exports.TextGenerationConfigSchema.optional(),
    credentials: exports.CredentialsSchema.optional(),
    messages: z.array(exports.MessageSchema),
    reasoning: exports.ReasoningConfigSchema.optional(),
    responseFormat: exports.ResponseFormatSchema.optional(),
    toolChoice: exports.ToolChoiceSchema.optional(),
    tools: z.array(exports.ToolSchema).optional(),
});
/**
 * Synthesize Speech Request
 *
 * TTS Interface Request Wrapper SynthesizeSpeechRequest for TTSInterface.SynthesizeSpeech
 */
exports.SynthesizeSpeechRequestSchema = z.object({
    config: exports.SpeechSynthesisConfigSchema.optional(),
    credentials: exports.CredentialsSchema.optional(),
    text: z.string(),
    voice: exports.VoiceSchema,
});
/**
 * TTS Creation Config
 *
 * TTSCreationConfig encapsulates the configuration for creating a TTS model.
 */
exports.TTSCreationConfigSchema = z.object({
    remoteConfig: exports.RemoteTtsConfigSchema.optional(),
});
/**
 * Configure Telemetry Request
 *
 * Telemetry Interface Request Wrapper ConfigureTelemetryRequest for telemetry.configureTelemetry
 */
exports.ConfigureTelemetryRequestSchema = z.object({
    config: exports.TelemetryConfigSchema.optional(),
    simpleConfig: exports.SimpleTelemetryConfigSchema.optional(),
});
function isDeviceType(obj) {
    return exports.DeviceTypeSchema.safeParse(obj).success;
}
function isFileType(obj) {
    return exports.FileTypeSchema.safeParse(obj).success;
}
function isImageDetail(obj) {
    return exports.ImageDetailSchema.safeParse(obj).success;
}
function isEffort(obj) {
    return exports.EffortSchema.safeParse(obj).success;
}
function isResponseFormat(obj) {
    return exports.ResponseFormatSchema.safeParse(obj).success;
}
function isElevenLabsOutputFormat(obj) {
    return exports.ElevenLabsOutputFormatSchema.safeParse(obj).success;
}
function isElevenLabsTextNormalization(obj) {
    return exports.ElevenLabsTextNormalizationSchema.safeParse(obj).success;
}
function isTimestampType(obj) {
    return exports.TimestampTypeSchema.safeParse(obj).success;
}
function isTurnStatus(obj) {
    return exports.TurnStatusSchema.safeParse(obj).success;
}
function isExporterType(obj) {
    return exports.ExporterTypeSchema.safeParse(obj).success;
}
function isLogLevel(obj) {
    return exports.LogLevelSchema.safeParse(obj).success;
}
function isMetricKind(obj) {
    return exports.MetricKindSchema.safeParse(obj).success;
}
function isMetricValueType(obj) {
    return exports.MetricValueTypeSchema.safeParse(obj).success;
}
function isSpanKind(obj) {
    return exports.SpanKindSchema.safeParse(obj).success;
}
function isSpanStatus(obj) {
    return exports.SpanStatusSchema.safeParse(obj).success;
}
function isCredentials(obj) {
    return exports.CredentialsSchema.safeParse(obj).success;
}
function isEmbedResponse(obj) {
    return exports.EmbedResponseSchema.safeParse(obj).success;
}
function isEmbedRequest(obj) {
    return exports.EmbedRequestSchema.safeParse(obj).success;
}
function isDeviceSelection(obj) {
    return exports.DeviceSelectionSchema.safeParse(obj).success;
}
function isRemoteTextEmbedderConfig(obj) {
    return exports.RemoteTextEmbedderConfigSchema.safeParse(obj).success;
}
function isFile(obj) {
    return exports.FileSchema.safeParse(obj).success;
}
function isKnowledgeRecords(obj) {
    return exports.KnowledgeRecordsSchema.safeParse(obj).success;
}
function isCompileKnowledgeResponse(obj) {
    return exports.CompileKnowledgeResponseSchema.safeParse(obj).success;
}
function isRetrievalConfig(obj) {
    return exports.RetrievalConfigSchema.safeParse(obj).success;
}
function isGetKnowledgeResponse(obj) {
    return exports.GetKnowledgeResponseSchema.safeParse(obj).success;
}
function isParsingConfig(obj) {
    return exports.ParsingConfigSchema.safeParse(obj).success;
}
function isLanguage(obj) {
    return exports.LanguageSchema.safeParse(obj).success;
}
function isRemoveKnowledgeRequest(obj) {
    return exports.RemoveKnowledgeRequestSchema.safeParse(obj).success;
}
function isLogitBias(obj) {
    return exports.LogitBiasSchema.safeParse(obj).success;
}
function isImageUrl(obj) {
    return exports.ImageUrlSchema.safeParse(obj).success;
}
function isToolCall(obj) {
    return exports.ToolCallSchema.safeParse(obj).success;
}
function isReasoningConfig(obj) {
    return exports.ReasoningConfigSchema.safeParse(obj).success;
}
function isFunctionChoice(obj) {
    return exports.FunctionChoiceSchema.safeParse(obj).success;
}
function isTool(obj) {
    return exports.ToolSchema.safeParse(obj).success;
}
function isGetModelsRequest(obj) {
    return exports.GetModelsRequestSchema.safeParse(obj).success;
}
function isLLMBenchmarks(obj) {
    return exports.LLMBenchmarksSchema.safeParse(obj).success;
}
function isModelId(obj) {
    return exports.ModelIdSchema.safeParse(obj).success;
}
function isLLMPerformance(obj) {
    return exports.LLMPerformanceSchema.safeParse(obj).success;
}
function isLLMPricing(obj) {
    return exports.LLMPricingSchema.safeParse(obj).success;
}
function isLLMSpec(obj) {
    return exports.LLMSpecSchema.safeParse(obj).success;
}
function isToolCallResult(obj) {
    return exports.ToolCallResultSchema.safeParse(obj).success;
}
function isToolList(obj) {
    return exports.ToolListSchema.safeParse(obj).success;
}
function isMCPHttpConfig(obj) {
    return exports.MCPHttpConfigSchema.safeParse(obj).success;
}
function isParagonApiConfig(obj) {
    return exports.ParagonApiConfigSchema.safeParse(obj).success;
}
function isMCPStdioConfig(obj) {
    return exports.MCPStdioConfigSchema.safeParse(obj).success;
}
function isClassThreshold(obj) {
    return exports.ClassThresholdSchema.safeParse(obj).success;
}
function isClassifyTextResponse(obj) {
    return exports.ClassifyTextResponseSchema.safeParse(obj).success;
}
function isEntityMatch(obj) {
    return exports.EntityMatchSchema.safeParse(obj).success;
}
function isIntent(obj) {
    return exports.IntentSchema.safeParse(obj).success;
}
function isEmbeddingMatcherConfig(obj) {
    return exports.EmbeddingMatcherConfigSchema.safeParse(obj).success;
}
function isKeywordGroup(obj) {
    return exports.KeywordGroupSchema.safeParse(obj).success;
}
function isKeywordMatch(obj) {
    return exports.KeywordMatchSchema.safeParse(obj).success;
}
function isDictionaryRule(obj) {
    return exports.DictionaryRuleSchema.safeParse(obj).success;
}
function isTextClassifierCreationConfig(obj) {
    return exports.TextClassifierCreationConfigSchema.safeParse(obj).success;
}
function isDeviceInfo(obj) {
    return exports.DeviceInfoSchema.safeParse(obj).success;
}
function isGoogleSerperApiClientCreationConfig(obj) {
    return exports.GoogleSerperApiClientCreationConfigSchema.safeParse(obj).success;
}
function isGoogleSerperApiSearchRequest(obj) {
    return exports.GoogleSerperApiSearchRequestSchema.safeParse(obj).success;
}
function isKnowledgeGraph(obj) {
    return exports.KnowledgeGraphSchema.safeParse(obj).success;
}
function isSitelink(obj) {
    return exports.SitelinkSchema.safeParse(obj).success;
}
function isPeopleAlsoAsk(obj) {
    return exports.PeopleAlsoAskSchema.safeParse(obj).success;
}
function isTopStory(obj) {
    return exports.TopStorySchema.safeParse(obj).success;
}
function isAcousticEchoCancellationConfig(obj) {
    return exports.AcousticEchoCancellationConfigSchema.safeParse(obj).success;
}
function isVoiceActivityDetectionConfig(obj) {
    return exports.VoiceActivityDetectionConfigSchema.safeParse(obj).success;
}
function isAudioChunkTimestamp(obj) {
    return exports.AudioChunkTimestampSchema.safeParse(obj).success;
}
function isTurnDetectionConfig(obj) {
    return exports.TurnDetectionConfigSchema.safeParse(obj).success;
}
function isSpeechRecognitionConfig(obj) {
    return exports.SpeechRecognitionConfigSchema.safeParse(obj).success;
}
function isLocalSttConfig(obj) {
    return exports.LocalSttConfigSchema.safeParse(obj).success;
}
function isRemoteSttConfig(obj) {
    return exports.RemoteSttConfigSchema.safeParse(obj).success;
}
function isStreamSpeechRecognitionConfig(obj) {
    return exports.StreamSpeechRecognitionConfigSchema.safeParse(obj).success;
}
function isStreamSpeechRecognitionResponse(obj) {
    return exports.StreamSpeechRecognitionResponseSchema.safeParse(obj).success;
}
function isLocalStreamingSttConfig(obj) {
    return exports.LocalStreamingSttConfigSchema.safeParse(obj).success;
}
function isRemoteStreamingSttConfig(obj) {
    return exports.RemoteStreamingSttConfigSchema.safeParse(obj).success;
}
function isElevenLabsQueryParams(obj) {
    return exports.ElevenLabsQueryParamsSchema.safeParse(obj).success;
}
function isElevenLabsRequestParams(obj) {
    return exports.ElevenLabsRequestParamsSchema.safeParse(obj).success;
}
function isSpeechSynthesisInferenceConfig(obj) {
    return exports.SpeechSynthesisInferenceConfigSchema.safeParse(obj).success;
}
function isSpeechSynthesisPostprocessingConfig(obj) {
    return exports.SpeechSynthesisPostprocessingConfigSchema.safeParse(obj).success;
}
function isVoice(obj) {
    return exports.VoiceSchema.safeParse(obj).success;
}
function isLocalTurnDetectorConfig(obj) {
    return exports.LocalTurnDetectorConfigSchema.safeParse(obj).success;
}
function isTurnPrediction(obj) {
    return exports.TurnPredictionSchema.safeParse(obj).success;
}
function isLocalVadConfig(obj) {
    return exports.LocalVadConfigSchema.safeParse(obj).success;
}
function isBatchProcessingOptions(obj) {
    return exports.BatchProcessingOptionsSchema.safeParse(obj).success;
}
function isMetricConfig(obj) {
    return exports.MetricConfigSchema.safeParse(obj).success;
}
function isTracerConfig(obj) {
    return exports.TracerConfigSchema.safeParse(obj).success;
}
function isSimpleTelemetryConfig(obj) {
    return exports.SimpleTelemetryConfigSchema.safeParse(obj).success;
}
function isSpanEvent(obj) {
    return exports.SpanEventSchema.safeParse(obj).success;
}
function isLogMessageRequest(obj) {
    return exports.LogMessageRequestSchema.safeParse(obj).success;
}
function isCounterMetric(obj) {
    return exports.CounterMetricSchema.safeParse(obj).success;
}
function isGaugeMetric(obj) {
    return exports.GaugeMetricSchema.safeParse(obj).success;
}
function isHistogramMetric(obj) {
    return exports.HistogramMetricSchema.safeParse(obj).success;
}
function isEmbedBatchRequest(obj) {
    return exports.EmbedBatchRequestSchema.safeParse(obj).success;
}
function isEmbedBatchResponse(obj) {
    return exports.EmbedBatchResponseSchema.safeParse(obj).success;
}
function isLocalTextEmbedderConfig(obj) {
    return exports.LocalTextEmbedderConfigSchema.safeParse(obj).success;
}
function isCompileKnowledgeRequest(obj) {
    return exports.CompileKnowledgeRequestSchema.safeParse(obj).success;
}
function isGetKnowledgeRequest(obj) {
    return exports.GetKnowledgeRequestSchema.safeParse(obj).success;
}
function isKnowledgeCompileConfig(obj) {
    return exports.KnowledgeCompileConfigSchema.safeParse(obj).success;
}
function isRemoteKnowledgeConfig(obj) {
    return exports.RemoteKnowledgeConfigSchema.safeParse(obj).success;
}
function isTextGenerationConfig(obj) {
    return exports.TextGenerationConfigSchema.safeParse(obj).success;
}
function isContentItem(obj) {
    return exports.ContentItemSchema.safeParse(obj).success;
}
function isToolChoice(obj) {
    return exports.ToolChoiceSchema.safeParse(obj).success;
}
function isGenerateContentRequest(obj) {
    return exports.GenerateContentRequestSchema.safeParse(obj).success;
}
function isModelInfo(obj) {
    return exports.ModelInfoSchema.safeParse(obj).success;
}
function isLocalLlmConfig(obj) {
    return exports.LocalLlmConfigSchema.safeParse(obj).success;
}
function isRemoteLlmConfig(obj) {
    return exports.RemoteLlmConfigSchema.safeParse(obj).success;
}
function isMCPSessionCreationConfig(obj) {
    return exports.MCPSessionCreationConfigSchema.safeParse(obj).success;
}
function isTextClassifierConfig(obj) {
    return exports.TextClassifierConfigSchema.safeParse(obj).success;
}
function isExtractEntitiesResponse(obj) {
    return exports.ExtractEntitiesResponseSchema.safeParse(obj).success;
}
function isLLMMatcherConfig(obj) {
    return exports.LLMMatcherConfigSchema.safeParse(obj).success;
}
function isKeywordMatcherResponse(obj) {
    return exports.KeywordMatcherResponseSchema.safeParse(obj).success;
}
function isEntity(obj) {
    return exports.EntitySchema.safeParse(obj).success;
}
function isDevice(obj) {
    return exports.DeviceSchema.safeParse(obj).success;
}
function isOrganicResult(obj) {
    return exports.OrganicResultSchema.safeParse(obj).success;
}
function isLocalAecFilterConfig(obj) {
    return exports.LocalAecFilterConfigSchema.safeParse(obj).success;
}
function isDetectSilenceRequest(obj) {
    return exports.DetectSilenceRequestSchema.safeParse(obj).success;
}
function isAudioChunk(obj) {
    return exports.AudioChunkSchema.safeParse(obj).success;
}
function isDetectVoiceActivityRequest(obj) {
    return exports.DetectVoiceActivityRequestSchema.safeParse(obj).success;
}
function isFilterAudioRequest(obj) {
    return exports.FilterAudioRequestSchema.safeParse(obj).success;
}
function isRecognizeSpeechRequest(obj) {
    return exports.RecognizeSpeechRequestSchema.safeParse(obj).success;
}
function isSTTCreationConfig(obj) {
    return exports.STTCreationConfigSchema.safeParse(obj).success;
}
function isStreamRecognizeSpeechRequest(obj) {
    return exports.StreamRecognizeSpeechRequestSchema.safeParse(obj).success;
}
function isStreamingSTTCreationConfig(obj) {
    return exports.StreamingSTTCreationConfigSchema.safeParse(obj).success;
}
function isElevenLabsSpeechSynthesisConfig(obj) {
    return exports.ElevenLabsSpeechSynthesisConfigSchema.safeParse(obj).success;
}
function isInworldSpeechSynthesisConfig(obj) {
    return exports.InworldSpeechSynthesisConfigSchema.safeParse(obj).success;
}
function isTurnDetectorCreationConfig(obj) {
    return exports.TurnDetectorCreationConfigSchema.safeParse(obj).success;
}
function isVADCreationConfig(obj) {
    return exports.VADCreationConfigSchema.safeParse(obj).success;
}
function isLoggerConfig(obj) {
    return exports.LoggerConfigSchema.safeParse(obj).success;
}
function isMetricsConfig(obj) {
    return exports.MetricsConfigSchema.safeParse(obj).success;
}
function isSpanData(obj) {
    return exports.SpanDataSchema.safeParse(obj).success;
}
function isRecordMetricRequest(obj) {
    return exports.RecordMetricRequestSchema.safeParse(obj).success;
}
function isTextEmbedderCreationConfig(obj) {
    return exports.TextEmbedderCreationConfigSchema.safeParse(obj).success;
}
function isLocalKnowledgeConfig(obj) {
    return exports.LocalKnowledgeConfigSchema.safeParse(obj).success;
}
function isMessage(obj) {
    return exports.MessageSchema.safeParse(obj).success;
}
function isGetModelsResponse(obj) {
    return exports.GetModelsResponseSchema.safeParse(obj).success;
}
function isLLMCreationConfig(obj) {
    return exports.LLMCreationConfigSchema.safeParse(obj).success;
}
function isMCPClientCreationConfig(obj) {
    return exports.MCPClientCreationConfigSchema.safeParse(obj).success;
}
function isClassifyTextRequest(obj) {
    return exports.ClassifyTextRequestSchema.safeParse(obj).success;
}
function isIntentMatcherConfig(obj) {
    return exports.IntentMatcherConfigSchema.safeParse(obj).success;
}
function isNerCreationConfig(obj) {
    return exports.NerCreationConfigSchema.safeParse(obj).success;
}
function isGoogleSerperApiSearchResponse(obj) {
    return exports.GoogleSerperApiSearchResponseSchema.safeParse(obj).success;
}
function isAECFilterCreationConfig(obj) {
    return exports.AECFilterCreationConfigSchema.safeParse(obj).success;
}
function isDetectTurnCompletionRequest(obj) {
    return exports.DetectTurnCompletionRequestSchema.safeParse(obj).success;
}
function isSpeechSynthesisConfig(obj) {
    return exports.SpeechSynthesisConfigSchema.safeParse(obj).success;
}
function isRemoteTtsConfig(obj) {
    return exports.RemoteTtsConfigSchema.safeParse(obj).success;
}
function isTelemetryConfig(obj) {
    return exports.TelemetryConfigSchema.safeParse(obj).success;
}
function isDequeueSpanTracesResponse(obj) {
    return exports.DequeueSpanTracesResponseSchema.safeParse(obj).success;
}
function isKnowledgeCreationConfig(obj) {
    return exports.KnowledgeCreationConfigSchema.safeParse(obj).success;
}
function isGenerateContentChatRequest(obj) {
    return exports.GenerateContentChatRequestSchema.safeParse(obj).success;
}
function isSynthesizeSpeechRequest(obj) {
    return exports.SynthesizeSpeechRequestSchema.safeParse(obj).success;
}
function isTTSCreationConfig(obj) {
    return exports.TTSCreationConfigSchema.safeParse(obj).success;
}
function isConfigureTelemetryRequest(obj) {
    return exports.ConfigureTelemetryRequestSchema.safeParse(obj).success;
}
