// @inworld/runtime/build/common/data_types/type_mapping.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphTypesTypeGuards = exports.GraphTypes = void 0;
exports.getTypeName = getTypeName;
exports.isJsNonNullishSerializable = isJsNonNullishSerializable;
/**
 * @internal
 */
const inworld_1 = require("../../internal/error/inworld");
const stream_1 = require("../stream");
const audio_1 = require("./api/audio");
const classification_result_1 = require("./api/classification_result");
const custom_1 = require("./api/custom");
const goal_advancement_request_1 = require("./api/goal_advancement_request");
const goal_advancement_response_1 = require("./api/goal_advancement_response");
const knowledge_records_1 = require("./api/knowledge_records");
const list_tools_response_1 = require("./api/list_tools_response");
const llm_chat_request_1 = require("./api/llm_chat_request");
const llm_chat_response_1 = require("./api/llm_chat_response");
const matched_intents_1 = require("./api/matched_intents");
const matched_keywords_1 = require("./api/matched_keywords");
const multimodal_content_1 = require("./api/multimodal_content");
const safety_result_1 = require("./api/safety_result");
const tool_call_response_1 = require("./api/tool_call_response");
const tts_request_1 = require("./api/tts_request");
const graph_types_1 = require("./graph_types");
const GraphTypes = __importStar(require("./graph_types"));
exports.GraphTypes = GraphTypes;
const graph_types_2 = require("./graph_types");
/**
 * Represents a custom-typed pointer to any structure.
 * This is a virtual TypeScript type that provides type safety without runtime
 * overhead. At runtime, this will just be the raw object, but TypeScript will
 * understand the structure.
 *
 * @template T - The TypeScript type of the data structure
 *
 * @example
 * ```typescript
 * function process(input: Custom<{internalFields: string}>) {
 *   // TypeScript understands the structure
 *   console.log(input.internalFields);
 * }
 * ```
 */
/**
 * Type guards for CustomNodeInputTypes to enable easy type checking and
 * switch-case patterns. These functions help distinguish between different
 * types in graph output streams.
 */
var GraphTypesTypeGuards;
(function (GraphTypesTypeGuards) {
    function isString(value) {
        return typeof value === 'string';
    }
    GraphTypesTypeGuards.isString = isString;
    function isCustom(value) {
        if (value instanceof custom_1.CustomData) {
            return true;
        }
        return typeof value === 'object' && value !== null && !hasKnownType(value);
    }
    GraphTypesTypeGuards.isCustom = isCustom;
    function isClassificationResult(value) {
        return (value instanceof classification_result_1.ClassificationResult ||
            (isNapiObject(value) && value.type === 'ClassificationResult'));
    }
    GraphTypesTypeGuards.isClassificationResult = isClassificationResult;
    function isContent(value) {
        return (value instanceof graph_types_1.Content ||
            (isNapiObject(value) && value.type === 'Content'));
    }
    GraphTypesTypeGuards.isContent = isContent;
    function isGoalAdvancement(value) {
        return (value instanceof goal_advancement_response_1.GoalAdvancementResponse ||
            (isNapiObject(value) && value.type === 'GoalAdvancement'));
    }
    GraphTypesTypeGuards.isGoalAdvancement = isGoalAdvancement;
    function isGoalAdvancementResponse(value) {
        return isGoalAdvancement(value);
    }
    GraphTypesTypeGuards.isGoalAdvancementResponse = isGoalAdvancementResponse;
    function isGoalAdvancementRequest(value) {
        return (value instanceof goal_advancement_request_1.GoalAdvancementRequest ||
            (isNapiObject(value) && value.type === 'GoalAdvancementRequest'));
    }
    GraphTypesTypeGuards.isGoalAdvancementRequest = isGoalAdvancementRequest;
    function isKnowledgeRecords(value) {
        return (value instanceof knowledge_records_1.KnowledgeRecords ||
            (isNapiObject(value) && value.type === 'KnowledgeRecords'));
    }
    GraphTypesTypeGuards.isKnowledgeRecords = isKnowledgeRecords;
    function isListToolsResponse(value) {
        return (value instanceof list_tools_response_1.ListToolsResponse ||
            (isNapiObject(value) && value.type === 'ListToolsResponse'));
    }
    GraphTypesTypeGuards.isListToolsResponse = isListToolsResponse;
    function isLLMChatRequest(value) {
        return (value instanceof llm_chat_request_1.LLMChatRequest ||
            (isNapiObject(value) && value.type === 'LLMChatRequest'));
    }
    GraphTypesTypeGuards.isLLMChatRequest = isLLMChatRequest;
    function isLLMChatResponse(value) {
        return (value instanceof llm_chat_response_1.LLMChatResponse ||
            (isNapiObject(value) && value.type === 'LLMChatResponse'));
    }
    GraphTypesTypeGuards.isLLMChatResponse = isLLMChatResponse;
    function isTTSRequest(value) {
        return (value instanceof tts_request_1.TTSRequest ||
            (isNapiObject(value) && value.type === 'TTSRequest'));
    }
    GraphTypesTypeGuards.isTTSRequest = isTTSRequest;
    function isMatchedIntents(value) {
        return (value instanceof matched_intents_1.MatchedIntents ||
            (isNapiObject(value) && value.type === 'MatchedIntents'));
    }
    GraphTypesTypeGuards.isMatchedIntents = isMatchedIntents;
    function isMatchedKeywords(value) {
        return (value instanceof matched_keywords_1.MatchedKeywords ||
            (isNapiObject(value) && value.type === 'MatchedKeywords'));
    }
    GraphTypesTypeGuards.isMatchedKeywords = isMatchedKeywords;
    function isSafetyResult(value) {
        return (value instanceof safety_result_1.SafetyResult ||
            (isNapiObject(value) && value.type === 'SafetyResult'));
    }
    GraphTypesTypeGuards.isSafetyResult = isSafetyResult;
    function isToolCallResponse(value) {
        return (value instanceof tool_call_response_1.ToolCallResponse ||
            (isNapiObject(value) && value.type === 'ToolCallResponse'));
    }
    GraphTypesTypeGuards.isToolCallResponse = isToolCallResponse;
    function isTextStream(value) {
        return (value instanceof stream_1.TextStream ||
            (isNapiObject(value) && value.type === 'TextStream'));
    }
    GraphTypesTypeGuards.isTextStream = isTextStream;
    function isContentStream(value) {
        return (value instanceof stream_1.ContentStream ||
            (isNapiObject(value) && value.type === 'ContentStream'));
    }
    GraphTypesTypeGuards.isContentStream = isContentStream;
    function isTTSOutputStream(value) {
        return (value instanceof stream_1.TTSOutputStream ||
            (isNapiObject(value) && value.type === 'TTSOutputStream'));
    }
    GraphTypesTypeGuards.isTTSOutputStream = isTTSOutputStream;
    function isSpeechChunkStream(value) {
        return (value instanceof stream_1.SpeechChunkStream ||
            (isNapiObject(value) && value.type === 'SpeechChunkStream'));
    }
    GraphTypesTypeGuards.isSpeechChunkStream = isSpeechChunkStream;
    function isAudioChunkStream(value) {
        return (value instanceof stream_1.AudioChunkStream ||
            (isNapiObject(value) && value.type === 'AudioChunkStream'));
    }
    GraphTypesTypeGuards.isAudioChunkStream = isAudioChunkStream;
    function isAudio(value) {
        return (value instanceof audio_1.Audio || (isNapiObject(value) && value.type === 'Audio'));
    }
    GraphTypesTypeGuards.isAudio = isAudio;
    function isMultimodalContent(value) {
        return (value instanceof multimodal_content_1.MultimodalContent ||
            (isNapiObject(value) && value.type === 'MultimodalContent'));
    }
    GraphTypesTypeGuards.isMultimodalContent = isMultimodalContent;
    function isGraphError(value) {
        return (value instanceof graph_types_2.GraphError ||
            (isNapiObject(value) && value.type === 'Error'));
    }
    GraphTypesTypeGuards.isGraphError = isGraphError;
    /**
     * Helper function to check if a value is a NAPI-formatted object with a type
     * field
     * @internal
     */
    function isNapiObject(value) {
        return (typeof value === 'object' &&
            value !== null &&
            typeof value.type === 'string');
    }
    /**
     * Helper function to check if a value is one of the known typed classes or
     * NAPI objects
     * @internal
     */
    function hasKnownType(value) {
        const isKnownInstance = value instanceof classification_result_1.ClassificationResult ||
            value instanceof graph_types_1.Content ||
            value instanceof goal_advancement_response_1.GoalAdvancementResponse ||
            value instanceof knowledge_records_1.KnowledgeRecords ||
            value instanceof list_tools_response_1.ListToolsResponse ||
            value instanceof llm_chat_request_1.LLMChatRequest ||
            value instanceof llm_chat_response_1.LLMChatResponse ||
            value instanceof tts_request_1.TTSRequest ||
            value instanceof matched_intents_1.MatchedIntents ||
            value instanceof matched_keywords_1.MatchedKeywords ||
            value instanceof safety_result_1.SafetyResult ||
            value instanceof tool_call_response_1.ToolCallResponse ||
            value instanceof stream_1.TextStream ||
            value instanceof stream_1.ContentStream ||
            value instanceof stream_1.TTSOutputStream ||
            value instanceof stream_1.SpeechChunkStream ||
            value instanceof stream_1.AudioChunkStream ||
            value instanceof audio_1.Audio ||
            value instanceof multimodal_content_1.MultimodalContent ||
            value instanceof custom_1.CustomData;
        const isKnownNapiType = isNapiObject(value) &&
            [
                'ClassificationResult',
                'Content',
                'GoalAdvancement',
                'KnowledgeRecords',
                'ListToolsResponse',
                'LLMChatRequest',
                'LLMChatResponse',
                'TTSRequest',
                'MatchedIntents',
                'MatchedKeywords',
                'SafetyResult',
                'MemoryState',
                'ToolCallResponse',
                'TextStream',
                'ContentStream',
                'TTSOutputStream',
                'SpeechChunkStream',
                'AudioChunkStream',
                'Audio',
                'MultimodalContent',
            ].includes(value.type);
        return isKnownInstance || isKnownNapiType;
    }
})(GraphTypesTypeGuards || (exports.GraphTypesTypeGuards = GraphTypesTypeGuards = {}));
/**
 * Helper function to get a human-readable type name for CustomNodeInputTypes
 */
function getTypeName(value) {
    if (GraphTypesTypeGuards.isString(value))
        return 'string';
    if (typeof value === 'object' &&
        value !== null &&
        'type' in value &&
        typeof value.type === 'string') {
        const napiType = value.type;
        switch (napiType) {
            case 'ClassificationResult':
                return 'ClassificationResult';
            case 'Content':
                return 'Content';
            case 'GoalAdvancement':
                return 'GoalAdvancement';
            case 'KnowledgeRecords':
                return 'KnowledgeRecords';
            case 'ListToolsResponse':
                return 'ListToolsResponse';
            case 'LLMChatRequest':
                return 'LLMChatRequest';
            case 'LLMChatResponse':
                return 'LLMChatResponse';
            case 'TTSRequest':
                return 'TTSRequest';
            case 'MatchedIntents':
                return 'MatchedIntents';
            case 'MatchedKeywords':
                return 'MatchedKeywords';
            case 'SafetyResult':
                return 'SafetyResult';
            case 'MemoryState':
                return 'MemoryState';
            case 'ToolCallResponse':
                return 'ToolCallResponse';
            case 'TextStream':
                return 'TextStream';
            case 'ContentStream':
                return 'ContentStream';
            case 'TTSOutputStream':
                return 'TTSOutputStream';
            case 'SpeechChunkStream':
                return 'SpeechChunkStream';
            case 'AudioChunkStream':
                return 'AudioChunkStream';
            default:
                return napiType;
        }
    }
    // Legacy class instance checks (fallback)
    if (GraphTypesTypeGuards.isClassificationResult(value))
        return 'ClassificationResult';
    if (GraphTypesTypeGuards.isContent(value))
        return 'Content';
    if (GraphTypesTypeGuards.isGoalAdvancement(value))
        return 'GoalAdvancement';
    if (GraphTypesTypeGuards.isKnowledgeRecords(value))
        return 'KnowledgeRecords';
    if (GraphTypesTypeGuards.isListToolsResponse(value))
        return 'ListToolsResponse';
    if (GraphTypesTypeGuards.isLLMChatRequest(value))
        return 'LLMChatRequest';
    if (GraphTypesTypeGuards.isLLMChatResponse(value))
        return 'LLMChatResponse';
    if (GraphTypesTypeGuards.isTTSRequest(value))
        return 'TTSRequest';
    if (GraphTypesTypeGuards.isMatchedIntents(value))
        return 'MatchedIntents';
    if (GraphTypesTypeGuards.isMatchedKeywords(value))
        return 'MatchedKeywords';
    if (GraphTypesTypeGuards.isSafetyResult(value))
        return 'SafetyResult';
    if (GraphTypesTypeGuards.isToolCallResponse(value))
        return 'ToolCallResponse';
    if (GraphTypesTypeGuards.isTextStream(value))
        return 'TextStream';
    if (GraphTypesTypeGuards.isContentStream(value))
        return 'ContentStream';
    if (GraphTypesTypeGuards.isTTSOutputStream(value))
        return 'TTSOutputStream';
    if (GraphTypesTypeGuards.isSpeechChunkStream(value))
        return 'SpeechChunkStream';
    if (GraphTypesTypeGuards.isAudioChunkStream(value))
        return 'AudioChunkStream';
    if (GraphTypesTypeGuards.isCustom(value))
        return 'Custom';
    if (GraphTypesTypeGuards.isGraphError(value))
        return 'Error';
    else {
        throw new inworld_1.InworldError(`Unknown type: ${value}`);
    }
}
/**
 * Type guard to check if a value is a non-nullish serializable JS type
 */
function isJsNonNullishSerializable(value) {
    if (value === null || value === undefined)
        return false;
    const type = typeof value;
    if (type === 'string' ||
        type === 'number' ||
        type === 'boolean' ||
        type === 'bigint') {
        return true;
    }
    if (Array.isArray(value)) {
        return true;
    }
    return type === 'object';
}
