// @inworld/runtime/build/internal/generated/pb-bindings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PROTO_TYPE_FULL_NAMES = void 0;
exports.createProtoTypeBindings = createProtoTypeBindings;
exports.PROTO_TYPE_FULL_NAMES = {
    AECFilterCreationConfig: 'ai.inworld.runtime.v1.library.speech.AECFilterCreationConfig',
    Audio: 'ai.inworld.runtime.v1.graphs.schema.Audio',
    AudioFrame: 'ai.inworld.runtime.v1.library.speech.AudioChunk',
    ClassificationResult: 'ai.inworld.runtime.v1.graphs.schema.ClassificationResult',
    ClassifyTextRequest: 'ai.inworld.runtime.v1.library.nlu.ClassifyTextRequest',
    ConfigsConfigureTelemetryRequest: 'ai.inworld.runtime.v1.library.telemetry.ConfigureTelemetryRequest',
    ConfigsCounterMetric: 'ai.inworld.runtime.v1.library.telemetry.CounterMetric',
    ConfigsCreateEnvRequest: 'ai.inworld.runtime.v1.graphs.spec.CreateEnvRequest',
    ConfigsCreateGraphRequest: 'ai.inworld.runtime.v1.graphs.spec.CreateGraphRequest',
    ConfigsCredentials: 'ai.inworld.runtime.v1.library.core.Credentials',
    ConfigsDequeueSpanTracesResponse: 'ai.inworld.runtime.v1.library.telemetry.DequeueSpanTracesResponse',
    ConfigsEmbedRequest: 'ai.inworld.runtime.v1.library.embeddings.EmbedRequest',
    ConfigsExecutionContext: 'ai.inworld.runtime.v1.graphs.execution.ExecutionContext',
    ConfigsFeatureFlagConfig: 'ai.inworld.runtime.v1.graphs.spec.FeatureFlagConfig',
    ConfigsGaugeMetric: 'ai.inworld.runtime.v1.library.telemetry.GaugeMetric',
    ConfigsGraphCreationConfig: 'ai.inworld.runtime.v1.graphs.spec.GraphCreationConfig',
    ConfigsHistogramMetric: 'ai.inworld.runtime.v1.library.telemetry.HistogramMetric',
    ConfigsLoggerConfig: 'ai.inworld.runtime.v1.library.telemetry.LoggerConfig',
    ConfigsLogMessageRequest: 'ai.inworld.runtime.v1.library.telemetry.LogMessageRequest',
    ConfigsMetricConfig: 'ai.inworld.runtime.v1.library.telemetry.MetricConfig',
    ConfigsMetricsConfig: 'ai.inworld.runtime.v1.library.telemetry.MetricsConfig',
    ConfigsRecordMetricRequest: 'ai.inworld.runtime.v1.library.telemetry.RecordMetricRequest',
    ConfigsSimpleTelemetryConfig: 'ai.inworld.runtime.v1.library.telemetry.SimpleTelemetryConfig',
    ConfigsSpanData: 'ai.inworld.runtime.v1.library.telemetry.SpanData',
    ConfigsSpanEvent: 'ai.inworld.runtime.v1.library.telemetry.SpanEvent',
    ConfigsStartExecutionRequest: 'ai.inworld.runtime.v1.graphs.execution.StartExecutionRequest',
    ConfigsStreamRecognizeSpeechRequest: 'ai.inworld.runtime.v1.library.speech.StreamRecognizeSpeechRequest',
    ConfigsStreamSpeechRecognitionConfig: 'ai.inworld.runtime.v1.library.speech.StreamSpeechRecognitionConfig',
    ConfigsStreamSpeechRecognitionResponse: 'ai.inworld.runtime.v1.library.speech.StreamSpeechRecognitionResponse',
    ConfigsTelemetryConfig: 'ai.inworld.runtime.v1.library.telemetry.TelemetryConfig',
    ConfigsThreadPoolConfig: 'ai.inworld.runtime.v1.graphs.spec.ThreadPoolConfig',
    ConfigsTracerConfig: 'ai.inworld.runtime.v1.library.telemetry.TracerConfig',
    ConfigsUserContext: 'ai.inworld.runtime.v1.graphs.execution.UserContext',
    ContentData: 'ai.inworld.runtime.v1.graphs.schema.ContentData',
    DataContainer: 'ai.inworld.runtime.v1.graphs.schema.DataContainer',
    DataCustomData: 'ai.inworld.runtime.v1.graphs.schema.CustomData',
    DataStreamWithMetadata: 'ai.inworld.runtime.v1.graphs.schema.DataStreamWithMetadata',
    DetectSilenceRequest: 'ai.inworld.runtime.v1.library.speech.DetectSilenceRequest',
    DetectTurnCompletionRequest: 'ai.inworld.runtime.v1.library.speech.DetectTurnCompletionRequest',
    DetectVoiceActivityRequest: 'ai.inworld.runtime.v1.library.speech.DetectVoiceActivityRequest',
    Device: 'ai.inworld.runtime.v1.library.platform.Device',
    EmbedBatchRequest: 'ai.inworld.runtime.v1.library.embeddings.EmbedBatchRequest',
    EmbedBatchResponse: 'ai.inworld.runtime.v1.library.embeddings.EmbedBatchResponse',
    EmbedResponse: 'ai.inworld.runtime.v1.library.embeddings.EmbedResponse',
    EntityMatch: 'ai.inworld.runtime.v1.library.nlu.EntityMatch',
    Error: 'ai.inworld.runtime.v1.graphs.schema.Error',
    ExtractEntitiesResponse: 'ai.inworld.runtime.v1.library.nlu.ExtractEntitiesResponse',
    File: 'ai.inworld.runtime.v1.library.knowledge.File',
    FilterAudioRequest: 'ai.inworld.runtime.v1.library.speech.FilterAudioRequest',
    GenerateContentChatRequest: 'ai.inworld.runtime.v1.library.llm.GenerateContentChatRequest',
    GenerateContentRequest: 'ai.inworld.runtime.v1.library.llm.GenerateContentRequest',
    GoalAdvancementRequest: 'ai.inworld.runtime.v1.graphs.schema.GoalAdvancementRequest',
    GoalAdvancementResponse: 'ai.inworld.runtime.v1.graphs.schema.GoalAdvancementResponse',
    GoogleAnnotation: 'google.protobuf.GeneratedCodeInfo.Annotation',
    GoogleDeclaration: 'google.protobuf.ExtensionRangeOptions.Declaration',
    GoogleDescriptorProto: 'google.protobuf.DescriptorProto',
    GoogleDuration: 'google.protobuf.Duration',
    GoogleEditionDefault: 'google.protobuf.FieldOptions.EditionDefault',
    GoogleEnumDescriptorProto: 'google.protobuf.EnumDescriptorProto',
    GoogleEnumOptions: 'google.protobuf.EnumOptions',
    GoogleEnumReservedRange: 'google.protobuf.EnumDescriptorProto.EnumReservedRange',
    GoogleEnumValueDescriptorProto: 'google.protobuf.EnumValueDescriptorProto',
    GoogleEnumValueOptions: 'google.protobuf.EnumValueOptions',
    GoogleExtensionRange: 'google.protobuf.DescriptorProto.ExtensionRange',
    GoogleExtensionRangeOptions: 'google.protobuf.ExtensionRangeOptions',
    GoogleFeatureSet: 'google.protobuf.FeatureSet',
    GoogleFeatureSetDefaults: 'google.protobuf.FeatureSetDefaults',
    GoogleFeatureSetEditionDefault: 'google.protobuf.FeatureSetDefaults.FeatureSetEditionDefault',
    GoogleFeatureSupport: 'google.protobuf.FieldOptions.FeatureSupport',
    GoogleFieldDescriptorProto: 'google.protobuf.FieldDescriptorProto',
    GoogleFieldOptions: 'google.protobuf.FieldOptions',
    GoogleFileDescriptorProto: 'google.protobuf.FileDescriptorProto',
    GoogleFileDescriptorSet: 'google.protobuf.FileDescriptorSet',
    GoogleFileOptions: 'google.protobuf.FileOptions',
    GoogleGeneratedCodeInfo: 'google.protobuf.GeneratedCodeInfo',
    GoogleListValue: 'google.protobuf.ListValue',
    GoogleLocation: 'google.protobuf.SourceCodeInfo.Location',
    GoogleMessageOptions: 'google.protobuf.MessageOptions',
    GoogleMethodDescriptorProto: 'google.protobuf.MethodDescriptorProto',
    GoogleMethodOptions: 'google.protobuf.MethodOptions',
    GoogleNamePart: 'google.protobuf.UninterpretedOption.NamePart',
    GoogleOneofDescriptorProto: 'google.protobuf.OneofDescriptorProto',
    GoogleOneofOptions: 'google.protobuf.OneofOptions',
    GoogleProtobufAny: 'google.protobuf.Any',
    GoogleReservedRange: 'google.protobuf.DescriptorProto.ReservedRange',
    GoogleServiceDescriptorProto: 'google.protobuf.ServiceDescriptorProto',
    GoogleServiceOptions: 'google.protobuf.ServiceOptions',
    GoogleSourceCodeInfo: 'google.protobuf.SourceCodeInfo',
    GoogleStruct: 'google.protobuf.Struct',
    GoogleTimestamp: 'google.protobuf.Timestamp',
    GoogleUninterpretedOption: 'google.protobuf.UninterpretedOption',
    GoogleValue: 'google.protobuf.Value',
    GoogleVisibilityFeature: 'google.protobuf.FeatureSet.VisibilityFeature',
    GraphsAddonConfig: 'ai.inworld.runtime.v1.graphs.spec.AddonConfig',
    GraphsAgentAction: 'ai.inworld.runtime.v1.graphs.schema.AgentAction',
    GraphsAgentInfo: 'ai.inworld.runtime.v1.graphs.schema.AgentInfo',
    GraphsAgentSpeech: 'ai.inworld.runtime.v1.graphs.schema.AgentSpeech',
    GraphsBeliefState: 'ai.inworld.runtime.v1.graphs.schema.BeliefState',
    GraphsComponent: 'ai.inworld.runtime.v1.graphs.spec.Component',
    GraphsComponentCreationConfig: 'ai.inworld.runtime.v1.graphs.spec.ComponentCreationConfig',
    GraphsCondition: 'ai.inworld.runtime.v1.graphs.schema.Condition',
    GraphsCooldownConfig: 'ai.inworld.runtime.v1.graphs.execution.CooldownConfig',
    GraphsCustomComponentCreationConfig: 'ai.inworld.runtime.v1.graphs.spec.CustomComponentCreationConfig',
    GraphsCustomNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.execution.CustomNodeExecutionConfig',
    GraphsEmbeddingMatcherNodeCreationConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.EmbeddingMatcherNodeCreationConfig',
    GraphsEmbeddingMatcherNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.EmbeddingMatcherNodeExecutionConfig',
    GraphsEvent: 'ai.inworld.runtime.v1.graphs.schema.Event',
    GraphsFallbackConfig: 'ai.inworld.runtime.v1.graphs.execution.FallbackConfig',
    GraphsGoalAdvancementNodeCreationConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.GoalAdvancementNodeCreationConfig',
    GraphsGoalAdvancementNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.GoalAdvancementNodeExecutionConfig',
    GraphsGoalForEvaluation: 'ai.inworld.runtime.v1.graphs.schema.GoalForEvaluation',
    GraphsGoogleSerperApiSearchNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.GoogleSerperApiSearchNodeExecutionConfig',
    GraphsGoogleSerperApiSearchRequest: 'ai.inworld.runtime.v1.graphs.schema.GoogleSerperApiSearchRequest',
    GraphsGoogleSerperApiSearchResponse: 'ai.inworld.runtime.v1.graphs.schema.GoogleSerperApiSearchResponse',
    GraphsGraphEdge: 'ai.inworld.runtime.v1.graphs.spec.GraphEdge',
    GraphsGraphNode: 'ai.inworld.runtime.v1.graphs.spec.GraphNode',
    GraphsGraphParameter: 'ai.inworld.runtime.v1.graphs.spec.GraphParameter',
    GraphsGraphSpec: 'ai.inworld.runtime.v1.graphs.spec.GraphSpec',
    GraphsIntentNodeCreationConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.IntentNodeCreationConfig',
    GraphsIntentNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.IntentNodeExecutionConfig',
    GraphsIntentsEmbeddingMatcherNodeCreationConfig: 'ai.inworld.runtime.v1.graphs.builtin.graphs.IntentsEmbeddingMatcherNodeCreationConfig',
    GraphsIntentsEmbeddingMatcherNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.graphs.IntentsEmbeddingMatcherNodeExecutionConfig',
    GraphsIntentsLLMPromptVariablesBuilderNodeCreationConfig: 'ai.inworld.runtime.v1.graphs.builtin.graphs.IntentsLLMPromptVariablesBuilderNodeCreationConfig',
    GraphsIntentsLLMPromptVariablesBuilderNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.graphs.IntentsLLMPromptVariablesBuilderNodeExecutionConfig',
    GraphsIntentsOutputNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.graphs.IntentsOutputNodeExecutionConfig',
    GraphsIntentsStrictMatcherNodeCreationConfig: 'ai.inworld.runtime.v1.graphs.builtin.graphs.IntentsStrictMatcherNodeCreationConfig',
    GraphsInteractionEvent: 'ai.inworld.runtime.v1.graphs.schema.InteractionEvent',
    GraphsKeywordMatcherNodeCreationConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.KeywordMatcherNodeCreationConfig',
    GraphsKnowledgeGraph: 'ai.inworld.runtime.v1.graphs.schema.GoogleSerperApiSearchResponse.KnowledgeGraph',
    GraphsKnowledgeNodeCreationConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.KnowledgeNodeCreationConfig',
    GraphsKnowledgeNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.KnowledgeNodeExecutionConfig',
    GraphsLLMChatAttempt: 'ai.inworld.runtime.v1.graphs.schema.LLMChatAttempt',
    GraphsLLMChatDefaults: 'ai.inworld.runtime.v1.graphs.schema.LLMChatDefaults',
    GraphsLLMChatMetadata: 'ai.inworld.runtime.v1.graphs.schema.LLMChatMetadata',
    GraphsLLMChatNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.LLMChatNodeExecutionConfig',
    GraphsLLMChatRoute: 'ai.inworld.runtime.v1.graphs.schema.LLMChatRoute',
    GraphsLLMChatRoutingNodeCreationConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.LLMChatRoutingNodeCreationConfig',
    GraphsLLMChatRoutingNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.LLMChatRoutingNodeExecutionConfig',
    GraphsLLMChatRoutingStrategy: 'ai.inworld.runtime.v1.graphs.schema.LLMChatRoutingStrategy',
    GraphsLLMChatWeightedLoadBalancingStrategy: 'ai.inworld.runtime.v1.graphs.schema.LLMChatWeightedLoadBalancingStrategy',
    GraphsLLMChatWeightedRoute: 'ai.inworld.runtime.v1.graphs.schema.LLMChatWeightedRoute',
    GraphsLLMCompletionNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.LLMCompletionNodeExecutionConfig',
    GraphsLLMCompletionRequest: 'ai.inworld.runtime.v1.graphs.schema.LLMCompletionRequest',
    GraphsLLMSelectionConfig: 'ai.inworld.runtime.v1.graphs.schema.LLMSelectionConfig',
    GraphsLoggerConfig: 'ai.inworld.runtime.v1.graphs.spec.LoggerConfig',
    GraphsMainGraph: 'ai.inworld.runtime.v1.graphs.spec.MainGraph',
    GraphsMatchedPhrases: 'ai.inworld.runtime.v1.graphs.schema.MatchedPhrases',
    GraphsMCPCallToolNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.MCPCallToolNodeExecutionConfig',
    GraphsMCPListToolsNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.MCPListToolsNodeExecutionConfig',
    GraphsNodeCreationConfig: 'ai.inworld.runtime.v1.graphs.spec.NodeCreationConfig',
    GraphsNodeErrorConfig: 'ai.inworld.runtime.v1.graphs.execution.NodeErrorConfig',
    GraphsNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.execution.NodeExecutionConfig',
    GraphsOrganicResult: 'ai.inworld.runtime.v1.graphs.schema.GoogleSerperApiSearchResponse.OrganicResult',
    GraphsPeopleAlsoAsk: 'ai.inworld.runtime.v1.graphs.schema.GoogleSerperApiSearchResponse.PeopleAlsoAsk',
    GraphsPhraseMatch: 'ai.inworld.runtime.v1.graphs.schema.PhraseMatch',
    GraphsProcessContext: 'ai.inworld.runtime.v1.graphs.execution.ProcessContext',
    GraphsQueryEvent: 'ai.inworld.runtime.v1.graphs.schema.GoalAdvancementRequest.QueryEvent',
    GraphsRandomCannedTextNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.RandomCannedTextNodeExecutionConfig',
    GraphsRealtimeAgentInputSlicerNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.graphs.RealtimeAgentInputSlicerNodeExecutionConfig',
    GraphsRegularNode: 'ai.inworld.runtime.v1.graphs.spec.RegularNode',
    GraphsResponseEvent: 'ai.inworld.runtime.v1.graphs.schema.GoalAdvancementRequest.ResponseEvent',
    GraphsRetryConfig: 'ai.inworld.runtime.v1.graphs.execution.RetryConfig',
    GraphsSitelink: 'ai.inworld.runtime.v1.graphs.schema.GoogleSerperApiSearchResponse.OrganicResult.Sitelink',
    GraphsSortCriteria: 'ai.inworld.runtime.v1.graphs.schema.SortCriteria',
    GraphsSTTNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.STTNodeExecutionConfig',
    GraphsSTTRequest: 'ai.inworld.runtime.v1.graphs.schema.STTRequest',
    GraphsSubgraph: 'ai.inworld.runtime.v1.graphs.spec.Subgraph',
    GraphsSubgraphNode: 'ai.inworld.runtime.v1.graphs.spec.SubgraphNode',
    GraphsTelemetryConfig: 'ai.inworld.runtime.v1.graphs.spec.TelemetryConfig',
    GraphsTextChunkingNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.TextChunkingNodeExecutionConfig',
    GraphsTextClassifierNodeCreationConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.TextClassifierNodeCreationConfig',
    GraphsTextClassifierNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.TextClassifierNodeExecutionConfig',
    GraphsTimestamp: 'ai.inworld.runtime.v1.graphs.schema.TTSOutput.Timestamp',
    GraphsTopStory: 'ai.inworld.runtime.v1.graphs.schema.GoogleSerperApiSearchResponse.TopStory',
    GraphsTransformNodeCreationConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.TransformNodeCreationConfig',
    GraphsTriggerEvent: 'ai.inworld.runtime.v1.graphs.schema.GoalAdvancementRequest.TriggerEvent',
    GraphsTTSNodeExecutionConfig: 'ai.inworld.runtime.v1.graphs.builtin.nodes.TTSNodeExecutionConfig',
    GraphsTTSOutput: 'ai.inworld.runtime.v1.graphs.schema.TTSOutput',
    GraphsUsage: 'ai.inworld.runtime.v1.graphs.schema.Usage',
    IntentMatch: 'ai.inworld.runtime.v1.graphs.schema.IntentMatch',
    Json: 'ai.inworld.runtime.v1.graphs.schema.Json',
    KeywordGroup: 'ai.inworld.runtime.v1.library.nlu.KeywordGroup',
    KeywordMatch: 'ai.inworld.runtime.v1.graphs.schema.KeywordMatch',
    KeywordMatcherResponse: 'ai.inworld.runtime.v1.library.nlu.KeywordMatcherResponse',
    KnowledgeCompilationRequest: 'ai.inworld.runtime.v1.library.knowledge.CompileKnowledgeRequest',
    KnowledgeCreationConfig: 'ai.inworld.runtime.v1.library.knowledge.KnowledgeCreationConfig',
    KnowledgeGetRequest: 'ai.inworld.runtime.v1.library.knowledge.GetKnowledgeRequest',
    KnowledgeRecords: 'ai.inworld.runtime.v1.graphs.schema.KnowledgeRecords',
    LibraryAcousticEchoCancellationConfig: 'ai.inworld.runtime.v1.library.speech.AcousticEchoCancellationConfig',
    LibraryAudioChunkTimestamp: 'ai.inworld.runtime.v1.library.speech.AudioChunkTimestamp',
    LibraryBatchProcessingOptions: 'ai.inworld.runtime.v1.library.telemetry.BatchProcessingOptions',
    LibraryClassThreshold: 'ai.inworld.runtime.v1.library.nlu.ClassThreshold',
    LibraryCompileKnowledgeResponse: 'ai.inworld.runtime.v1.library.knowledge.CompileKnowledgeResponse',
    LibraryContentItem: 'ai.inworld.runtime.v1.library.llm.ContentItem',
    LibraryDeviceInfo: 'ai.inworld.runtime.v1.library.platform.DeviceInfo',
    LibraryDeviceSelection: 'ai.inworld.runtime.v1.library.platform.DeviceSelection',
    LibraryDictionaryRule: 'ai.inworld.runtime.v1.library.nlu.DictionaryRule',
    LibraryElevenLabsQueryParams: 'ai.inworld.runtime.v1.library.speech.ElevenLabsQueryParams',
    LibraryElevenLabsRequestParams: 'ai.inworld.runtime.v1.library.speech.ElevenLabsRequestParams',
    LibraryElevenLabsSpeechSynthesisConfig: 'ai.inworld.runtime.v1.library.speech.ElevenLabsSpeechSynthesisConfig',
    LibraryEmbeddingMatcherConfig: 'ai.inworld.runtime.v1.library.nlu.EmbeddingMatcherConfig',
    LibraryEntity: 'ai.inworld.runtime.v1.library.nlu.Entity',
    LibraryFunctionChoice: 'ai.inworld.runtime.v1.library.llm.FunctionChoice',
    LibraryGetKnowledgeResponse: 'ai.inworld.runtime.v1.library.knowledge.GetKnowledgeResponse',
    LibraryGetModelsRequest: 'ai.inworld.runtime.v1.library.llm.GetModelsRequest',
    LibraryGetModelsResponse: 'ai.inworld.runtime.v1.library.llm.GetModelsResponse',
    LibraryGoogleSerperAPIClientCreationConfig: 'ai.inworld.runtime.v1.library.serp.GoogleSerperAPIClientCreationConfig',
    LibraryGoogleSerperApiSearchRequest: 'ai.inworld.runtime.v1.library.serp.GoogleSerperApiSearchRequest',
    LibraryGoogleSerperApiSearchResponse: 'ai.inworld.runtime.v1.library.serp.GoogleSerperApiSearchResponse',
    LibraryImageUrl: 'ai.inworld.runtime.v1.library.llm.ImageUrl',
    LibraryIntent: 'ai.inworld.runtime.v1.library.nlu.Intent',
    LibraryIntentMatcherConfig: 'ai.inworld.runtime.v1.library.nlu.IntentMatcherConfig',
    LibraryInworldSpeechSynthesisConfig: 'ai.inworld.runtime.v1.library.speech.InworldSpeechSynthesisConfig',
    LibraryKeywordMatch: 'ai.inworld.runtime.v1.library.nlu.KeywordMatch',
    LibraryKnowledgeCompileConfig: 'ai.inworld.runtime.v1.library.knowledge.KnowledgeCompileConfig',
    LibraryKnowledgeGraph: 'ai.inworld.runtime.v1.library.serp.GoogleSerperApiSearchResponse.KnowledgeGraph',
    LibraryKnowledgeRecords: 'ai.inworld.runtime.v1.library.knowledge.KnowledgeRecords',
    LibraryLanguage: 'ai.inworld.runtime.v1.library.core.Language',
    LibraryLLMBenchmarks: 'ai.inworld.runtime.v1.library.llm.LLMBenchmarks',
    LibraryLlmMatcherConfig: 'ai.inworld.runtime.v1.library.nlu.LlmMatcherConfig',
    LibraryLLMPerformance: 'ai.inworld.runtime.v1.library.llm.LLMPerformance',
    LibraryLLMPricing: 'ai.inworld.runtime.v1.library.llm.LLMPricing',
    LibraryLLMSpec: 'ai.inworld.runtime.v1.library.llm.LLMSpec',
    LibraryLocalAECFilterConfig: 'ai.inworld.runtime.v1.library.speech.LocalAECFilterConfig',
    LibraryLocalKnowledgeConfig: 'ai.inworld.runtime.v1.library.knowledge.LocalKnowledgeConfig',
    LibraryLocalLLMConfig: 'ai.inworld.runtime.v1.library.llm.LocalLLMConfig',
    LibraryLocalStreamingSTTConfig: 'ai.inworld.runtime.v1.library.speech.LocalStreamingSTTConfig',
    LibraryLocalSTTConfig: 'ai.inworld.runtime.v1.library.speech.LocalSTTConfig',
    LibraryLocalTextEmbedderConfig: 'ai.inworld.runtime.v1.library.embeddings.LocalTextEmbedderConfig',
    LibraryLocalTurnDetectorConfig: 'ai.inworld.runtime.v1.library.speech.LocalTurnDetectorConfig',
    LibraryLocalVADConfig: 'ai.inworld.runtime.v1.library.speech.LocalVADConfig',
    LibraryLogitBias: 'ai.inworld.runtime.v1.library.llm.LogitBias',
    LibraryMCPHttpConfig: 'ai.inworld.runtime.v1.library.mcp.MCPHttpConfig',
    LibraryMCPSessionCreationConfig: 'ai.inworld.runtime.v1.library.mcp.MCPSessionCreationConfig',
    LibraryMCPStdioConfig: 'ai.inworld.runtime.v1.library.mcp.MCPStdioConfig',
    LibraryModelId: 'ai.inworld.runtime.v1.library.llm.ModelId',
    LibraryModelInfo: 'ai.inworld.runtime.v1.library.llm.ModelInfo',
    LibraryOrganicResult: 'ai.inworld.runtime.v1.library.serp.GoogleSerperApiSearchResponse.OrganicResult',
    LibraryParagonApiConfig: 'ai.inworld.runtime.v1.library.mcp.ParagonApiConfig',
    LibraryParsingConfig: 'ai.inworld.runtime.v1.library.knowledge.ParsingConfig',
    LibraryPeopleAlsoAsk: 'ai.inworld.runtime.v1.library.serp.GoogleSerperApiSearchResponse.PeopleAlsoAsk',
    LibraryReasoningConfig: 'ai.inworld.runtime.v1.library.llm.ReasoningConfig',
    LibraryRemoteKnowledgeConfig: 'ai.inworld.runtime.v1.library.knowledge.RemoteKnowledgeConfig',
    LibraryRemoteLLMConfig: 'ai.inworld.runtime.v1.library.llm.RemoteLLMConfig',
    LibraryRemoteStreamingSTTConfig: 'ai.inworld.runtime.v1.library.speech.RemoteStreamingSTTConfig',
    LibraryRemoteSTTConfig: 'ai.inworld.runtime.v1.library.speech.RemoteSTTConfig',
    LibraryRemoteTextEmbedderConfig: 'ai.inworld.runtime.v1.library.embeddings.RemoteTextEmbedderConfig',
    LibraryRemoteTTSConfig: 'ai.inworld.runtime.v1.library.speech.RemoteTTSConfig',
    LibraryRemoveKnowledgeRequest: 'ai.inworld.runtime.v1.library.knowledge.RemoveKnowledgeRequest',
    LibraryRetrievalConfig: 'ai.inworld.runtime.v1.library.knowledge.RetrievalConfig',
    LibrarySitelink: 'ai.inworld.runtime.v1.library.serp.GoogleSerperApiSearchResponse.OrganicResult.Sitelink',
    LibrarySpeechRecognitionConfig: 'ai.inworld.runtime.v1.library.speech.SpeechRecognitionConfig',
    LibrarySpeechSynthesisConfig: 'ai.inworld.runtime.v1.library.speech.SpeechSynthesisConfig',
    LibrarySpeechSynthesisInferenceConfig: 'ai.inworld.runtime.v1.library.speech.SpeechSynthesisInferenceConfig',
    LibrarySpeechSynthesisPostprocessingConfig: 'ai.inworld.runtime.v1.library.speech.SpeechSynthesisPostprocessingConfig',
    LibraryTextClassifierConfig: 'ai.inworld.runtime.v1.library.nlu.TextClassifierConfig',
    LibraryTextGenerationConfig: 'ai.inworld.runtime.v1.library.llm.TextGenerationConfig',
    LibraryToolCallResult: 'ai.inworld.runtime.v1.library.llm.ToolCallResult',
    LibraryToolChoice: 'ai.inworld.runtime.v1.library.llm.ToolChoice',
    LibraryTopStory: 'ai.inworld.runtime.v1.library.serp.GoogleSerperApiSearchResponse.TopStory',
    LibraryTurnDetectionConfig: 'ai.inworld.runtime.v1.library.speech.TurnDetectionConfig',
    LibraryVoice: 'ai.inworld.runtime.v1.library.speech.Voice',
    ListToolsResponse: 'ai.inworld.runtime.v1.graphs.schema.ListToolsResponse',
    LLMChatRequest: 'ai.inworld.runtime.v1.graphs.schema.LLMChatRequest',
    LLMChatResponse: 'ai.inworld.runtime.v1.graphs.schema.LLMChatResponse',
    LLMChatRoutingRequest: 'ai.inworld.runtime.v1.graphs.schema.LLMChatRoutingRequest',
    LLMCompletionResponse: 'ai.inworld.runtime.v1.graphs.schema.LLMCompletionResponse',
    LLMCreationConfig: 'ai.inworld.runtime.v1.library.llm.LLMCreationConfig',
    LLMMessage: 'ai.inworld.runtime.v1.library.llm.Message',
    MatchedIntents: 'ai.inworld.runtime.v1.graphs.schema.MatchedIntents',
    MatchedKeywords: 'ai.inworld.runtime.v1.graphs.schema.MatchedKeywords',
    MCPClientCreationConfig: 'ai.inworld.runtime.v1.library.mcp.MCPClientCreationConfig',
    MultimodalContent: 'ai.inworld.runtime.v1.graphs.schema.MultimodalContent',
    NERCreationConfig: 'ai.inworld.runtime.v1.library.nlu.NERCreationConfig',
    RecognizeSpeechRequest: 'ai.inworld.runtime.v1.library.speech.RecognizeSpeechRequest',
    SafetyResult: 'ai.inworld.runtime.v1.graphs.schema.SafetyResult',
    StreamingSTTCreationConfig: 'ai.inworld.runtime.v1.library.speech.StreamingSTTCreationConfig',
    STTCreationConfig: 'ai.inworld.runtime.v1.library.speech.STTCreationConfig',
    SynthesizeSpeechRequest: 'ai.inworld.runtime.v1.library.speech.SynthesizeSpeechRequest',
    Text: 'ai.inworld.runtime.v1.graphs.schema.Text',
    TextClassifierCreationConfig: 'ai.inworld.runtime.v1.library.nlu.TextClassifierCreationConfig',
    TextClassifierResponse: 'ai.inworld.runtime.v1.library.nlu.ClassifyTextResponse',
    TextEmbedderCreationConfig: 'ai.inworld.runtime.v1.library.embeddings.TextEmbedderCreationConfig',
    Tool: 'ai.inworld.runtime.v1.library.llm.Tool',
    ToolCall: 'ai.inworld.runtime.v1.library.llm.ToolCall',
    ToolCallRequest: 'ai.inworld.runtime.v1.graphs.schema.ToolCallRequest',
    ToolCallResponse: 'ai.inworld.runtime.v1.graphs.schema.ToolCallResponse',
    ToolCallResult: 'ai.inworld.runtime.v1.graphs.schema.ToolCallResult',
    ToolList: 'ai.inworld.runtime.v1.library.llm.ToolList',
    TTSCreationConfig: 'ai.inworld.runtime.v1.library.speech.TTSCreationConfig',
    TTSRequest: 'ai.inworld.runtime.v1.graphs.schema.TTSRequest',
    TurnDetectorCreationConfig: 'ai.inworld.runtime.v1.library.speech.TurnDetectorCreationConfig',
    TurnPrediction: 'ai.inworld.runtime.v1.library.speech.TurnPrediction',
    VADCreationConfig: 'ai.inworld.runtime.v1.library.speech.VADCreationConfig',
    VoiceActivityDetectionConfig: 'ai.inworld.runtime.v1.library.speech.VoiceActivityDetectionConfig',
};
function createProtoTypeBindings(root) {
    return {
        AECFilterCreationConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.AECFilterCreationConfig'),
        Audio: root.lookupType('ai.inworld.runtime.v1.graphs.schema.Audio'),
        AudioFrame: root.lookupType('ai.inworld.runtime.v1.library.speech.AudioChunk'),
        ClassificationResult: root.lookupType('ai.inworld.runtime.v1.graphs.schema.ClassificationResult'),
        ClassifyTextRequest: root.lookupType('ai.inworld.runtime.v1.library.nlu.ClassifyTextRequest'),
        ConfigsConfigureTelemetryRequest: root.lookupType('ai.inworld.runtime.v1.library.telemetry.ConfigureTelemetryRequest'),
        ConfigsCounterMetric: root.lookupType('ai.inworld.runtime.v1.library.telemetry.CounterMetric'),
        ConfigsCreateEnvRequest: root.lookupType('ai.inworld.runtime.v1.graphs.spec.CreateEnvRequest'),
        ConfigsCreateGraphRequest: root.lookupType('ai.inworld.runtime.v1.graphs.spec.CreateGraphRequest'),
        ConfigsCredentials: root.lookupType('ai.inworld.runtime.v1.library.core.Credentials'),
        ConfigsDequeueSpanTracesResponse: root.lookupType('ai.inworld.runtime.v1.library.telemetry.DequeueSpanTracesResponse'),
        ConfigsEmbedRequest: root.lookupType('ai.inworld.runtime.v1.library.embeddings.EmbedRequest'),
        ConfigsExecutionContext: root.lookupType('ai.inworld.runtime.v1.graphs.execution.ExecutionContext'),
        ConfigsFeatureFlagConfig: root.lookupType('ai.inworld.runtime.v1.graphs.spec.FeatureFlagConfig'),
        ConfigsGaugeMetric: root.lookupType('ai.inworld.runtime.v1.library.telemetry.GaugeMetric'),
        ConfigsGraphCreationConfig: root.lookupType('ai.inworld.runtime.v1.graphs.spec.GraphCreationConfig'),
        ConfigsHistogramMetric: root.lookupType('ai.inworld.runtime.v1.library.telemetry.HistogramMetric'),
        ConfigsLoggerConfig: root.lookupType('ai.inworld.runtime.v1.library.telemetry.LoggerConfig'),
        ConfigsLogMessageRequest: root.lookupType('ai.inworld.runtime.v1.library.telemetry.LogMessageRequest'),
        ConfigsMetricConfig: root.lookupType('ai.inworld.runtime.v1.library.telemetry.MetricConfig'),
        ConfigsMetricsConfig: root.lookupType('ai.inworld.runtime.v1.library.telemetry.MetricsConfig'),
        ConfigsRecordMetricRequest: root.lookupType('ai.inworld.runtime.v1.library.telemetry.RecordMetricRequest'),
        ConfigsSimpleTelemetryConfig: root.lookupType('ai.inworld.runtime.v1.library.telemetry.SimpleTelemetryConfig'),
        ConfigsSpanData: root.lookupType('ai.inworld.runtime.v1.library.telemetry.SpanData'),
        ConfigsSpanEvent: root.lookupType('ai.inworld.runtime.v1.library.telemetry.SpanEvent'),
        ConfigsStartExecutionRequest: root.lookupType('ai.inworld.runtime.v1.graphs.execution.StartExecutionRequest'),
        ConfigsStreamRecognizeSpeechRequest: root.lookupType('ai.inworld.runtime.v1.library.speech.StreamRecognizeSpeechRequest'),
        ConfigsStreamSpeechRecognitionConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.StreamSpeechRecognitionConfig'),
        ConfigsStreamSpeechRecognitionResponse: root.lookupType('ai.inworld.runtime.v1.library.speech.StreamSpeechRecognitionResponse'),
        ConfigsTelemetryConfig: root.lookupType('ai.inworld.runtime.v1.library.telemetry.TelemetryConfig'),
        ConfigsThreadPoolConfig: root.lookupType('ai.inworld.runtime.v1.graphs.spec.ThreadPoolConfig'),
        ConfigsTracerConfig: root.lookupType('ai.inworld.runtime.v1.library.telemetry.TracerConfig'),
        ConfigsUserContext: root.lookupType('ai.inworld.runtime.v1.graphs.execution.UserContext'),
        ContentData: root.lookupType('ai.inworld.runtime.v1.graphs.schema.ContentData'),
        DataContainer: root.lookupType('ai.inworld.runtime.v1.graphs.schema.DataContainer'),
        DataCustomData: root.lookupType('ai.inworld.runtime.v1.graphs.schema.CustomData'),
        DataStreamWithMetadata: root.lookupType('ai.inworld.runtime.v1.graphs.schema.DataStreamWithMetadata'),
        DetectSilenceRequest: root.lookupType('ai.inworld.runtime.v1.library.speech.DetectSilenceRequest'),
        DetectTurnCompletionRequest: root.lookupType('ai.inworld.runtime.v1.library.speech.DetectTurnCompletionRequest'),
        DetectVoiceActivityRequest: root.lookupType('ai.inworld.runtime.v1.library.speech.DetectVoiceActivityRequest'),
        Device: root.lookupType('ai.inworld.runtime.v1.library.platform.Device'),
        EmbedBatchRequest: root.lookupType('ai.inworld.runtime.v1.library.embeddings.EmbedBatchRequest'),
        EmbedBatchResponse: root.lookupType('ai.inworld.runtime.v1.library.embeddings.EmbedBatchResponse'),
        EmbedResponse: root.lookupType('ai.inworld.runtime.v1.library.embeddings.EmbedResponse'),
        EntityMatch: root.lookupType('ai.inworld.runtime.v1.library.nlu.EntityMatch'),
        Error: root.lookupType('ai.inworld.runtime.v1.graphs.schema.Error'),
        ExtractEntitiesResponse: root.lookupType('ai.inworld.runtime.v1.library.nlu.ExtractEntitiesResponse'),
        File: root.lookupType('ai.inworld.runtime.v1.library.knowledge.File'),
        FilterAudioRequest: root.lookupType('ai.inworld.runtime.v1.library.speech.FilterAudioRequest'),
        GenerateContentChatRequest: root.lookupType('ai.inworld.runtime.v1.library.llm.GenerateContentChatRequest'),
        GenerateContentRequest: root.lookupType('ai.inworld.runtime.v1.library.llm.GenerateContentRequest'),
        GoalAdvancementRequest: root.lookupType('ai.inworld.runtime.v1.graphs.schema.GoalAdvancementRequest'),
        GoalAdvancementResponse: root.lookupType('ai.inworld.runtime.v1.graphs.schema.GoalAdvancementResponse'),
        GoogleAnnotation: root.lookupType('google.protobuf.GeneratedCodeInfo.Annotation'),
        GoogleDeclaration: root.lookupType('google.protobuf.ExtensionRangeOptions.Declaration'),
        GoogleDescriptorProto: root.lookupType('google.protobuf.DescriptorProto'),
        GoogleDuration: root.lookupType('google.protobuf.Duration'),
        GoogleEditionDefault: root.lookupType('google.protobuf.FieldOptions.EditionDefault'),
        GoogleEnumDescriptorProto: root.lookupType('google.protobuf.EnumDescriptorProto'),
        GoogleEnumOptions: root.lookupType('google.protobuf.EnumOptions'),
        GoogleEnumReservedRange: root.lookupType('google.protobuf.EnumDescriptorProto.EnumReservedRange'),
        GoogleEnumValueDescriptorProto: root.lookupType('google.protobuf.EnumValueDescriptorProto'),
        GoogleEnumValueOptions: root.lookupType('google.protobuf.EnumValueOptions'),
        GoogleExtensionRange: root.lookupType('google.protobuf.DescriptorProto.ExtensionRange'),
        GoogleExtensionRangeOptions: root.lookupType('google.protobuf.ExtensionRangeOptions'),
        GoogleFeatureSet: root.lookupType('google.protobuf.FeatureSet'),
        GoogleFeatureSetDefaults: root.lookupType('google.protobuf.FeatureSetDefaults'),
        GoogleFeatureSetEditionDefault: root.lookupType('google.protobuf.FeatureSetDefaults.FeatureSetEditionDefault'),
        GoogleFeatureSupport: root.lookupType('google.protobuf.FieldOptions.FeatureSupport'),
        GoogleFieldDescriptorProto: root.lookupType('google.protobuf.FieldDescriptorProto'),
        GoogleFieldOptions: root.lookupType('google.protobuf.FieldOptions'),
        GoogleFileDescriptorProto: root.lookupType('google.protobuf.FileDescriptorProto'),
        GoogleFileDescriptorSet: root.lookupType('google.protobuf.FileDescriptorSet'),
        GoogleFileOptions: root.lookupType('google.protobuf.FileOptions'),
        GoogleGeneratedCodeInfo: root.lookupType('google.protobuf.GeneratedCodeInfo'),
        GoogleListValue: root.lookupType('google.protobuf.ListValue'),
        GoogleLocation: root.lookupType('google.protobuf.SourceCodeInfo.Location'),
        GoogleMessageOptions: root.lookupType('google.protobuf.MessageOptions'),
        GoogleMethodDescriptorProto: root.lookupType('google.protobuf.MethodDescriptorProto'),
        GoogleMethodOptions: root.lookupType('google.protobuf.MethodOptions'),
        GoogleNamePart: root.lookupType('google.protobuf.UninterpretedOption.NamePart'),
        GoogleOneofDescriptorProto: root.lookupType('google.protobuf.OneofDescriptorProto'),
        GoogleOneofOptions: root.lookupType('google.protobuf.OneofOptions'),
        GoogleProtobufAny: root.lookupType('google.protobuf.Any'),
        GoogleReservedRange: root.lookupType('google.protobuf.DescriptorProto.ReservedRange'),
        GoogleServiceDescriptorProto: root.lookupType('google.protobuf.ServiceDescriptorProto'),
        GoogleServiceOptions: root.lookupType('google.protobuf.ServiceOptions'),
        GoogleSourceCodeInfo: root.lookupType('google.protobuf.SourceCodeInfo'),
        GoogleStruct: root.lookupType('google.protobuf.Struct'),
        GoogleTimestamp: root.lookupType('google.protobuf.Timestamp'),
        GoogleUninterpretedOption: root.lookupType('google.protobuf.UninterpretedOption'),
        GoogleValue: root.lookupType('google.protobuf.Value'),
        GoogleVisibilityFeature: root.lookupType('google.protobuf.FeatureSet.VisibilityFeature'),
        GraphsAddonConfig: root.lookupType('ai.inworld.runtime.v1.graphs.spec.AddonConfig'),
        GraphsAgentAction: root.lookupType('ai.inworld.runtime.v1.graphs.schema.AgentAction'),
        GraphsAgentInfo: root.lookupType('ai.inworld.runtime.v1.graphs.schema.AgentInfo'),
        GraphsAgentSpeech: root.lookupType('ai.inworld.runtime.v1.graphs.schema.AgentSpeech'),
        GraphsBeliefState: root.lookupType('ai.inworld.runtime.v1.graphs.schema.BeliefState'),
        GraphsComponent: root.lookupType('ai.inworld.runtime.v1.graphs.spec.Component'),
        GraphsComponentCreationConfig: root.lookupType('ai.inworld.runtime.v1.graphs.spec.ComponentCreationConfig'),
        GraphsCondition: root.lookupType('ai.inworld.runtime.v1.graphs.schema.Condition'),
        GraphsCooldownConfig: root.lookupType('ai.inworld.runtime.v1.graphs.execution.CooldownConfig'),
        GraphsCustomComponentCreationConfig: root.lookupType('ai.inworld.runtime.v1.graphs.spec.CustomComponentCreationConfig'),
        GraphsCustomNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.execution.CustomNodeExecutionConfig'),
        GraphsEmbeddingMatcherNodeCreationConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.EmbeddingMatcherNodeCreationConfig'),
        GraphsEmbeddingMatcherNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.EmbeddingMatcherNodeExecutionConfig'),
        GraphsEvent: root.lookupType('ai.inworld.runtime.v1.graphs.schema.Event'),
        GraphsFallbackConfig: root.lookupType('ai.inworld.runtime.v1.graphs.execution.FallbackConfig'),
        GraphsGoalAdvancementNodeCreationConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.GoalAdvancementNodeCreationConfig'),
        GraphsGoalAdvancementNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.GoalAdvancementNodeExecutionConfig'),
        GraphsGoalForEvaluation: root.lookupType('ai.inworld.runtime.v1.graphs.schema.GoalForEvaluation'),
        GraphsGoogleSerperApiSearchNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.GoogleSerperApiSearchNodeExecutionConfig'),
        GraphsGoogleSerperApiSearchRequest: root.lookupType('ai.inworld.runtime.v1.graphs.schema.GoogleSerperApiSearchRequest'),
        GraphsGoogleSerperApiSearchResponse: root.lookupType('ai.inworld.runtime.v1.graphs.schema.GoogleSerperApiSearchResponse'),
        GraphsGraphEdge: root.lookupType('ai.inworld.runtime.v1.graphs.spec.GraphEdge'),
        GraphsGraphNode: root.lookupType('ai.inworld.runtime.v1.graphs.spec.GraphNode'),
        GraphsGraphParameter: root.lookupType('ai.inworld.runtime.v1.graphs.spec.GraphParameter'),
        GraphsGraphSpec: root.lookupType('ai.inworld.runtime.v1.graphs.spec.GraphSpec'),
        GraphsIntentNodeCreationConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.IntentNodeCreationConfig'),
        GraphsIntentNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.IntentNodeExecutionConfig'),
        GraphsIntentsEmbeddingMatcherNodeCreationConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.graphs.IntentsEmbeddingMatcherNodeCreationConfig'),
        GraphsIntentsEmbeddingMatcherNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.graphs.IntentsEmbeddingMatcherNodeExecutionConfig'),
        GraphsIntentsLLMPromptVariablesBuilderNodeCreationConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.graphs.IntentsLLMPromptVariablesBuilderNodeCreationConfig'),
        GraphsIntentsLLMPromptVariablesBuilderNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.graphs.IntentsLLMPromptVariablesBuilderNodeExecutionConfig'),
        GraphsIntentsOutputNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.graphs.IntentsOutputNodeExecutionConfig'),
        GraphsIntentsStrictMatcherNodeCreationConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.graphs.IntentsStrictMatcherNodeCreationConfig'),
        GraphsInteractionEvent: root.lookupType('ai.inworld.runtime.v1.graphs.schema.InteractionEvent'),
        GraphsKeywordMatcherNodeCreationConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.KeywordMatcherNodeCreationConfig'),
        GraphsKnowledgeGraph: root.lookupType('ai.inworld.runtime.v1.graphs.schema.GoogleSerperApiSearchResponse.KnowledgeGraph'),
        GraphsKnowledgeNodeCreationConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.KnowledgeNodeCreationConfig'),
        GraphsKnowledgeNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.KnowledgeNodeExecutionConfig'),
        GraphsLLMChatAttempt: root.lookupType('ai.inworld.runtime.v1.graphs.schema.LLMChatAttempt'),
        GraphsLLMChatDefaults: root.lookupType('ai.inworld.runtime.v1.graphs.schema.LLMChatDefaults'),
        GraphsLLMChatMetadata: root.lookupType('ai.inworld.runtime.v1.graphs.schema.LLMChatMetadata'),
        GraphsLLMChatNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.LLMChatNodeExecutionConfig'),
        GraphsLLMChatRoute: root.lookupType('ai.inworld.runtime.v1.graphs.schema.LLMChatRoute'),
        GraphsLLMChatRoutingNodeCreationConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.LLMChatRoutingNodeCreationConfig'),
        GraphsLLMChatRoutingNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.LLMChatRoutingNodeExecutionConfig'),
        GraphsLLMChatRoutingStrategy: root.lookupType('ai.inworld.runtime.v1.graphs.schema.LLMChatRoutingStrategy'),
        GraphsLLMChatWeightedLoadBalancingStrategy: root.lookupType('ai.inworld.runtime.v1.graphs.schema.LLMChatWeightedLoadBalancingStrategy'),
        GraphsLLMChatWeightedRoute: root.lookupType('ai.inworld.runtime.v1.graphs.schema.LLMChatWeightedRoute'),
        GraphsLLMCompletionNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.LLMCompletionNodeExecutionConfig'),
        GraphsLLMCompletionRequest: root.lookupType('ai.inworld.runtime.v1.graphs.schema.LLMCompletionRequest'),
        GraphsLLMSelectionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.schema.LLMSelectionConfig'),
        GraphsLoggerConfig: root.lookupType('ai.inworld.runtime.v1.graphs.spec.LoggerConfig'),
        GraphsMainGraph: root.lookupType('ai.inworld.runtime.v1.graphs.spec.MainGraph'),
        GraphsMatchedPhrases: root.lookupType('ai.inworld.runtime.v1.graphs.schema.MatchedPhrases'),
        GraphsMCPCallToolNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.MCPCallToolNodeExecutionConfig'),
        GraphsMCPListToolsNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.MCPListToolsNodeExecutionConfig'),
        GraphsNodeCreationConfig: root.lookupType('ai.inworld.runtime.v1.graphs.spec.NodeCreationConfig'),
        GraphsNodeErrorConfig: root.lookupType('ai.inworld.runtime.v1.graphs.execution.NodeErrorConfig'),
        GraphsNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.execution.NodeExecutionConfig'),
        GraphsOrganicResult: root.lookupType('ai.inworld.runtime.v1.graphs.schema.GoogleSerperApiSearchResponse.OrganicResult'),
        GraphsPeopleAlsoAsk: root.lookupType('ai.inworld.runtime.v1.graphs.schema.GoogleSerperApiSearchResponse.PeopleAlsoAsk'),
        GraphsPhraseMatch: root.lookupType('ai.inworld.runtime.v1.graphs.schema.PhraseMatch'),
        GraphsProcessContext: root.lookupType('ai.inworld.runtime.v1.graphs.execution.ProcessContext'),
        GraphsQueryEvent: root.lookupType('ai.inworld.runtime.v1.graphs.schema.GoalAdvancementRequest.QueryEvent'),
        GraphsRandomCannedTextNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.RandomCannedTextNodeExecutionConfig'),
        GraphsRealtimeAgentInputSlicerNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.graphs.RealtimeAgentInputSlicerNodeExecutionConfig'),
        GraphsRegularNode: root.lookupType('ai.inworld.runtime.v1.graphs.spec.RegularNode'),
        GraphsResponseEvent: root.lookupType('ai.inworld.runtime.v1.graphs.schema.GoalAdvancementRequest.ResponseEvent'),
        GraphsRetryConfig: root.lookupType('ai.inworld.runtime.v1.graphs.execution.RetryConfig'),
        GraphsSitelink: root.lookupType('ai.inworld.runtime.v1.graphs.schema.GoogleSerperApiSearchResponse.OrganicResult.Sitelink'),
        GraphsSortCriteria: root.lookupType('ai.inworld.runtime.v1.graphs.schema.SortCriteria'),
        GraphsSTTNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.STTNodeExecutionConfig'),
        GraphsSTTRequest: root.lookupType('ai.inworld.runtime.v1.graphs.schema.STTRequest'),
        GraphsSubgraph: root.lookupType('ai.inworld.runtime.v1.graphs.spec.Subgraph'),
        GraphsSubgraphNode: root.lookupType('ai.inworld.runtime.v1.graphs.spec.SubgraphNode'),
        GraphsTelemetryConfig: root.lookupType('ai.inworld.runtime.v1.graphs.spec.TelemetryConfig'),
        GraphsTextChunkingNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.TextChunkingNodeExecutionConfig'),
        GraphsTextClassifierNodeCreationConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.TextClassifierNodeCreationConfig'),
        GraphsTextClassifierNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.TextClassifierNodeExecutionConfig'),
        GraphsTimestamp: root.lookupType('ai.inworld.runtime.v1.graphs.schema.TTSOutput.Timestamp'),
        GraphsTopStory: root.lookupType('ai.inworld.runtime.v1.graphs.schema.GoogleSerperApiSearchResponse.TopStory'),
        GraphsTransformNodeCreationConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.TransformNodeCreationConfig'),
        GraphsTriggerEvent: root.lookupType('ai.inworld.runtime.v1.graphs.schema.GoalAdvancementRequest.TriggerEvent'),
        GraphsTTSNodeExecutionConfig: root.lookupType('ai.inworld.runtime.v1.graphs.builtin.nodes.TTSNodeExecutionConfig'),
        GraphsTTSOutput: root.lookupType('ai.inworld.runtime.v1.graphs.schema.TTSOutput'),
        GraphsUsage: root.lookupType('ai.inworld.runtime.v1.graphs.schema.Usage'),
        IntentMatch: root.lookupType('ai.inworld.runtime.v1.graphs.schema.IntentMatch'),
        Json: root.lookupType('ai.inworld.runtime.v1.graphs.schema.Json'),
        KeywordGroup: root.lookupType('ai.inworld.runtime.v1.library.nlu.KeywordGroup'),
        KeywordMatch: root.lookupType('ai.inworld.runtime.v1.graphs.schema.KeywordMatch'),
        KeywordMatcherResponse: root.lookupType('ai.inworld.runtime.v1.library.nlu.KeywordMatcherResponse'),
        KnowledgeCompilationRequest: root.lookupType('ai.inworld.runtime.v1.library.knowledge.CompileKnowledgeRequest'),
        KnowledgeCreationConfig: root.lookupType('ai.inworld.runtime.v1.library.knowledge.KnowledgeCreationConfig'),
        KnowledgeGetRequest: root.lookupType('ai.inworld.runtime.v1.library.knowledge.GetKnowledgeRequest'),
        KnowledgeRecords: root.lookupType('ai.inworld.runtime.v1.graphs.schema.KnowledgeRecords'),
        LibraryAcousticEchoCancellationConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.AcousticEchoCancellationConfig'),
        LibraryAudioChunkTimestamp: root.lookupType('ai.inworld.runtime.v1.library.speech.AudioChunkTimestamp'),
        LibraryBatchProcessingOptions: root.lookupType('ai.inworld.runtime.v1.library.telemetry.BatchProcessingOptions'),
        LibraryClassThreshold: root.lookupType('ai.inworld.runtime.v1.library.nlu.ClassThreshold'),
        LibraryCompileKnowledgeResponse: root.lookupType('ai.inworld.runtime.v1.library.knowledge.CompileKnowledgeResponse'),
        LibraryContentItem: root.lookupType('ai.inworld.runtime.v1.library.llm.ContentItem'),
        LibraryDeviceInfo: root.lookupType('ai.inworld.runtime.v1.library.platform.DeviceInfo'),
        LibraryDeviceSelection: root.lookupType('ai.inworld.runtime.v1.library.platform.DeviceSelection'),
        LibraryDictionaryRule: root.lookupType('ai.inworld.runtime.v1.library.nlu.DictionaryRule'),
        LibraryElevenLabsQueryParams: root.lookupType('ai.inworld.runtime.v1.library.speech.ElevenLabsQueryParams'),
        LibraryElevenLabsRequestParams: root.lookupType('ai.inworld.runtime.v1.library.speech.ElevenLabsRequestParams'),
        LibraryElevenLabsSpeechSynthesisConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.ElevenLabsSpeechSynthesisConfig'),
        LibraryEmbeddingMatcherConfig: root.lookupType('ai.inworld.runtime.v1.library.nlu.EmbeddingMatcherConfig'),
        LibraryEntity: root.lookupType('ai.inworld.runtime.v1.library.nlu.Entity'),
        LibraryFunctionChoice: root.lookupType('ai.inworld.runtime.v1.library.llm.FunctionChoice'),
        LibraryGetKnowledgeResponse: root.lookupType('ai.inworld.runtime.v1.library.knowledge.GetKnowledgeResponse'),
        LibraryGetModelsRequest: root.lookupType('ai.inworld.runtime.v1.library.llm.GetModelsRequest'),
        LibraryGetModelsResponse: root.lookupType('ai.inworld.runtime.v1.library.llm.GetModelsResponse'),
        LibraryGoogleSerperAPIClientCreationConfig: root.lookupType('ai.inworld.runtime.v1.library.serp.GoogleSerperAPIClientCreationConfig'),
        LibraryGoogleSerperApiSearchRequest: root.lookupType('ai.inworld.runtime.v1.library.serp.GoogleSerperApiSearchRequest'),
        LibraryGoogleSerperApiSearchResponse: root.lookupType('ai.inworld.runtime.v1.library.serp.GoogleSerperApiSearchResponse'),
        LibraryImageUrl: root.lookupType('ai.inworld.runtime.v1.library.llm.ImageUrl'),
        LibraryIntent: root.lookupType('ai.inworld.runtime.v1.library.nlu.Intent'),
        LibraryIntentMatcherConfig: root.lookupType('ai.inworld.runtime.v1.library.nlu.IntentMatcherConfig'),
        LibraryInworldSpeechSynthesisConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.InworldSpeechSynthesisConfig'),
        LibraryKeywordMatch: root.lookupType('ai.inworld.runtime.v1.library.nlu.KeywordMatch'),
        LibraryKnowledgeCompileConfig: root.lookupType('ai.inworld.runtime.v1.library.knowledge.KnowledgeCompileConfig'),
        LibraryKnowledgeGraph: root.lookupType('ai.inworld.runtime.v1.library.serp.GoogleSerperApiSearchResponse.KnowledgeGraph'),
        LibraryKnowledgeRecords: root.lookupType('ai.inworld.runtime.v1.library.knowledge.KnowledgeRecords'),
        LibraryLanguage: root.lookupType('ai.inworld.runtime.v1.library.core.Language'),
        LibraryLLMBenchmarks: root.lookupType('ai.inworld.runtime.v1.library.llm.LLMBenchmarks'),
        LibraryLlmMatcherConfig: root.lookupType('ai.inworld.runtime.v1.library.nlu.LlmMatcherConfig'),
        LibraryLLMPerformance: root.lookupType('ai.inworld.runtime.v1.library.llm.LLMPerformance'),
        LibraryLLMPricing: root.lookupType('ai.inworld.runtime.v1.library.llm.LLMPricing'),
        LibraryLLMSpec: root.lookupType('ai.inworld.runtime.v1.library.llm.LLMSpec'),
        LibraryLocalAECFilterConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.LocalAECFilterConfig'),
        LibraryLocalKnowledgeConfig: root.lookupType('ai.inworld.runtime.v1.library.knowledge.LocalKnowledgeConfig'),
        LibraryLocalLLMConfig: root.lookupType('ai.inworld.runtime.v1.library.llm.LocalLLMConfig'),
        LibraryLocalStreamingSTTConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.LocalStreamingSTTConfig'),
        LibraryLocalSTTConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.LocalSTTConfig'),
        LibraryLocalTextEmbedderConfig: root.lookupType('ai.inworld.runtime.v1.library.embeddings.LocalTextEmbedderConfig'),
        LibraryLocalTurnDetectorConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.LocalTurnDetectorConfig'),
        LibraryLocalVADConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.LocalVADConfig'),
        LibraryLogitBias: root.lookupType('ai.inworld.runtime.v1.library.llm.LogitBias'),
        LibraryMCPHttpConfig: root.lookupType('ai.inworld.runtime.v1.library.mcp.MCPHttpConfig'),
        LibraryMCPSessionCreationConfig: root.lookupType('ai.inworld.runtime.v1.library.mcp.MCPSessionCreationConfig'),
        LibraryMCPStdioConfig: root.lookupType('ai.inworld.runtime.v1.library.mcp.MCPStdioConfig'),
        LibraryModelId: root.lookupType('ai.inworld.runtime.v1.library.llm.ModelId'),
        LibraryModelInfo: root.lookupType('ai.inworld.runtime.v1.library.llm.ModelInfo'),
        LibraryOrganicResult: root.lookupType('ai.inworld.runtime.v1.library.serp.GoogleSerperApiSearchResponse.OrganicResult'),
        LibraryParagonApiConfig: root.lookupType('ai.inworld.runtime.v1.library.mcp.ParagonApiConfig'),
        LibraryParsingConfig: root.lookupType('ai.inworld.runtime.v1.library.knowledge.ParsingConfig'),
        LibraryPeopleAlsoAsk: root.lookupType('ai.inworld.runtime.v1.library.serp.GoogleSerperApiSearchResponse.PeopleAlsoAsk'),
        LibraryReasoningConfig: root.lookupType('ai.inworld.runtime.v1.library.llm.ReasoningConfig'),
        LibraryRemoteKnowledgeConfig: root.lookupType('ai.inworld.runtime.v1.library.knowledge.RemoteKnowledgeConfig'),
        LibraryRemoteLLMConfig: root.lookupType('ai.inworld.runtime.v1.library.llm.RemoteLLMConfig'),
        LibraryRemoteStreamingSTTConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.RemoteStreamingSTTConfig'),
        LibraryRemoteSTTConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.RemoteSTTConfig'),
        LibraryRemoteTextEmbedderConfig: root.lookupType('ai.inworld.runtime.v1.library.embeddings.RemoteTextEmbedderConfig'),
        LibraryRemoteTTSConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.RemoteTTSConfig'),
        LibraryRemoveKnowledgeRequest: root.lookupType('ai.inworld.runtime.v1.library.knowledge.RemoveKnowledgeRequest'),
        LibraryRetrievalConfig: root.lookupType('ai.inworld.runtime.v1.library.knowledge.RetrievalConfig'),
        LibrarySitelink: root.lookupType('ai.inworld.runtime.v1.library.serp.GoogleSerperApiSearchResponse.OrganicResult.Sitelink'),
        LibrarySpeechRecognitionConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.SpeechRecognitionConfig'),
        LibrarySpeechSynthesisConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.SpeechSynthesisConfig'),
        LibrarySpeechSynthesisInferenceConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.SpeechSynthesisInferenceConfig'),
        LibrarySpeechSynthesisPostprocessingConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.SpeechSynthesisPostprocessingConfig'),
        LibraryTextClassifierConfig: root.lookupType('ai.inworld.runtime.v1.library.nlu.TextClassifierConfig'),
        LibraryTextGenerationConfig: root.lookupType('ai.inworld.runtime.v1.library.llm.TextGenerationConfig'),
        LibraryToolCallResult: root.lookupType('ai.inworld.runtime.v1.library.llm.ToolCallResult'),
        LibraryToolChoice: root.lookupType('ai.inworld.runtime.v1.library.llm.ToolChoice'),
        LibraryTopStory: root.lookupType('ai.inworld.runtime.v1.library.serp.GoogleSerperApiSearchResponse.TopStory'),
        LibraryTurnDetectionConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.TurnDetectionConfig'),
        LibraryVoice: root.lookupType('ai.inworld.runtime.v1.library.speech.Voice'),
        ListToolsResponse: root.lookupType('ai.inworld.runtime.v1.graphs.schema.ListToolsResponse'),
        LLMChatRequest: root.lookupType('ai.inworld.runtime.v1.graphs.schema.LLMChatRequest'),
        LLMChatResponse: root.lookupType('ai.inworld.runtime.v1.graphs.schema.LLMChatResponse'),
        LLMChatRoutingRequest: root.lookupType('ai.inworld.runtime.v1.graphs.schema.LLMChatRoutingRequest'),
        LLMCompletionResponse: root.lookupType('ai.inworld.runtime.v1.graphs.schema.LLMCompletionResponse'),
        LLMCreationConfig: root.lookupType('ai.inworld.runtime.v1.library.llm.LLMCreationConfig'),
        LLMMessage: root.lookupType('ai.inworld.runtime.v1.library.llm.Message'),
        MatchedIntents: root.lookupType('ai.inworld.runtime.v1.graphs.schema.MatchedIntents'),
        MatchedKeywords: root.lookupType('ai.inworld.runtime.v1.graphs.schema.MatchedKeywords'),
        MCPClientCreationConfig: root.lookupType('ai.inworld.runtime.v1.library.mcp.MCPClientCreationConfig'),
        MultimodalContent: root.lookupType('ai.inworld.runtime.v1.graphs.schema.MultimodalContent'),
        NERCreationConfig: root.lookupType('ai.inworld.runtime.v1.library.nlu.NERCreationConfig'),
        RecognizeSpeechRequest: root.lookupType('ai.inworld.runtime.v1.library.speech.RecognizeSpeechRequest'),
        SafetyResult: root.lookupType('ai.inworld.runtime.v1.graphs.schema.SafetyResult'),
        StreamingSTTCreationConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.StreamingSTTCreationConfig'),
        STTCreationConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.STTCreationConfig'),
        SynthesizeSpeechRequest: root.lookupType('ai.inworld.runtime.v1.library.speech.SynthesizeSpeechRequest'),
        Text: root.lookupType('ai.inworld.runtime.v1.graphs.schema.Text'),
        TextClassifierCreationConfig: root.lookupType('ai.inworld.runtime.v1.library.nlu.TextClassifierCreationConfig'),
        TextClassifierResponse: root.lookupType('ai.inworld.runtime.v1.library.nlu.ClassifyTextResponse'),
        TextEmbedderCreationConfig: root.lookupType('ai.inworld.runtime.v1.library.embeddings.TextEmbedderCreationConfig'),
        Tool: root.lookupType('ai.inworld.runtime.v1.library.llm.Tool'),
        ToolCall: root.lookupType('ai.inworld.runtime.v1.library.llm.ToolCall'),
        ToolCallRequest: root.lookupType('ai.inworld.runtime.v1.graphs.schema.ToolCallRequest'),
        ToolCallResponse: root.lookupType('ai.inworld.runtime.v1.graphs.schema.ToolCallResponse'),
        ToolCallResult: root.lookupType('ai.inworld.runtime.v1.graphs.schema.ToolCallResult'),
        ToolList: root.lookupType('ai.inworld.runtime.v1.library.llm.ToolList'),
        TTSCreationConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.TTSCreationConfig'),
        TTSRequest: root.lookupType('ai.inworld.runtime.v1.graphs.schema.TTSRequest'),
        TurnDetectorCreationConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.TurnDetectorCreationConfig'),
        TurnPrediction: root.lookupType('ai.inworld.runtime.v1.library.speech.TurnPrediction'),
        VADCreationConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.VADCreationConfig'),
        VoiceActivityDetectionConfig: root.lookupType('ai.inworld.runtime.v1.library.speech.VoiceActivityDetectionConfig'),
    };
}
