// @inworld/runtime/build/primitives/speech/index.js
"use strict";
/**
 * Speech primitives module.
 * Exports STT, StreamingSTT, TTS, VAD, TurnDetector, and AECFilter implementations.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VAD = exports.TurnDetector = exports.TTS = exports.STT = exports.StreamingSTT = exports.AECFilter = void 0;
__exportStar(require("./types"), exports);
// Implemented primitives
var aec_filter_1 = require("./aec_filter");
Object.defineProperty(exports, "AECFilter", { enumerable: true, get: function () { return aec_filter_1.AECFilter; } });
var streaming_stt_1 = require("./streaming_stt");
Object.defineProperty(exports, "StreamingSTT", { enumerable: true, get: function () { return streaming_stt_1.StreamingSTT; } });
var stt_1 = require("./stt");
Object.defineProperty(exports, "STT", { enumerable: true, get: function () { return stt_1.STT; } });
var tts_1 = require("./tts");
Object.defineProperty(exports, "TTS", { enumerable: true, get: function () { return tts_1.TTS; } });
var turn_detector_1 = require("./turn_detector");
Object.defineProperty(exports, "TurnDetector", { enumerable: true, get: function () { return turn_detector_1.TurnDetector; } });
var vad_1 = require("./vad");
Object.defineProperty(exports, "VAD", { enumerable: true, get: function () { return vad_1.VAD; } });
