// @inworld/runtime/build/primitives/speech/stt.js
"use strict";
/**
 * Speech-to-Text (STT) primitive using N-API addon.
 */
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = Object.create((typeof AsyncIterator === "function" ? AsyncIterator : Object).prototype), verb("next"), verb("throw"), verb("return", awaitReturn), i[Symbol.asyncIterator] = function () { return this; }, i;
    function awaitReturn(f) { return function (v) { return Promise.resolve(v).then(f, reject); }; }
    function verb(n, f) { if (g[n]) { i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; if (f) i[n] = f(i[n]); } }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.STT = void 0;
const node_api_1 = require("../../internal/node_api");
const pb_helpers_1 = require("../../internal/pb_helpers");
const errors_1 = require("../errors");
/**
 * Speech-to-Text (STT) class.
 * Converts spoken audio into text transcriptions.
 * Supports both local models and cloud-based APIs.
 *
 * @example
 * ```typescript
 * // Create STT instance with cloud API
 * const stt = await STT.create({
 *   remoteConfig: {
 *     apiKey: process.env.INWORLD_API_KEY,
 *     defaultConfig: { languageCode: 'en-US' },
 *     defaultTimeout: { seconds: 30 }
 *   }
 * });
 *
 * // Recognize speech from audio
 * const stream = await stt.recognizeSpeech(audioChunk, { languageCode: 'en-US' });
 *
 * let transcription = '';
 * for await (const textChunk of stream) {
 *   transcription += textChunk;
 *   console.log('Partial:', textChunk);
 * }
 * console.log('Final:', transcription);
 * ```
 */
class STT {
    /**
     * Creates a new STT instance (internal constructor).
     * Use `STT.create()` static method instead.
     *
     * @param {any} sttInterface - Addon STTInterface instance
     * @internal
     */
    constructor(sttInterface) {
        this.sttInterface = sttInterface;
    }
    /**
     * Creates a new STT instance with the specified configuration.
     *
     * @param {STTCreationConfig} config - STT creation configuration
     * @returns {Promise<STT>} Promise resolving to STT instance
     * @throws {STTError} If STT creation fails
     *
     * @example
     * ```typescript
     * // Local model
     * const stt = await STT.create({
     *   localConfig: {
     *     modelPath: 'models/stt/whisper-base',
     *     device: { type: 'CPU', index: 0 },
     *     defaultConfig: { languageCode: 'en-US' }
     *   }
     * });
     *
     * // Cloud API
     * const stt = await STT.create({
     *   remoteConfig: {
     *     apiKey: process.env.INWORLD_API_KEY,
     *     defaultConfig: { languageCode: 'en-US' },
     *     defaultTimeout: { seconds: 30 }
     *   }
     * });
     * ```
     */
    static async create(config) {
        try {
            const env = (0, node_api_1.getInworldAddonEnv)();
            const speech = env.speech;
            if (!speech) {
                throw new errors_1.STTError('Speech namespace not available in addon environment');
            }
            const factory = speech.createSTTFactory();
            const configPb = pb_helpers_1.PbHelper.Speech.createSTTCreationConfig(config);
            const sttInterface = await factory.createSTT(configPb);
            return new STT(sttInterface);
        }
        catch (error) {
            if (error instanceof errors_1.STTError) {
                throw error;
            }
            throw new errors_1.STTError(`Failed to create STT instance: ${error.message}`, error.stack);
        }
    }
    /**
     * Recognizes speech from audio and returns a stream of text results.
     * The stream yields partial transcriptions as they become available.
     *
     * @param {AudioChunk} audio - Audio chunk to transcribe
     * @param {RecognizeSpeechConfig} config - Optional recognition configuration
     * @returns {Promise<AsyncIterable<string>>} Stream of text transcriptions
     * @throws {STTError} If speech recognition fails
     *
     * @example
     * ```typescript
     * const audioChunk = {
     *   data: audioData, // Float32Array
     *   sampleRate: 16000
     * };
     *
     * const stream = await stt.recognizeSpeech(audioChunk, {
     *   languageCode: 'en-US'
     * });
     *
     * let fullTranscription = '';
     * for await (const textChunk of stream) {
     *   fullTranscription += textChunk;
     *   console.log('Received:', textChunk);
     * }
     *
     * console.log('Complete transcription:', fullTranscription);
     * ```
     */
    async recognizeSpeech(audio, config) {
        try {
            if (!audio) {
                throw new errors_1.STTError('Audio chunk is required');
            }
            if (!audio.data || audio.data.length === 0) {
                throw new errors_1.STTError('Audio data cannot be empty');
            }
            if (!audio.sampleRate || audio.sampleRate <= 0) {
                throw new errors_1.STTError('Valid sample rate is required');
            }
            const requestPb = pb_helpers_1.PbHelper.Speech.createRecognizeSpeechRequest(audio, config);
            // Returns an async iterable of protobuf messages
            const pbStream = await this.sttInterface.recognizeSpeech(requestPb);
            // Transform the protobuf stream into a string stream
            return this.transformPbStreamToText(pbStream);
        }
        catch (error) {
            if (error instanceof errors_1.STTError) {
                throw error;
            }
            throw new errors_1.STTError(`Failed to recognize speech: ${error.message}`, error.stack);
        }
    }
    /**
     * Transforms protobuf stream to text stream.
     * @param {AsyncIterable<any>} pbStream - Protobuf message stream
     * @returns {AsyncIterable<string>} Text stream
     * @private
     */
    transformPbStreamToText(pbStream) {
        return __asyncGenerator(this, arguments, function* transformPbStreamToText_1() {
            var _a, e_1, _b, _c;
            try {
                try {
                    for (var _d = true, pbStream_1 = __asyncValues(pbStream), pbStream_1_1; pbStream_1_1 = yield __await(pbStream_1.next()), _a = pbStream_1_1.done, !_a; _d = true) {
                        _c = pbStream_1_1.value;
                        _d = false;
                        const pbMessage = _c;
                        const text = pb_helpers_1.PbHelper.Speech.unpackText(pbMessage);
                        if (text) {
                            yield yield __await(text);
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (!_d && !_a && (_b = pbStream_1.return)) yield __await(_b.call(pbStream_1));
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            catch (error) {
                throw new errors_1.STTError(`Failed to process STT stream: ${error.message}`, error.stack);
            }
        });
    }
    /**
     * Convenience method to recognize speech and return complete transcription.
     * Waits for the entire stream to complete before returning.
     *
     * @param {AudioChunk} audio - Audio chunk to transcribe
     * @param {RecognizeSpeechConfig} config - Optional recognition configuration
     * @returns {Promise<string>} Complete transcription
     * @throws {STTError} If speech recognition fails
     *
     * @example
     * ```typescript
     * const audioChunk = {
     *   data: audioData,
     *   sampleRate: 16000
     * };
     *
     * const transcription = await stt.recognizeSpeechComplete(audioChunk);
     * console.log('Transcription:', transcription);
     * ```
     */
    async recognizeSpeechComplete(audio, config) {
        var _a, e_2, _b, _c;
        const stream = await this.recognizeSpeech(audio, config);
        let fullText = '';
        try {
            for (var _d = true, stream_1 = __asyncValues(stream), stream_1_1; stream_1_1 = await stream_1.next(), _a = stream_1_1.done, !_a; _d = true) {
                _c = stream_1_1.value;
                _d = false;
                const textChunk = _c;
                fullText += textChunk;
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (!_d && !_a && (_b = stream_1.return)) await _b.call(stream_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return fullText;
    }
}
exports.STT = STT;
