// @inworld/runtime/build/primitives/speech/tts.js
"use strict";
/**
 * Text-to-Speech (TTS) primitive using N-API addon.
 */
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = Object.create((typeof AsyncIterator === "function" ? AsyncIterator : Object).prototype), verb("next"), verb("throw"), verb("return", awaitReturn), i[Symbol.asyncIterator] = function () { return this; }, i;
    function awaitReturn(f) { return function (v) { return Promise.resolve(v).then(f, reject); }; }
    function verb(n, f) { if (g[n]) { i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; if (f) i[n] = f(i[n]); } }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TTS = void 0;
const node_api_1 = require("../../internal/node_api");
const pb_helpers_1 = require("../../internal/pb_helpers");
const errors_1 = require("../errors");
/**
 * Text-to-Speech (TTS) class.
 * Synthesizes natural-sounding speech from text.
 * Supports both local models and cloud-based APIs.
 *
 * @example
 * ```typescript
 * // Create TTS instance with cloud API
 * const tts = await TTS.create({
 *   remoteConfig: {
 *     apiKey: process.env.INWORLD_API_KEY,
 *     defaultConfig: {
 *       voice: 'en-US-Neural2-A',
 *       sampleRateHz: 24000
 *     },
 *     defaultTimeout: { seconds: 30 }
 *   }
 * });
 *
 * // Synthesize speech
 * const stream = await tts.synthesizeSpeech({
 *   voice: 'en-US-Neural2-A',
 *   text: 'Hello, world!'
 * });
 *
 * for await (const audioChunk of stream) {
 *   // Play or save audio chunk
 *   console.log('Received audio:', audioChunk.data.length, 'samples');
 * }
 * ```
 */
class TTS {
    /**
     * Creates a new TTS instance (internal constructor).
     * Use `TTS.create()` static method instead.
     *
     * @param {any} ttsInterface - Addon TTSInterface instance
     * @internal
     */
    constructor(ttsInterface) {
        this.ttsInterface = ttsInterface;
    }
    /**
     * Creates a new TTS instance with the specified configuration.
     *
     * @param {TTSCreationConfig} config - TTS creation configuration
     * @returns {Promise<TTS>} Promise resolving to TTS instance
     * @throws {TTSError} If TTS creation fails
     *
     * @example
     * ```typescript
     * // Local model
     * const tts = await TTS.create({
     *   localConfig: {
     *     modelPath: 'models/tts/torch-tts',
     *     device: { type: 'CPU', index: 0 },
     *     defaultConfig: {
     *       voice: 'default',
     *       sampleRateHz: 22050
     *     }
     *   }
     * });
     *
     * // Cloud API
     * const tts = await TTS.create({
     *   remoteConfig: {
     *     apiKey: process.env.INWORLD_API_KEY,
     *     defaultConfig: {
     *       voice: 'en-US-Neural2-A',
     *       sampleRateHz: 24000
     *     },
     *     defaultTimeout: { seconds: 30 }
     *   }
     * });
     * ```
     */
    static async create(config) {
        try {
            const env = (0, node_api_1.getInworldAddonEnv)();
            const speech = env.speech;
            if (!speech) {
                throw new errors_1.TTSError('Speech namespace not available in addon environment');
            }
            const factory = speech.createTTSFactory();
            const configPb = pb_helpers_1.PbHelper.Speech.createTTSCreationConfig(config);
            const ttsInterface = await factory.createTTS(configPb);
            return new TTS(ttsInterface);
        }
        catch (error) {
            if (error instanceof errors_1.TTSError) {
                throw error;
            }
            throw new errors_1.TTSError(`Failed to create TTS instance: ${error.message}`, error.stack);
        }
    }
    /**
     * Synthesizes speech from text and returns a stream of audio chunks.
     * The stream yields audio data as it becomes available.
     *
     * @param {SynthesizeSpeechRequest} request - Synthesis request with voice and text
     * @returns {Promise<AsyncIterable<AudioChunk>>} Stream of audio chunks
     * @throws {TTSError} If speech synthesis fails
     *
     * @example
     * ```typescript
     * const stream = await tts.synthesizeSpeech({
     *   voice: 'en-US-Neural2-A',
     *   text: 'Hello, world! How are you today?'
     * });
     *
     * const audioChunks: Float32Array[] = [];
     * for await (const chunk of stream) {
     *   audioChunks.push(chunk.data);
     *   console.log('Received chunk:', chunk.data.length, 'samples at', chunk.sampleRate, 'Hz');
     * }
     *
     * // Concatenate all chunks
     * const totalLength = audioChunks.reduce((sum, arr) => sum + arr.length, 0);
     * const fullAudio = new Float32Array(totalLength);
     * let offset = 0;
     * for (const chunk of audioChunks) {
     *   fullAudio.set(chunk, offset);
     *   offset += chunk.length;
     * }
     * ```
     */
    async synthesizeSpeech(request) {
        try {
            if (!request) {
                throw new errors_1.TTSError('Synthesis request is required');
            }
            if (!request.text || request.text.trim().length === 0) {
                throw new errors_1.TTSError('Text cannot be empty');
            }
            if (!request.voice) {
                throw new errors_1.TTSError('Voice is required');
            }
            const requestPb = pb_helpers_1.PbHelper.Speech.createSynthesizeSpeechRequest(request);
            // Returns an async iterable of protobuf AudioChunk messages
            const pbStream = await this.ttsInterface.synthesizeSpeech(requestPb);
            // Transform the protobuf stream into AudioChunk stream
            return this.transformPbStreamToAudio(pbStream);
        }
        catch (error) {
            if (error instanceof errors_1.TTSError) {
                throw error;
            }
            throw new errors_1.TTSError(`Failed to synthesize speech: ${error.message}`, error.stack);
        }
    }
    /**
     * Transforms protobuf stream to audio chunk stream.
     * @param {AsyncIterable<any>} pbStream - Protobuf message stream
     * @returns {AsyncIterable<AudioChunk>} Audio chunk stream
     * @private
     */
    transformPbStreamToAudio(pbStream) {
        return __asyncGenerator(this, arguments, function* transformPbStreamToAudio_1() {
            var _a, e_1, _b, _c;
            try {
                try {
                    for (var _d = true, pbStream_1 = __asyncValues(pbStream), pbStream_1_1; pbStream_1_1 = yield __await(pbStream_1.next()), _a = pbStream_1_1.done, !_a; _d = true) {
                        _c = pbStream_1_1.value;
                        _d = false;
                        const pbMessage = _c;
                        const audioChunk = pb_helpers_1.PbHelper.Speech.unpackAudioFrame(pbMessage);
                        if (audioChunk && audioChunk.data && audioChunk.data.length > 0) {
                            yield yield __await(audioChunk);
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (!_d && !_a && (_b = pbStream_1.return)) yield __await(_b.call(pbStream_1));
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            catch (error) {
                throw new errors_1.TTSError(`Failed to process TTS stream: ${error.message}`, error.stack);
            }
        });
    }
    /**
     * Convenience method to synthesize speech and return complete audio.
     * Waits for the entire stream to complete and concatenates all chunks.
     *
     * @param {SynthesizeSpeechRequest} request - Synthesis request
     * @returns {Promise<AudioChunk>} Complete audio as single chunk
     * @throws {TTSError} If speech synthesis fails
     *
     * @example
     * ```typescript
     * const audio = await tts.synthesizeSpeechComplete({
     *   voice: 'en-US-Neural2-A',
     *   text: 'Hello, world!'
     * });
     *
     * console.log('Generated audio:', audio.data.length, 'samples');
     * console.log('Sample rate:', audio.sampleRate, 'Hz');
     *
     * // Save to file or play
     * ```
     */
    async synthesizeSpeechComplete(request) {
        var _a, e_2, _b, _c;
        const stream = await this.synthesizeSpeech(request);
        const chunks = [];
        let sampleRate = 0;
        try {
            for (var _d = true, stream_1 = __asyncValues(stream), stream_1_1; stream_1_1 = await stream_1.next(), _a = stream_1_1.done, !_a; _d = true) {
                _c = stream_1_1.value;
                _d = false;
                const chunk = _c;
                chunks.push(chunk);
                if (sampleRate === 0) {
                    sampleRate = chunk.sampleRate;
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (!_d && !_a && (_b = stream_1.return)) await _b.call(stream_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
        if (chunks.length === 0) {
            throw new errors_1.TTSError('No audio chunks received from TTS');
        }
        // Concatenate all audio chunks
        const totalLength = chunks.reduce((sum, chunk) => sum + chunk.data.length, 0);
        const fullAudio = new Float32Array(totalLength);
        let offset = 0;
        for (const chunk of chunks) {
            fullAudio.set(chunk.data, offset);
            offset += chunk.data.length;
        }
        return {
            data: fullAudio,
            sampleRate: sampleRate || 22050, // Default sample rate
        };
    }
    /**
     * Synthesizes speech from text with simpler API (uses defaults from config).
     *
     * @param {string} text - Text to synthesize
     * @param {string} voice - Optional voice override
     * @returns {Promise<AsyncIterable<AudioChunk>>} Stream of audio chunks
     * @throws {TTSError} If speech synthesis fails
     *
     * @example
     * ```typescript
     * const stream = await tts.speak('Hello, world!');
     *
     * for await (const chunk of stream) {
     *   // Process audio chunk
     * }
     * ```
     */
    async speak(text, voice) {
        if (!voice) {
            // Will use default voice from config
            voice = 'default';
        }
        return this.synthesizeSpeech({
            voice,
            text,
        });
    }
}
exports.TTS = TTS;
