// @inworld/runtime/build/primitives/speech/turn_detector.js
"use strict";
/**
 * Turn Detector primitive using N-API addon.
 * Detects when a speaker has completed their conversational turn.
 */
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = Object.create((typeof AsyncIterator === "function" ? AsyncIterator : Object).prototype), verb("next"), verb("throw"), verb("return", awaitReturn), i[Symbol.asyncIterator] = function () { return this; }, i;
    function awaitReturn(f) { return function (v) { return Promise.resolve(v).then(f, reject); }; }
    function verb(n, f) { if (g[n]) { i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; if (f) i[n] = f(i[n]); } }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TurnDetector = void 0;
const node_api_1 = require("../../internal/node_api");
const pb_helpers_1 = require("../../internal/pb_helpers");
const errors_1 = require("../errors");
/**
 * Turn Detector class.
 * Analyzes audio to detect when a speaker has finished their turn in a conversation.
 * Uses machine learning models to predict turn completion probability.
 *
 * @example
 * ```typescript
 * // Create turn detector with local model
 * const turnDetector = await TurnDetector.create({
 *   localConfig: {
 *     modelPath: 'models/turn_detection/pipecat_smart_turn_v3.onnx',
 *     device: { type: 'CPU', index: 0 },
 *     defaultConfig: { threshold: 0.7 }
 *   }
 * });
 *
 * // Detect turn completion in audio
 * const audioChunk = {
 *   data: new Float32Array(audioBuffer),
 *   sampleRate: 16000
 * };
 *
 * const result = await turnDetector.detectTurnCompletion(audioChunk);
 *
 * if (result.prediction === 1 && result.probability > 0.7) {
 *   console.log(`Turn completed with ${(result.probability * 100).toFixed(1)}% confidence`);
 *   // Stop speaking, allow other party to respond
 * }
 * ```
 */
class TurnDetector {
    /**
     * Creates a new TurnDetector instance (internal constructor).
     * Use `TurnDetector.create()` static method instead.
     *
     * @param {any} turnDetectorInterface - Addon TurnDetectorInterface instance
     * @internal
     */
    constructor(turnDetectorInterface) {
        this.turnDetectorInterface = turnDetectorInterface;
    }
    /**
     * Creates a new TurnDetector instance with the specified configuration.
     * Currently only supports local (on-device) models.
     *
     * @param {TurnDetectorCreationConfig} config - Turn detector configuration
     * @returns {Promise<TurnDetector>} Promise resolving to TurnDetector instance
     * @throws {TurnDetectorError} If turn detector creation fails
     *
     * @example
     * ```typescript
     * const turnDetector = await TurnDetector.create({
     *   localConfig: {
     *     modelPath: 'models/turn_detection/pipecat_smart_turn_v3.onnx',
     *     device: { type: 'CPU', index: 0 },
     *     defaultConfig: {
     *       threshold: 0.7  // Confidence threshold for turn completion
     *     }
     *   }
     * });
     * ```
     */
    static async create(config) {
        try {
            const env = (0, node_api_1.getInworldAddonEnv)();
            const speech = env.speech;
            if (!speech) {
                throw new errors_1.TurnDetectorError('Speech namespace not available in addon environment');
            }
            const factory = speech.createTurnDetectorFactory();
            const configPb = pb_helpers_1.PbHelper.Speech.createTurnDetectorCreationConfig(config);
            const turnDetectorInterface = await factory.createTurnDetector(configPb);
            return new TurnDetector(turnDetectorInterface);
        }
        catch (error) {
            if (error instanceof errors_1.TurnDetectorError) {
                throw error;
            }
            throw new errors_1.TurnDetectorError(`Failed to create TurnDetector instance: ${error.message}`, error.stack);
        }
    }
    /**
     * Detects whether a turn has been completed in the provided audio.
     * Returns a prediction (0 or 1) and probability (0-1).
     *
     * @param {AudioChunk} audio - Audio chunk to analyze
     * @param {TurnDetectorConfig} config - Optional detection config
     * @returns {Promise<TurnDetectionResult>} Turn detection result
     * @throws {TurnDetectorError} If detection fails
     *
     * @example
     * ```typescript
     * const audioChunk = {
     *   data: new Float32Array(samples),
     *   sampleRate: 16000
     * };
     *
     * const result = await turnDetector.detectTurnCompletion(audioChunk);
     *
     * console.log('Prediction:', result.prediction);  // 0 (not complete) or 1 (complete)
     * console.log('Probability:', result.probability);  // 0.0 to 1.0
     *
     * if (result.prediction === 1 && result.probability > 0.8) {
     *   console.log('High confidence turn completion detected');
     * }
     * ```
     */
    async detectTurnCompletion(audio, config) {
        try {
            if (!audio || !audio.data) {
                throw new errors_1.TurnDetectorError('Audio data is required');
            }
            if (!audio.sampleRate || audio.sampleRate <= 0) {
                throw new errors_1.TurnDetectorError('Valid sample rate is required');
            }
            const requestPb = pb_helpers_1.PbHelper.Speech.createDetectTurnCompletionRequest(audio, config);
            const result = await this.turnDetectorInterface.detectTurnCompletion(requestPb);
            // Result is already unpacked as a plain object
            return {
                prediction: result.prediction,
                probability: result.probability,
            };
        }
        catch (error) {
            if (error instanceof errors_1.TurnDetectorError) {
                throw error;
            }
            throw new errors_1.TurnDetectorError(`Failed to detect turn completion: ${error.message}`, error.stack);
        }
    }
    /**
     * Checks if a turn is completed based on prediction and threshold.
     *
     * @param {AudioChunk} audio - Audio chunk to analyze
     * @param {number} threshold - Probability threshold (default: 0.7)
     * @returns {Promise<boolean>} True if turn is complete
     * @throws {TurnDetectorError} If detection fails
     *
     * @example
     * ```typescript
     * const isTurnComplete = await turnDetector.isTurnComplete(audioChunk, 0.8);
     *
     * if (isTurnComplete) {
     *   console.log('Turn is complete, ready for response');
     *   // Switch to listening mode
     * }
     * ```
     */
    async isTurnComplete(audio, threshold = 0.7) {
        const result = await this.detectTurnCompletion(audio, { threshold });
        return result.prediction === 1 && result.probability >= threshold;
    }
    /**
     * Continuously monitors audio chunks for turn completion.
     * Useful for real-time conversation systems.
     *
     * @param {AsyncIterable<AudioChunk>} audioStream - Stream of audio chunks
     * @param {TurnDetectorConfig} config - Detection config
     * @returns {AsyncGenerator<TurnDetectionResult>} Stream of detection results
     * @throws {TurnDetectorError} If detection fails
     *
     * @example
     * ```typescript
     * async function monitorTurns(audioStream) {
     *   for await (const result of turnDetector.monitorTurns(audioStream, { threshold: 0.7 })) {
     *     console.log(`Prediction: ${result.prediction}, Probability: ${result.probability}`);
     *
     *     if (result.prediction === 1 && result.probability > 0.8) {
     *       console.log('Turn completed!');
     *       break;  // Stop monitoring
     *     }
     *   }
     * }
     * ```
     */
    monitorTurns(audioStream, config) {
        return __asyncGenerator(this, arguments, function* monitorTurns_1() {
            var _a, e_1, _b, _c;
            try {
                try {
                    for (var _d = true, audioStream_1 = __asyncValues(audioStream), audioStream_1_1; audioStream_1_1 = yield __await(audioStream_1.next()), _a = audioStream_1_1.done, !_a; _d = true) {
                        _c = audioStream_1_1.value;
                        _d = false;
                        const audioChunk = _c;
                        const result = yield __await(this.detectTurnCompletion(audioChunk, config));
                        yield yield __await(result);
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (!_d && !_a && (_b = audioStream_1.return)) yield __await(_b.call(audioStream_1));
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            catch (error) {
                if (error instanceof errors_1.TurnDetectorError) {
                    throw error;
                }
                throw new errors_1.TurnDetectorError(`Failed to monitor turns: ${error.message}`, error.stack);
            }
        });
    }
    /**
     * Processes multiple audio chunks and returns their detection results.
     *
     * @param {AudioChunk[]} audioChunks - Array of audio chunks
     * @param {TurnDetectorConfig} config - Detection config
     * @returns {Promise<TurnDetectionResult[]>} Array of detection results
     * @throws {TurnDetectorError} If detection fails
     *
     * @example
     * ```typescript
     * const chunks = [chunk1, chunk2, chunk3];
     * const results = await turnDetector.detectBatch(chunks);
     *
     * results.forEach((result, i) => {
     *   console.log(`Chunk ${i + 1}: Prediction=${result.prediction}, Prob=${result.probability}`);
     * });
     * ```
     */
    async detectBatch(audioChunks, config) {
        if (!Array.isArray(audioChunks)) {
            throw new errors_1.TurnDetectorError('Audio chunks must be an array');
        }
        return Promise.all(audioChunks.map((chunk) => this.detectTurnCompletion(chunk, config)));
    }
    /**
     * Gets the maximum probability from recent audio chunks.
     * Useful for determining the strongest turn completion signal.
     *
     * @param {AudioChunk[]} audioChunks - Recent audio chunks
     * @param {TurnDetectorConfig} config - Detection config
     * @returns {Promise<TurnDetectionResult>} Result with highest probability
     * @throws {TurnDetectorError} If detection fails
     *
     * @example
     * ```typescript
     * const recentChunks = [chunk1, chunk2, chunk3, chunk4, chunk5];
     * const strongest = await turnDetector.getStrongestSignal(recentChunks);
     *
     * if (strongest.probability > 0.9) {
     *   console.log('Very strong turn completion signal detected');
     * }
     * ```
     */
    async getStrongestSignal(audioChunks, config) {
        const results = await this.detectBatch(audioChunks, config);
        if (results.length === 0) {
            throw new errors_1.TurnDetectorError('No audio chunks provided');
        }
        return results.reduce((max, current) => current.probability > max.probability ? current : max);
    }
}
exports.TurnDetector = TurnDetector;
