// @inworld/runtime/benchmarks/k6-load-tests/basic-memory-test.js
import { check } from 'k6';
import http from 'k6/http';
import { Rate } from 'k6/metrics';
import ws from 'k6/ws';

import {
  config,
  generateTestUser,
  getLoadUrl,
  getRandomMessage,
  getUnloadUrl,
  getWebSocketUrl,
} from './config.js';

export const options = {
  thresholds: config.THRESHOLDS,
  stages: config.LOAD_STAGES.ramp_up,
};

const errors = new Rate('errors');
const userKeys = new Set();

export default function () {
  const user = generateTestUser();
  const loadResponse = loadSession(user);
  if (!loadResponse || !loadResponse.agent) {
    console.error(`Failed to load session for user ${user.name}`);
    return;
  }
  userKeys.add(user.key);
  simulateWebSocketConversation(user, loadResponse.agent);
}

function loadSession(user) {
  const payload = JSON.stringify({
    userName: user.name,
    agent: config.DEFAULT_AGENT,
  });

  const params = {
    headers: {
      'Content-Type': 'application/json',
    },
  };

  const response = http.post(
    `${getLoadUrl()}?key=${user.key}`,
    payload,
    params,
  );

  check(response, {
    'load session status is 200': (r) => r.status === 200,
    'load response contains agent': (r) => r.json('agent') !== undefined,
  });

  return response.status === 200 ? response.json() : null;
}

function simulateWebSocketConversation(user, agent) {
  const url = `${getWebSocketUrl()}?key=${user.key}`;

  const params = {
    tags: {
      user_id: user.id,
      agent_name: agent.name,
    },
  };

  const response = ws.connect(url, params, function (socket) {
    socket.on('open', function open() {
      console.log(`User ${user.name} connected to WebSocket`);
    });
    socket.on('message', function message(data) {
      const packet = JSON.parse(data);

      check(packet, {
        'received valid packet': (p) => p.type !== undefined,
        'no error packets': (p) => p.type !== 'ERROR',
      });

      if (packet.type === 'ERROR') {
        errors.add(1);
      }
    });

    socket.on('error', function error(e) {
      console.error(`WebSocket error for ${user.name}:`, e);
    });

    const messageCount = 5;
    let messagesSent = 0;
    for (let i = 0; i < messageCount; i++) {
      socket.setTimeout(
        () => {
          if (socket.readyState === ws.READY_STATE_OPEN) {
            const message = {
              type: 'TEXT',
              text: getRandomMessage(),
            };
            socket.send(JSON.stringify(message));
            messagesSent++;
            console.log(`User ${user.name} sent: ${message.text}`);
          }
        },
        (i + 1) * 2000,
      ); // Send message every 2 seconds, starting at 2s
    }

    socket.setTimeout(
      () => {
        socket.close();

        if (messageCount !== messagesSent) {
          throw Error('Should be equal');
        }
      },
      (messageCount + 2) * 2000,
    );
  });

  check(response, {
    'websocket connection successful': (r) => r && r.status === 101,
  });
}

function unloadSession(key) {
  const response = http.post(`${getUnloadUrl()}?key=${key}`, '{}', {
    headers: {
      'Content-Type': 'application/json',
    },
  });

  check(response, {
    'unload session status is 200': (r) => r.status === 200,
    'unload response contains message': (r) => r.json('message') !== undefined,
  });
}

export function teardown() {
  console.log('Load test completed');
  userKeys.forEach((key) => {
    unloadSession(key);
  });
}
