// @inworld/runtime/benchmarks/k6-load-tests/config.js
// Configuration for K6 load testing
export const config = {
  // Server configuration
  HOST: __ENV.HOST || 'localhost',
  PORT: __ENV.PORT || '4000',

  // Test configuration
  VUS: __ENV.VUS || 2,
  DURATION: __ENV.DURATION || '30s',

  // Test data
  DEFAULT_AGENT: {
    name: 'TestAgent',
    description: 'A friendly AI assistant created for load testing',
    motivation: 'To help users and respond quickly during load tests',
  },

  // Message templates
  TEXT_MESSAGES: [
    'Hello, how are you?',
    'Can you help me with something?',
    'What is the weather like today?',
    'Tell me a joke',
    'How does artificial intelligence work?',
    'What can you help me with?',
    'That is interesting, tell me more',
    'Thank you for your help',
    'Can you explain that in more detail?',
    'What other topics can we discuss?',
  ],

  // Audio simulation (base64 encoded silence for testing)
  SILENCE_AUDIO_B64:
    'UklGRiQAAABXQVZFZm10IBAAAAABAAEARKwAAIhYAQACABAAZGF0YQAAAAA=',

  // Test thresholds
  THRESHOLDS: {
    http_req_duration: ['p(95)<1000'], // 95% of HTTP requests must complete below 1s
    http_req_failed: ['rate<0.05'], // Error rate must be below 5%
    ws_connecting: ['p(95)<500'], // WebSocket connection time
    ws_session_duration: ['p(95)<20000'], // WebSocket session duration (20s for full conversations)
  },

  // Test stages for different load patterns
  LOAD_STAGES: {
    ramp_up: [
      { duration: '10s', target: 5 },
      { duration: '30s', target: 10 },
      { duration: '20s', target: 20 },
      { duration: '10s', target: 0 },
    ],
    stress: [
      { duration: '30s', target: 50 },
      { duration: '1m', target: 100 },
      { duration: '30s', target: 50 },
      { duration: '10s', target: 0 },
    ],
    spike: [
      { duration: '10s', target: 10 },
      { duration: '5s', target: 100 },
      { duration: '10s', target: 10 },
      { duration: '5s', target: 0 },
    ],
  },
};

// Helper functions
export function getRandomMessage() {
  return config.TEXT_MESSAGES[
    Math.floor(Math.random() * config.TEXT_MESSAGES.length)
  ];
}

export function getLoadUrl() {
  return `http://${config.HOST}:${config.PORT}/load`;
}

export function getUnloadUrl() {
  return `http://${config.HOST}:${config.PORT}/unload`;
}

export function getWebSocketUrl() {
  return `ws://${config.HOST}:${config.PORT}/session`;
}

export function generateTestUser() {
  const userId = Math.random().toString(36).substr(2, 9);
  return {
    id: userId,
    name: `LoadTestUser_${userId}`,
    key: `session_${userId}_${Date.now()}`,
  };
}
